---
title: "Tree-sequence processing and statistics"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Tree-sequence processing and statistics}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
env_present <- "slendr-env" %in% reticulate::conda_list()$name

knitr::opts_chunk$set(
  collapse = FALSE,
  comment = "#>",
  fig.width = 6,
  fig.height = 4,
  dpi = 80,
  eval = Sys.which("slim") != "" && env_present
)
```

In this vignette, we will show how to specify sampling events to record individuals in the tree-sequence output file (a procedure which is called "remembering" of individuals in the SLiM context) and how to perform simple analyses using *slendr*'s interface to the [*tskit*](https://tskit.dev) Python library. We will demonstrate these features on a simulation of Neanderthal introgression into anatomically modern humans. Specifically, we will show how to estimate the amount of Neanderthal ancestry using $f$-statistics calculated directly on the tree-sequence data structure generated by a *slendr* model, all entirely from R.

### Model of Neanderthal introgression into Eurasians

First, let's set up a simple *non-spatial* model of Neanderthal introgression using *slendr.* This is essentially the same procedure which we have shown in another vignette introducing [non-spatial *slendr* models](vignette-04-nonspatial-models.html). This is no different from a spatial model, except that we left out the `map` argument in calling `population()`.

```{r}
library(slendr)

library(ggplot2)
library(dplyr)

set.seed(314159)

# create the ancestor of everyone and a chimpanzee outgroup
# (we set both N = 1 to reduce the computational time for this model)
anc <- population("ANC", time = 6.5e6, N = 1)
chimp <- population("CH", parent = anc, time = 6e6, N = 1)

# two populations of anatomically modern humans: Africans and Europeans
afr <- population("AFR", parent = anc, time = 600e3, N = 10000)
eur <- population("EUR", parent = afr, time = 70e3, N = 5000)

# Neanderthal population splitting at 600 ky ago from modern humans
# (becomes extinct by 40 ky ago)
nea <- population("NEA", parent = anc, time = 600e3, N = 1000, remove = 40e3)

# 3% Neanderthal introgression into Europeans between 55-50 ky ago
gf <- geneflow(from = nea, to = eur, rate = 0.03, start = 55000, end = 50000)

model <- compile(
  populations = list(anc, chimp, nea, afr, eur), geneflow = gf,
  generation_time = 30,
  dir = file.path(tempdir(), "introgression"), overwrite = TRUE
)
```

Here's our toy model visualized as a graph. Not particularly illuminating in this simple example, but it's always worth keeping in mind that such graph is embedded within every *slendr* model and can be always invoked to make sure the model you're setting up is correct:

```{r, introgression_graph, fig.width = 6, fig.height = 4}
plot_graph(model) + ggplot2::theme(legend.position = "none")
```

## Scheduling of sampling events

Now that we have defined a model, how do we sample data from it? Ideally, we would like to to schedule sampling events at a given time, sampling a defined number of individuals from a given population. This is why *slendr* provides a function `sampling()` which serves to define such sampling schedule automatically and enforces that only populations which are already (i.e. after their appearance in the simulation) or still (before they are removed from the simulation) present will be sampled from.

In our example, we want to sample two Neanderthal individuals (the older one being the [Altai Neanderthal](https://www.nature.com/articles/nature12886), the younger one [Vindija Neanderthal](https://www.science.org/doi/10.1126/science.aao1887)). These two genomes are what we [need](https://www.pnas.org/content/116/5/1639) to estimate Neanderthal ancestry proportion using a so-called $f_4$-ratio statistic (more on that below):

```{r}
nea_samples <- sampling(model, times = c(70000, 40000), list(nea, 1))
nea_samples
```

As you can see, the `sampling()` function simply accepts the vector of times at which remembering should be schedule, and then a list of pairs `(<slendr population>, <number of individuals>)` encoding from which populations should how many individuals be remembered at time points given in the `times` vector.

Next, we want to sample some present-day individuals: an outgroup representing a chimpanzee, and a couple of Africans and Europeans:

```{r}
present_samples <- sampling(model, times = 0, list(chimp, 1), list(afr, 5), list(eur, 10))
present_samples
```

As you can see above, the `sampling()` function returns a plain old data frame with a very simple structure with three columns: time, population name, and the number of individuals. This means that you can define sampling events using whatever input data you might already have available (such as radiocarbon-dated ancient DNA samples from an Excel sheet from some publication). For instance, there has been a [lot](https://www.nature.com/articles/nature17993) of [interest](https://www.pnas.org/content/116/5/1639) to estimate the trajectory of Neanderthal ancestry in Europe over time using ancient DNA data from anatomically modern human individuals (also called early modern humans, EMH) across the last couple of tens of thousands of years. We can simulate something close to the available [EMH ancient DNA data set](https://www.nature.com/articles/nature17993) over the last 50 thousand years by running doing this:

```{r}
emh_samples <- sampling(model, times = runif(n = 40, min = 10000, max = 40000), list(eur, 1))
emh_samples
```

This samples a single ancient European individuals at randomly chosen times between 40 and 10 ky ago.

One nice feature of the `sampling()` function is that it only schedules sampling events for a population, if that population is present in the simulation at a given time. This makes it possible to simply take a wide time range for sampling, specify all populations and sizes of the samples, and let the function generate sampling events only for populations present at each time. If for some reason a stricter control over sampling is required, this behavior can be switched off by setting `strict = TRUE` like this:

```{r, eval = FALSE}
# this attempts to sample a Neanderthal individual at a point when Neanderthals
# are already extinct, resulting in an error
sampling(model, times = 10000, list(nea, 1), strict = TRUE)
```

    Error: Cannot schedule sampling for 'NEA' at time 10000 because the population will not be present in the simulation at that point. Consider running this function with `strict = FALSE` which will automatically retain only valid sampling events.

Now that we already have the `model` object ready, we can simulate data from it, sampling individuals according to our sampling schedule. We do this by calling the `slim()` function as usual and specifying the sampling events with the `sampling =` argument. Note that we bind the individual sampling schedule data frames using the `rbind` function provided by base R (as we show above, the sampling schedule really is just a data frame and we can manipulate it as such):

```{r}
slim(
  model, sequence_length = 10e6, recombination_rate = 1e-8,
  sampling = rbind(nea_samples, present_samples, emh_samples),
  method = "batch", verbose = FALSE, random_seed = 314159
)
```

## R interface for *tskit* and *pyslim*

When we execute the model, *slendr* will instruct SLiM to save the output of the simulation as a tree sequence file. By default, the file is stored in the model directory:

```{r}
ts_file <- file.path(model$path, "output_slim.trees")
file.exists(ts_file)
```

Tree-sequences are one of the most revolutionary developments in population genetics in the last couple of decades for a number of reasons. One of them is the possibility to store extremely large data sets succinctly by encoding the entire evolutionary history of a sample of individuals as a series of correlated tree genealogies along the genome.

Going into too much detail on this topic is clearly beyond the scope of this tutorial, especially because everything is explain much better [elsewhere](https://tskit.dev/learn.html). Instead, what we will demonstrate in the rest of this vignette is how you can access and manipulate tree-sequence outputs generated by *slendr* models and perform various statistics on them using Python modules [*tskit*](https://tskit.dev/tskit/docs/stable/) and [*pyslim*](https://pyslim.readthedocs.io/en/latest/index.html) directly from *slendr,* without having to leave R! The key is a magical R package [*reticulate*](https://rstudio.github.io/reticulate/index.html) which creates a seamless binding of Python modules with R. This means that even if you don't know Python, *slendr* allows you to do do quite a lot with tree-sequences in R.

Of course, if you are a proficient Python user, it needs to be said that once you have the tree-sequence file generated by *slendr* & SLiM, you can easily perform every conceivable analysis directly using *tskit*. The intention here is to show how you can continue working on the tree-sequence files in R even after you have run the entire *slendr* simulation.

First, in order to be able to interface with *tskit* and *pyslim* using the *reticulate* package, you will need a working Python environment with the required Python modules `pyslim`, `tskit` and `msprime` already installed. I personally manage Python installations and separate package environments using [pyenv](https://github.com/pyenv/pyenv) and [pyenv-virtualenv](https://github.com/pyenv/pyenv-virtualenv). However, this is only my own personal preference. If you prefer using conda or any [other Python management solution](https://xkcd.com/1987/), feel free to use that. In any case, I recommend using tooling which is supported by [reticulate](https://rstudio.github.io/reticulate/index.html) itself, because it will be easy to find support should you run into any trouble.

Because setting up Python environments can be quite a hassle, _slendr_ provides a single function `setup_env()`. If you call it without any arguments, _slendr_ will automatically download, install, and setup a **completely separate** Python environment just for _slendr_ and activate it in the background. Even if you're an advanced Python user, I recommend going through this route because it will make sure that all dependencies are at their right versions that have been tested in _slendr_.

(If you insist on managing your own Python environment, that's OK too. Take a look at [this guide](https://CRAN.R-project.org/package=reticulate/vignettes/versions.html). You will be interested in the function `reticulate::user_virtualenv()` or `reticulate::use_condaenv()`.)

Let's now activate the _slendr_ Python environment (potentially installing it if it's not present yet):

```{r}
setup_env()
```

## Loading and processing tree-sequence output files

With the technicalities out of the way, we can now load the tree-sequence file saved by SLiM using the *slendr* function `ts_load()`. Optionally, we can also instruct this function to simplify the tree-sequence to only the individuals that we explicitly sampled (recall the sampling schedule we set up with the `sampling()` function above). What *slendr* does here is provide an R-friendly interface to *tskit* and *pyslim*, essentially reproducing steps in [this tutorial](https://tskit.dev/pyslim/docs/latest/tutorial.html) in a single R command. Note that we have to provide the `model` object generated by `compile()` above in order to have all model annotation information for the simulated tree-sequence data (we have to do this only once, and only during loading):

```{r}
ts <- ts_load(model)
ts
```

We can see that the tree-sequence data is very large and contains many more individuals than we need (much more than we explicitly specified for sampling)&mdash; a lot more than what we scheduled for sampling! To avoid this issue, `ts_load()` accepts an optional argument `simplify` which triggers the [simplification process](https://tskit.dev/pyslim/docs/latest/tutorial.html#simplification) to narrow down the set of recorded individuals just to those that we explicitly remembered:

```{r}
ts <- ts_load(model, simplify = TRUE)
ts
```

By default, simplification in `ts_load()` simplified the tree-sequence down to only remembered individuals (i.e. those we explicitly scheduled for sampling). We can also narrow down the simplification to a defined set of individuals using the `simplify_to = ` argument. Internally, simplification is implemented in a dedicated function `ts_simplify()` which we can always call explicitly, like this:

```{r}
ts_small <- ts_simplify(ts, simplify_to = c("CH_1", "NEA_1", "NEA_2", "AFR_1", "AFR_2", "EUR_20", "EUR_50"))
ts_small
```

Similarly, *slendr* provides a function `ts_recapitate()` which performs [recapitation]https://tskit.dev/pyslim/docs/latest/tutorial.html#recapitation). Again, we can do this in one go, by specifying `recapitate = TRUE` in the call to `ts_load()` after specifying a couple of additional arguments required for recapitation (see the pyslim documentation in the [recapitation](https://tskit.dev/pyslim/docs/latest/tutorial.html#recapitation) section for more detail):

```{r, eval = FALSE}
ts <- ts_load(model, recapitate = TRUE, simplify = TRUE,
              recombination_rate = 1e-8, Ne = 10000)
```

If we were to run this code chunk, we would get a message informing that the recapitation was not necessary because all genealogies in the tree-sequence are already coalesced (i.e. have each only one root node). We can do this check ourselves by calling another *slendr* function `ts_coalesced()` which verifies that all trees in the sequence are fully coalesced:

```{r}
ts_coalesced(ts)
```

You might have noticed that we did not simulate any mutations during the SLiM run. This is for computational efficiency. Luckily, the tree-sequence contains the complete history of a sample of individuals which makes it very easy to sprinkle mutations on the genealogies after this simulation is over. We can add mutations a given rate by running:

```{r}
ts <- ts_mutate(ts, mutation_rate = 1e-8, random_seed = 314159)
ts
```

Even this step can be performed in one go throught the `ts_load` function like this:

```{r, eval = FALSE}
ts <- ts_load(model, recapitate = TRUE, simplify = TRUE, random_seed = 314159,
              recombination_rate = 1e-8, Ne = 10000, mutation_rate = 1e-8)
```

Having done that, we can calculate some basic statistics on our simulated data.

However, before we do that, we would first like to note that everything that we do in the rest of this vignette (i.e. whenever we call a function with the prefix `ts_*()` in *slendr*), we are interfacing with the *tskit* Python module under the hood. Our goal is to capture most of the analyses one might want to perform on tree-sequences in R and wrap them in a neat interface indistinguishable from any other R function---this is, after all, the reason why *reticulate* has been created in the first place (making various Python data science modules appear as if they were regular R packages).

## Visualisation of trees and tree-sequences

-----

**Note:** The visualisation of trees and tree-sequences in *slendr* is currently still work-in-progress. For a much more comprehensive overview of SVG plotting in *tskit* see the [official tutorial](https://tskit.dev/tutorials/viz.html) on the subject.

-----

We have spend a lot of time talking about how powerful the *tskit* library is. Another great example is tree(-sequence) visualisation. *tskit* [implements](https://tskit.dev/tutorials/viz.html) visualisation of data in an [SVG format](https://en.wikipedia.org/wiki/Scalable_Vector_Graphics) which we can use for quick interactive plotting directly from the R console. First, we introduce a function `ts_tree()` which allows us to extract one tree from the tree-sequence (either an *i*-th tree in the sequence, or a tree overlapping an *i*-th position of the simulated genome, depending on the value of its `mode` argument):

```{r}
# take the 1st tree in the sequence (mode = "index" by default)
tree <- ts_tree(ts_small, i = 1)
tree
```

We can then visualize the tree by calling:

```{r, tree}
ts_draw(tree, width = 1000, height = 600)
```

If we specify `labels = TRUE` and provide the tree-sequence object itself (because that carries all the required model metadata such as individual names), we get a slightly more readable figure. We can also provide keyword arguments specific to *tskit* `draw_svg()` function, such as `time_scale` here:

```{r, tree_labels}
ts_draw(tree, width = 1000, height = 600, labels = TRUE, ts = ts_small, time_scale = "rank")
```


## Accessing *tskit* functionality directly

As we mentioned in the previous section, the goal of this vignette is to show how to use *slendr* to perform the main tree-sequence operations using a convenient R interface to *tskit*. However, you should always keep in mind that you are not restricted to a subset of *tskit* functionality that *slendr* translated to R (i.e. [all functions](../reference/index.html) with the prefix `ts_`). Thanks to the incredible R package [*reticulate*](https://rstudio.github.io/reticulate/index.html), you can access Python methods and object variables directly, using the `$` operator.

As an example, instead of calling the function `ts_coalesced()` on the tree-sequence as we did above, we could check that all trees are coalesced by running the following snippet instead (note that this is very inefficient and we're only doing the operation for the first one hundred trees):

```{r}
# iterate over all trees in the tree-sequence and check if each
# has only one root (i.e. is fully coalesced) - note that Python
# lists are 0-based, which is something we need to take care of
all(sapply(seq_len(ts$num_trees)[1:100],
           function(i) ts$at_index(i - 1)$num_roots == 1))
```

We believe it makes sense to use the R interface whenever possible (even if only because it makes many operations a little bit more convenient). However, if there is some functionality in *slendr* missing, you can always resort to accessing the Python objects directly as we have just demonstrated. You can verify that all [methods and attributes](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence) of a Python tree-sequence object is still accessible in R:

```{r}
names(ts)
```

In fact, you will recognize some of the elements in the output above from examples involving `ts_` functions in this vignette! In short, there is no blackbox---*slendr* only provides a slightly more convenient layer over *tskit* for R users.

## Calculating *f*-statistics

In addition to being a revolutionary breakthrough in terms of computation efficiency, many statistics that we are often interested in population genetics are a natural consequence of having a direct access to tree sequence genealogies, simply because those genealogies capture the true demographic history of a sample. Again, we can't go into too much detail here but we encourage you to take a look at a paper by [Ralph *et al.*](https://doi.org/10.1534/genetics.120.303253) on the duality between statistics expressed in terms of branch lengths and the traditional summaries based on samples of genetic variation.

For instance, we have functions such as `ts_f2()`, `ts_f3()`, `ts_f4()` and `ts_f4ratio()` which calculate the well-known set of Patterson's $f$-statistics:

```{r}
# f2 is a measure of the branch length connecting A and B
ts_f2(ts, A = "EUR_1", B = "AFR_1")

# f4 is a measure of the drift shared between A and B after their split from C
ts_f3(ts, A = "EUR_1", B = "AFR_1", C = "CH_1")

# this value should be very close to zero (no introgression in Africans)
ts_f4(ts, "AFR_1", "AFR_2", "NEA_1", "CH_1", mode = "branch")

# this value should be significantly negative (many more ABBA sites
# compared to BABA site due to the introgression into Europeans)
ts_f4(ts, "AFR_1", "EUR_1", "NEA_1", "CH_1", mode = "branch")
```

These functions accept a `mode =` argument, specifying whether the statistics should be calculated using mutation site patterns (`mode = "site"`, the default), branch lengths (`mode = "branch"`), or for each node (`mode = "node"`), as well as the `windows` argument, similarly to other "multiway" statistics implemented by tskit. See the [relevant](https://tskit.dev/tskit/docs/stable/stats.html#sec-stats-mode) [sections](https://tskit.dev/tskit/docs/stable/stats.html#sec-stats-sample-sets-multi-way) of the official *tskit* documentation for more on this topic.

Note that in the previous chunk we refered to individuals by their names (not numeric IDs of nodes as you would do it with *tskit* in Python). We allow this for readability and to make it easier to see which individuals are which based on the specified sampling schedule (the names are assigned to individuals based on the order of their sampling). We can get an overview of the individuals scheduled for sampling (i.e. permanently remembered) and their names with a helper function `ts_samples()`:

```{r}
ts_samples(ts)
```

That said, if you would like to run some statistics on nodes rather than on individuals, you can do it simply by using integer IDs instead of character names in each function's interface.

## Estimating Neanderthal ancestry proportions

Let's try to put these new tools to practice and estimate the proportion of Neanderthal ancestry in Africans and Europeans in our simulated data. We can do this using the Patterson's $f_4$-ratio statistic implemented in the `ts_f4ratio()` function in *slendr* (you can find more information about this particular version of the statistic [in this paper](https://www.pnas.org/content/116/5/1639)):

```{r}
# first get a table of simulated African and European individuals in the tree-sequence
inds <- ts_samples(ts) %>% dplyr::filter(pop %in% c("AFR", "EUR"))

# estimate the amounts of Neanderthal ancestry in these individuals and add
# these values to the table
inds$ancestry <- ts_f4ratio(ts, X = inds$name, "NEA_1", "NEA_2", "AFR_1", "CH_1")$alpha
```

If we now summarise the inferred Neanderthal distribution in both populations, we see that there is no Neanderthal ancestry in Africans (as expected by our model--Africans did not receive a Neanderthal introgression pulse) but there is a small proportion of Neanderthal ancestry in Europeans (consistent with the 3% introgression pulse we simulated between):

```{r, ts_nea_distributions}
ggplot(inds, aes(pop, ancestry, fill = pop)) +
  geom_boxplot() +
  geom_jitter() +
  labs(y = "Neanderthal ancestry proportion", x = "") +
  theme(legend.position = "none") +
  coord_cartesian(ylim = c(0, 0.1))
```

This is exactly as we specified in the model configuration above, suggesting that our simulations work as they should. You can see that there is quite a bit of noise but that's because we simulated only a small amount of sequence.

We can also plot the trajectory of Neanderthal ancestry in Europe during the time-window for which we have simulated ancient and present-day DNA samples:

```{r, ts_nea_trajectory}
dplyr::filter(inds, pop == "EUR") %>%
  ggplot(aes(time, ancestry)) +
  geom_point() +
  geom_smooth(method = "lm", linetype = 2, color = "red", size = 0.5) +
  xlim(40000, 0) + coord_cartesian(ylim = c(0, 0.1)) +
  labs(x = "time [years ago]", y = "Neanderthal ancestry proportion")
```

Again, this is a result consistent with empirical estimates of Neanderthal ancestry using [ancient DNA data](https://www.pnas.org/content/116/5/1639).

## ADMIXTOOLS analyses

In case you would like to verify some *f*-statistics results using the venerable [ADMIXTOOLS](https://doi.org/10.1534/genetics.112.145037) software (see the linked paper which formally introduced these statistics in the first place), you can convert the tree-sequence data to a file format called EIGENSTRAT using the `ts_eigenstrat()` function. The file conversion is internally handled by the R package [*admixr*](https://bodkan.net/admixr/) and returns an `EIGENSTRAT` object which ties all individual `EIGENSTRAT` file components together (see the [tutorial](https://bodkan.net/admixr/articles/tutorial.html) to *admixr* for an extensive overview). *admixr* is an R package for running automated ADMIXTOOLS analyses entirely from R and makes these types of analyses very convenient.

```{r}
snps <- ts_eigenstrat(ts, prefix = file.path(tempdir(), "eigenstrat", "data"))
```

Running an *admixr* analysis is then as easy as plugging the object into an *admixr* function. For instance, we can estimate the proportion of Neanderthal ancestry in a couple of individuals $X$ like this (*admixr* calls this proportion `alpha`):

```{r}
library(admixr)

f4ratio(data = snps, X = c("EUR_1", "EUR_2", "AFR_2"),
        A = "NEA_1", B = "NEA_2", C = "AFR_1", O = "CH_1")
```

In fact, lets compare the values obtained by both *tskit* and *admixr*/ADMIXTOOLS for all individuals:

```{r, ts_vs_admixr}
europeans <- inds[inds$pop == "EUR", ]$name

# tskit result
result_ts <- ts_f4ratio(ts, X = europeans, A = "NEA_1", B = "NEA_2", C = "AFR_1", O = "CH_1") %>% select(alpha_ts = alpha)

# result obtained by admixr/ADMIXTOOLS
result_admixr <- f4ratio(snps, X = europeans, A = "NEA_1", B = "NEA_2", C = "AFR_1", O = "CH_1") %>% select(alpha_admixr = alpha)

bind_cols(result_admixr, result_ts) %>%
  ggplot(aes(alpha_ts, alpha_admixr)) +
  geom_point() +
  geom_abline(slope = 1, linetype = 2, color = "red", size = 0.5) +
  labs(x = "f4-ratio statistic calculated with admixr/ADMIXTOOLS",
       y = "f4-ratio statistic calculated with tskit")
```

The correspondence between the two looks good! 🎉 Again, note that the large amount of variance around the expected value of 3% ancestry is due to an extremely small amount of sequence data simulated here.

## VCF output

In case you need to process simulated data in some other software, you can use the function `ts_vcf()` to save the simulated genotypes in a VCF format:

```{r}
ts_vcf(ts, path = file.path(tempdir(), "output.vcf.gz"))
```

You can also specify only a subset of individuals to be saved in the VCF:

```{r}
ts_vcf(ts, path = file.path(tempdir(), "output_subset.vcf.gz"),
       individuals = c("CH_1", "NEA_1", "EUR_1", "AFR_1"))
```

## Other statistics

What follows is a very brief overview of other statistics which are implemented in *tskit* and for which *slendr* provides an easy-to-use R interface. As you will see, the goal of these functions is to get you to a result using a single function call, making them very convenient for quick interactive exploratory analyses on the simulated data right in the R console.

We will continue to use our simulated Neanderthal introgression tree-sequence data for these examples.

### $F_{st}$

The $F_{st}$ statistic is implemented by the function `ts_fst()`.

If a single genome-wide $F_{st}$ is to be calculated (i.e. not a window-based calculation), the `ts_fst()` returns a simple three-column data frame

```{r}
ts_fst(ts, sample_sets = list(afr = c("AFR_1", "AFR_2", "AFR_3"), eur = c("EUR_1", "EUR_2")))
```

In case a non-named list of sample sets was provided, set names are generated automatically:

```{r}
ts_fst(ts, sample_sets = list(c("AFR_1", "AFR_2", "AFR_3"), c("EUR_1", "EUR_2")))
```

Of course, this is much less readable and we encourage you to name the sample sets appropriately.

In case more than two sample sets are specified, all pairwise statistics are computed:

```{r}
ts_fst(ts, sample_sets = list(afr = c("AFR_1", "AFR_2", "AFR_3"),
                              eur = c("EUR_1", "EUR_2"),
                              nea = c("NEA_1", "NEA_2")))
```

As with many other statistics implemented by [*tskit*](https://tskit.dev/tskit/docs/stable/python-api.html)*,* `ts_fst()` accepts a `windows` argument, specifying the breakpoints between windows. In this case, the `Fst` column in the resulting data frame is a so called "list-column", with each item in the column being a vector of $F_{st}$ values, one per each window. List-columns can be a little confusing for new R users, but we highly encourage you to get used to them as they allow extremely consise and elegant handling of structured data within normal data frames (you can start with [this](https://jennybc.github.io/purrr-tutorial/ls13_list-columns.html) introduction).

```{r}
# define breakpoints between 20 windows
breakpoints <- seq(0, ts$sequence_length, length.out = 21)

# calculate window-based Fst statistic
win_fst <- ts_fst(
  ts, windows = breakpoints,
  sample_sets = list(afr = c("AFR_1", "AFR_2", "AFR_3"),
                     eur = c("EUR_1", "EUR_2"),
                     nea = c("NEA_1", "NEA_2"))
)

# we get 20 values for each parwise calculation
win_fst
```

For instance, here are window-based $F_st$ values for the `afr-vs-eur` calculation (first row of the table above):

```{r}
win_fst[1, ]$Fst
```

### Tajima's $D$

The function `ts_tajima()` has nearly the same interface as `ts_fst()` shown above.

If a non-window version is to be calculated, we get a single genome-wide values for each sample set (named or non-named list of character vectors with individual names):

```{r}
ts_tajima(ts, list(afr = c("AFR_1", "AFR_2", "AFR_3"), eur = c("EUR_1", "EUR_2")))
```

For window-based version, the function returns the `D` column as a list column of vectors with $i$-th element being the Tajima's D value for the $i$-th window:

```{r}
ts_tajima(ts, list(afr = c("AFR_1", "AFR_2"), eur = c("EUR_1", "EUR_2")), windows = breakpoints)
```

### Diversity

We can calculate diversity within given groups of individuals with the function `ts_diversity()`. For instance, even in our extremely simplified example, we would expect the highest levels of diversity in Africans, followed by Europeans, Neanderthals and the "degenerate" single individual outgroup "chimpanzee". Is this true? Let's find out.

First we extract individuals from all populations, creating a list of character vectors for each group (which is what functions such as `ts_diversity()` expects as an input):

```{r}
# get sampled individuals from all populations
sample_sets <- ts_samples(ts) %>%
  split(., .$pop) %>%
  lapply(function(pop) pop$name)

sample_sets
```

Now we can calculate diversity in each population and sort the results in an increasing order of diversity:

```{r}
ts_diversity(ts, sample_sets) %>% dplyr::arrange(diversity)
```

Great! This matches our expectations. We simulated chimp "population" as only one individual, so we expect essentially no diversity after millions of years of evolution.

### Divergence

We can calculate pairwise divergence between groups of individuals using the function `ts_divergence()`. Given our model, we would expect the lowest divergence between the two modern human groups `AFR` and `EUR`, then between Neanderthals and the two modern humans, and all three groups (`AFR`, `EUR` and `NEA`) should have equal, much deeper divergence from the outgroup chimpanzee `CH`.

```{r}
ts_divergence(ts, sample_sets) %>% arrange(divergence)
```

After sorting the table based on the value in the divergence column, we can see the results fit our expectations.

## More information

These were only a couple of examples of statistical functions implemented in *tskit* for which we provide a native R interface in *slendr*. You can find more tree-sequence statistics in the [reference manual](../reference/index.html#section-tree-sequence-statistics) on the project website. Not [all statistics](https://tskit.dev/tskit/docs/stable/stats.html#available-statistics) from the *tskit* library are implemented, but we intend to expand the selection provided by *slendr* in the near future. If there is some functionality that you would like to use in your project missing in *slendr*, please don't hesitate to let us now by creating an issue on our [GitHub page](https://github.com/bodkan/slendr/issues).

Finally, if you would like to see more examples of *tskit* interface in action, take a look at the vignette which describes [switching between the SLiM and *msprime* back ends](vignette-07-backends.html) of the *slendr* package.
