---
title: "Tree-sequences, tskit interface, f-statistics"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Tree-sequences, tskit interface, f-statistics}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 6,
  fig.height = 4,
  dpi = 80,
  eval = TRUE
)

# for some reason knitr/reticulate sets this to default, sabotaging
# our efforts to make reticulate use our custom environment -- this
# appears to solve the problem
Sys.unsetenv("RETICULATE_PYTHON")

devtools::load_all(".")
```

In this vignette, we will show how to specify sampling events to record individuals in the tree-sequence output file (a procedure which is called "remembering" of individuals in the SLiM context) and how to perform simple analyses using *slendr*'s interface to the [*tskit*](http://tskit.dev) Python library. Specifically, we will demonstrate these features on a model of Neanderthal introgression into anatomically modern humans by showing how to estimate the amount of Neanderthal ancestry using $f$-statistics calculated directly on the tree-sequence data structure entirely from R.

### Model of Neanderthal introgression into Eurasians

First, let's set up a simple *non-spatial* model of Neanderthal introgression using *slendr.* This is essentially the same procedure which we have shown in another vignette introducing [non-spatial *slendr* models](vignette-04-nonspatial-models.html).

```{r}
library(slendr)

# create the ancestor of everyone, and a chimpanzee outgroup
# (both N = 1 to reduce the computational time for this model, as this
# doesn't affect the results for this demonstration)
anc <- population("ANC", time = 6.5e6, N = 1)
chimp <- population("CH", parent = anc, time = 6e6, N = 1)

# two populations of anatomically modern humans: Africans and Europeans
afr <- population("AFR", parent = anc, time = 600e3, N = 10000)
eur <- population("EUR", parent = afr, time = 70e3, N = 5000)

# Neanderthal population splitting at 0.5 Mya from the common ancestor with
# anatomically modern humans (and extinct by 40 kya)
nea <- population("NEA", parent = anc, time = 600e3, N = 1000, remove = 40e3)

# 3% Neanderthal introgression into Europeans 50-45 kya
gf <- geneflow(from = nea, to = eur, rate = 0.03, start = 55000, end = 50000)

model <- compile(
  populations = list(anc, chimp, nea, afr, eur), geneflow = gf,
  generation_time = 30,
  dir = file.path("/tmp/", "introgression"), overwrite = TRUE
  #dir = file.path(tempdir(), "introgression"), overwrite = TRUE
)
```

Here's our toy model visualized as a graph. Not particularly illuminating in this simple example, but it's always worth keeping in mind that such graph is embedded within every *slendr* model and can be always invoked to make sure the model you're setting up is correct:

```{r, introgression_graph, fig.width = 6, fig.height = 4}
graph(model) + ggplot2::theme(legend.position = "none")
```

## Scheduling of sampling events

Now that we have defined a model, how do we sample data from it? Ideally, we would like to to schedule sampling events at a given time, sampling a defined number of individuals from a given population. This is why *slendr* provides a function `sampling()` which allows defining such sampling schedule automatically while, at the same time, enforcing that only populations which are already (i.e. after their appearance in the simulation) or still (before they are removed from the simulation) will be sampled from.

In our example, we want to sample two Neanderthal individuals (the older one being the [Altai Neanderthal](), then younger one [Vindija Neanderthal]()). This is what is required to estimate Neanderthal ancestyr proportion using a so-called $f_4$-ratio statistics (more on that below):

```{r}
nea_samples <- sampling(times = c(70000, 40000), list(nea, 1))
nea_samples
```

The `sampling()` function simply accepts the vector of times at which remembering should be schedule, and then a list of pairs `(<slendr population>, <number of individuals>)` encoding from which populations should how many individuals be remembered at times given in the `times` vector. Easy.

Next, we want to sample some present-day individuals: an outgroup representing a chimpanzee, and a couple of Africans and Europeans:

```{r}
present_samples <- sampling(times = 0, list(chimp, 1), list(afr, 5), list(eur, 5))
present_samples
```

As you can see above, the `sampling()` function returns a plain old data frame with a very simple structure with three columns: time, population name, and the number of individuals. This means that you can define sampling events using whatever input data you might already have available (such as radiocarbon-dated ancient DNA samples from an Excel sheet from some publication). For instance, there has been a [lot](https://www.nature.com/articles/nature17993) of [interest](https://www.pnas.org/content/116/5/1639) to estimate the trajectory of Neanderthal ancestry in Europe over time using ancient DNA data from anatomically modern human individuals (also called early modern humans, EMH) across the last couple of tens of thousands of years. We can simulate something close to the available [EMH ancient DNA data set](https://www.nature.com/articles/nature17993) over the last 50 thousand years by running something like this:

```{r}
emh_samples <- sampling(times = runif(n = 50, min = 10000, max = 40000), list(eur, 1))
emh_samples
```

This samples a single ancient European individuals at randomly chosen times between 40 and 10 ky ago.

One nice feature of the `sampling()` function is that it only schedules sampling events for populations if that population is present in the simulation at a given time. This makes it possible to simply specify the whole time range for sampling, specify all populations and sizes of the samples, and let the function generate sampling events only for populations present at each time. If for some reason a stricter control over sampling is required, this behavior can be switched off by setting `strict = TRUE` like this:

```{r, eval = FALSE}
# this attempts to sample a Neanderthal individual at a point when Neanderthals
# are already extinct, resulting in an error
sampling(times = 10000, list(nea, 1), strict = TRUE)
```

    Error: Cannot schedule sampling for 'NEA' at time 10000 because the population will not be present in the simulation at that point. Consider running this function with `strict = FALSE` which will automatically retain only valid sampling events.

Now that we already have the `model` object ready, we can simulate data from it, sampling individuals according to our sampling schedule. We do this by calling the `slim()` function as usual, but this time we set `ts_recording = TRUE` (switching of tree-sequence recording in SLiM) and we specify the sampling events with the `sampling =` argument. Note that we bind the individual sampling schedule data frames using the `rbind` function provided by base R (as we show above, the sampling schedule really is just a data frame and we can manipulate it as such):

```{r}
slim(
  model, seq_length = 10e6, recomb_rate = 1e-8,
  ts_recording = TRUE,
  sampling = rbind(nea_samples, present_samples, emh_samples),
  method = "batch", verbose = TRUE
)
```

## R interface for *tskit*, *pyslim* and *msprime*

The result of switching on `ts_recording = TRUE` in the `slim()` call above is that SLiM will save the output of the simulation as a tree sequence file. By default, the file is stored in the model directory:

```{r}
ts_file <- file.path(model$path, "output_tree_seq.trees")
file.exists(ts_file)
```

Tree-sequences are one of the most revolutionary developments in population genetics in the last couple of decades for a number of reasons. One of them is the possibility to store extremely large data sets succinctly by encoding the entire evolutionary history of a sample of individuals as a series of correlated tree genealogies along the genome.

Going into too much detail on this topic is clearly beyond the scope of this tutorial, especially because everything is explain much better [elsewhere](https://tskit.dev/learn.html). Instead, what we will demonstrate in the rest of this vignette is how you can access and manipulate tree-sequence outputs generated by *slendr* models and perform various statistics on them using Python modules [tskit](https://tskit.dev/tskit/docs/stable/) and [pyslim](https://pyslim.readthedocs.io/en/latest/index.html) (used to access and manipulate tree-sequence files produced by SLiM) from *slendr,* without having to leave R! The key is a magical R package [reticulate](https://rstudio.github.io/reticulate/index.html) which creates seamless binding Python modules with R. This means that even if you don't know Python, *slendr* allows you to do do quite a lot with tree-sequences in R.

Of course, if you are a proficient Python user, it needs to be said that once you have the tree-sequence file generated by *slendr* & SLiM, you can easily perform every conceivable analysis directly using *tskit*. The intention here is to show how you can continue working on the tree-sequence files in R even after you have run the entire *slendr* simulation.

First, in order to be able to interface with *tskit* and *pyslim* using the *reticulate* package, you will need a working Python environment with the required Python modules already installed. I personally manage Python installations and separate package environments using [pyenv](https://github.com/pyenv/pyenv) and [pyenv-virtualenv](https://github.com/pyenv/pyenv-virtualenv). However, this is only my own personal preference. If you prefer using conda or any [other Python management solution](https://xkcd.com/1987/), feel free to use that. In any case, I recommend using tooling which is supported by [reticulate](https://rstudio.github.io/reticulate/index.html) itself, because it will be easy to find support should you run into any trouble.

For completeness, here is how I have [just]{.ul} installed *tskit* and *pyslim* into a brand new environment on my Mac using pyenv:

``` {.bash}
# 1. install the most recent version of Python (enforcing the compilation of
# Python as a shared library is important for the reticulate package to work,
# at least on a Mac)
env PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 3.9.6

# create a separate Python environment dedicated to our R/Python interface
# and call it 'retipy'
pyenv virtualenv 3.9.6 retipy

# install required Python modules into the environment
pyenv activate retipy
pip install tskit pyslim
```

Now we can instruct the *reticulate* package to use the newly created Python environment:

```{r}
reticulate::use_virtualenv("~/.pyenv/versions/retipy", required = TRUE)
```

We can make sure that *reticulate* is indeed using our environment by calling:

```{r}
reticulate::py_config()
```

## Loading and processing tree-sequence output files

With the technicalities out of the way, we can now load the tree-sequence file saved by SLiM using the *slendr* function `ts_load()`. Optionally, we can also instruct this function to simplify the tree-sequence to only the individuals that we explicitly sampled (using the sampling schedule set up by `sampling()` above). What *slendr* does here is provide an R-friendly interface to *tskit* and *pyslim*, essentially reproducing steps in [this tutorial](https://pyslim.readthedocs.io/en/stable/tutorial.html) in a single step. Note that we have to provide the `model` object generated by `compile()` above:

```{r}
ts <- ts_load(ts_file, model)
ts
```

We can see that the tree-sequence data is very large and contains many more individuals than we need (i.e. that we explicitly specified for sampling): there are in total `r ts$num_individuals` individuals in the tree-sequence which is a lot more than the `r sum(rbind(nea_samples, present_samples, emh_samples)$n)` we scheduled for sampling! To avoid this issue, `ts_load()` accepts an optional argument `simplify` which triggers the [simplification process](https://pyslim.readthedocs.io/en/latest/tutorial.html#simplification) to narrow down the set of recorded individuals just to those that we explicitly remembered:

```{r}
ts <- ts_load(ts_file, model, simplify = TRUE)
```

Similarly, *slendr* provides a function `ts_recapitate()` which performs [recapitation](https://pyslim.readthedocs.io/en/latest/tutorial.html#recapitation). Again, we can do this in one go, by specifying `recapitate = TRUE` in the call to `ts_load()` after specifying a couple of additional arguments required for recapitation (see the pyslim documentation in the [recapitation](https://pyslim.readthedocs.io/en/latest/tutorial.html#recapitation) section for more detail):

```{r, eval = FALSE}
ts <- ts_load(ts_file, model, recapitate = TRUE, simplify = TRUE,
              recombination_rate = 1e-8, Ne = 10000)
```

In this case we get a message informing that the recapitation was not necessary because all genealogies in the tree-sequence are already coalesced (i.e. have each only one root node). We can do this check ourselves by another *slendr* function `ts_coalesced()`:

```{r}
ts_coalesced(ts)
```

For computational efficiency, we did not simulate any mutations during the SLiM run. In order to be able to calculate statistics from genetic variation, we can add mutations to the tree-sequence at a given rate by running:

```{r}
ts <- ts_mutate(ts, mutation_rate = 1e-8)
```

Having done that, we can calculate some basic statistics on our simulated data. We note that everything that we do here (every function with the prefix `ts_*()` in *slendr*) is, in fact, interfacing with the *tskit* Python module. Our goal is to capture the most basic of analyses one might want to perform on tree-sequences in R and wrap them in a neat interface indistinguishable from any other R function---this is, after all, the reason why *reticulate* has been created in the first place (making various Python data science modules appear as if they were regular R packages).

That said, [*slendr* doesn't try to capture everything that is possible to do in *tskit*]{.ul}. I personally love R and prefer to use it for as much of my work as possible, but for anything that does not involve calculating population genetics statistics on simulated tree genealogies it is better to use *tskit* directly. In other words, if you're developing a novel population genetic inference method on which operates on tree-sequence data, you should probably use *tskit* directly (either its C API or the Python interface), not R. This is why we haven't implemented *tskit*'s functionality for iterating over trees in the tree-sequence, traversing nodes, etc. It's likely that the number of *tskit* functions implemented in *slendr* will slowly grow (especially those related to population genetics statistics) but this is where we currently draw the line: *slendr* is an R package for simulating data (tree-sequences or other) in SLiM and for calculating statistics on it, nothing less, nothing more.

## Calculating *f*-statistics

In addition to being a revolutionary breakthrough in terms of computation efficiency, many population genetics statistics we are often interested in are a natural consequence of having a direct access to tree sequence genealogies simply because those capture the true demographic history of a sample. Again, we can't go into too much detail here but I encourage you to take a look at a paper by [Ralph *et al.*](https://www.genetics.org/content/215/3/779) on the duality between statistics expressed in terms of branch lengths and the traditional summaries based on samples of genetic variation.

For instance, we have functions such as `ts_f2()`, `ts_f3()`, `ts_f4()` and `ts_f4ratio()` which calculate the well-known set of Patterson's $f$-statistics:

```{r}
ts_f2(ts, "EUR1", "EUR2")

ts_f3(ts, "EUR1", "EUR2", "CH1")
ts_f3(ts, "EUR1", "AFR1", "CH1")
ts_f3(ts, "EUR1", "NEA1", "CH1")

ts_f4(ts, "AFR1", "AFR2", "NEA1", "CH1")
ts_f4(ts, "AFR1", "EUR1", "NEA1", "CH1")
```

These functions accept a `mode =` argument, specifying whether the statistics should be calculated using mutation site patterns (`mode = "site"`, the default) or branch lengths (`mode = "branch"`). See the [relevant section](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.write_vcf) of the official *tskit* documentation for more on this topic.

## Estimating Neanderthal ancestry proportions

Let's try to put these new tools to practice and estimate the proportion of Neanderthal ancestry in Africans and Europeans in our simulated data. We can do this using the Patterson's $f_4$-ratio statistic implemented in the `ts_f4ratio()` function in *slendr* (you can find more information about this particular version of the statistic [in this paper](https://www.pnas.org/content/116/5/1639)):

```{r}
# get a table of simulated African and European individuals in the tree-sequence
inds <- ts_individuals(ts) %>% dplyr::filter(pop %in% c("AFR", "EUR"))

# estimate the amounts of Neanderthal ancestry in these individuals
inds$ancestry <- ts_f4ratio(ts, X = inds$name, "NEA1", "NEA2", "AFR1", "CH1")$alpha
```

If we now summarise the inferred Neanderthal distribution in both populations, we see that there is no Neanderthal ancestry in Africans (as expected by our model--Africans did not receive a Neanderthal introgression pulse) but there is around 3% Neanderthal ancestry in Europeans (consistent with the 3% introgression pulse we simulated between 50-45 thousand years [ky] ago):

```{r}
ggplot(inds, aes(pop, ancestry, fill = pop)) +
  geom_boxplot() +
  geom_jitter() +
  labs(y = "Neanderthal ancestry proportion", x = "") +
  theme(legend.position = "none") +
  coord_cartesian(ylim = c(0, 0.1))
```

This is exactly as we specified in the model configuration above, suggesting that our simulations work as they should. You can see that there is quite a bit of noise but that's because we simulated only 10Mb of sequence.

We can also plot the trajectory of Neanderthal ancestry in Europe over the last 50 thousand years:

```{r}
inds[inds$pop == "EUR", ] %>%
  ggplot(aes(time, ancestry)) +
  geom_point() +
  geom_smooth(method = "lm", linetype = 2, color = "red", size = 0.5) +
  xlim(40000, 0) + coord_cartesian(ylim = c(0, 0.1)) +
  labs(x = "time [years ago]", y = "Neanderthal ancestry proportion")
```

Again, this is a result consistent with empirical estimates of Neanderthal ancestry using [ancient DNA data](https://www.pnas.org/content/116/5/1639).

## ADMIXTOOLS analyses

In case you would like to verify some *f*-statistics results using the venerable [ADMIXTOOLS](https://academic.oup.com/genetics/article/192/3/1065/5935193) software (see the linked paper which formally introduced these statistics in the first place), you can convert the tree-sequence data to a file format called EIGENSTRAT using the `ts_eigenstrat()` function. The file conversion is internally handled by the R package [*admixr*](http://bodkan.net/admixr) and returns an `EIGENSTRAT` object which ties all individual `EIGENSTRAT` file components together (see the [tutorial](https://bodkan.net/admixr/articles/tutorial.html) to *admixr* for an extensive overview). *admixr* is an R package for running automated ADMIXTOOLS analyses entirely from R and makes these types of analyses very convenient.

```{r}
snps <- ts_eigenstrat(ts, prefix = file.path(tempdir(), "eigenstrat", "data"))
snps
```

Running an *admixr* analysis is then as easy as plugging the object into an *admixr* function. For instance, we can estimate the proportion of Neanderthal ancestry in a couple of individuals $X$ like this (*admixr* calls this proportion `alpha`):

```{r}
library(admixr)

f4ratio(data = snps, X = c("EUR1", "EUR2", "AFR2"),
        A = "NEA1", B = "NEA2", C = "AFR1", O = "CH1")
```

In fact, lets compare the values obtained by both *tskit* and *admixr*/ADMIXTOOLS for all individuals:

```{r}
# tskit result
result_ts <- ts_f4ratio(ts, X = inds$name, A = "NEA1", B = "NEA2", C = "AFR1", O = "CH1") %>% dplyr::select(alpha_ts = alpha)

# result obtained by admixr/ADMIXTOOLS
result_admixr <- f4ratio(snps, X = inds$name, A = "NEA1", B = "NEA2", C = "AFR1", O = "CH1") %>% dplyr::select(alpha_admixr = alpha)

dplyr::bind_cols(result_ts, result_admixr) %>%
  ggplot(aes(alpha_ts, alpha_admixr)) +
  geom_point() +
  geom_abline(slope = 1, linetype = 2, color = "red", size = 0.5) +
  labs(x = "f4-ratio statistic calculated with admixr/ADMIXTOOLS",
       y = "f4-ratio statistic calculated with tskit")
```

The correspondence between the two looks pretty much perfect! 🎉 Again, not that the large amount of variance around the expected value of 3% ancestry is due to an extremely small data simulated here (only 10 Mb of sequence).

## VCF output

In case you need to process simulated data in some other software, you can use the function `ts_vcf()` to save the simulated genotypes in a VCF format:

```{r}
ts_vcf(ts, vcf = file.path(tempdir(), "output.vcf.gz"))
```

You can also specify only a subset of individuals to be saved in the VCF:

```{r}
ts_vcf(ts, vcf = file.path(tempdir(), "output_subset.vcf.gz"),
       individuals = c("CH1", "NEA1", "EUR1", "AFR1"))
```
