<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing non-slendr tree sequences • slendr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Analyzing non-slendr tree sequences">
<meta property="og:description" content="slendr">
<meta property="og:image" content="https://www.slendr.net/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slendr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-00-installation.html">Installation instructions</a>
    </li>
    <li>
      <a href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    </li>
    <li>
      <a href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    </li>
    <li>
      <a href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    </li>
    <li>
      <a href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    </li>
    <li>
      <a href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a>
    </li>
    <li>
      <a href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a>
    </li>
    <li>
      <a href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a>
    </li>
    <li>
      <a href="../articles/vignette-08-nonslendr-tskit.html">Analyzing non-slendr tree sequences</a>
    </li>
    <li>
      <a href="../articles/vignette-09-paper.html">Examples from the slendr paper</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/slendr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/dr_bodkan" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyzing non-slendr tree sequences</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/HEAD/vignettes/vignette-08-nonslendr-tskit.Rmd" class="external-link"><code>vignettes/vignette-08-nonslendr-tskit.Rmd</code></a></small>
      <div class="hidden name"><code>vignette-08-nonslendr-tskit.Rmd</code></div>

    </div>

    
    
<p>In previous vignettes we have demonstrated the <em>slendr</em> R
interface for defining and executing msprime and SLiM population genetic
models. We have also provided an overview for its interface to the
tree-sequence processing library <a href="https://tskit.dev" class="external-link">tskit</a>.</p>
<p>However, although its model-definition interface is quite convenient,
<em>slendr</em> cannot (and never will) support <em>every</em> possible
msprime or SLiM. model The array of features provided by these
simulation frameworks is simply to big and implementing an R interface
to every single one of them would not make much sense.</p>
<p>That said, the tree-sequence outputs produced by “pure”
(i.e. non-slendr) msprime and SLiM scripts are no different from those
generated by <em>slendr</em> models themselves. For users who would
rather use their own simulation scripts but find <em>slendr</em>’s tskit
R interface (not its simulation interface) appealing, the R package
provides a possibility to load, process, visualize and analyze standard
tree sequences which were not generated by <em>slendr</em> itself.</p>
<p>In this vignette we give a brief overview of how this works, using
example tree sequences produced by two very simple simulation scripts
written in pure SLiM and msprime (i.e. scripts which don’t utilize
<em>slendr</em>’s spatial features, symbolic names of simulated
individuals, and automated translation of time units). We won’t be going
into detail explaining how those scripts work, because we assume this
functionality will be used by those already familiar with msprime or
SLiM. Similarly, we won’t cover the R-tskit functionality of
<em>slendr</em> in detail either, simply because the contents of this
vignette is already covered by <a href="index.html">other tutorials
provided by <em>slendr</em></a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>

<span class="va">SEED</span> <span class="op">&lt;-</span> <span class="fl">42</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span></code></pre></div>
<div class="section level4">
<h4 id="non-spatial-slim-tree-sequences">Non-spatial SLiM tree sequences<a class="anchor" aria-label="anchor" href="#non-spatial-slim-tree-sequences"></a>
</h4>
<p>Consider the following SLiM script, which creates a couple of
populations (with different <span class="math inline">\(N_e\)</span>)
splitting from an ancestral population <code>p1</code> (lets save it to
<code>/tmp/nonspatial.slim</code>):</p>
<pre><code>initialize() {
    setSeed(42);
    initializeTreeSeq();
    initializeMutationRate(0);
    initializeMutationType("m1", 0.5, "f", 0.0);
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 1e6);
    initializeRecombinationRate(1e-8);
}

1 {
    sim.addSubpop("p1", 10);
}

1000 {
    sim.addSubpopSplit("p2", 500, p1);
}

3000 {
    sim.addSubpopSplit("p3", 2500, p1);
}

5000 {
    sim.addSubpopSplit("p4", 10000, p1);
}

6000 late() {
    sim.treeSeqOutput("/tmp/nonspatial-slim.trees");
}</code></pre>
<p>After we run this script in SLiM, we can use <em>slendr</em> to load
the output tree sequence (saved to
<code>/tmp/nonspatial-slim.trees</code>), simplify it, and overlay
mutations on it using the <a href="vignette-05-tree-sequences.html#loading-and-processing-tree-sequence-output-files">standard
functionality</a> originally developed for <em>slendr</em> tree
sequences. Note that this is the same command we would use for loading
<em>slendr</em> tree sequences, except we direct the
<code><a href="../reference/ts_load.html">ts_load()</a></code> function straight to the tree-sequence output
file rather than using the <code>ts_load(&lt;model-object&gt;)</code>
format used when working with standard <em>slendr</em> simulations. This
way, <em>slendr</em></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="st">"/tmp/nonspatial-slim.trees"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="../reference/ts_simplify.html">ts_simplify</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span>mutation_rate <span class="op">=</span> <span class="fl">1e-7</span>, random_seed <span class="op">=</span> <span class="va">SEED</span><span class="op">)</span></code></pre></div>
<p>We can extract information about individual’s names, nodes,
population assignments, etc. just as with any <em>slendr</em> tree
sequence with the function <code><a href="../reference/ts_nodes.html">ts_nodes()</a></code>. As with standard
<em>slendr</em> models, this function loads the “raw” node and
individual <a href="https://tskit.dev/tutorials/tables_and_editing.html" class="external-link">tree-sequences
tables</a>, performs a couple of join operations, and presents the whole
thing in a nice unified form for interactive data analysis (which can
also include spatial information—see below):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_nodes.html">ts_nodes</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sampled</span><span class="op">)</span>
<span class="va">data</span>
<span class="co">#&gt; slendr 'nodes' object </span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; times are expressed in a forward time direction</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; summary of the table data contents:</span>
<span class="co">#&gt;   p1 - 10 'sampled', 0 'remembered', 0 'retained', 10 'alive' individuals</span>
<span class="co">#&gt;   p2 - 500 'sampled', 0 'remembered', 0 'retained', 500 'alive' individuals</span>
<span class="co">#&gt;   p3 - 2500 'sampled', 0 'remembered', 0 'retained', 2500 'alive' individuals</span>
<span class="co">#&gt;   p4 - 10000 'sampled', 0 'remembered', 0 'retained', 10000 'alive' individuals</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; total:</span>
<span class="co">#&gt;   - 13010 'sampled' individuals</span>
<span class="co">#&gt;   - 0 'remembered' individuals</span>
<span class="co">#&gt;   - 0 'retained' individuals</span>
<span class="co">#&gt;   - 13010 'alive' individuals</span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; oldest sampled individual: 0 time units 'before present' </span>
<span class="co">#&gt; youngest sampled individual: 0 time units 'before present' </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; oldest node: 0 time units 'before present' </span>
<span class="co">#&gt; youngest node: 0 time units 'before present' </span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; overview of the underlying table object:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 26,020 × 10</span></span>
<span class="co">#&gt;    pop   node_id  time time_tskit sampled remembered retained alive pedigree_id</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> p1          0     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>000</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> p1          1     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>000</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> p1          2     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>001</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> p1          3     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>001</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> p1          4     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>002</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> p1          5     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>002</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> p1          6     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>003</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> p1          7     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>003</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> p1          8     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>004</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> p1          9     0          0 TRUE    FALSE      FALSE    TRUE     20<span style="text-decoration: underline;">073</span>004</span>
<span class="co">#&gt; <span style="color: #949494;"># … with 26,010 more rows, and 1 more variable: ind_id &lt;dbl&gt;</span></span></code></pre></div>
<p>Moving on to tskit statistics, we can use the data table above to
extract a list of nodes belonging to each population (this is what
various tskit tree-sequence statistics operate on, and <em>slendr</em>
follows that design). Here we are computing the nucleotide diversity in
each of the four populations using the <code><a href="../reference/ts_diversity.html">ts_diversity()</a></code>
function, by first creating a list of lists with node IDs
(i.e. chromosomes) of individuals in each population:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">node_id</span>, <span class="va">data</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span>

<span class="co"># compute nucleotide diversity in each population</span>
<span class="co"># (any other ts_*() tskit R interface function should work)</span>
<span class="fu"><a href="../reference/ts_diversity.html">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">sample_sets</span><span class="op">)</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 2</span></span>
<span class="co">#&gt;   set    diversity</span>
<span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> p1    0.000<span style="text-decoration: underline;">007</span>55</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> p2    0.000<span style="text-decoration: underline;">241</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">3</span> p3    0.000<span style="text-decoration: underline;">463</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">4</span> p4    0.000<span style="text-decoration: underline;">201</span></span></code></pre></div>
<p>Just as with <em>slendr</em> tree sequences (as demonstrated in our
<a href="vignette-09-paper.html#example-4-figure-5">paper</a>) we can
get a individual trees too, extracted in the in the phylogenetic format
provided by the <a href="https://cran.r-project.org/web/packages/ape/index.html" class="external-link">ape</a> R
package. Here we first simplify the tree sequence even further to just
10 nodes to make things manageable:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">node_id</span>, <span class="fl">10</span><span class="op">)</span>
<span class="va">ts_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_simplify.html">ts_simplify</a></span><span class="op">(</span><span class="va">ts</span>, simplify_to <span class="op">=</span> <span class="va">samples</span><span class="op">)</span>

<span class="co"># extract the 42nd tree in the genealogy to an R 'phylo' format</span>
<span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_phylo.html">ts_phylo</a></span><span class="op">(</span><span class="va">ts_small</span>, <span class="fl">42</span><span class="op">)</span>
<span class="co">#&gt; Starting checking the validity of tree...</span>
<span class="co">#&gt; Found number of tips: n = 10 </span>
<span class="co">#&gt; Found number of nodes: m = 9 </span>
<span class="co">#&gt; Done.</span>
<span class="va">tree</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Phylogenetic tree with 10 tips and 9 internal nodes.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Tip labels:</span>
<span class="co">#&gt;   1, 0, 9, 4, 8, 7, ...</span>
<span class="co">#&gt; Node labels:</span>
<span class="co">#&gt;   45, 28, 22, 31, 35, 36, ...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Rooted; includes branch lengths.</span></code></pre></div>
<p>Once we have that R tree object, we can use packages like <a href="https://yulab-smu.top/treedata-book/" class="external-link">ggtree</a> to visualize the
tree (any other phylogenetic package would work too). Note that because
nodes of ‘ape phylo’ trees must conform to a strict format (they must be
labelled <code>1...N</code>), we will extract the information about the
node IDs in the tskit tree-sequence data to be able to plot them in the
tree.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/treedata-book/" class="external-link">ggtree</a></span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'ggtree':</span>
<span class="co">#&gt;   method      from </span>
<span class="co">#&gt;   identify.gg ggfun</span>
<span class="co">#&gt; ggtree v3.2.1  For help: https://yulab-smu.top/treedata-book/</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; If you use ggtree in published research, please cite the most appropriate paper(s):</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 1. Guangchuang Yu. Using ggtree to visualize data on tree-like structures. Current Protocols in Bioinformatics. 2020, 69:e96. doi:10.1002/cpbi.96</span>
<span class="co">#&gt; 2. Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution. 2018, 35(12):3041-3043. doi:10.1093/molbev/msy194</span>
<span class="co">#&gt; 3. Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'ggtree'</span>
<span class="co">#&gt; The following object is masked from 'package:magrittr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     inset</span>

<span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_nodes.html">ts_nodes</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>node <span class="op">=</span> <span class="va">phylo_id</span>, tskit_id <span class="op">=</span> <span class="va">node_id</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">tree</span>, branch.length <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/ggtree/man/attacher.html" class="external-link">%&lt;+%</a></span> <span class="va">labels</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_label</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">tskit_id</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-08-nonslendr-tskit_files/figure-html/nonslendr_tree-1.png" width="640"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tree</span>, show.tip.label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">tiplabels</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="msprime-non-slendr-tree-sequences">msprime (non-slendr) tree sequences<a class="anchor" aria-label="anchor" href="#msprime-non-slendr-tree-sequences"></a>
</h4>
<p>The same as above applies also to msprime tree sequences (which is
really not that surprising, given that it’s all tskit under the
hood).</p>
<p>We can start with Python:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> msprime</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>ts <span class="op">=</span> msprime.sim_ancestry(<span class="dv">100</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>ts.dump(<span class="st">"/tmp/msprime.trees"</span>)</span></code></pre></div>
<p>And then we can proceed with loading the msprime tree sequence into R
and analyze it with the slendr functionality:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="st">"/tmp/msprime.trees"</span><span class="op">)</span>

<span class="fu"><a href="../reference/ts_nodes.html">ts_nodes</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span>
<span class="co">#&gt; slendr 'nodes' object </span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; times are expressed in a backward time direction</span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; overview of the underlying table object:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 199 × 7</span></span>
<span class="co">#&gt;    pop   ind_id node_id  time time_tskit sampled pop_id</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 0         <span style="color: #BB0000;">NA</span>       0     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 0         <span style="color: #BB0000;">NA</span>       1     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 0         <span style="color: #BB0000;">NA</span>       2     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 0         <span style="color: #BB0000;">NA</span>       3     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 0         <span style="color: #BB0000;">NA</span>       4     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 0         <span style="color: #BB0000;">NA</span>       5     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 0         <span style="color: #BB0000;">NA</span>       6     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 0         <span style="color: #BB0000;">NA</span>       7     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 0         <span style="color: #BB0000;">NA</span>       8     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 0         <span style="color: #BB0000;">NA</span>       9     0          0 TRUE         0</span>
<span class="co">#&gt; <span style="color: #949494;"># … with 189 more rows</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="spatial-slim-non-slendr-tree-sequences">Spatial SLiM (non-slendr) tree sequences<a class="anchor" aria-label="anchor" href="#spatial-slim-non-slendr-tree-sequences"></a>
</h4>
<p>Furthermore, the generalized interface also supports
<em>slendr</em>’s <a href="vignette-06-locations.html">spatial
tree-sequence features</a>, with all the <a href="vignette-09-paper.html#example-4-figure-5">bells and
whistles</a>.</p>
<p>For instance, lets take the following spatial SLiM script (modified
from the SLiM manual) and execute it with SLiM the usual way:</p>
<pre><code>initialize() {
    setSeed(42);
      initializeSLiMOptions(keepPedigrees=T, dimensionality="xy");
      initializeTreeSeq();
      initializeMutationRate(1e-7);
      initializeMutationType("m1", 0.5, "f", 0.0);
      initializeGenomicElementType("g1", m1, 1.0);
      initializeGenomicElement(g1, 0, 1e6);
      initializeRecombinationRate(1e-8);
}
1 early() {
    sim.addSubpop("p1", 500);

    // initial positions are random in ([0,1], [0,1])
    p1.individuals.x = runif(p1.individualCount);
      p1.individuals.y = runif(p1.individualCount);
}
modifyChild() {
      // draw a child position near the first parent, within bounds
      do child.x = parent1.x + rnorm(1, 0, 0.02);
      while ((child.x &lt; 0.0) | (child.x &gt; 1.0));

      do child.y = parent1.y + rnorm(1, 0, 0.02);
      while ((child.y &lt; 0.0) | (child.y &gt; 1.0));

      return T;
}
1: late() {
    sim.treeSeqRememberIndividuals(sim.subpopulations.individuals, permanent = F);
}

10000 late() {
    sim.treeSeqOutput("/tmp/spatial-slim.trees");
}</code></pre>
<p>We can then load and simplify the output tree sequence in just as we
did above in this vignette (or anywhere in the <em>slendr</em>
documentation):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="st">"/tmp/spatial-slim.trees"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/ts_simplify.html">ts_simplify</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Finally, we can access the spatio-temporal data embedded in the
output tree sequence in the standard <em>slendr</em> way (note the
spatial <a href="https://r-spatial.github.io/sf/" class="external-link"><em>sf</em></a> column
<code>location</code> with the <code>POINT</code> data type):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_nodes.html">ts_nodes</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span>
<span class="va">data</span>
<span class="co">#&gt; slendr 'nodes' object </span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; times are expressed in a forward time direction</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; summary of the table data contents:</span>
<span class="co">#&gt;   p1 - 500 'sampled', 0 'remembered', 500 'retained', 500 'alive' individuals</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; total:</span>
<span class="co">#&gt;   - 500 'sampled' individuals</span>
<span class="co">#&gt;   - 0 'remembered' individuals</span>
<span class="co">#&gt;   - 895 'retained' individuals</span>
<span class="co">#&gt;   - 500 'alive' individuals</span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; oldest sampled individual: 0 time units 'before present' </span>
<span class="co">#&gt; youngest sampled individual: 0 time units 'before present' </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; oldest node: 8463 time units 'before present' </span>
<span class="co">#&gt; youngest node: 0 time units 'before present' </span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; overview of the underlying sf object:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,955 × 11</span></span>
<span class="co">#&gt;    pop   node_id  time time_tskit               location sampled remembered</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;POINT&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> p1          0     0          0   (0.904485 0.0777489) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> p1          1     0          0   (0.904485 0.0777489) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> p1          2     0          0   (0.668828 0.5246381) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> p1          3     0          0   (0.668828 0.5246381) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> p1          4     0          0  (0.7707701 0.1518462) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> p1          5     0          0  (0.7707701 0.1518462) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> p1          6     0          0  (0.7617792 0.2616453) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> p1          7     0          0  (0.7617792 0.2616453) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> p1          8     0          0 (0.9807475 0.04223008) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> p1          9     0          0 (0.9807475 0.04223008) TRUE    FALSE     </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 1,945 more rows, and 4 more variables: retained &lt;lgl&gt;, alive &lt;lgl&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   pedigree_id &lt;dbl&gt;, ind_id &lt;dbl&gt;</span></span></code></pre></div>
<p>Because we get the tree sequence converted to the spatial <a href="https://r-spatial.github.io/sf/" class="external-link"><em>sf</em></a> data format, we
can use standard geospatial packages to use any spatial data analysis
methods that those packages provide.</p>
<p>To briefly demonstrate what this means, we can trivially plot the
location of each recorded node:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">time</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-08-nonslendr-tskit_files/figure-html/nonslendr_locations-1.png" width="640"></p>
<p>We can also collect spatio-temporal ancestry information of a
particular node (i.e. the times and locations of all of its ancestors
all the way to the root, with each “link” in the plot signifying
parent-child edge somewhere along the tree sequence) and plot it on a 2D
surface (x and y dimensions [0, 1]). The plot is a little chaotic, but
hopefully conveys the idea (the “focal node” 0 is highlighted in red).
This is essentially the same plot we have in <a href="vignette-09-paper.html#example-4-figure-5">the last figure of our
paper</a>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ancestral_links</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_ancestors.html">ts_ancestors</a></span><span class="op">(</span><span class="va">ts</span>, <span class="fl">0</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">ancestral_links</span>, size <span class="op">=</span> <span class="fl">0.5</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="va">parent_time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_set_geometry</a></span><span class="op">(</span><span class="va">ancestral_links</span>, <span class="st">"parent_location"</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">parent_time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">data</span><span class="op">$</span><span class="va">node_id</span> <span class="op">==</span> <span class="fl">0</span>, <span class="op">]</span>, size <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-08-nonslendr-tskit_files/figure-html/nonslendr_ancestries-1.png" width="640"></p>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In this vignette we gave a brief overview of using <em>slendr</em>’s
<a href="../reference/index.html#tree-sequence-loading-and-processing">R-tskit
interface</a> for loading, processing, and analyzing “pure”
non-<em>slendr</em> tree sequences produced by msprime and SLiM
scripts.</p>
<p>Although we have only touched upon the most basic features of its
R-tskit interface for standard tree sequences, it is important to note
that as far as <em>slendr</em> is concerned, it does not matter how a
tree sequence was produced, as long as it conforms to the tskit
specification. This means that regardless of the source of your tree
sequence data, you should be able to use <em>slendr</em>’s tskit
functionality to run your analyses.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a895c93f0e9eda05eb10b29a9921d586',
    indexName: 'slendr',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
