<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="slendr">
<title>Extracting true ancestry tracts • slendr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extracting true ancestry tracts">
<meta property="og:description" content="slendr">
<meta property="og:image" content="https://www.slendr.net/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">slendr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.1.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--articles">
    <span class="fa fa-book"></span>
     
    Articles
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--articles">
    <a class="dropdown-item" href="../articles/vignette-00-installation.html">Installation instructions</a>
    <a class="dropdown-item" href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    <a class="dropdown-item" href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    <a class="dropdown-item" href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    <a class="dropdown-item" href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    <a class="dropdown-item" href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a>
    <a class="dropdown-item" href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a>
    <a class="dropdown-item" href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a>
    <a class="dropdown-item" href="../articles/vignette-08-nonslendr-tskit.html">Analyzing non-slendr tree sequences</a>
    <a class="dropdown-item" href="../articles/vignette-09-paper.html">Examples from the slendr paper</a>
    <a class="dropdown-item" href="../articles/vignette-10-tracts.html">Extracting true ancestry tracts</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bodkan/slendr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/fleventy5">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Extracting true ancestry tracts</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/HEAD/vignettes/vignette-10-tracts.Rmd" class="external-link"><code>vignettes/vignette-10-tracts.Rmd</code></a></small>
      <div class="d-none name"><code>vignette-10-tracts.Rmd</code></div>
    </div>

    
    
<p><strong>Please note that the <em>tspop</em> support in
<em>slendr</em> implemented in the <code><a href="../reference/ts_tracts.html">ts_tracts()</a></code> function is
extremely experimental, only minimally tested, and its functionality is
expected to change quite a bit. Please wait for the release of the next
major version of <em>slendr</em> (which should include a more developed
<code><a href="../reference/ts_tracts.html">ts_tracts()</a></code>) before you use this in your own
work.</strong></p>
<p><em>slendr</em> now includes an experimental, use-it-at-your-own-risk
interface to an exciting <a href="https://academic.oup.com/bioinformaticsadvances/advance-article/doi/10.1093/bioadv/vbad163/7429395" class="external-link">new
algorithm</a> for extracting true tracts of ancestry as implemented in
the Python module <a href="https://tspop.readthedocs.io/en/latest/index.html" class="external-link"><em>tspop</em></a>.</p>
<p>The interface is implemented in an R function
<code><a href="../reference/ts_tracts.html">ts_tracts()</a></code> and this vignette describes its use on a simple
toy model of Neanderthal and Denisovan introgression into modern
humans.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/init_env.html">init_env</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="demographic-model">Demographic model<a class="anchor" aria-label="anchor" href="#demographic-model"></a>
</h3>
<p>Let’s imagine the following demographic model of Neanderthal
introgression into the ancestors of all non-Africans (represented by
“EUR” and “PAP” populations, approximating European and Papuan people
living today), followed by Denisovan introgression into the ancestors of
Papuans:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anc_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"ancestor_all"</span>, time <span class="op">=</span> <span class="fl">700e3</span>, N <span class="op">=</span> <span class="fl">10000</span>, remove <span class="op">=</span> <span class="fl">640e3</span><span class="op">)</span></span>
<span><span class="va">afr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"AFR"</span>, parent <span class="op">=</span> <span class="va">anc_all</span>, time <span class="op">=</span> <span class="fl">650e3</span>, N <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span><span class="va">anc_arch</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"ancestor_archaics"</span>, parent <span class="op">=</span> <span class="va">anc_all</span>, time <span class="op">=</span> <span class="fl">650e3</span>, N <span class="op">=</span> <span class="fl">10000</span>, remove <span class="op">=</span> <span class="fl">390e3</span><span class="op">)</span></span>
<span><span class="va">nea</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"NEA"</span>, parent <span class="op">=</span> <span class="va">anc_arch</span>, time <span class="op">=</span> <span class="fl">400e3</span>, N <span class="op">=</span> <span class="fl">2000</span>, remove <span class="op">=</span> <span class="fl">30e3</span><span class="op">)</span></span>
<span><span class="va">den</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"DEN"</span>, parent <span class="op">=</span> <span class="va">anc_arch</span>, time <span class="op">=</span> <span class="fl">400e3</span>, N <span class="op">=</span> <span class="fl">2000</span>, remove <span class="op">=</span> <span class="fl">30e3</span><span class="op">)</span></span>
<span><span class="va">nonafr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"nonAFR"</span>, parent <span class="op">=</span> <span class="va">afr</span>, time <span class="op">=</span> <span class="fl">100e3</span>, N <span class="op">=</span> <span class="fl">3000</span>, remove <span class="op">=</span> <span class="fl">39e3</span><span class="op">)</span></span>
<span><span class="va">eur</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"EUR"</span>, parent <span class="op">=</span> <span class="va">nonafr</span>, time <span class="op">=</span> <span class="fl">45e3</span>, N <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span><span class="va">pap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"PAP"</span>, parent <span class="op">=</span> <span class="va">nonafr</span>, time <span class="op">=</span> <span class="fl">45e3</span>, N <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">nea</span>, to <span class="op">=</span> <span class="va">nonafr</span>, rate <span class="op">=</span> <span class="fl">0.03</span>, start <span class="op">=</span> <span class="fl">55000</span>, end <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">den</span>, to <span class="op">=</span> <span class="va">pap</span>, rate <span class="op">=</span> <span class="fl">0.07</span>, start <span class="op">=</span> <span class="fl">35000</span>, end <span class="op">=</span> <span class="fl">30000</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span></span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">anc_all</span>, <span class="va">afr</span>, <span class="va">anc_arch</span>, <span class="va">nea</span>, <span class="va">den</span>, <span class="va">nonafr</span>, <span class="va">eur</span>, <span class="va">pap</span><span class="op">)</span>,</span>
<span>  gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>  generation_time <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  serialize <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_model.html">plot_model</a></span><span class="op">(</span></span>
<span>  <span class="va">model</span>, sizes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR"</span>, <span class="st">"EUR"</span>, <span class="st">"nonAFR"</span>, <span class="st">"PAP"</span>, <span class="st">"ancestor_all"</span>, <span class="st">"DEN"</span>, <span class="st">"ancestor_archaics"</span>, <span class="st">"NEA"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-10-tracts_files/figure-html/introgression_model-1.png" width="360"></p>
</div>
<div class="section level3">
<h3 id="tree-sequence-simulation">Tree sequence simulation<a class="anchor" aria-label="anchor" href="#tree-sequence-simulation"></a>
</h3>
<p>Let’s now simulate a 50Mb tree sequence from this model, recording 50
diploid individuals from EUR and PAP populations:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">eur</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">pap</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">100e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, samples <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="fl">42</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="extracting-ancestry-tracts">Extracting ancestry tracts<a class="anchor" aria-label="anchor" href="#extracting-ancestry-tracts"></a>
</h3>
<p>In order to extract tracts of Neanderthal and Denisovan ancestry, we
can use <em>slendr</em>’s new function <code><a href="../reference/ts_tracts.html">ts_tracts()</a></code> which
serves as a simplified R-friendly interface to the Python method <a href="https://tspop.readthedocs.io/en/latest/tspop.html#tspop.get_pop_ancestry" class="external-link"><code>tspop.get_pop_ancestry()</code></a>.
An important piece of information used by the function is a so-called
“census time”, which records the time of recording of the “ancestral
population” identity of each node ancestral to each subsegment in our
sample set. Please see the excellent <a href="https://tspop.readthedocs.io/en/latest/ideas.html" class="external-link">vignette of
<em>tspop</em></a> for more information on the inner workings of the
algorithm.</p>
<p>In our case, let’s extract the ancestry tracts corresponding to
ancestral nodes present at 55 thousand years ago – this time corresponds
to the moment of the start of the archaic introgression:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nea_tracts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_tracts.html">ts_tracts</a></span><span class="op">(</span><span class="va">ts</span>, census <span class="op">=</span> <span class="fl">55000</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; PopAncestry summary</span></span>
<span><span class="co">#&gt; Number of ancestral populations:     3</span></span>
<span><span class="co">#&gt; Number of sample chromosomes:        200</span></span>
<span><span class="co">#&gt; Number of ancestors:             44052</span></span>
<span><span class="co">#&gt; Total length of genomes:         20000000000.000000</span></span>
<span><span class="co">#&gt; Ancestral coverage:          20000000000.000000</span></span>
<span><span class="va">den_tracts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_tracts.html">ts_tracts</a></span><span class="op">(</span><span class="va">ts</span>, census <span class="op">=</span> <span class="fl">35000</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; PopAncestry summary</span></span>
<span><span class="co">#&gt; Number of ancestral populations:     3</span></span>
<span><span class="co">#&gt; Number of sample chromosomes:        200</span></span>
<span><span class="co">#&gt; Number of ancestors:             55373</span></span>
<span><span class="co">#&gt; Total length of genomes:         20000000000.000000</span></span>
<span><span class="co">#&gt; Ancestral coverage:          20000000000.000000</span></span>
<span></span>
<span><span class="va">tracts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">nea_tracts</span>, <span class="va">den_tracts</span><span class="op">)</span></span></code></pre></div>
<p>This is what a table with all ancestry tracts looks like. As we would
expect, we see a column indicating a name of each individual, the left
and right coordinates of each tract in each individual, as well as the
population name of the source of each ancestry tract:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tracts</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 17,002 × 8</span></span></span>
<span><span class="co">#&gt;    name  node_id pop   source_pop     left    right length source_pop_id</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> EUR_1       0 EUR   NEA          <span style="text-decoration: underline;">781</span>887   <span style="text-decoration: underline;">852</span>668  <span style="text-decoration: underline;">70</span>781             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> EUR_1       0 EUR   NEA         1<span style="text-decoration: underline;">451</span>389  1<span style="text-decoration: underline;">463</span>837  <span style="text-decoration: underline;">12</span>448             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> EUR_1       0 EUR   NEA         1<span style="text-decoration: underline;">596</span>995  1<span style="text-decoration: underline;">601</span>441   <span style="text-decoration: underline;">4</span>446             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> EUR_1       0 EUR   NEA         1<span style="text-decoration: underline;">629</span>857  1<span style="text-decoration: underline;">689</span>656  <span style="text-decoration: underline;">59</span>799             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> EUR_1       0 EUR   NEA         3<span style="text-decoration: underline;">203</span>711  3<span style="text-decoration: underline;">339</span>565 <span style="text-decoration: underline;">135</span>854             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> EUR_1       0 EUR   NEA         3<span style="text-decoration: underline;">850</span>629  3<span style="text-decoration: underline;">947</span>923  <span style="text-decoration: underline;">97</span>294             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> EUR_1       0 EUR   NEA         6<span style="text-decoration: underline;">190</span>700  6<span style="text-decoration: underline;">205</span>627  <span style="text-decoration: underline;">14</span>927             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> EUR_1       0 EUR   NEA        10<span style="text-decoration: underline;">394</span>211 10<span style="text-decoration: underline;">426</span>667  <span style="text-decoration: underline;">32</span>456             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> EUR_1       0 EUR   NEA        13<span style="text-decoration: underline;">044</span>685 13<span style="text-decoration: underline;">180</span>258 <span style="text-decoration: underline;">135</span>573             3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> EUR_1       0 EUR   NEA        14<span style="text-decoration: underline;">858</span>956 14<span style="text-decoration: underline;">912</span>068  <span style="text-decoration: underline;">53</span>112             3</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 16,992 more rows</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="summaries-of-ancestral-proportions">Summaries of ancestral proportions<a class="anchor" aria-label="anchor" href="#summaries-of-ancestral-proportions"></a>
</h3>
<p>When we summarise the ancestry proportions in target EUR and PAP
populations, we see that the EUR population only carries about ~3% of
Neanderthal ancestry and that this is also true for the PAP population.
However, we also see that Papuans carry about 7% of Denisovan ancestry.
This is consistent with our model, but also with the expectation from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5772775/" class="external-link">empirical
data</a>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary</span> <span class="op">&lt;-</span> <span class="va">tracts</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">node_id</span>, <span class="va">pop</span>, <span class="va">source_pop</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>prop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">length</span><span class="op">)</span> <span class="op">/</span> <span class="fl">100e6</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'name', 'node_id', 'pop'. You can override</span></span>
<span><span class="co">#&gt; using the `.groups` argument.</span></span>
<span></span>
<span><span class="va">summary</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">pop</span>, <span class="va">source_pop</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">prop</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">source_pop</span>, <span class="va">pop</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'pop'. You can override using the `.groups`</span></span>
<span><span class="co">#&gt; argument.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   pop [2]</span></span></span>
<span><span class="co">#&gt;   pop   source_pop `mean(prop)`</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> EUR   NEA              0.030<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> PAP   NEA              0.034<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> PAP   DEN              0.073<span style="text-decoration: underline;">9</span></span></span></code></pre></div>
<p>Let’s visualize these proportions at an individual level:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">source_pop</span>, <span class="va">prop</span>, color <span class="op">=</span> <span class="va">source_pop</span>, fill <span class="op">=</span> <span class="va">source_pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.03</span>, <span class="fl">0.08</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"ancestry proportion"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">pop</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Ancestry proportions in each individual"</span>,</span>
<span>          <span class="st">"(vertical lines represent 3% and 7% baseline expectations"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-10-tracts_files/figure-html/anc_prop_summary-1.png" width="360"></p>
</div>
<div class="section level3">
<h3 id="chromosome-painting-of-ancestry-tracts">“Chromosome painting” of ancestry tracts<a class="anchor" aria-label="anchor" href="#chromosome-painting-of-ancestry-tracts"></a>
</h3>
<p>Because the <code>tracts</code> object contains the coordinates of
every single ancestry segment in each of the simulated individuals, we
can “paint” each chromosome with each of the two archaic human
ancestries:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tracts</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>chrom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">name</span>, <span class="st">" (node"</span>, <span class="va">node_id</span>, <span class="st">")"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">left</span>, xend <span class="op">=</span> <span class="va">right</span>, y <span class="op">=</span> <span class="va">chrom</span>, yend <span class="op">=</span> <span class="va">chrom</span>, color <span class="op">=</span> <span class="va">source_pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span>linewidth <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"position [bp]"</span>, y <span class="op">=</span> <span class="st">"haplotype"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"True ancestry tracts along each chromosome"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">pop</span> <span class="op">~</span> <span class="va">.</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-10-tracts_files/figure-html/chrom_painting-1.png" width="720"></p>
<p>By lining up NEA &amp; DEN ancestry tracts in both EUR and PAP
populations, we can see how the common origin of Neanderthal ancestry in
both non-African populations manifests in a significant overlap of NEA
tracts between both populations.</p>
</div>
<div class="section level3">
<h3 id="average-tract-lengths">Average tract lengths:<a class="anchor" aria-label="anchor" href="#average-tract-lengths"></a>
</h3>
<p>Let’s compute simple summaries of tract lengths in the simulated
data, and compare them to theoretical expectations.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tracts</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">pop</span>, <span class="va">source_pop</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">length</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; `summarise()` has grouped output by 'pop'. You can override using the `.groups`</span></span>
<span><span class="co">#&gt; argument.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   pop [2]</span></span></span>
<span><span class="co">#&gt;   pop   source_pop `mean(length)`</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> EUR   NEA                <span style="text-decoration: underline;">65</span>765.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> PAP   NEA                <span style="text-decoration: underline;">69</span>204.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> PAP   DEN               <span style="text-decoration: underline;">100</span>317.</span></span></code></pre></div>
<p>Theoretical expectations (from Racimo and Slatkin 2015, Box 1)</p>
<ul>
<li>Neanderthal tracts:</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">0.03</span></span>
<span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fl">52500</span> <span class="op">/</span> <span class="fl">30</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span></span>
<span></span>
<span><span class="va">mean_nea</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">/</span> <span class="op">(</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">m</span><span class="op">)</span> <span class="op">*</span> <span class="va">r</span> <span class="op">*</span> <span class="op">(</span><span class="va">t</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mean_nea</span></span>
<span><span class="co">#&gt; [1] 58943.84</span></span></code></pre></div>
<ul>
<li>Denisovan tracts:</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">0.07</span></span>
<span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fl">37500</span> <span class="op">/</span> <span class="fl">30</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span></span>
<span></span>
<span><span class="va">mean_den</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">/</span> <span class="op">(</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="va">m</span><span class="op">)</span> <span class="op">*</span> <span class="va">r</span> <span class="op">*</span> <span class="op">(</span><span class="va">t</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mean_den</span></span>
<span><span class="co">#&gt; [1] 86090.38</span></span></code></pre></div>
<p>As we can see, our simulations are not that far of from the
theoretical expectations, giving us confidence that our simulations (and
the ancestry tract extraction algorithm) are working as expected.</p>
</div>
<div class="section level3">
<h3 id="distribution-of-ancestry-tract-lengths">Distribution of ancestry tract lengths<a class="anchor" aria-label="anchor" href="#distribution-of-ancestry-tract-lengths"></a>
</h3>
<p>Finally, let’s plot the distributions of lengths of each of the
ancestry tracts. The case of archaic human introgression is very well
studied so it’s perhaps not that exciting to look at these figures. That
said, in less well studied species, it might be interesting to use these
kinds of simulations for inference of introgression times and
proportions via Approximate Bayesian Computation or by another
method:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">expectation_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR"</span>, <span class="st">"PAP"</span>, <span class="st">"PAP"</span><span class="op">)</span>,</span>
<span>  source_pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NEA"</span>, <span class="st">"NEA"</span>, <span class="st">"DEN"</span><span class="op">)</span>,</span>
<span>  length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mean_nea</span>, <span class="va">mean_nea</span>, <span class="va">mean_den</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_densities</span> <span class="op">&lt;-</span> <span class="va">tracts</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">length</span>, color <span class="op">=</span> <span class="va">source_pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">expectation_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">length</span>, color <span class="op">=</span> <span class="va">source_pop</span><span class="op">)</span>,</span>
<span>             linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">pop</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Distribution of tract lengths per different ancestries"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p_densities</span>, <span class="va">p_densities</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-10-tracts_files/figure-html/tract_lengths-1.png" width="360"></p>
</div>
<div class="section level2">
<h2 id="pure-msprime-tree-sequence">Pure <em>msprime</em> tree sequence<a class="anchor" aria-label="anchor" href="#pure-msprime-tree-sequence"></a>
</h2>
<p>Finally, as a sanity check, let’s use the pure <em>msprime</em>
simulation example from the <a href="https://tspop.readthedocs.io/en/latest/basicusage.html" class="external-link">official
<em>tspop</em> documentation</a> to test that <code><a href="../reference/ts_tracts.html">ts_tracts()</a></code>
behaves as expected even on a standard <em>msprime</em> tree-sequence
object.</p>
<p>First, let’s run the simulation code exactly as it is:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> msprime</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>pop_size <span class="op">=</span> <span class="dv">500</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>sequence_length <span class="op">=</span> <span class="fl">1e7</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>seed <span class="op">=</span> <span class="dv">98765</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>rho <span class="op">=</span> <span class="fl">3e-8</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Make the Demography object.</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>demography <span class="op">=</span> msprime.Demography()</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>demography.add_population(name<span class="op">=</span><span class="st">"RED"</span>, initial_size<span class="op">=</span>pop_size)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Population(initial_size=500, growth_rate=0, name='RED', description='', extra_metadata={}, default_sampling_time=None, initially_active=None, id=0)</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>demography.add_population(name<span class="op">=</span><span class="st">"BLUE"</span>, initial_size<span class="op">=</span>pop_size)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Population(initial_size=500, growth_rate=0, name='BLUE', description='', extra_metadata={}, default_sampling_time=None, initially_active=None, id=1)</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>demography.add_population(name<span class="op">=</span><span class="st">"ADMIX"</span>, initial_size<span class="op">=</span>pop_size)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Population(initial_size=500, growth_rate=0, name='ADMIX', description='', extra_metadata={}, default_sampling_time=None, initially_active=None, id=2)</span></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>demography.add_population(name<span class="op">=</span><span class="st">"ANC"</span>, initial_size<span class="op">=</span>pop_size)</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Population(initial_size=500, growth_rate=0, name='ANC', description='', extra_metadata={}, default_sampling_time=None, initially_active=None, id=3)</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>demography.add_admixture(</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    time<span class="op">=</span><span class="dv">100</span>, derived<span class="op">=</span><span class="st">"ADMIX"</span>, ancestral<span class="op">=</span>[<span class="st">"RED"</span>, <span class="st">"BLUE"</span>], proportions<span class="op">=</span>[<span class="fl">0.5</span>, <span class="fl">0.5</span>]</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Admixture(time=100, derived='ADMIX', ancestral=['RED', 'BLUE'], proportions=[0.5, 0.5])</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>demography.add_census(time<span class="op">=</span><span class="fl">100.01</span>) <span class="co"># Census is here!</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; CensusEvent(time=100.01)</span></span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>demography.add_population_split(</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>    time<span class="op">=</span><span class="dv">1000</span>, derived<span class="op">=</span>[<span class="st">"RED"</span>, <span class="st">"BLUE"</span>], ancestral<span class="op">=</span><span class="st">"ANC"</span></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PopulationSplit(time=1000, derived=['RED', 'BLUE'], ancestral='ANC')</span></span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate.</span></span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>ts <span class="op">=</span> msprime.sim_ancestry(</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>    samples<span class="op">=</span>{<span class="st">"RED"</span>: <span class="dv">0</span>, <span class="st">"BLUE"</span>: <span class="dv">0</span>, <span class="st">"ADMIX"</span> : <span class="dv">2</span>},</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>    demography<span class="op">=</span>demography,</span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>    random_seed<span class="op">=</span>seed,</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a>    sequence_length<span class="op">=</span>sequence_length,</span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a>    recombination_rate<span class="op">=</span>rho</span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>Let’s save the <em>msprime</em> tree sequence to disk so that we can
load it into R (i.e., approximating what you might want to do should you
want to use <code><a href="../reference/ts_tracts.html">ts_tracts()</a></code> without running a <em>slendr</em>
simulation first):</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>ts.dump(<span class="st">"/tmp/sim.trees"</span>)</span></code></pre></div>
<p>Now let’s move to R again, load the tree sequence into
<em>slendr</em> and extract ancestry tracts from it using
<code><a href="../reference/ts_tracts.html">ts_tracts()</a></code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="st">"/tmp/sim.trees"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">squashed_tracts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_tracts.html">ts_tracts</a></span><span class="op">(</span><span class="va">sim_ts</span>, census <span class="op">=</span> <span class="fl">100.01</span>, squashed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; PopAncestry summary</span></span>
<span><span class="co">#&gt; Number of ancestral populations:     2</span></span>
<span><span class="co">#&gt; Number of sample chromosomes:        4</span></span>
<span><span class="co">#&gt; Number of ancestors:             118</span></span>
<span><span class="co">#&gt; Total length of genomes:         40000000.000000</span></span>
<span><span class="co">#&gt; Ancestral coverage:          40000000.000000</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">squashed_tracts</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt;   sample    left   right population</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      0       0  <span style="text-decoration: underline;">419</span>848          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>      0  <span style="text-decoration: underline;">419</span>848  <span style="text-decoration: underline;">483</span>009          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>      0  <span style="text-decoration: underline;">483</span>009 1<span style="text-decoration: underline;">475</span>765          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>      0 1<span style="text-decoration: underline;">475</span>765 2<span style="text-decoration: underline;">427</span>904          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>      0 2<span style="text-decoration: underline;">427</span>904 3<span style="text-decoration: underline;">635</span>390          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>      0 3<span style="text-decoration: underline;">635</span>390 4<span style="text-decoration: underline;">606</span>954          1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">squashed_tracts</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">#&gt;   sample    left    right population</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      3 7<span style="text-decoration: underline;">134</span>130  7<span style="text-decoration: underline;">362</span>300          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>      3 7<span style="text-decoration: underline;">362</span>300  7<span style="text-decoration: underline;">369</span>409          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>      3 7<span style="text-decoration: underline;">369</span>409  7<span style="text-decoration: underline;">596</span>783          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>      3 7<span style="text-decoration: underline;">596</span>783  8<span style="text-decoration: underline;">289</span>015          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>      3 8<span style="text-decoration: underline;">289</span>015  8<span style="text-decoration: underline;">918</span>727          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>      3 8<span style="text-decoration: underline;">918</span>727 10<span style="text-decoration: underline;">000</span>000          0</span></span></code></pre></div>
<p>By setting <code>squashed = FALSE</code>, we get the full,
un-squashed ancestry segments, each with its appropriate ancestral node
ID:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_tracts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_tracts.html">ts_tracts</a></span><span class="op">(</span><span class="va">sim_ts</span>, census <span class="op">=</span> <span class="fl">100.01</span>, squashed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; PopAncestry summary</span></span>
<span><span class="co">#&gt; Number of ancestral populations:     2</span></span>
<span><span class="co">#&gt; Number of sample chromosomes:        4</span></span>
<span><span class="co">#&gt; Number of ancestors:             118</span></span>
<span><span class="co">#&gt; Total length of genomes:         40000000.000000</span></span>
<span><span class="co">#&gt; Ancestral coverage:          40000000.000000</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">full_tracts</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   sample   left   right ancestor population</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      0      0   <span style="text-decoration: underline;">33</span>027       74          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>      0  <span style="text-decoration: underline;">33</span>027  <span style="text-decoration: underline;">155</span>453       33          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>      0 <span style="text-decoration: underline;">155</span>453  <span style="text-decoration: underline;">290</span>542       46          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>      0 <span style="text-decoration: underline;">290</span>542  <span style="text-decoration: underline;">419</span>848       18          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>      0 <span style="text-decoration: underline;">419</span>848  <span style="text-decoration: underline;">483</span>009       83          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>      0 <span style="text-decoration: underline;">483</span>009 1<span style="text-decoration: underline;">475</span>765       28          0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">full_tracts</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   sample    left    right ancestor population</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      3 8<span style="text-decoration: underline;">477</span>625  8<span style="text-decoration: underline;">672</span>850       94          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>      3 8<span style="text-decoration: underline;">672</span>850  8<span style="text-decoration: underline;">849</span>756       95          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>      3 8<span style="text-decoration: underline;">849</span>756  8<span style="text-decoration: underline;">918</span>727      131          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>      3 8<span style="text-decoration: underline;">918</span>727  9<span style="text-decoration: underline;">165</span>035       44          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>      3 9<span style="text-decoration: underline;">165</span>035  9<span style="text-decoration: underline;">176</span>562       47          0</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>      3 9<span style="text-decoration: underline;">176</span>562 10<span style="text-decoration: underline;">000</span>000       58          0</span></span></code></pre></div>
<p>By comparing the two tables above to the <em>pandas</em> data frames
in the <a href="https://tspop.readthedocs.io/en/latest/basicusage.html" class="external-link"><em>tspop</em>
documentation</a>, we can see that we obtained the same results.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
