<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simulating data with SLiM and msprime backends • slendr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Simulating data with SLiM and msprime backends">
<meta property="og:description" content="slendr">
<meta property="og:image" content="https://www.slendr.net/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slendr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    </li>
    <li>
      <a href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    </li>
    <li>
      <a href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    </li>
    <li>
      <a href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    </li>
    <li>
      <a href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a>
    </li>
    <li>
      <a href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a>
    </li>
    <li>
      <a href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/slendr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/fleventy5">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette-07-backends_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simulating data with SLiM and msprime backends</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/master/vignettes/vignette-07-backends.Rmd"><code>vignettes/vignette-07-backends.Rmd</code></a></small>
      <div class="hidden name"><code>vignette-07-backends.Rmd</code></div>

    </div>

    
    
<p>Even though the most distinguishing feature of the <em>slendr</em> R package is programming spatial population genetic models across real and abstract geographic landscapes, having the ability to run “standard”, non-spatial models directly from an R environment could be often quite useful. This is why <em>slendr</em> supports also non-spatial SLiM models, which can be executed by <em>slendr</em> simply by excluding the map component from a SLiM simulation (briefly discussed in <a href="vignette-04-nonspatial-models.html">this vignette</a>).</p>
<p>In some cases, the forward nature of a SLiM simulation (meaning that many individuals need to be tracked in each generation) makes the simulations too slow. For instance, running an <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Approximate Bayesian Computation (ABC)</a> analysis often requires running thousands or even millions of simulations. In cases like this, the faster the simulations the better.</p>
<p>For these situations, the <em>slendr</em> R package provides an alternative back end for simulating non-spatial population models using the Python coalescent simulation framework <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>. Unlike forward time simulators, coalescent simulations only reconstruct the genealogical history of a sample of individuals of interest backwards in time, which makes it much more computational efficient.</p>
<p>The goal of this vignette is to demonstrate that a single <em>slendr</em> model configuration can be run using the built-in SLiM and msprime backend engines <em>without any changes</em>!</p>
<p>To start off, let’s load the <em>slendr</em> package together with some standard R data analysis packages, and instruct <em>reticulate</em> where to look for a Python environment with all necessary Python dependencies (see <a href="vignette-05-tree-sequences.html#r-interface-for-tskit-and-pyslim-1">this</a> section of another vignette for instructions how to setup the Python environment the easiest way possible):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr">slendr</a></span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>

<span class="va">seed</span> <span class="op">&lt;-</span> <span class="fl">42</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">seed</span><span class="op">)</span>

<span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/use_python.html">use_virtualenv</a></span><span class="op">(</span><span class="st">"~/.venvs/retipy"</span>, required <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># let's make sure reticulate found the correct Python environment</span>
<span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/py_config.html">py_config</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>#&gt; python:         /Users/martin_petr/.venvs/retipy/bin/python
#&gt; libpython:      /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/config-3.9-darwin/libpython3.9.dylib
#&gt; pythonhome:     /Users/martin_petr/.venvs/retipy:/Users/martin_petr/.venvs/retipy
#&gt; version:        3.9.0 (v3.9.0:9cf6752276, Oct  5 2020, 11:29:23)  [Clang 6.0 (clang-600.0.57)]
#&gt; numpy:          /Users/martin_petr/.venvs/retipy/lib/python3.9/site-packages/numpy
#&gt; numpy_version:  1.21.4
#&gt; tskit:          /Users/martin_petr/.venvs/retipy/lib/python3.9/site-packages/tskit
#&gt; 
#&gt; NOTE: Python version was forced by RETICULATE_PYTHON</code></pre>
<div id="detecting-gene-flow-from-msprime-and-slim-tree-sequences" class="section level2">
<h2 class="hasAnchor">
<a href="#detecting-gene-flow-from-msprime-and-slim-tree-sequences" class="anchor"></a>Detecting gene flow from msprime and SLiM tree sequences</h2>
<div id="defining-a-model" class="section level3">
<h3 class="hasAnchor">
<a href="#defining-a-model" class="anchor"></a>Defining a model</h3>
<p>We now define up a simple <em>non-spatial</em> model of gene flow between populations (also known as population admixture or introgression). This will involve essentially the same procedure which we have shown in another vignette introducing <a href="vignette-04-nonspatial-models.html">non-spatial <em>slendr</em> models</a>. Note that this is no different from how we would normally specify a spatial model, except that we left out the <code>map</code> argument in making the <code><a href="../reference/population.html">population()</a></code> calls.</p>
<p>To demonstrate additional features of the <em>slendr</em> interface to the <a href="https://tskit.dev/tskit/docs/stable/introduction.html"><em>tskit</em></a> Python library, we will perform gene flow detection test using the so-called <span class="math inline">\(f_4\)</span> and <span class="math inline">\(f_4\)</span>-ratio statistics (very briefly introduces in <a href="vignette-05-tree-sequences.html">this vignette</a>). To make things a little more interesting, we will define two population models: one model without any gene flow, and another which includes gene flow. In defining this model and expressing the <span class="math inline">\(f\)</span>-statistics we will use the same nomenclature used in the first study by <a href="https://academic.oup.com/genetics/article/192/3/1065/5935193">Patterson <em>et al.</em></a> that described <span class="math inline">\(f_4-ratio\)</span> statistic in the first place. if you’re not familiar with these statistical tests of gene flow, we recommend that you take a look at the relevant sections of the linked paper.</p>
<p><img src="../reference/figures/f4_tree.jpg" style="width:90.0%"></p>
<p>Let’s start by first defining the populations and their splits as established in the figure above:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seq_len</span> <span class="op">&lt;-</span> <span class="fl">250e6</span> <span class="co"># amount of sequence to simulate</span>
<span class="va">rec_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span> <span class="co"># uniform recombination rate</span>
<span class="va">mut_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span> <span class="co"># mutation rate</span>

<span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"outgroup"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"b"</span>, parent <span class="op">=</span> <span class="va">o</span>, time <span class="op">=</span> <span class="fl">500</span>, N <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="va">c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"c"</span>, parent <span class="op">=</span> <span class="va">b</span>, time <span class="op">=</span> <span class="fl">1000</span>, N <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"x1"</span>, parent <span class="op">=</span> <span class="va">c</span>, time <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span>
<span class="va">x2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"x2"</span>, parent <span class="op">=</span> <span class="va">c</span>, time <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span>
<span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"a"</span>, parent <span class="op">=</span> <span class="va">b</span>, time <span class="op">=</span> <span class="fl">1500</span>, N <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p>Note that the Ne of the populations <code>x1</code> and <code>x2</code> is set to be much higher than the rest. We set the <span class="math inline">\(N_e\)</span> of the other populations to lower values to speed up the forward SLiM simulations (they won’t affect the results we’re interested in anyway). Higher values of the <span class="math inline">\(N_e\)</span> of <code>x1</code> and <code>x2</code> populations will ensure that the effect of the drift acting on those two populations will be much smaller. Because we will simulate and later measure the proportion of ancestry from the population <code>b</code> to <code>x1</code>, this will ensure that the ancestry proportion will not drift too far away from the expectation and it will make the interesting patterns stand out more clearly.</p>
<p>(Of course, this is all done for demonstration purposes only and to speed up the SLiM simulations by making the <span class="math inline">\(N_e\)</span> of the other populations smaller. In practice, we would be running these kinds of simulations only using the <em>msprime</em> back end.)</p>
</div>
</div>
<div id="compiling-the-model-and-simulating-data" class="section level2">
<h2 class="hasAnchor">
<a href="#compiling-the-model-and-simulating-data" class="anchor"></a>Compiling the model and simulating data</h2>
<p>We will now use those populations to compile two models: the first model will be without any geneflow (left panel in the figure wo above), the second model will include 10% gene flow from <code>b</code> to <code>x1</code> (right panel in the figure above). We will also schedule sampling of a couple of individuals at the very end of the simulation: 50 individuals each from populations <code>x1</code> and <code>x2</code> to capture a bit more of the natural variation of <code>b</code> ancestry in <code>x1</code>, and only one individual from the rest.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># no gene flow model</span>
<span class="va">model_nogf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile.html">compile</a></span><span class="op">(</span>populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="va">b</span>, <span class="va">x1</span>, <span class="va">x2</span>, <span class="va">c</span>, <span class="va">o</span><span class="op">)</span>, generation_time <span class="op">=</span> <span class="fl">1</span>, sim_length <span class="op">=</span> <span class="fl">2200</span><span class="op">)</span>

<span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampling.html">sampling</a></span><span class="op">(</span><span class="va">model_nogf</span>, times <span class="op">=</span> <span class="fl">2200</span>,
                    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">b</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">x1</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">x2</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">c</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">o</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>

<span class="co"># model with gene flow</span>
<span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneflow.html">geneflow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">b</span>, to <span class="op">=</span> <span class="va">x1</span>, start <span class="op">=</span> <span class="fl">2100</span>, end <span class="op">=</span> <span class="fl">2150</span>, rate <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>

<span class="va">model_gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile.html">compile</a></span><span class="op">(</span>populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="va">b</span>, <span class="va">x1</span>, <span class="va">x2</span>, <span class="va">c</span>, <span class="va">o</span><span class="op">)</span>, geneflow <span class="op">=</span> <span class="va">gf</span>, generation_time <span class="op">=</span> <span class="fl">1</span>, sim_length <span class="op">=</span> <span class="fl">2200</span><span class="op">)</span> 

<span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampling.html">sampling</a></span><span class="op">(</span><span class="va">model_gf</span>, times <span class="op">=</span> <span class="fl">2200</span>,
                    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">b</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">x1</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">x2</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">c</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">o</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Now we will run both models (with and without gene flow) in the two <em>slendr</em> backends, SLiM and msprime:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># model without gene flow</span>
<span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model_nogf</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, sampling <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span>
<span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span><span class="va">model_nogf</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, sampling <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span>

<span class="co"># model with b -&gt; x1 gene flow</span>
<span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model_gf</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, sampling <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span>
<span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span><span class="va">model_gf</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, sampling <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></code></pre></div>
<p>Note that we are using exactly the same model configuration object in both simulation runs! In fact, even the function interface looks nearly exactly the same. It doesn’t matter what the specific details of your demographic models are, <em>slendr</em> will interpret them correctly regardless of which back end simulation engine you choose to use.</p>
<p>Because <em>slendr</em> saves all output files directly to the model directory by default, we can detect two distinct tree sequence output files, one generated by <em>slendr</em>’s SLiM backend, another by its msprime backend:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"cd"</span>, <span class="va">model_nogf</span><span class="op">$</span><span class="va">path</span>, <span class="st">"; ls -l *.trees"</span><span class="op">)</span>, intern <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></code></pre></div>
<pre><code>#&gt; -rw-r--r-- 1 martin_petr staff  12320116 Feb  2 15:02 output_msprime.trees
#&gt; -rw-r--r-- 1 martin_petr staff 162311700 Feb  2 15:01 output_slim.trees</code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"cd"</span>, <span class="va">model_gf</span><span class="op">$</span><span class="va">path</span>, <span class="st">"; ls -l *.trees"</span><span class="op">)</span>, intern <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></code></pre></div>
<pre><code>#&gt; -rw-r--r-- 1 martin_petr staff  12012340 Feb  2 15:03 output_msprime.trees
#&gt; -rw-r--r-- 1 martin_petr staff 160840780 Feb  2 15:03 output_slim.trees</code></pre>
<p>However, note that the files have very different sizes. In particular, the <code>output_msprime.trees</code> much smaller. This is because being a coalescent simulator, msprime only saves information about nodes which are coalescent nodes somewhere (and some time) in the tree sequence.</p>
<div id="comparing-the-outputs-of-msprime-and-slim-runs-of-a-slendr-model" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-the-outputs-of-msprime-and-slim-runs-of-a-slendr-model" class="anchor"></a>Comparing the outputs of msprime and SLiM runs of a <em>slendr</em> model</h3>
<p>Having run our introgression models with the two simulation back ends, let’s load both SLiM and msprime tree sequence outputs and compare their contents (hoping they will be the same!).</p>
<p>First, we will load the tree sequences generated by msprime and <em>slendr</em>/SLiM into Python. We can load the SLiM tree sequence output using <a href="https://www.slendr.net/articles/vignette-05-tree-sequences.html#loading-and-processing-tree-sequence-output-files-1">dedicated tskit functionality</a> in R (and we do so below). However, although we can run <em>slendr</em> model with an msprime back end, for most of the time of existence of <em>slendr</em> the tree sequences from SLiM and msprime were not 100% interchangeable. Until this is resolved (and it will be resolved soon), we will demonstrate how to repeat the same f4 an f4-ratio analysis we will perform below using the slendr/tskit interface using a “raw” <em>reticulate</em> interface to the Python tskit and msprime modules.</p>
<p>Although mildly annoying (the ultimate goal is to be able to analyse tree sequence data with slendr’s R interface regardless of which simulator produced it), it will be instructive to show how this can be done.</p>
<p>First a couple of necessary msprime-adjacent custom functions:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># msprime alternative of slendr's ts_load() function for tree sequences</span>
<span class="co"># generated by the msprime backend</span>
<span class="va">load_msprime_ts</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">model</span>, <span class="va">mut_rate</span>, <span class="va">seed</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">path</span>, <span class="st">"output_msprime.trees"</span><span class="op">)</span>

  <span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu">slendr</span><span class="fu">:::</span><span class="va">tskit</span><span class="op">$</span><span class="fu">load</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/path.expand.html">path.expand</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span><span class="op">)</span>
  <span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu">slendr</span><span class="fu">:::</span><span class="va">msp</span><span class="op">$</span><span class="fu">sim_mutations</span><span class="op">(</span><span class="va">ts</span>, rate <span class="op">=</span> <span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span>
  <span class="va">ts</span>
<span class="op">}</span>

<span class="co"># This function will pull out node IDs of each individual (i.e. its two</span>
<span class="co"># chromosome IDs) from the tree sequence table metadata</span>
<span class="va">get_individuals</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">pop_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">ts</span><span class="op">$</span><span class="va">num_populations</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>
    pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">pop_ids</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">ts</span><span class="op">$</span><span class="fu">population</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">name</span><span class="op">)</span>,
    nodes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">pop_ids</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">ts</span><span class="op">$</span><span class="fu">samples</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>
      pairs <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">nodes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html">map2</a></span><span class="op">(</span><span class="va">n</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span>,
                                          <span class="va">n</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span>,
                                          <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">.x</span>, <span class="va">.y</span><span class="op">)</span><span class="op">)</span>
      <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">pairs</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">nodes</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">pop</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>ind <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">pop</span>, <span class="st">"_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># An f4 and f4-ratio calculating functions with an interface analogous to those</span>
<span class="co"># which are already part of the slendr library:</span>
<span class="va">msprime_f4</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span>, <span class="va">w</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">z</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">individuals</span> <span class="op">&lt;-</span> <span class="fu">get_individuals</span><span class="op">(</span><span class="va">ts</span><span class="op">)</span>

  <span class="va">w_nodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">individuals</span>, <span class="va">ind</span> <span class="op">==</span> <span class="va">w</span><span class="op">)</span><span class="op">$</span><span class="va">pairs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
  <span class="va">x_nodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">individuals</span>, <span class="va">ind</span> <span class="op">==</span> <span class="va">x</span><span class="op">)</span><span class="op">$</span><span class="va">pairs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
  <span class="va">y_nodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">individuals</span>, <span class="va">ind</span> <span class="op">==</span> <span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">pairs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
  <span class="va">z_nodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">individuals</span>, <span class="va">ind</span> <span class="op">==</span> <span class="va">z</span><span class="op">)</span><span class="op">$</span><span class="va">pairs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>

  <span class="va">ts</span><span class="op">$</span><span class="fu">f4</span><span class="op">(</span>sample_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">w_nodes</span>, <span class="va">x_nodes</span>, <span class="va">y_nodes</span>, <span class="va">z_nodes</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">msprime_f4ratio</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">ts</span>, <span class="va">x</span>, <span class="va">a</span>, <span class="va">b</span>, <span class="va">c</span>, <span class="va">o</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">num</span> <span class="op">&lt;-</span> <span class="fu">msprime_f4</span><span class="op">(</span><span class="va">ts</span>, <span class="va">a</span>, <span class="va">o</span>, <span class="va">x</span>, <span class="va">c</span><span class="op">)</span>
  <span class="va">den</span> <span class="op">&lt;-</span> <span class="fu">msprime_f4</span><span class="op">(</span><span class="va">ts</span>, <span class="va">a</span>, <span class="va">o</span>, <span class="va">b</span>, <span class="va">c</span><span class="op">)</span>
  <span class="va">num</span> <span class="op">/</span> <span class="va">den</span>
<span class="op">}</span></code></pre></div>
<p>Now let’s load SLiM and msprime tree sequence outputs from our models:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># SLiM outputs -- we can use built-in slendr functions for those</span>
<span class="va">slim_nogf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="va">model_nogf</span>, recapitate <span class="op">=</span> <span class="cn">TRUE</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span>, mutate <span class="op">=</span> <span class="cn">TRUE</span>,
                     Ne <span class="op">=</span> <span class="fl">10</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, mutation_rate <span class="op">=</span> <span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span>

<span class="va">slim_gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="va">model_gf</span>, recapitate <span class="op">=</span> <span class="cn">TRUE</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span>, mutate <span class="op">=</span> <span class="cn">TRUE</span>,
                   Ne <span class="op">=</span> <span class="fl">10</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, mutation_rate <span class="op">=</span> <span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span>

<span class="co"># msprime outputs -- for the moment, we have to use msprime-specific</span>
<span class="co"># functions defined above</span>
<span class="va">msprime_nogf</span> <span class="op">&lt;-</span> <span class="fu">load_msprime_ts</span><span class="op">(</span><span class="va">model_nogf</span>, <span class="va">mut_rate</span>, <span class="va">seed</span><span class="op">)</span>
<span class="va">msprime_gf</span> <span class="op">&lt;-</span> <span class="fu">load_msprime_ts</span><span class="op">(</span><span class="va">model_gf</span>, <span class="va">mut_rate</span>, <span class="va">seed</span><span class="op">)</span></code></pre></div>
<p>As we can see below, the tree sequence summary of the data loaded and processed by <em>slendr</em>’s <code><a href="../reference/ts_load.html">ts_load()</a></code> function (data produced by the SLiM backend) is very similar to what we got “manually” for the msprime-produced tree sequence using the custom defined functions. This is somehow obvious but it is a nice and cheap sanity check indicating that the tree sequence data structure produced by the two backends from the same demographic model is almost the same.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">slim_nogf</span></code></pre></div>
<pre><code>#&gt; ╔═══════════════════════════╗
#&gt; ║TreeSequence               ║
#&gt; ╠═══════════════╤═══════════╣
#&gt; ║Trees          │      79863║
#&gt; ╟───────────────┼───────────╢
#&gt; ║Sequence Length│  250000000║
#&gt; ╟───────────────┼───────────╢
#&gt; ║Time Units     │generations║
#&gt; ╟───────────────┼───────────╢
#&gt; ║Sample Nodes   │        208║
#&gt; ╟───────────────┼───────────╢
#&gt; ║Total Size     │   14.5 MiB║
#&gt; ╚═══════════════╧═══════════╝
#&gt; ╔═══════════╤══════╤═════════╤════════════╗
#&gt; ║Table      │Rows  │Size     │Has Metadata║
#&gt; ╠═══════════╪══════╪═════════╪════════════╣
#&gt; ║Edges      │158738│  4.8 MiB│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Individuals│ 21028│  2.0 MiB│         Yes║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Migrations │     0│  8 Bytes│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Mutations  │ 95424│  3.4 MiB│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Nodes      │ 21539│800.0 KiB│         Yes║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Populations│     6│  2.7 KiB│         Yes║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Provenances│     4│ 35.2 KiB│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Sites      │ 95410│  2.3 MiB│          No║
#&gt; ╚═══════════╧══════╧═════════╧════════════╝</code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">msprime_nogf</span></code></pre></div>
<pre><code>#&gt; ╔═══════════════════════════╗
#&gt; ║TreeSequence               ║
#&gt; ╠═══════════════╤═══════════╣
#&gt; ║Trees          │      81304║
#&gt; ╟───────────────┼───────────╢
#&gt; ║Sequence Length│  250000000║
#&gt; ╟───────────────┼───────────╢
#&gt; ║Time Units     │generations║
#&gt; ╟───────────────┼───────────╢
#&gt; ║Sample Nodes   │        208║
#&gt; ╟───────────────┼───────────╢
#&gt; ║Total Size     │   18.7 MiB║
#&gt; ╚═══════════════╧═══════════╝
#&gt; ╔═══════════╤══════╤═════════╤════════════╗
#&gt; ║Table      │Rows  │Size     │Has Metadata║
#&gt; ╠═══════════╪══════╪═════════╪════════════╣
#&gt; ║Edges      │324243│  9.9 MiB│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Individuals│   104│  2.9 KiB│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Migrations │     0│  8 Bytes│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Mutations  │ 94911│  3.3 MiB│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Nodes      │ 26533│725.5 KiB│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Populations│     6│414 Bytes│         Yes║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Provenances│     2│  3.8 KiB│          No║
#&gt; ╟───────────┼──────┼─────────┼────────────╢
#&gt; ║Sites      │ 94896│  2.3 MiB│          No║
#&gt; ╚═══════════╧══════╧═════════╧════════════╝</code></pre>
</div>
<div id="comparing-results-of-population-genetics-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-results-of-population-genetics-statistics" class="anchor"></a>Comparing results of population genetics statistics</h3>
<p>Having loaded both msprime and SLiM tree sequences generated by the same model, let’s see if we can get comparable results when we calculate some statistics of interest.</p>
<p>Let’s first extract all individuals from populations <code>x1</code> and <code>x2</code> and estimate the values of <span class="math inline">\(f_4(c, x1 \textrm{ or } x2; b, o)\)</span> (which should be significantly negative for individuals in the population <code>x1</code> due to it’s ancestry coming from <code>b</code> in the introgression models but consistent with zero for <code>x2</code>) and <span class="math inline">\(f_4\textrm{-ratio}(a, b, \textrm{ or } x2; c, o)\)</span>, which estimates the <em>proportion</em> of <code>b</code>-like ancestry in <code>x1</code> and <code>x2</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract vector of names of the "test individuals" in populations `x1` and `x2`</span>
<span class="va">X_individuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_samples.html">ts_samples</a></span><span class="op">(</span><span class="va">slim_gf</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">pop</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span>
<span class="va">X_individuals</span></code></pre></div>
<pre><code>#&gt;   [1] "x1_1"  "x1_2"  "x1_3"  "x1_4"  "x1_5"  "x1_6"  "x1_7"  "x1_8"  "x1_9" 
#&gt;  [10] "x1_10" "x1_11" "x1_12" "x1_13" "x1_14" "x1_15" "x1_16" "x1_17" "x1_18"
#&gt;  [19] "x1_19" "x1_20" "x1_21" "x1_22" "x1_23" "x1_24" "x1_25" "x1_26" "x1_27"
#&gt;  [28] "x1_28" "x1_29" "x1_30" "x1_31" "x1_32" "x1_33" "x1_34" "x1_35" "x1_36"
#&gt;  [37] "x1_37" "x1_38" "x1_39" "x1_40" "x1_41" "x1_42" "x1_43" "x1_44" "x1_45"
#&gt;  [46] "x1_46" "x1_47" "x1_48" "x1_49" "x1_50" "x2_1"  "x2_2"  "x2_3"  "x2_4" 
#&gt;  [55] "x2_5"  "x2_6"  "x2_7"  "x2_8"  "x2_9"  "x2_10" "x2_11" "x2_12" "x2_13"
#&gt;  [64] "x2_14" "x2_15" "x2_16" "x2_17" "x2_18" "x2_19" "x2_20" "x2_21" "x2_22"
#&gt;  [73] "x2_23" "x2_24" "x2_25" "x2_26" "x2_27" "x2_28" "x2_29" "x2_30" "x2_31"
#&gt;  [82] "x2_32" "x2_33" "x2_34" "x2_35" "x2_36" "x2_37" "x2_38" "x2_39" "x2_40"
#&gt;  [91] "x2_41" "x2_42" "x2_43" "x2_44" "x2_45" "x2_46" "x2_47" "x2_48" "x2_49"
#&gt; [100] "x2_50"</code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate f4-statistics on individuals of `x1` and `x2` populations using data</span>
<span class="co"># from the two models (a model with no gene flow and a gene flow model) -- we use</span>
<span class="co"># map_dfr to iterate across all individuals from `X_individuals` and binding all</span>
<span class="co"># resulting data frames into a single data frame</span>
<span class="va">df_slim_f4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span>
  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span><span class="op">(</span><span class="va">X_individuals</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ts_f2.html">ts_f4</a></span><span class="op">(</span><span class="va">slim_nogf</span>, <span class="st">"c_1"</span>, <span class="va">.x</span>, <span class="st">"b_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"no gene flow"</span><span class="op">)</span>,
  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span><span class="op">(</span><span class="va">X_individuals</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ts_f2.html">ts_f4</a></span><span class="op">(</span><span class="va">slim_gf</span>, <span class="st">"c_1"</span>, <span class="va">.x</span>, <span class="st">"b_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"gene flow"</span><span class="op">)</span>
<span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">f4</span>, <span class="va">model</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"SLiM backend"</span><span class="op">)</span>

<span class="co"># compute the proportions of `b` ancestry in `x1` (expected 10%) and `x2`</span>
<span class="co"># (expected 0% because this population did not receive any gene flow from `b`)</span>
<span class="va">df_slim_f4ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span>
  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span><span class="op">(</span><span class="va">X_individuals</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ts_f2.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">slim_nogf</span>, <span class="va">.x</span>, <span class="st">"a_1"</span>, <span class="st">"b_1"</span>, <span class="st">"c_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"no gene flow"</span><span class="op">)</span>,
  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span><span class="op">(</span><span class="va">X_individuals</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ts_f2.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">slim_gf</span>, <span class="va">.x</span>, <span class="st">"a_1"</span>, <span class="st">"b_1"</span>, <span class="st">"c_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"gene flow"</span><span class="op">)</span>
<span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">alpha</span>, <span class="va">model</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"SLiM backend"</span><span class="op">)</span></code></pre></div>
<p>Now, let’s repeat the same analysis on tree sequences produced by the <em>msprime</em> backend from the same two <em>slendr</em> demographic models (again, we use variants of <em>slendr</em> functions used in the previous code block because <em>msprime</em> tree sequences are not 100% compatible with SLiM tree sequences as of time of this writing).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Calculate f4 and f4-ratio statistics from the msprime sequences and format the</span>
<span class="co"># result in a way that's compatible with ts_f4() and ts_f4ratio() outputs of</span>
<span class="co"># standard slendr population genetics functions</span>

<span class="va">df_msprime_f4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">X_individuals</span>,
         f4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">X_individuals</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu">msprime_f4</span><span class="op">(</span><span class="va">msprime_nogf</span>, <span class="st">"c_1"</span>, <span class="va">i</span>, <span class="st">"b_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span>,
         model <span class="op">=</span> <span class="st">"no gene flow"</span><span class="op">)</span>,
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">X_individuals</span>,
         f4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">X_individuals</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu">msprime_f4</span><span class="op">(</span><span class="va">msprime_gf</span>, <span class="st">"c_1"</span>, <span class="va">i</span>, <span class="st">"b_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span>,
         model <span class="op">=</span> <span class="st">"gene flow"</span><span class="op">)</span>
<span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"msprime backend"</span><span class="op">)</span>

<span class="va">df_msprime_f4ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>
    X <span class="op">=</span> <span class="va">X_individuals</span>,
    alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">X_individuals</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu">msprime_f4ratio</span><span class="op">(</span><span class="va">msprime_nogf</span>, <span class="va">i</span>, <span class="st">"a_1"</span>, <span class="st">"b_1"</span>, <span class="st">"c_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span>,
    model <span class="op">=</span> <span class="st">"no gene flow"</span>
  <span class="op">)</span>,
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>
    X <span class="op">=</span> <span class="va">X_individuals</span>,
    alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">X_individuals</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu">msprime_f4ratio</span><span class="op">(</span><span class="va">msprime_gf</span>, <span class="va">i</span>, <span class="st">"a_1"</span>, <span class="st">"b_1"</span>, <span class="st">"c_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span>,
    model <span class="op">=</span> <span class="st">"gene flow"</span>
  <span class="op">)</span>
<span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"msprime backend"</span><span class="op">)</span></code></pre></div>
<p>Finally we can proceed with plotting both SLiM and msprime backend results of the two models and compare them together. As noted above, despite the fact that SLiM is a forward simulator and msprime is a coalescent simulator, population genetic expectations from both the gene flow and no gene flow models should match very closely (barring a level of uncertainty expected due to randomness of the population genetic processes at play).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df_f4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">df_slim_f4</span>, <span class="va">df_msprime_f4</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"x1_"</span>, <span class="va">X</span><span class="op">)</span>,
                             <span class="st">"x1 (received gene flow)"</span>,
                             <span class="st">"x2 (no gene flow)"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df_f4</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">f4</span>, fill <span class="op">=</span> <span class="va">population</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">simulator</span> <span class="op">~</span> <span class="va">model</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"number of individuals"</span>, x <span class="op">=</span> <span class="st">"f4 statistic"</span>,
       title <span class="op">=</span> <span class="st">"f4(c, x1 or x2; b, outgroup)"</span>,
       subtitle <span class="op">=</span> <span class="st">"f4 ~0 is consistent with no gene flow, negative value indicates gene flow with 'b'"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-07-backends_files/figure-html/msprime_slim_f4_distributions-1.png" width="640"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df_f4ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">df_slim_f4ratio</span>, <span class="va">df_msprime_f4ratio</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"x1_"</span>, <span class="va">X</span><span class="op">)</span>,
                             <span class="st">"x1 (received gene flow)"</span>,
                             <span class="st">"x2 (no gene flow)"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df_f4ratio</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">alpha</span>, fill <span class="op">=</span> <span class="va">population</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">simulator</span> <span class="op">~</span> <span class="va">model</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.1</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"number of individuals"</span>, x <span class="op">=</span> <span class="st">"ancestry proportion (f4-ratio statistic)"</span>,
       title <span class="op">=</span> <span class="st">"f4-ratio estimate of 'b' ancestry calculated from simulated data"</span>,
       subtitle <span class="op">=</span> <span class="st">"f4-ratio = f4(a, outgroup; x1 or x2, c) / f4(a, outgroup; b, c)"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-07-backends_files/figure-html/msprime_slim_f4ratio_distributions-1.png" width="640"></p>
<p>Things seem to be looking good and behave as expected! A couple of observations:</p>
<ul>
<li>In the model without gene flow, the <span class="math inline">\(f_4\)</span> statistic value is zero for both <code>x1</code> and <code>x2</code> populations. This is expected, because <span class="math inline">\(f_4\)</span> ~0 is consistent with the relationship between lineages conforming to a standard phylogenetic tree (no admixture edges). Essentially, <span class="math inline">\(f_4\)</span> is a hypothesis test of “tree-ness” of the data.</li>
<li>In the model which includes gene flow from <code>b</code> to <code>x1</code>, the <span class="math inline">\(f_4\)</span> value is significantly negative. This means that a simple tree hypothesis can be rejected and that gene flow must have occurred between <code>b</code> and <code>x1</code>. Note that this test does not reveal the <em>direction</em> of gene flow, only it’s presence.</li>
<li>Consistently with the <span class="math inline">\(f_4\)</span> results, the <span class="math inline">\(f_4-ratio\)</span> estimates match our expectations. Specifically, the statistic estimates around 10% ancestry of <code>b</code> in <code>x1</code> in the gene flow model and 0% ancestry in all other cases.</li>
<li>Most importantly, the results obtained from the SLiM and msprime back ends are nearly exactly the same. This means that (for non-spatial models) you can use SLiM and <em>msprime</em> backend interchangeably.</li>
</ul>
</div>
</div>
<div id="computing-allele-frequency-spectra" class="section level2">
<h2 class="hasAnchor">
<a href="#computing-allele-frequency-spectra" class="anchor"></a>Computing allele frequency spectra</h2>
<p>Let’s look at one more example. Imagine the following five demographic models describing the changes in the population size of some abstract population:</p>
<p><img src="../reference/figures/demographic_models.jpg" style="width:90.0%"></p>
<p>These models and the population size change events could be, depending on the specifics of our project, defined using different time units. If you work with radiocarbon-dated samples and ancient DNA you might be used to thinking about population history in units of “years before present”—in this case, the “present” time (bottom of the figure) would be at time 0 and earlier events would be specified as “5000 years before present”, etc.</p>
<p>On the other hand, if you were interested in modelling some theoretical population, you might not want to concern yourselves with this and would be quite happy starting our simulation at “generation 1” and continue until some specified “generation X”.</p>
<p>The <em>slendr</em> package makes all of this extremely easy, because you can specify times in whatever direction and units you want, as long as they are kept consistent for all populations and events in your model. This is a rather unique feature among popgen simulation software as most programs either make you express your model in “generations in forward direction”, or you have to explicitly convert time in generations going backwards.</p>
<p>In the previous section of the vignette we have shown that SLiM and msprime can be both used as simulation backend. In the remainder of this vignette we will present additional analysis, demonstrating the fact that you can express time of your models in a forward or backward direction and still run those models in <em>slendr</em>’s SLiM (forward simulator) or <em>msprime</em> (backward simulator).</p>
<p>Specifically, we will define all five demographic models from the figure above twice, in different orientations of time – forward or backward. We will then compute allele frequency spectrum from each simulation (five models, each forward or backward) and verify that both forward and backward pairs give the same result regardless of whether we use the SLiM back end or the <em>msprime</em> backend.</p>
<div id="model-definition" class="section level3">
<h3 class="hasAnchor">
<a href="#model-definition" class="anchor"></a>Model definition</h3>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">1000</span>
<span class="va">N_factor</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co"># by what factor should Ne change</span>

<span class="va">seq_len</span> <span class="op">&lt;-</span> <span class="fl">250e6</span>
<span class="va">rec_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span>
<span class="va">mut_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span></code></pre></div>
<p>First, let’s define the five population histories in a forward time direction:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># constant Ne model</span>
<span class="va">forward_const</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"const"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span>

<span class="co"># decreasing step Ne model</span>
<span class="va">forward_decr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"decr"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span>, map <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">/</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span>

<span class="co"># increasing step Ne model</span>
<span class="va">forward_incr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"inc"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">*</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span>

<span class="co"># exponential increase in size</span>
<span class="va">forward_exp_incr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"exp_inc"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">2000</span>, end <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">*</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span>

<span class="co"># exponential decrease in size</span>
<span class="va">forward_exp_decr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"exp_decr"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">2000</span>, end <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">/</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span></code></pre></div>
<p>Now let’s flip the flow of time around and define the population histories in units of “time ago”:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># constant Ne model</span>
<span class="va">backward_const</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"const"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span>

<span class="co"># decreasing step Ne model</span>
<span class="va">backward_decr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"decr"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">/</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span>

<span class="co"># increasing step Ne model</span>
<span class="va">backward_incr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"inc"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">*</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span>

<span class="co"># exponential increase in size</span>
<span class="va">backward_exp_incr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"exp_inc"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">3000</span>, end <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">*</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span>

<span class="co"># exponential decrease in size</span>
<span class="va">backward_exp_decr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"exp_decr"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">3000</span>, end <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">/</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span></code></pre></div>
<p>For convenience, we will write a helper function which will, for a given population history object (ten objects just above), compile a <em>slendr</em> model, run it in both SLiM and <em>msprime</em> back ends, load the tree sequence, and compute allele frequency spectrum (AFS) from it. This will save us lots of code repetition (always a good thing), minimize the chance of copy-pasting bugs (definitely a good thing!) and compare the AFS patterns across different models and the two <em>slendr</em> simulation back ends (they should be the same according to population genetics theory).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">compile_run_afs</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">model_name</span>, <span class="va">pop</span>, <span class="va">seed</span> <span class="op">=</span> <span class="fl">42</span><span class="op">)</span> <span class="op">{</span>
  <span class="co"># maximum length of the simulation (necessary for forward models which start</span>
  <span class="co"># in generation 1)</span>
  <span class="va">sim_length</span> <span class="op">&lt;-</span> <span class="fl">5000</span> 

  <span class="co"># define sampling times given the direction of time</span>
  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="../reference/split_time.html">split_time</a></span><span class="op">(</span><span class="va">pop</span><span class="op">)</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">sampling_time</span> <span class="op">&lt;-</span> <span class="va">sim_length</span>
    <span class="va">direction</span> <span class="op">&lt;-</span> <span class="st">"forward"</span>
  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>
    <span class="va">sampling_time</span> <span class="op">&lt;-</span> <span class="fl">0</span>
    <span class="va">direction</span> <span class="op">&lt;-</span> <span class="st">"backward"</span>
  <span class="op">}</span>

  <span class="co"># compile model</span>
  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile.html">compile</a></span><span class="op">(</span><span class="va">pop</span>, generation_time <span class="op">=</span> <span class="fl">15</span>, direction <span class="op">=</span> <span class="va">direction</span>, sim_length <span class="op">=</span> <span class="va">sim_length</span><span class="op">)</span>

  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampling.html">sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="va">sampling_time</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">pop</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span>

  <span class="co"># run the model in SLiM</span>
  <span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>,
       sampling <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

  <span class="co"># run the same model in msprim</span>
  <span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>,
          sampling <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

  <span class="co"># load the SLiM tree sequence</span>
  <span class="va">slim_ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span>
    <span class="va">model</span>, recapitate <span class="op">=</span> <span class="cn">TRUE</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span>, mutate <span class="op">=</span> <span class="cn">TRUE</span>,
    Ne <span class="op">=</span> <span class="va">N</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, mutation_rate <span class="op">=</span> <span class="va">mut_rate</span>,
    random_seed <span class="op">=</span> <span class="va">seed</span>
  <span class="op">)</span>

  <span class="co"># load the msprime tree sequence</span>
  <span class="va">msprime_ts</span> <span class="op">&lt;-</span> <span class="fu">load_msprime_ts</span><span class="op">(</span><span class="va">model</span>, <span class="va">mut_rate</span>, <span class="va">seed</span><span class="op">)</span>

  <span class="co"># compute the AFS from the SLiM and msprime tree sequences and bind the</span>
  <span class="co"># results (derived allele counts per frequency bin) in a data frame</span>
  <span class="co"># (note that we have to again treat both tree sequence objects a little</span>
  <span class="co"># differently)</span>
  <span class="co"># we remove the first element because it doesn't contain useful data</span>
  <span class="co"># (https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.allele_frequency_spectrum)</span>
  <span class="va">msprime_afs</span> <span class="op">&lt;-</span> <span class="va">msprime_ts</span><span class="op">$</span><span class="fu">allele_frequency_spectrum</span><span class="op">(</span>polarised <span class="op">=</span> <span class="cn">TRUE</span>, span_normalise <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>
  <span class="va">slim_afs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_afs.html">ts_afs</a></span><span class="op">(</span><span class="va">slim_ts</span>, polarised <span class="op">=</span> <span class="cn">TRUE</span>, span_normalise <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>

  <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"msprime"</span>, model <span class="op">=</span> <span class="va">model_name</span>, f <span class="op">=</span> <span class="va">msprime_afs</span><span class="op">)</span>,
    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"SLiM"</span>, model <span class="op">=</span> <span class="va">model_name</span>, f <span class="op">=</span> <span class="va">slim_afs</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">simulator</span>, <span class="va">model</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span>, direction <span class="op">=</span> <span class="va">direction</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>Now that this function doesn’t do anything special. It simply wraps a standard <em>slendr</em> simulation pipeline (1. create population(s); 2. compile a model object; 3. simulate the model; 4. load the results; 5. analyse the data) into a single function.</p>
<p>Now that we have the whole pipeline, we can generate all of our simulated data and bind all the result into a single data frame:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">afs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span>
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"constant"</span>, <span class="va">forward_const</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"constant"</span>, <span class="va">backward_const</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"step contraction"</span>, <span class="va">forward_decr</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"step contraction"</span>, <span class="va">backward_decr</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"step increase"</span>, <span class="va">forward_incr</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"step increase"</span>, <span class="va">backward_incr</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"exponential decrease"</span>, <span class="va">forward_exp_decr</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"exponential decrease"</span>, <span class="va">backward_exp_decr</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"exponential increase"</span>, <span class="va">forward_exp_incr</span><span class="op">)</span>,
  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"exponential increase"</span>, <span class="va">backward_exp_incr</span><span class="op">)</span>
<span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span>
    <span class="va">model</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"step contraction"</span>, <span class="st">"constant"</span>, <span class="st">"step increase"</span>,
                      <span class="st">"exponential decrease"</span>, <span class="st">"exponential increase"</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p>Finally, we can plot the allele frequency spectra from all models, for the two time directions, and for the two simulation back ends:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">afs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">n</span>, <span class="va">f</span>, color <span class="op">=</span> <span class="va">direction</span>, linetype <span class="op">=</span> <span class="va">simulator</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">model</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"number of derived alleles"</span>, y <span class="op">=</span> <span class="st">"frequency"</span>,
       title <span class="op">=</span> <span class="st">"Site frequency spectra obtained from five demographic models"</span>,
       subtitle <span class="op">=</span> <span class="st">"Each model was specified in forward or backward direction of time and executed by
two different backend scripts in slendr (SLiM and msprime)"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span><span class="st">"direction of\ntime in slendr"</span><span class="op">)</span>,
         linetype <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span><span class="st">"slendr backend\nengine used"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_linetype_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">100</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-07-backends_files/figure-html/msprime_slim_afs-1.png" width="640"></p>
<p>Again, a couple of things to note:</p>
<ul>
<li><p>The two models that involve population contraction (be it in a single step or an exponential collapse), we see a decrease in low frequency variants. On the other hand, models with population expansion show a high proportion of low frequency variants. The model with constant population size shows intermediate pattern. This is a nice validation that the simulation works as expected as it is exactly what is predicted by population genetics theory.</p></li>
<li><p>Both forward and backward time models give exactly the same result. This is not unexpected as anything else would be a bug in <em>slendr</em>. However, although unsurprising, we hope that it makes it clear that you are free to use whatever time unit specification you want to encode <em>slendr</em> models.</p></li>
<li><p>For non-spatial models, results obtained from <em>msprime</em> and SLiM back ends are consistent with one another. Again, this makes sense because although they are completely different software, population genetic theory governing the shape of allele frequency spectra applies equally to both. In fact, the same example as we have just implemented is an important part of <em>slendr</em> unit test suite which verifies that the simulations created with <em>slendr</em> are correct.</p></li>
</ul>
</div>
</div>
<div id="conclusion" class="section level2">
<h2 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h2>
<p>This example shows that if you have a standard, non-spatial demographic model and would like to implement and simulate it in R, <em>slendr</em> gives you a way to do so efficiently using it’s <em>msprime</em> back end in addition to the SLiM back end used throughout most of the documentation.</p>
</div>
<div id="faq-can-i-load-tree-sequences-generated-by-msprime-backend-with-slendr" class="section level2">
<h2 class="hasAnchor">
<a href="#faq-can-i-load-tree-sequences-generated-by-msprime-backend-with-slendr" class="anchor"></a>FAQ: Can I load tree sequences generated by msprime backend with <em>slendr</em>?</h2>
<p>The <code><a href="../reference/ts_load.html">ts_load()</a></code> and other <em>slendr</em> functions for working with tree sequences currently assume that the data was generated by SLiM. As such, loading and handling msprime tree sequences is not supported at the moment. This vignette demonstrates how easy it is to access msprime tree sequences from R using the <em>reticulate</em> Python-R interface. That said, SLiM and msprime tree sequence interfaces will be soon made compatible with one another and the need for the special treatment of msprime tree sequence data will go away soon.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a895c93f0e9eda05eb10b29a9921d586',
    indexName: 'slendr',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
