<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="slendr">
<title>Extending models with custom SLiM code • slendr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extending models with custom SLiM code">
<meta property="og:description" content="slendr">
<meta property="og:image" content="https://www.slendr.net/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">slendr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.1.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--articles">
    <span class="fa fa-book"></span>
     
    Articles
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--articles">
    <a class="dropdown-item" href="../articles/vignette-00-installation.html">Installation instructions</a>
    <a class="dropdown-item" href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    <a class="dropdown-item" href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    <a class="dropdown-item" href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    <a class="dropdown-item" href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    <a class="dropdown-item" href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a>
    <a class="dropdown-item" href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a>
    <a class="dropdown-item" href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a>
    <a class="dropdown-item" href="../articles/vignette-08-nonslendr-tskit.html">Analyzing non-slendr tree sequences</a>
    <a class="dropdown-item" href="../articles/vignette-09-paper.html">Examples from the slendr paper</a>
    <a class="dropdown-item" href="../articles/vignette-10-tracts.html">Extracting true ancestry tracts</a>
    <a class="dropdown-item" href="../articles/vignette-11-extensions.html">Selection simulations using slendr</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bodkan/slendr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/fleventy5">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Extending models with custom SLiM code</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/HEAD/vignettes/vignette-11-extensions.Rmd" class="external-link"><code>vignettes/vignette-11-extensions.Rmd</code></a></small>
      <div class="d-none name"><code>vignette-11-extensions.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p><em>slendr</em> has been designed specifically for the purpose of
making it as easy as possible to program population genetic simulations
of spatio-temporal demographic histories using SLiM. Our primary goal
was to have the means to simulate arbitrarily complex spatial scenarios
and simulate data which could be used for development of new spatial
inference population genetic methods (and benchmarking of current
methods). This functionality is briefly described in <a href="">vignette
#1</a> and in more detail in <a href="">vignette #6</a>.</p>
<p>After the spatial simulation features have been implemented, it
turned out to be trivial to support also “traditional”, non-spatial
demographic models, such as those described in <a href="">vignette
#4</a> and <a href="">vignette #5</a> – tree-like models with population
divergences, gene-flow events, and popularion resize events, like you
can see in any evolutionary biology textbook. Because non-spatial models
can be often implemented (and executed) more efficient in a coalescent
setting, we added the possibility to run any compiled <em>slendr</em>
model not through SLiM but through a back-end script implemented in
<em>msprime</em>, hidden behind the <em>slendr</em> function
<code><a href="../reference/msprime.html">msprime()</a></code>.</p>
<p>Throughout all this time, <em>slendr</em> models were purely neutral,
without much planning to extend simulations toward non-neutral
scenarios. However, <strong>with the easy specification of non-spatial
demographic models and their simulation with the <em>slendr</em>/SLiM
engine, many users have been asking about the possibility to simulate
non-neutral scenarios or, more generally, customization of the
simplified genome architecture assumed by <em>slendr</em> by default (a
single chromosome with uniform recombination and purely neutral
mutations).</strong></p>
<p>This vignette describes how this can be done with
<em>slendr</em>.</p>
<p><strong>Why even do this if SLiM can do these things on its own? Fair
question!</strong> After all, SLiM can obviously simulate nearly any
kind of conceivable evolutionary model.</p>
<p>One motivation for supporting non-neutral customized genomic
architecture and mutation models in <em>slendr</em> is from users who
find it easier to program complex “base” demographic models in R (such
as models with complex history of population divergences, gene-flow
events, resizes), let <em>slendr</em> take care of demographic history .
You only have to provide custom SLiM code for non-neutral
evolution.**</p>
</div>
<div class="section level3">
<h3 id="disclaimers-and-caveats">Disclaimers and caveats<a class="anchor" aria-label="anchor" href="#disclaimers-and-caveats"></a>
</h3>
<p>If you want to use the <em>slendr</em>/SLiM extension functionality
described in this vignette, you should be aware of the following
restrictions and guidelines:</p>
<ul>
<li><p>The extension functionality is primarily design to support
complex selection models using customized SLiM snippets while still
using <em>slendr</em> for handling the basic demographic modeling
scaffold (population splits, resizes, gene flows, etc.). Because the
handling of these demographic events remains unchanged even for
<code><a href="../reference/slim.html">slim()</a></code> models, the coalescent <code><a href="../reference/msprime.html">msprime()</a></code> engine
of <em>slendr</em> (which implements the coalescent, neutral counterpart
of <em>slendr</em> models) has not changed and there is no means to
customize it.</p></li>
<li><p>It is assumed that the <a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.slim" class="external-link">SLiM
script bundled with <em>slendr</em></a> stays unmodified. Any extension
<em>SLiM</em> code customization snippets will be added on top of it –
customizing mutation types, genomic element types, recombination maps,
custom fitness callbacks, etc. Functionality involving demographic
events (splits, gene-flow, etc.) will stay in place as it is.
User-provided custom code is not allowed to change how this
functionality operates.</p></li>
<li><p>Related to the above point – <em>slendr</em> models assume (and
will always assume) Wright-Fisher (WF) dynamics. Although perhaps not
necessarily a problem for simulating non-spatial selection models
(i.e. <em>slendr</em> models which don’t include a map), adding
selection to spatial models in SLiM WF setting is unlikely to lead to
very meaningful results. If you need non-WF models, you will have to use
pure SLiM. <em>slendr</em> will not be useful for you.</p></li>
<li><p>The built-in SLiM engine of <em>slendr</em> provides several
utility Eidos functions which make it easy to extend the engine with
custom callbacks, and give the possibility to refer to some
<em>slendr</em>-specific model information. See the list in the section
below.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="slendr-slim-api">
<em>slendr</em> / SLiM “API”<a class="anchor" aria-label="anchor" href="#slendr-slim-api"></a>
</h3>
<p>Below we will describe the following slendr / SLiM “API” functions
and constants you can use to customize the default, SLiM back-end script
that comes bundled with <em>slendr</em>:</p>
<ul>
<li><p><code><a href="../reference/population.html">population()</a></code></p></li>
<li><p><code>tick()</code> and <code>model_time()</code></p></li>
<li><p><code>save_state()</code> and <code>reset_state()</code></p></li>
<li><p><code>write_log()</code></p></li>
<li><p><code>SIMULATION_START</code> and
<code>SIMULATION_END</code></p></li>
<li><p><code>SEQUENCE_LENGTH</code></p></li>
<li><p><code>OUTPUT_DIR</code></p></li>
</ul>
<p>Let’s first load all required R libraries before we dive in:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/init_env.html">init_env</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">seed</span> <span class="op">&lt;-</span> <span class="fl">42</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="referring-to-populations-population-eidos-function">Referring to populations: <code>population()</code> Eidos
function<a class="anchor" aria-label="anchor" href="#referring-to-populations-population-eidos-function"></a>
</h4>
<p>As a recap, when we program a <em>slendr</em> model, we can refer to
populations using symbolic names such as “AFR”, “EUR”, etc. in our R
scripts in the following way:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">afr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"AFR"</span>, time <span class="op">=</span> <span class="fl">100000</span>, N <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span></span>
<span><span class="va">eur</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"EUR"</span>, time <span class="op">=</span> <span class="fl">60000</span>, N <span class="op">=</span> <span class="fl">2000</span>, parent <span class="op">=</span> <span class="va">afr</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># &lt;... compile model and simulate a tree sequence `ts` ...&gt;</span></span></code></pre></div>
<p>When we then want to analyze the tree-sequence output of
<em>slendr</em> models, we can refer to populations (or individuals
sampled from those populations) using the same symbolic names like this
(not with integer node IDs internally used by <em>tskit</em>):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute heterozygosity in the individual "EUR"</span></span>
<span><span class="fu"><a href="../reference/ts_diversity.html">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"EUR_1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute genetic divergence between selected Africans and Europeans</span></span>
<span><span class="va">afr_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"AFR_3"</span><span class="op">)</span></span>
<span><span class="va">eur_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span>, <span class="st">"EUR_3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ts_divergence.html">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>afr <span class="op">=</span> <span class="va">afr_samples</span>, eur <span class="op">=</span> <span class="va">eur_samples</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Along the same lines, when you’re customizing a <em>slendr</em>/SLiM
simulation, you can get a <code>Subpopulation</code> SLiM object
corresponding to a symbolic name of a <em>slendr</em> population (as
used by the R side of things) using the Eidos function
<code><a href="../reference/population.html">population()</a></code>. For instance, let’s assume we compiled a toy
model of human demography from the <a href="">vignette #4</a> and ran it
in the SLiMgui using <code>slim(..., method = "gui")</code>. We can then
open an Eidos console in the GUI and type the following:</p>
<pre><code>&gt; // get a SLiM object corresponding to the "AFR" population
&gt; population("AFR")
Subpopulation&lt;p0&gt;
&gt; 
&gt; // get the number of genomes in the population
&gt; length(population("AFR").genomes)
6000
&gt; 
&gt; // get both "AFR" and "OOA" subpopulation objects
&gt; population(c("AFR", "OOA"))
Subpopulation&lt;p0&gt; Subpopulation&lt;p1&gt;</code></pre>
<p>This way, if our population is named “AFR” or “OOA”, when we write a
SLiM extension code for <em>slendr</em> we don’t have to know whether
this or that population is called <code>p0</code> or <code>p2</code> on
the SLiM side, etc. We can use the names of populations as programmed on
the R side of slendr.</p>
<p>Additionally, the <code><a href="../reference/population.html">population()</a></code> Eidos function provides
some helpful error checking. For instance, if we try to get a SLiM
<code>Subpopulation</code> object corresponding to a <em>slendr</em>
population which will exist at some point but doesn’t yet exist in a
running SLiM simulation at a particular time, we will get an informative
error:</p>
<pre><code>&gt; population("YAM")
The following populations not present in tick 81 (slendr model time 97600): YAM</code></pre>
<p>If we make a typo and try to access a population which doesn’t exist
in a <em>slendr</em> model at all, we get another informative error
message:</p>
<pre><code>&gt; population("asdf")
Not all provided population identifiers are present in the model.
Check your code to make sure that all of these are defined: asdf</code></pre>
<p><strong>How is this function useful in practice, though?</strong>
Because the purpose of this vignette is to show how to extend
<em>slendr</em> to non-neutral scenarios, we could for instance select a
random chromosome from a given population using the
<code><a href="../reference/population.html">population()</a></code> function like this (maybe to add a beneficial
mutation). We use this bit of code in the more elaborate complete
examples below.</p>
<pre><code>target_genome = sample(population("AFR").genomes, 1);

target_genome.addNewMutation(m0, selectionCoeff = 0.05, position = 1000000);</code></pre>
</div>
<div class="section level4">
<h4 id="referring-to-times-tick-eidos-function">Referring to times: <code>tick()</code> Eidos function<a class="anchor" aria-label="anchor" href="#referring-to-times-tick-eidos-function"></a>
</h4>
<p>Another useful feature of <em>slendr</em> is its ability to use
arbitrary time units in model definition. For instance, when creating a
“EUR” population in the snippet above, we wrote this:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">afr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"AFR"</span>, time <span class="op">=</span> <span class="fl">90000</span>, N <span class="op">=</span> <span class="fl">20000</span><span class="op">)</span></span>
<span><span class="va">eur</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"EUR"</span>, time <span class="op">=</span> <span class="fl">60000</span>, N <span class="op">=</span> <span class="fl">2000</span>, parent <span class="op">=</span> <span class="va">afr</span><span class="op">)</span></span></code></pre></div>
<p>The numbers 100000 and 60000 in the split time are not meaningful by
themselves, but if we compile and run the <em>slendr</em> model like
this (we run it in SLiMgui here to be able to demonstrate various Eidos
functions):</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span>populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="va">eur</span><span class="op">)</span>, generation_time <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">100000</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, method <span class="op">=</span> <span class="st">"gui"</span><span class="op">)</span></span></code></pre></div>
<p>We can then interpret 90000 and 60000 as “years before present”,
which can be quite convenient when we’re building models using
radiocarbon-dated or fossil-dated ages, rather than using traditional
units of generations (forward in time as in SLiM, backwards in time as
in _msprime). Any time we later refer to a particular time, either in a
model visualization or during tree-sequence, we can use these “natural
units” without having to convert years before present into generations
forward in time.</p>
<p>Similarly to referring to <em>slendr</em> population symbolic names
in a consistent way between R and SLiM itself, we can also refer to
times of events using <em>slendr</em>-specific time units using the
function <code>tick()</code>. This function takes in a time in
<em>slendr</em> time units – years before present, generations backwards
in time, whatever you used in your <em>slendr</em> R script, and
translates those to SLiM’s “ticks”. For instance, in our example model
of modern human history, we could get the tick number corresponding to
the time of 40 thousand years ago by calling:</p>
<pre><code>&gt; tick(40000)
1668</code></pre>
<p>Similarly to the consistency check performed by the
<code><a href="../reference/population.html">population()</a></code> Eidos function, <code>tick()</code> also makes
sure that the time given lies within the time window expected for the
running simulation. For instance, our example model only starts at 100
thousand years ago, so if we try to get the tick number corresponding to
a million years ago, we get this:</p>
<pre><code>&gt; tick(1e6)
Some of the times fall outside of the range of the slendr model:
  - oldest possible event: 90000
  - youngest possible event: 0

The offending times were: 1000000</code></pre>
<p>Indeed, the tick-based time boundaries of the model are:</p>
<pre><code>&gt; // the very first time point of the simulation
&gt; tick(90e3)
1
&gt; // the very last time point of the simulation
&gt; tick(0)
3001</code></pre>
<p>As with <code><a href="../reference/population.html">population()</a></code>, we can also perform the
<em>slendr</em>-time-to-tick conversion in a vectorized manner:</p>
<pre><code>&gt; tick(c(90000, 0))
1 3001</code></pre>
<p>Naturally, if our model does not use any special time units (for
instance, if all times are encoded in generations, even forward in
time), the <code>tick()</code> function becomes an identity
function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"pop"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">simple_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span><span class="va">pop</span>, generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">simple_model</span>, sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, method <span class="op">=</span> <span class="st">"gui"</span><span class="op">)</span></span></code></pre></div>
<p>Indeed, if we pop up the Eidos console in SLiMgui for this model, we
can verify this:</p>
<pre><code>&gt; // no conversion needed for models using times of generations
&gt; tick(c(1, 1001))
1 1001</code></pre>
<p>Still, the <code>tick()</code> function can be useful even for
<em>slendr</em> models which use the same time units as SLiM itself
(generations forward in time) because it provides useful boundary
checking. For instance, this is what happens if we try to get the tick
number in a model above (which only runs from generation 1 to generation
1 + 1000):</p>
<pre><code>&gt; tick(c(0, 1e6))
Some of the times fall outside of the range of the slendr model:
  - oldest possible event: 1
  - youngest possible event: 1001

The offending times were: 0, 1000000</code></pre>
<p>Additionally, <strong>the <code>tick()</code> function automatically
takes care of offsetting the tick count when a burn-in period was
specified for the simulation.</strong></p>
<p>Below we’ll see that thanks to the possibility of using arbitrary
integer expressions in SLiM 4.2, we can use the <code>tick()</code>
Eidos function for easy and straightforward scheduling of custom
callbacks.</p>
</div>
<div class="section level4">
<h4 id="referring-to-model-times-model_time-eidos-function">Referring to model times: <code>model_time()</code> Eidos
function<a class="anchor" aria-label="anchor" href="#referring-to-model-times-model_time-eidos-function"></a>
</h4>
<p>This function is an inverse to the <code>tick()</code> function. For
instance, if we want to write out an output with a time stamp using
times of the <em>slendr</em> model (like years before present) and not
tick numbers.</p>
<p>As an example, this gives us the model time (in years ago)
corresponding to the first tick:</p>
<pre><code>&gt; model_time(1)
90000</code></pre>
<p>This gives us model time corresponding to the very last tick of the
simulation (i.e. the present-day at “0 years before present”):</p>
<pre><code>&gt; model_time(3001)
0</code></pre>
</div>
<div class="section level4">
<h4 id="logging-write_log-eidos-function">Logging: <code>write_log()</code> Eidos function<a class="anchor" aria-label="anchor" href="#logging-write_log-eidos-function"></a>
</h4>
<p><code>write_log()</code> is a tiny helper function provided by
<em>slendr</em>’s SLiM back-end script which serves to print out a given
string to the SLiM log output together with the appropriate tick number
at that time. This produces output of the following kind:</p>
<pre><code>&gt; write_log("hello from the current event")
tick 696: hello from the current event</code></pre>
</div>
<div class="section level4">
<h4 id="saving-and-re-starting-simulation-state-save_state-and-reset_state">Saving and re-starting simulation state: <code>save_state()</code>
and <code>reset_state()</code><a class="anchor" aria-label="anchor" href="#saving-and-re-starting-simulation-state-save_state-and-reset_state"></a>
</h4>
<p>Whenever we’re dealing with simulations of, say, trajectories of
beneficial alleles, we usually have to take care of situations in which
the allele of interest gets lost before the simulation finishes running.
In such cases, we often need to reset the simulation to a state just
before the mutation was added and try again. Section 9.2 of the
venerable SLiM manual (“Making sweeps conditional on fixation”) is a
great example on how to solve this with base SLiM.</p>
<p>Although it would be quite easy to use the same approach in the
<em>slendr</em>/SLiM extension snippets described here, there’s one
issue with this approach: <code>sim.outputFull()</code> and
<code>sim.readFromPopulationFile()</code> do not preserve some important
<em>slendr</em> tags which are assigned using
<code>&lt;Subpopulation&gt;.{set,get}Value()</code> methods. Doing so
would require poking into <em>slendr</em>’s SLiM codebase, which would
be confusing for any user.</p>
<p>To circumvent the problem, <em>slendr</em>’s SLiM back-end script
provides the following functions:</p>
<ul>
<li>
<code>save_state()</code>: saves the full state of the SLiM
simulation (just as <code>sim.outputFull()</code> does), while also
saving those few <em>slendr</em>-specific values;</li>
<li>
<code>reset_state()</code>: restores the full SLiM simulation state
(including <em>slendr</em> specific tags and values), and
<strong>chooses a new random seed</strong>. The latter is performed
because restarting a simulation using the <code>save_state()</code> -
<code>reset_state()</code> tandem is practically always done in order to
change the outcome of a simulation.</li>
</ul>
<p>To demonstrate how these two functions might be used in practice,
let’s say we wanted to build on the (by default purely neutral!) model
of African and Eurasian history from <a href="">this vignette</a>
discussed above, and say that we want to add a beneficial mutation to
the “EUR” population at time 15 ky ago. We could utilize the few bits of
Eidos code introduced so far to define the following <em>slendr</em>
extension snippet (for now let’s ignore the question of how to actually
plug this into a <em>slendr</em> simulation):</p>
<pre><code>function (void) add_mutation(s pop_name, f selection_coef) {
  // sample the first target carrier chromosome of the new mutation...
  target = sample(population(pop_name).genomes, 1);
  // ... and add the mutation to it
  mutation = target.addNewDrawnMutation(m0, position = 1);

  defineGlobal("BACKGROUND", target.mutations);
  defineGlobal("FOCAL", mutation);

  write_log("adding beneficial mutation to population " + pop_name);
}

tick(15000) late() {
  // save simulation state in case we need to restart if the mutation is lost
  save_state();

  add_mutation("pop", s);
}

tick(15000):SIMULATION_END late() {
  segregating = sim.countOfMutationsOfType(m0) &gt; 0;
  fixed = sum(sim.substitutions.mutationType == m0) == 1;
  
  // the mutation is not segregating and is not fixed either -- we must restart
  if (!segregating &amp; !fixed) {
    write_log("mutation lost -- restarting");
    
    reset_state();
    
    add_mutation("EUR");
  }
}</code></pre>
<p><strong>Note that for simplicity we don’t define our own mutation
types, and simply re-use the <code>m0</code> mutation type used by
<em>slendr</em> by default. This is unlikely to be very useful for more
complex simulations and we will look at a more general solution
below.</strong> The point of this example is to show the use of
<code>save_state()</code> and <code>reset_state()</code> in
practice.</p>
</div>
<div class="section level4">
<h4 id="global-constants">Global constants<a class="anchor" aria-label="anchor" href="#global-constants"></a>
</h4>
<p>By inspecting the built-in SLiM simulation script of <em>slendr</em>,
you will find that it contains a number of global constants. Most of
them should be considered internal and users shouldn’t rely on them in
their code. However, one useful constant which you can see being used in
the snippet above is <code>SEQUENCE_LENGTH</code> – this is the total
amount of sequence simulated (i.e., the number passed as
<code>slim(&lt;model&gt;, sequence_length = &lt;SEQUENCE_LENGTH&gt;, ...)</code>
in your R code).</p>
<p>Of course, specifying your own <code>initialize() {...}</code>
callback with your own genomic elements will give you more flexibility
and you might want to skip specifying <code>sequence_length =</code>
entirely, as we also show below.</p>
<p>Another very useful pair of <em>slendr</em> constants are
<code>SIMULATION_START</code> and <code>SIMULATION_END</code>, which
specify at which time points (in ticks!) does a given <em>slendr</em>
model start and end. The latter can be particularly helpful for
customized output callbacks.</p>
<p>Even more relevant to writing customized SLiM extension scripts for
<em>slendr</em> are constants <code>TS_PATH</code> (the location to
where an output tree sequence will be written by <em>slendr</em>, if
needed) and <code>OUTPUT_DIR</code>, which is the path to the output
directory which can be specified via the <code>output_dir =</code>
argument of the <code><a href="../reference/slim.html">slim()</a></code> function. The latter is useful for
specifying where should all user-output files be saved, making it
possible to pick them up in downstream analyses.</p>
</div>
</div>
<div class="section level2">
<h2 id="practical-examples">Practical examples<a class="anchor" aria-label="anchor" href="#practical-examples"></a>
</h2>
<div class="section level3">
<h3 id="putting-it-all-together-running-a-customized-slendr-simulation">Putting it all together: running a customized <em>slendr</em>
simulation<a class="anchor" aria-label="anchor" href="#putting-it-all-together-running-a-customized-slendr-simulation"></a>
</h3>
<p>The above examples show how you can refer to a <em>slendr</em>
population with its symbolic name on the SLiM side using the
<code><a href="../reference/population.html">population()</a></code> Eidos function provided by the <em>slendr</em>
built-in SLiM script, and how you can convert <em>slendr</em>-specific
time units into SLiM’s internal ticks using the <code>tick()</code>
function. You’ve also learned how to save and reset SLiM state using
provided functions <code>save_state()</code> and
<code>reset_state()</code> and how to log outputs using the function
<code>write_log()</code>. We demonstrated all this by defining a
customized snippet of SLiM code which we now want to use in a full
<em>slendr</em> simulation run. Of course, the example we’ve chosen is
extremely trivial – you could, in principle, use any feature available
for SLiM (as long as it’s compatible with Wright-Fisher models, which
<em>slendr</em> currently assumes as a basis for its models).</p>
<p>How do we use this customization in practice? One way to do this
would be to edit the built-in <em>slendr</em> SLiM script and manually
add your own SLiM code, but that would be brittle and not reproducible.
There’s a much better way to do this that’s directly supported by
<em>slendr</em>’s R interface.</p>
<p>Let’s say that we defined the following model of modern human
demographic history in slendr. This is exactly the same example as the
one we show in <a href="">vignette #4</a>:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/init_env.html">init_env</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; The interface to all required Python modules has been activated.</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># African ancestral population</span></span>
<span><span class="va">afr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"AFR"</span>, time <span class="op">=</span> <span class="fl">90000</span>, N <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># first migrants out of Africa</span></span>
<span><span class="va">ooa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"OOA"</span>, parent <span class="op">=</span> <span class="va">afr</span>, time <span class="op">=</span> <span class="fl">60000</span>, N <span class="op">=</span> <span class="fl">500</span>, remove <span class="op">=</span> <span class="fl">23000</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fl">2000</span>, time <span class="op">=</span> <span class="fl">40000</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Eastern hunter-gatherers</span></span>
<span><span class="va">ehg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"EHG"</span>, parent <span class="op">=</span> <span class="va">ooa</span>, time <span class="op">=</span> <span class="fl">28000</span>, N <span class="op">=</span> <span class="fl">1000</span>, remove <span class="op">=</span> <span class="fl">6000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># European population</span></span>
<span><span class="va">eur</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"EUR"</span>, parent <span class="op">=</span> <span class="va">ehg</span>, time <span class="op">=</span> <span class="fl">25000</span>, N <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Anatolian farmers</span></span>
<span><span class="va">ana</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"ANA"</span>, time <span class="op">=</span> <span class="fl">28000</span>, N <span class="op">=</span> <span class="fl">3000</span>, parent <span class="op">=</span> <span class="va">ooa</span>, remove <span class="op">=</span> <span class="fl">4000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Yamnaya steppe population</span></span>
<span><span class="va">yam</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"YAM"</span>, time <span class="op">=</span> <span class="fl">8000</span>, N <span class="op">=</span> <span class="fl">500</span>, parent <span class="op">=</span> <span class="va">ehg</span>, remove <span class="op">=</span> <span class="fl">2500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define gene-flow events</span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">ana</span>, to <span class="op">=</span> <span class="va">yam</span>, rate <span class="op">=</span> <span class="fl">0.4</span>, start <span class="op">=</span> <span class="fl">7900</span>, end <span class="op">=</span> <span class="fl">7800</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">ana</span>, to <span class="op">=</span> <span class="va">eur</span>, rate <span class="op">=</span> <span class="fl">0.5</span>, start <span class="op">=</span> <span class="fl">6000</span>, end <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">yam</span>, to <span class="op">=</span> <span class="va">eur</span>, rate <span class="op">=</span> <span class="fl">0.65</span>, start <span class="op">=</span> <span class="fl">4000</span>, end <span class="op">=</span> <span class="fl">3500</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-11-extensions_files/figure-html/model_plot-1.png" width="480"></p>
<p>Let’s also assume we have the following “SLiM snippet file”:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">extension_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"extension_trajectory.txt"</span>, package <span class="op">=</span> <span class="st">"slendr"</span><span class="op">)</span></span></code></pre></div>
<pre><code>// Because we want to simulate non-neutral evolution, we have to provide a
// custom initialization callback -- slendr will use it to replace its default
// neutral genomic architecture (i.e. the initialize() {...} callback it uses
// by default for neutral simulations). Note that we can refer to slendr's
// constants SEQUENCE_LENGTH and RECOMBINATION_RATE, which will carry values
// passed through from R via slendr's slim() R function.
initialize() {
    // define some parameters of the model
    defineConstant("s", 0.1);
    defineConstant("onset_time", 15000);
    defineConstant("target_pop", "EUR");

    initializeMutationType("m1", 0.5, "f", s);

    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, SEQUENCE_LENGTH - 1);

    initializeMutationRate(0);
    initializeRecombinationRate(RECOMBINATION_RATE);

    defineConstant("output_file", "~/Desktop/output.tsv");
}

function (void) add_mutation(void) {
    // sample one target carrier of the new mutation...
    target = sample(population(target_pop).genomes, 1);
    // ... and add the mutation in the middle of it
    mut = target.addNewDrawnMutation(m1, position = asInteger(SEQUENCE_LENGTH / 2));

    // save the mutation for later reference
    defineGlobal("MUTATION", mut);

    write_log("adding beneficial mutation to population " + target_pop);

    // write the header of the output file
    writeFile(output_file, "time\tfrequency");
}

tick(onset_time) late() {
    // save simulation state in case we need to restart if the mutation is lost
    // (save_state() is a built-in function provided by slendr for customization)
    save_state();

    add_mutation();
}

tick(onset_time):SIMULATION_END late() {
    // the mutation is not segregating and is not fixed either -- we must restart
    if (!MUTATION.isSegregating &amp; !MUTATION.isFixed) {
        write_log("mutation lost -- restarting");

        // reload the simulation state from just before we added the beneficial
        // mutation above (reset_state() is another built-in slendr function)
        reset_state();

        add_mutation();
    }

    // compute the frequency of the mutation of interest
    frequency = population("EUR").genomes.mutationFrequenciesInGenomes();

    // save the current frequency to the output file
    writeFile(output_file,
              model_time(community.tick) + "\t" +
              frequency, append = T);
}</code></pre>
<p>We can include the extension snippet into the standard
<em>slendr</em> engine SLiM script by providing a path to it in
<code><a href="../reference/compile_model.html">compile_model()</a></code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span></span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="va">ooa</span>, <span class="va">ehg</span>, <span class="va">eur</span>, <span class="va">ana</span>, <span class="va">yam</span><span class="op">)</span>,</span>
<span>  gene_flow <span class="op">=</span> <span class="va">gf</span>, generation_time <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  extension <span class="op">=</span> <span class="va">extension_path</span>  <span class="co"># &lt;--- include the SLiM extension snippet</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can check that the extension snippet was really appended to the
built-in SLiM engine script by running <code><a href="../reference/slim.html">slim()</a></code> function and
setting <code>method = "gui"</code> (look towards the end of the
script!):</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">0</span>, ts <span class="op">=</span> <span class="cn">FALSE</span>, method <span class="op">=</span> <span class="st">"gui"</span><span class="op">)</span></span></code></pre></div>
<p>Note that we set <code>ts = FALSE</code> in the <code><a href="../reference/slim.html">slim()</a></code>
call. This way we can switch off generating of a tree sequence because
we don’t care about it in this example. We only want to save the
frequency trajectory of the beneficial allele as saved in
<code>~/Desktop/output.tsv</code>.</p>
<p>Checking the output file shows that the frequency trajectories were
indeed saved:</p>
<pre><code>$ head ~/Desktop/output.tsv 
time    frequency
15000   0.0001
14970   0.0005
14940   0.0007
14910   0.0008
14880   0.0005
14850   0.0007
14820   0.0009
14790   0.001
14760   0.0013</code></pre>
<p>Speaking of which: one thing that’s a little annoying about the <a href="https://github.com/bodkan/slendr/blob/selection-extension/inst/extdata/extension_trajectory.txt" class="external-link">extension
snippet</a> used in this example is that all the parameters are
hard-coded (selection coefficient, target population, and even the
output file path). If we wanted to run the <em>slendr</em> script for
different values of selection coefficient, or examining the behavior of
the model depending on in which population does the beneficial allele
arise, we’d have to create multiple copies of the extension script. We
can do better using <em>slendr</em>’s support for substitution or
“templating”.</p>
</div>
<div class="section level3">
<h3 id="putting-it-all-together-parametrizing-a-customized-slendr-simulation">Putting it all together: parametrizing a customized <em>slendr</em>
simulation<a class="anchor" aria-label="anchor" href="#putting-it-all-together-parametrizing-a-customized-slendr-simulation"></a>
</h3>
<p>For easy parametrization of customized <em>slendr</em> / SLiM models,
slendr provides a function <code><a href="../reference/substitute_values.html">substitute_values()</a></code>. Simply
speaking, rather than having to hard code values of parameters in your
extension SLiM snippet files, you can indicate that a parameter value
should be substituted in the file using a simple syntax
<code>{{parameter_name}}</code>.</p>
<p>Take a look at a <a href="https://github.com/bodkan/slendr/blob/selection-extension/inst/extdata/extension_trajectory_params.txt" class="external-link">new,
flexible version</a> of the snippet we used above and look for the
<code>{{...}}</code> template placeholders. This version has been
expanded to study the trajectory of the focal selected allele depending
on in which population it originated.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">extension_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"extension_trajectory_params.txt"</span>, package <span class="op">=</span> <span class="st">"slendr"</span><span class="op">)</span></span></code></pre></div>
<pre><code>// Define model constants (to be substituted) all in one place
// (each {{placeholder}} will be replaced by a value passed from R).
// Note that string constant template patterns are surrounded by "quotes"!
initialize() {
    defineConstant("s", {{s}});
    defineConstant("onset_time", {{onset_time}});
    defineConstant("target_pop", "{{target_pop}}");
    defineConstant("origin_pop", "{{origin_pop}}");

    // compose an output file based on given parameters
    defineConstant("output_file", OUTPUT_DIR + "/" +
                                  "traj_" + target_pop + "_" + origin_pop + ".tsv");
}
// Because we want to simulate non-neutral evolution, we have to provide a
// custom initialization callback -- slendr will use it to replace its default
// neutral genomic architecture (i.e. the initialize() {...} callback it uses
// by default for neutral simulations). Note that we can refer to slendr's
// constants SEQUENCE_LENGTH and RECOMBINATION_RATE, which will carry values
// passed through from R via slendr's slim() R function.
initialize() {
    initializeMutationType("m1", 0.5, "f", s);

    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, SEQUENCE_LENGTH - 1);

    initializeMutationRate(0);
    initializeRecombinationRate(RECOMBINATION_RATE);
}

function (void) add_mutation(void) {
    // sample one target carrier of the new mutation...
    target = sample(population(origin_pop).genomes, 1);
    // ... and add the mutation in the middle of it
    mut = target.addNewDrawnMutation(m1, position = asInteger(SEQUENCE_LENGTH / 2));

    // save the mutation for later reference
    defineGlobal("MUTATION", mut);

    write_log("adding beneficial mutation to population " + target_pop);

    writeFile(output_file, "time\tfreq_origin\tfreq_target");
}

tick(onset_time) late() {
    // save simulation state in case we need to restart if the mutation is lost
    save_state();

    add_mutation();
}

tick(onset_time):SIMULATION_END late() {
    // the mutation is not segregating and is not fixed either -- we must restart
    if (!MUTATION.isSegregating &amp; !MUTATION.isFixed) {
        write_log("mutation lost -- restarting");

        reset_state();

        add_mutation();
    }

    // compute the frequency of the mutation of interest and save it (if the
    // mutation is missing at this time, save its frequency as NA)
    freq_origin = "NA";
    freq_target = "NA";
    if (population(origin_pop, check = T))
      freq_origin = population(origin_pop).genomes.mutationFrequenciesInGenomes();
    if (population(target_pop, check = T))
      freq_target = population(target_pop).genomes.mutationFrequenciesInGenomes();

    writeFile(output_file,
              model_time(community.tick) + "\t" +
              freq_origin + "\t" +
              freq_target, append = T);
}</code></pre>
<p>Substitute values of parameters in a customization SLiM extension
script:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">extension</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/substitute_values.html">substitute_values</a></span><span class="op">(</span></span>
<span>  <span class="va">extension_path</span>,</span>
<span>  s <span class="op">=</span> <span class="fl">0.1</span>, onset_time <span class="op">=</span> <span class="fl">15000</span>,</span>
<span>  origin_pop <span class="op">=</span> <span class="st">"EUR"</span>, target_pop <span class="op">=</span> <span class="st">"EUR"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Missing a parameter gives an error immediately:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">extension</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/substitute_values.html">substitute_values</a></span><span class="op">(</span></span>
<span>  <span class="va">extension_path</span>,</span>
<span>  onset_time <span class="op">=</span> <span class="fl">15000</span>,</span>
<span>  origin_pop <span class="op">=</span> <span class="st">"EUR"</span>, target_pop <span class="op">=</span> <span class="st">"EUR"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Error: The extension script contains the following unsubstituted patterns: {{s}}</span></span></code></pre></div>
<p>Then plug the parametrized script into <code><a href="../reference/compile_model.html">compile_model()</a></code>
just as we did above:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span></span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="va">ooa</span>, <span class="va">ehg</span>, <span class="va">eur</span>, <span class="va">ana</span>, <span class="va">yam</span><span class="op">)</span>,</span>
<span>  gene_flow <span class="op">=</span> <span class="va">gf</span>, generation_time <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  extension <span class="op">=</span> <span class="va">extension</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">0</span>, output_dir <span class="op">=</span> <span class="va">output_dir</span>,</span>
<span>     ts <span class="op">=</span> <span class="cn">FALSE</span>, random_seed <span class="op">=</span> <span class="fl">42</span><span class="op">)</span></span></code></pre></div>
<p>We can leverage the flexibility of this solution to run another
version of this model, this time adding the mutation to a different
population. In fact, let’s check what happens when the beneficial allele
appears in EHG, ANA, and YAM populations. Note that this uses the same
extension snippet, and just substitutes different values to the
placeholder parameters:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">run_model</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">origin_pop</span>, <span class="va">onset_time</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">extension</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/substitute_values.html">substitute_values</a></span><span class="op">(</span></span>
<span>    <span class="va">extension_path</span>,</span>
<span>    s <span class="op">=</span> <span class="fl">0.1</span>, onset_time <span class="op">=</span> <span class="va">onset_time</span>,</span>
<span>    origin_pop <span class="op">=</span> <span class="va">origin_pop</span>, target_pop <span class="op">=</span> <span class="st">"EUR"</span></span>
<span>  <span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="va">ooa</span>, <span class="va">ehg</span>, <span class="va">eur</span>, <span class="va">ana</span>, <span class="va">yam</span><span class="op">)</span>,</span>
<span>    gene_flow <span class="op">=</span> <span class="va">gf</span>, generation_time <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    extension <span class="op">=</span> <span class="va">extension</span></span>
<span>  <span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">0</span>,</span>
<span>       output_dir <span class="op">=</span> <span class="va">output_dir</span>, ts <span class="op">=</span> <span class="cn">FALSE</span>, random_seed <span class="op">=</span> <span class="fl">42</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">run_model</span><span class="op">(</span>origin_pop <span class="op">=</span> <span class="st">"EUR"</span>, onset_time <span class="op">=</span> <span class="fl">15000</span><span class="op">)</span></span>
<span><span class="fu">run_model</span><span class="op">(</span>origin_pop <span class="op">=</span> <span class="st">"ANA"</span>, onset_time <span class="op">=</span> <span class="fl">15000</span><span class="op">)</span></span>
<span><span class="fu">run_model</span><span class="op">(</span>origin_pop <span class="op">=</span> <span class="st">"EHG"</span>, onset_time <span class="op">=</span> <span class="fl">15000</span><span class="op">)</span></span>
<span><span class="fu">run_model</span><span class="op">(</span>origin_pop <span class="op">=</span> <span class="st">"YAM"</span>, onset_time <span class="op">=</span> <span class="fl">8000</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize the different trajectories like this. Not that the
different allele frequency trajectories reflect the more complex
demographic history in the European population, particularly the
dilution of the frequency due to influx of ancestry from other
populations into Europeans, as shown in the demographic tree above. In
simulations in which the allele originated not directly in EUR but in
another population where EUR can trace its ancestry through ancient gene
flow, the trajectory of the allele leading to fixation is a bit delayed,
depending on when a particular gene flow happened. Still, because it’s
been under such a strong selection in all of them that it reached
fixation even prior to a gene flow and because the gene flow happened in
such a strong proportion, it reaches fixation much faster.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">load_traj</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">origin_pop</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">output_dir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"traj_EUR_"</span>, <span class="va">origin_pop</span>, <span class="st">".tsv"</span><span class="op">)</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="va">df</span><span class="op">$</span><span class="va">origin</span> <span class="op">&lt;-</span> <span class="va">origin_pop</span></span>
<span>  <span class="va">df</span><span class="op">$</span><span class="va">target</span> <span class="op">&lt;-</span> <span class="st">"EUR"</span></span>
<span>  <span class="va">df</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">traj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu">load_traj</span><span class="op">(</span><span class="st">"EUR"</span><span class="op">)</span>, <span class="fu">load_traj</span><span class="op">(</span><span class="st">"ANA"</span><span class="op">)</span>, <span class="fu">load_traj</span><span class="op">(</span><span class="st">"EHG"</span><span class="op">)</span>, <span class="fu">load_traj</span><span class="op">(</span><span class="st">"YAM"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">traj</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">freq_target</span>, linetype <span class="op">=</span> <span class="st">"EUR"</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">freq_origin</span>, color <span class="op">=</span> <span class="va">origin</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">15000</span>, <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Allele frequency in EUR given the origin in another population"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"years before present"</span>, y <span class="op">=</span> <span class="st">"allele frequency"</span>,</span>
<span>       color <span class="op">=</span> <span class="st">"frequency\nin original\npopulation"</span>,</span>
<span>       linetype <span class="op">=</span> <span class="st">"frequency\nin target\npopulation"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_linetype_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"solid"</span>, <span class="st">"dashed"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">origin</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning: Removed 419 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_line()`).</span></span></code></pre>
<p><img src="vignette-11-extensions_files/figure-html/trajectories-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="programming-slendrslim-extension-snippets-directly-in-r-scripts">Programming <em>slendr</em>/SLiM extension snippets directly in R
scripts<a class="anchor" aria-label="anchor" href="#programming-slendrslim-extension-snippets-directly-in-r-scripts"></a>
</h3>
<p>Thanks to the multiline string support implemented in R 4.2, we can
specify the SLiM extension code directly as an R string inside our R
script, instead of having to plug it in from an external file like we
did in the previous example. Including the SLiM code directly in this
way makes it a little easier to iterate during development.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extension "template" provided as a single string (this contains the same code</span></span>
<span><span class="co"># as the script used just above, except specified directly in R)</span></span>
<span><span class="va">extension_template</span> <span class="op">&lt;-</span> <span class="st">r"(</span></span>
<span><span class="st">// Because we want to simulate non-neutral evolution, we have to provide a</span></span>
<span><span class="st">// custom initialization callback -- slendr will use it to replace its default</span></span>
<span><span class="st">// neutral genomic architecture (i.e. the initialize() {...} callback it uses</span></span>
<span><span class="st">// by default for neutral simulations). Note that we can refer to slendr's</span></span>
<span><span class="st">// constants SEQUENCE_LENGTH and RECOMBINATION_RATE, which will carry values</span></span>
<span><span class="st">// passed through from R via slendr's slim() R function.</span></span>
<span><span class="st">initialize() {</span></span>
<span><span class="st">    initializeMutationType("m1", 0.5, "f", 0.0);</span></span>
<span><span class="st"></span></span>
<span><span class="st">    initializeGenomicElementType("g1", m1, 1.0);</span></span>
<span><span class="st">    initializeGenomicElement(g1, 0, SEQUENCE_LENGTH - 1);</span></span>
<span><span class="st"></span></span>
<span><span class="st">    initializeMutationRate(0);</span></span>
<span><span class="st">    initializeRecombinationRate(RECOMBINATION_RATE);</span></span>
<span><span class="st">}</span></span>
<span><span class="st"></span></span>
<span><span class="st">// Define model constants (to be substituted) all in one place</span></span>
<span><span class="st">// (each {{placeholder}} will be replaced by a value passed from R).</span></span>
<span><span class="st">// Note that string constant template patterns are surrounded by "quotes"!</span></span>
<span><span class="st">initialize() {</span></span>
<span><span class="st">    defineConstant("s", {{s}});</span></span>
<span><span class="st">    defineConstant("onset_time", {{onset_time}});</span></span>
<span><span class="st">    defineConstant("target_pop", "{{target_pop}}");</span></span>
<span><span class="st">    defineConstant("origin_pop", "{{origin_pop}}");</span></span>
<span><span class="st"></span></span>
<span><span class="st">    // compose an output file based on given parameters</span></span>
<span><span class="st">    defineConstant("output_file", OUTPUT_DIR + "/" +</span></span>
<span><span class="st">                                  "traj_" + target_pop + "_" + origin_pop + ".tsv");</span></span>
<span><span class="st">}</span></span>
<span><span class="st"></span></span>
<span><span class="st">function (void) add_mutation(void) {</span></span>
<span><span class="st">    // sample one target carrier of the new mutation...</span></span>
<span><span class="st">    target = sample(population(origin_pop).genomes, 1);</span></span>
<span><span class="st">    // ... and add the mutation in the middle of it</span></span>
<span><span class="st">    mut = target.addNewMutation(m1, s, position = asInteger(SEQUENCE_LENGTH / 2));</span></span>
<span><span class="st"></span></span>
<span><span class="st">    // save the mutation for later reference</span></span>
<span><span class="st">    defineGlobal("MUTATION", mut);</span></span>
<span><span class="st"></span></span>
<span><span class="st">    write_log("adding beneficial mutation to population " + target_pop);</span></span>
<span><span class="st"></span></span>
<span><span class="st">    writeFile(output_file, "tick\ttime\tfreq_origin\tfreq_target");</span></span>
<span><span class="st">}</span></span>
<span><span class="st"></span></span>
<span><span class="st">tick(onset_time) late() {</span></span>
<span><span class="st">    // save simulation state in case we need to restart if the mutation is lost</span></span>
<span><span class="st">    save_state();</span></span>
<span><span class="st"></span></span>
<span><span class="st">    add_mutation();</span></span>
<span><span class="st">}</span></span>
<span><span class="st"></span></span>
<span><span class="st">tick(onset_time):SIMULATION_END late() {</span></span>
<span><span class="st">    // the mutation is not segregating and is not fixed either -- we must restart</span></span>
<span><span class="st">    if (!MUTATION.isSegregating &amp; !MUTATION.isFixed) {</span></span>
<span><span class="st">        write_log("mutation lost -- restarting");</span></span>
<span><span class="st"></span></span>
<span><span class="st">        reset_state();</span></span>
<span><span class="st"></span></span>
<span><span class="st">        add_mutation();</span></span>
<span><span class="st">    }</span></span>
<span><span class="st"></span></span>
<span><span class="st">    // compute the frequency of the mutation of interest and save it (if the</span></span>
<span><span class="st">    // mutation is missing at this time, save its frequency as NA)</span></span>
<span><span class="st">    freq_origin = "NA";</span></span>
<span><span class="st">    freq_target = "NA";</span></span>
<span><span class="st">    if (population(origin_pop, check = T))</span></span>
<span><span class="st">      freq_origin = population(origin_pop).genomes.mutationFrequenciesInGenomes();</span></span>
<span><span class="st">    if (population(target_pop, check = T))</span></span>
<span><span class="st">      freq_target = population(target_pop).genomes.mutationFrequenciesInGenomes();</span></span>
<span><span class="st"></span></span>
<span><span class="st">    writeFile(output_file,</span></span>
<span><span class="st">              community.tick + "\t" +</span></span>
<span><span class="st">              model_time(community.tick) + "\t" +</span></span>
<span><span class="st">              freq_origin + "\t" +</span></span>
<span><span class="st">              freq_target, append = T);</span></span>
<span><span class="st">}</span></span>
<span><span class="st">)"</span></span></code></pre></div>
<p>Having defined the SLiM snippet in this way, the rest of our code
will work in exactly the same way as in the example just above (the only
thing that we changed in the code below is providing the SLiM extension
string directly instead of providing a path to a file):</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">run_model</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">origin_pop</span>, <span class="va">onset_time</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">extension</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/substitute_values.html">substitute_values</a></span><span class="op">(</span></span>
<span>    <span class="va">extension_template</span>, <span class="co"># &lt;--- template SLiM code string directly (not as a file!)</span></span>
<span>    s <span class="op">=</span> <span class="fl">0.1</span>, onset_time <span class="op">=</span> <span class="va">onset_time</span>,</span>
<span>    origin_pop <span class="op">=</span> <span class="va">origin_pop</span>, target_pop <span class="op">=</span> <span class="st">"EUR"</span></span>
<span>  <span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span></span>
<span>    populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="va">ooa</span>, <span class="va">ehg</span>, <span class="va">eur</span>, <span class="va">ana</span>, <span class="va">yam</span><span class="op">)</span>,</span>
<span>    gene_flow <span class="op">=</span> <span class="va">gf</span>, generation_time <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    extension <span class="op">=</span> <span class="va">extension</span></span>
<span>  <span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">1e6</span>, recombination_rate <span class="op">=</span> <span class="fl">0</span>,</span>
<span>       output_dir <span class="op">=</span> <span class="va">output_dir</span>, ts <span class="op">=</span> <span class="cn">FALSE</span>, random_seed <span class="op">=</span> <span class="fl">42</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">run_model</span><span class="op">(</span><span class="st">"EUR"</span>, onset_time <span class="op">=</span> <span class="fl">15000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">load_traj</span><span class="op">(</span><span class="st">"EUR"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   tick  time freq_origin freq_target origin target</span></span>
<span><span class="co">#&gt; 1 2501 15000       1e-04       1e-04    EUR    EUR</span></span>
<span><span class="co">#&gt; 2 2502 14970       3e-04       3e-04    EUR    EUR</span></span>
<span><span class="co">#&gt; 3 2503 14940       4e-04       4e-04    EUR    EUR</span></span>
<span><span class="co">#&gt; 4 2504 14910       5e-04       5e-04    EUR    EUR</span></span>
<span><span class="co">#&gt; 5 2505 14880       4e-04       4e-04    EUR    EUR</span></span>
<span><span class="co">#&gt; 6 2506 14850       4e-04       4e-04    EUR    EUR</span></span></code></pre>
<div class="section level4">
<h4 id="customizing-genomic-architecture-selective-sweep-simulations">Customizing genomic architecture (selective sweep simulations)<a class="anchor" aria-label="anchor" href="#customizing-genomic-architecture-selective-sweep-simulations"></a>
</h4>
<p><strong>Customization of <em>slendr</em> / SLiM models can extend
also to the genomic architecture itself.</strong> Because altering the
initialization procedure by providing a custom
<code>initialize() {...}</code> callback in the SLiM extension code
entirely overrides setting up a (by default just one) genomic element
type and recombination rate, users might want to set those up entirely
by themselves. This means that it’s possible to avoid the
<code>SEQUENCE_LENGTH</code> and <code>RECOMBINATION_RATE</code>
arguments used by the <em>slendr</em> back-end engine provided via the
<code>slim(..., sequence_length = ..., recombination_rate = ..., ...)</code>
arguments.</p>
<p>Let’s take the following <em>slendr</em> model of Neanderthal
introgression:</p>
<!-- ```{r} -->
<!-- library(slendr) -->
<!-- init_env() -->
<!-- # ancestor of Neanderthals and anatomically modern humans -->
<!-- # ancestor <- population("ancestor", time = 650e3, N = 10, remove = 599e3) -->
<!-- # two populations of anatomically modern humans: Africans and Europeans -->
<!-- # afr <- population("AFR", parent = ancestor, time = 600e3, N = 5000) -->
<!-- # eur <- population("EUR", parent = afr, time = 80e3, N = 5000) -->
<!-- eur <- population("EUR", time = 80e3, N = 10000) -->
<!-- # Neanderthal population splitting at 600 ky ago from modern humans -->
<!-- # nea <- population("NEA", parent = ancestor, time = 600e3, N = 1000, remove = 40e3) -->
<!-- nea <- population("NEA", time = 600e3, N = 1000, remove = 40e3) -->
<!-- # 3% Neanderthal introgression into Europeans between 55-50 ky ago -->
<!-- gf <- gene_flow(from = nea, to = eur, rate = 0.05, start = 55000, end = 54500) -->
<!-- ``` -->
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create the ancestor of everyone and a chimpanzee outgroup</span></span>
<span><span class="co"># (we set both N = 1 to reduce the computational time for this model)</span></span>
<span><span class="va">chimp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"CH"</span>, time <span class="op">=</span> <span class="fl">6.5e6</span>, N <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># two populations of anatomically modern humans: Africans and Europeans</span></span>
<span><span class="va">afr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"AFR"</span>, parent <span class="op">=</span> <span class="va">chimp</span>, time <span class="op">=</span> <span class="fl">6e6</span>, N <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span><span class="va">eur</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"EUR"</span>, parent <span class="op">=</span> <span class="va">afr</span>, time <span class="op">=</span> <span class="fl">70e3</span>, N <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Neanderthal population splitting at 600 ky ago from modern humans</span></span>
<span><span class="co"># (becomes extinct by 40 ky ago)</span></span>
<span><span class="va">nea</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"NEA"</span>, parent <span class="op">=</span> <span class="va">afr</span>, time <span class="op">=</span> <span class="fl">600e3</span>, N <span class="op">=</span> <span class="fl">1000</span>, remove <span class="op">=</span> <span class="fl">40e3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 5% Neanderthal introgression into Europeans between 55-50 ky ago</span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">nea</span>, to <span class="op">=</span> <span class="va">eur</span>, rate <span class="op">=</span> <span class="fl">0.05</span>, start <span class="op">=</span> <span class="fl">55000</span>, end <span class="op">=</span> <span class="fl">45000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span></span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">chimp</span>, <span class="va">nea</span>, <span class="va">afr</span>, <span class="va">eur</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>  generation_time <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-11-extensions_files/figure-html/introgression_model-1.png" width="360"></p>
<p>Now we put together SLiM extension code to set up a non-neutral
<em>slendr</em> simulation:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">extension</span> <span class="op">&lt;-</span> <span class="st">r"(</span></span>
<span><span class="st">initialize() {</span></span>
<span><span class="st">  // model parameters to be substitute_values()'d from R below</span></span>
<span><span class="st">  defineConstant("gene_length", {{gene_length}});</span></span>
<span><span class="st">  defineConstant("n_genes", {{n_genes}});</span></span>
<span><span class="st">  defineConstant("n_markers", {{n_markers}});</span></span>
<span><span class="st">  defineConstant("introgression_time", {{introgression_time}});</span></span>
<span><span class="st">  defineConstant("output_file", OUTPUT_DIR + "/{{output_file}}");</span></span>
<span><span class="st"></span></span>
<span><span class="st">  // total length of the genome to be simulated</span></span>
<span><span class="st">  defineConstant("total_length", n_genes * gene_length);</span></span>
<span><span class="st"></span></span>
<span><span class="st">  // positions of neutral Neanderthal markers along the genome</span></span>
<span><span class="st">  defineConstant("neutral_pos", seq(0, total_length - 1, by = gene_length / n_markers));</span></span>
<span><span class="st">  // positions of deleterious mutations in the center of each gene</span></span>
<span><span class="st">  defineConstant("selected_pos", seq(gene_length / 2, total_length - 1, by = gene_length));</span></span>
<span><span class="st">}</span></span>
<span><span class="st"></span></span>
<span><span class="st">// Because we want to simulate non-neutral evolution, we have to provide a</span></span>
<span><span class="st">// custom initialization callback -- slendr will use it to replace its default</span></span>
<span><span class="st">// neutral genomic architecture (i.e. the initialize() {...} callback it uses</span></span>
<span><span class="st">// by default for neutral simulations).</span></span>
<span><span class="st">initialize() {</span></span>
<span><span class="st">  initializeMutationType("m0", 0.5, "f", 0.0); // neutral Neanderthal marker mutation type</span></span>
<span><span class="st">  initializeMutationType("m1", 0.5, "f", {{s}}); // deleterious Neanderthal mutation type</span></span>
<span><span class="st">  initializeGenomicElementType("g1", m0, 1.0); // genomic type of 'genes'</span></span>
<span><span class="st"></span></span>
<span><span class="st">  genes = c(); // gene start-end coordinates</span></span>
<span><span class="st">  breakpoints = c(); // recombination breakpoints</span></span>
<span><span class="st">  rates = c(); // recombination rates</span></span>
<span><span class="st"></span></span>
<span><span class="st">  // compute coordinates of genes, as well as the recombination map breakpoints</span></span>
<span><span class="st">  // between each gene</span></span>
<span><span class="st">  start = 0;</span></span>
<span><span class="st">  for (i in seqLen(n_genes)) {</span></span>
<span><span class="st">    // end of the next gene</span></span>
<span><span class="st">    end = start + gene_length - 1;</span></span>
<span><span class="st">    genes = c(genes, start, end);</span></span>
<span><span class="st"></span></span>
<span><span class="st">    // uniform recombination within a gene, followed by a 0.5 recombination breakpoint</span></span>
<span><span class="st">    rates = c(rates, RECOMBINATION_RATE, 0.5);</span></span>
<span><span class="st">    breakpoints = c(breakpoints, end, end + 1);</span></span>
<span><span class="st"></span></span>
<span><span class="st">    // start of the following genes</span></span>
<span><span class="st">    start = end + 1;</span></span>
<span><span class="st">  }</span></span>
<span><span class="st"></span></span>
<span><span class="st">  // odd elements --&gt; starts of genes</span></span>
<span><span class="st">  gene_starts = integerMod(seqAlong(genes), 2) == 0;</span></span>
<span><span class="st">  // even elements --&gt; ends of genes</span></span>
<span><span class="st">  gene_ends = integerMod(seqAlong(genes), 2) == 1;</span></span>
<span><span class="st"></span></span>
<span><span class="st">  // set up all the genes at once</span></span>
<span><span class="st">  initializeGenomicElement(g1, genes[gene_starts], genes[gene_ends]);</span></span>
<span><span class="st"></span></span>
<span><span class="st">  // set up the recombination map</span></span>
<span><span class="st">  initializeRecombinationRate(rates, breakpoints);</span></span>
<span><span class="st"></span></span>
<span><span class="st">  // no mutation rate (we will add neutral variation after the SLiM run)</span></span>
<span><span class="st">  initializeMutationRate(0);</span></span>
<span><span class="st">}</span></span>
<span><span class="st"></span></span>
<span><span class="st">// Add Neanderthal-specific mutations one tick prior to the introgression</span></span>
<span><span class="st">tick(introgression_time) - 1 late() {</span></span>
<span><span class="st">  // get all Neanderthal chromosomes just prior to the introgression</span></span>
<span><span class="st">  target = population("NEA").genomes;</span></span>
<span><span class="st"></span></span>
<span><span class="st">  write_log("adding neutral Neanderthal markers");</span></span>
<span><span class="st"></span></span>
<span><span class="st">  mutations = target.addNewDrawnMutation(m0, position = asInteger(neutral_pos));</span></span>
<span><span class="st">  defineConstant("MARKERS", target.mutations);</span></span>
<span><span class="st"></span></span>
<span><span class="st">  write_log("adding deleterious Neanderthal mutation");</span></span>
<span><span class="st"></span></span>
<span><span class="st">  target.addNewDrawnMutation(m1, position = asInteger(selected_pos));</span></span>
<span><span class="st">}</span></span>
<span><span class="st"></span></span>
<span><span class="st">// At the end, output Neanderthal ancestry proportions along the genome to a file</span></span>
<span><span class="st">// (in our R analysis code we will compare the results of this with the</span></span>
<span><span class="st">// equivalent computation using a tree sequence)</span></span>
<span><span class="st">SIMULATION_END late() {</span></span>
<span><span class="st">  df = DataFrame(</span></span>
<span><span class="st">    "gene", asInteger(MARKERS.position / gene_length),</span></span>
<span><span class="st">    "pos", MARKERS.position,</span></span>
<span><span class="st">    "freq", sim.mutationFrequencies(population("EUR"), MARKERS)</span></span>
<span><span class="st">  );</span></span>
<span><span class="st">  writeFile(output_file, df.serialize(format = "tsv"));</span></span>
<span><span class="st">}</span></span>
<span><span class="st">)"</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/substitute_values.html">substitute_values</a></span><span class="op">(</span></span>
<span>    introgression_time <span class="op">=</span> <span class="fl">55000</span>, s <span class="op">=</span> <span class="op">-</span><span class="fl">0.001</span>,</span>
<span>    gene_length <span class="op">=</span> <span class="fl">5e6</span>, n_genes <span class="op">=</span> <span class="fl">200</span>, n_markers <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    output_file <span class="op">=</span> <span class="st">"output_frequencies.tsv"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>And put everything together by compiling a <em>slendr</em> model
(which now swaps <em>slendr</em>’s default neutrality for a non-neutral
genomic architecture, including deleterious mutations):</p>
<p>We will also record 10 Neanderthals and 100 Africans and Europeans
each in a tree sequence:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nea_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fl">50000</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">nea</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">modern_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">eur</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">nea_samples</span>, <span class="va">modern_samples</span><span class="op">)</span></span></code></pre></div>
<p>Using the compiled model, we can simulate a tree sequence using the
<code><a href="../reference/slim.html">slim()</a></code> function like in any standard <em>slendr</em>
workflow. However, note that we were able to skip
<code>sequence_length =</code> and <code>recombination_rate =</code>
arguments! This is because our non-neutral extension snippet already
customizes though, so there’s no reason for us to instruct
<code><a href="../reference/slim.html">slim()</a></code> on that front in any way:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, samples <span class="op">=</span> <span class="va">samples</span>, output_dir <span class="op">=</span> <span class="va">output_dir</span><span class="op">)</span></span></code></pre></div>
<p>It’s been suggested that Neanderthals had a significantly lower <span class="math inline">\(N_e\)</span> compared to anatomically modern
humans (AMH), which might have decreased efficacy of negative selection
against deleterious variants compared to AMH. Consequently, following
Neanderthal admixture into AMH, introgressed Neanderthal DNA in AMH
would have been under negative selection, particularly in regions of the
genome under stronger selective constraints.</p>
<p>The above implies that if we compute the proportion of Neanderthal
ancestry along a sample of European genomes, we should see a clear dip
in the amount of surviving Neanderthal DNA as a function of distance
from a locus under selection. Because our customized SLiM snippet saved
the allele frequency of each Neanderthal marker in Europeans, we can
just load this data and directly visualize it:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">freqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"~/Projects/introgression_data/output_frequencies.tsv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>pos <span class="op">=</span> <span class="va">pos</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">5e6</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">40000000</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (3): gene, pos, freq</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">freqs</span>, <span class="va">pos</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>freq <span class="op">=</span> <span class="fl">100</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">freq</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">pos</span>, <span class="va">freq</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">5e6</span> <span class="op">/</span> <span class="fl">2</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"coordinate along a gene"</span>, y <span class="op">=</span> <span class="st">"Neanderthal ancestry proportion [%]"</span>,</span>
<span>       title <span class="op">=</span> <span class="st">"Proportion of Neanderthal ancestry in Europeans along 5Mb independent genes"</span>,</span>
<span>       subtitle <span class="op">=</span> <span class="st">"(dashed line indicates introgressed deleterious Neanderthal allele)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-11-extensions_files/figure-html/introgression_freqs-1.png" width="480"></p>
<p>The hypothesis clearly stands against the data!</p>
<p>We shouldn’t forget that the core output data structure of
<em>slendr</em> is a tree sequence. This gives us much more flexibility
in terms of which statistics can we compute, and how effectively can
they be computed because we don’t need any mutations or genotype files
to calculate them.</p>
<p>For instance, as an alternative to the direct computation of allele
frequencies during the SLiM run (saved in
<code>output_frequencies.tsv</code>), we could use the simulated tree
sequence to quickly calculate the genetic divergence between Europeans
and a Neanderthal in windows along the simulated genome.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_genes</span> <span class="op">&lt;-</span> <span class="fl">200</span></span>
<span><span class="va">gene_length</span> <span class="op">&lt;-</span> <span class="fl">5e6</span></span>
<span><span class="va">window_length</span> <span class="op">&lt;-</span> <span class="fl">100e3</span></span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load a tree sequence and extract the names of recorded individuals</span></span>
<span><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"~/Projects/introgression_data/output.trees"</span>, <span class="va">model</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_names.html">ts_names</a></span><span class="op">(</span><span class="va">ts</span>, split <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $EUR</span></span>
<span><span class="co">#&gt;   [1] "EUR_1"   "EUR_2"   "EUR_3"   "EUR_4"   "EUR_5"   "EUR_6"   "EUR_7"  </span></span>
<span><span class="co">#&gt;   [8] "EUR_8"   "EUR_9"   "EUR_10"  "EUR_11"  "EUR_12"  "EUR_13"  "EUR_14" </span></span>
<span><span class="co">#&gt;  [15] "EUR_15"  "EUR_16"  "EUR_17"  "EUR_18"  "EUR_19"  "EUR_20"  "EUR_21" </span></span>
<span><span class="co">#&gt;  [22] "EUR_22"  "EUR_23"  "EUR_24"  "EUR_25"  "EUR_26"  "EUR_27"  "EUR_28" </span></span>
<span><span class="co">#&gt;  [29] "EUR_29"  "EUR_30"  "EUR_31"  "EUR_32"  "EUR_33"  "EUR_34"  "EUR_35" </span></span>
<span><span class="co">#&gt;  [36] "EUR_36"  "EUR_37"  "EUR_38"  "EUR_39"  "EUR_40"  "EUR_41"  "EUR_42" </span></span>
<span><span class="co">#&gt;  [43] "EUR_43"  "EUR_44"  "EUR_45"  "EUR_46"  "EUR_47"  "EUR_48"  "EUR_49" </span></span>
<span><span class="co">#&gt;  [50] "EUR_50"  "EUR_51"  "EUR_52"  "EUR_53"  "EUR_54"  "EUR_55"  "EUR_56" </span></span>
<span><span class="co">#&gt;  [57] "EUR_57"  "EUR_58"  "EUR_59"  "EUR_60"  "EUR_61"  "EUR_62"  "EUR_63" </span></span>
<span><span class="co">#&gt;  [64] "EUR_64"  "EUR_65"  "EUR_66"  "EUR_67"  "EUR_68"  "EUR_69"  "EUR_70" </span></span>
<span><span class="co">#&gt;  [71] "EUR_71"  "EUR_72"  "EUR_73"  "EUR_74"  "EUR_75"  "EUR_76"  "EUR_77" </span></span>
<span><span class="co">#&gt;  [78] "EUR_78"  "EUR_79"  "EUR_80"  "EUR_81"  "EUR_82"  "EUR_83"  "EUR_84" </span></span>
<span><span class="co">#&gt;  [85] "EUR_85"  "EUR_86"  "EUR_87"  "EUR_88"  "EUR_89"  "EUR_90"  "EUR_91" </span></span>
<span><span class="co">#&gt;  [92] "EUR_92"  "EUR_93"  "EUR_94"  "EUR_95"  "EUR_96"  "EUR_97"  "EUR_98" </span></span>
<span><span class="co">#&gt;  [99] "EUR_99"  "EUR_100"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $NEA</span></span>
<span><span class="co">#&gt; [1] "NEA_1"</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute coordinates of sliding windows along the genome</span></span>
<span><span class="va">windows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0</span>, to <span class="op">=</span> <span class="va">n_genes</span> <span class="op">*</span> <span class="va">gene_length</span> <span class="op">-</span> <span class="fl">1</span>, by <span class="op">=</span> <span class="va">window_length</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">windows</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] 0e+00 1e+05 2e+05 3e+05 4e+05 5e+05</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">windows</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] 999400000 999500000 999600000 999700000 999800000 999900000</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute divergence from the tree sequence in each window separately</span></span>
<span><span class="va">divergence</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_divergence.html">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">samples</span>, windows <span class="op">=</span> <span class="va">windows</span>, mode <span class="op">=</span> <span class="st">"branch"</span><span class="op">)</span><span class="op">$</span><span class="va">divergence</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute average divergence at each position in a gene, and a 95% C.I.</span></span>
<span><span class="va">div_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  pop <span class="op">=</span> <span class="st">"eur"</span>,</span>
<span>  pos <span class="op">=</span> <span class="va">windows</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="va">gene_length</span>,</span>
<span>  div <span class="op">=</span> <span class="va">divergence</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">pop</span>, <span class="va">pos</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span></span>
<span>    mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">div</span><span class="op">)</span>, n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, std <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">div</span><span class="op">)</span>,</span>
<span>    ci_low <span class="op">=</span> <span class="va">mean</span> <span class="op">-</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">std</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span>,</span>
<span>    ci_up <span class="op">=</span> <span class="va">mean</span> <span class="op">+</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">std</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">#&gt; `summarise()` has grouped output by 'pop'. You can override using the `.groups`</span></span>
<span><span class="co">#&gt; argument.</span></span></code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">div_df</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html" class="external-link">geom_ribbon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">pos</span>, ymin <span class="op">=</span> <span class="va">ci_low</span>, ymax <span class="op">=</span> <span class="va">ci_up</span><span class="op">)</span>, fill <span class="op">=</span> <span class="st">"grey70"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">pos</span>, <span class="va">mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span> <span class="op">-</span> <span class="va">window_length</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"coordinate along a gene"</span>, y <span class="op">=</span> <span class="st">"divergence to Neanderthal"</span>,</span>
<span>       title <span class="op">=</span> <span class="st">"Divergence of Europeans to a Neanderthal genome along 5Mb independent genes"</span>,</span>
<span>       subtitle <span class="op">=</span> <span class="st">"(dashed line indicates introgressed deleterious Neanderthal allele)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-11-extensions_files/figure-html/introgression_divergence-1.png" width="480"></p>
<p>Indeed, we clearly observe lower introgression levels (indicated by
an increased European-Neanderthal divergence) around loci under
selection and increased surviving introgression far from those loci,
because purifying selection has removed Neanderthal introgressed DNA
from regions under selective constraint.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
