<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Spatially annotated tree sequences • slendr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Spatially annotated tree sequences">
<meta property="og:description" content="slendr">
<meta property="og:image" content="https://www.slendr.net/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slendr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-00-installation.html">Installation instructions</a>
    </li>
    <li>
      <a href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    </li>
    <li>
      <a href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    </li>
    <li>
      <a href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    </li>
    <li>
      <a href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    </li>
    <li>
      <a href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a>
    </li>
    <li>
      <a href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a>
    </li>
    <li>
      <a href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a>
    </li>
    <li>
      <a href="../articles/vignette-08-nonslendr-tskit.html">Analyzing non-slendr tree sequences</a>
    </li>
    <li>
      <a href="../articles/vignette-09-paper.html">Examples from the slendr paper</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/slendr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/dr_bodkan" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Spatially annotated tree sequences</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/HEAD/vignettes/vignette-06-locations.Rmd" class="external-link"><code>vignettes/vignette-06-locations.Rmd</code></a></small>
      <div class="hidden name"><code>vignette-06-locations.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The main selling point of the <em>slendr</em> R package is
programming complex <em>spatially explicit</em> population genetic
models. Because we use SLiM as the simulation engine, we can store
simulated data efficiently in a tree sequence format which allows us to
run large population-scale simulations. In the previous vignettes, we
described how you can specify spatial population dynamics and how you
can access tree sequence data and calculate population genetic
statistics on it (focusing on <a href="vignette-05-tree-sequences.html">non-spatial models</a> for
simplicity). Now it’s time to show you how to work with simulated tree
sequence in a spatial context.</p>
</div>
<div class="section level2">
<h2 id="model-specification">Model specification<a class="anchor" aria-label="anchor" href="#model-specification"></a>
</h2>
<p>Let’s first load all required R libraries:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>

<span class="va">seed</span> <span class="op">&lt;-</span> <span class="fl">314159</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span><span class="op">)</span></code></pre></div>
<!-- Let's also set up the Python _slendr_ environment which we will use to analyse tree sequence data: -->
<!-- ```{r} -->
<!-- # activate slendr's Python interface, installing all dependencies if necessary -->
<!-- setup_env() -->
<!-- # check that we have the correct environment -->
<!-- check_env() -->
<!-- ``` -->
<p>We begin by specifying our spatial model. We will use the same
demographic model of modern human history in West Eurasia, which we
extensively discussed in the <a href="vignette-01-tutorial.html">introductory tutorial</a> and on the
main <a href="../index.html">landing page</a>. Here is a complete model
definition script, without further comments:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># simulated world map</span>
<span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/world.html">world</a></span><span class="op">(</span>
  xrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">15</span>, <span class="fl">60</span><span class="op">)</span>,
  yrange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">65</span><span class="op">)</span>,
  crs <span class="op">=</span> <span class="st">"EPSG:3035"</span>
<span class="op">)</span>

<span class="co"># couple of broad geographic regions</span>
<span class="va">africa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/region.html">region</a></span><span class="op">(</span><span class="st">"Africa"</span>, <span class="va">map</span>, polygon <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">18</span>, <span class="fl">20</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">40</span>, <span class="fl">20</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">33</span><span class="op">)</span>,
                                               <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">32</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">35</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>, <span class="fl">35</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">europe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/region.html">region</a></span><span class="op">(</span><span class="st">"Europe"</span>, <span class="va">map</span>, polygon <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">8</span>, <span class="fl">35</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">36</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">38</span><span class="op">)</span>,
                                               <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">35</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">25</span>, <span class="fl">35</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">33</span>, <span class="fl">45</span><span class="op">)</span>,
                                               <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">58</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">60</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">15</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">anatolia</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/region.html">region</a></span><span class="op">(</span><span class="st">"Anatolia"</span>, <span class="va">map</span>, polygon <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">28</span>, <span class="fl">35</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">40</span>, <span class="fl">35</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">42</span>, <span class="fl">40</span><span class="op">)</span>,
                                                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">43</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">27</span>, <span class="fl">40</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">25</span>, <span class="fl">38</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># define population histories</span>

<span class="co"># African ancestral population</span>
<span class="va">afr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span>
  <span class="st">"AFR"</span>, parent <span class="op">=</span> <span class="st">"ancestor"</span>, time <span class="op">=</span> <span class="fl">52000</span>, N <span class="op">=</span> <span class="fl">3000</span>,
  map <span class="op">=</span> <span class="va">map</span>, polygon <span class="op">=</span> <span class="va">africa</span>
<span class="op">)</span>

<span class="co"># population of the first migrants out of Africa</span>
<span class="va">ooa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span>
  <span class="st">"OOA"</span>, parent <span class="op">=</span> <span class="va">afr</span>, time <span class="op">=</span> <span class="fl">51000</span>, N <span class="op">=</span> <span class="fl">500</span>, remove <span class="op">=</span> <span class="fl">25000</span>,
  center <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">33</span>, <span class="fl">30</span><span class="op">)</span>, radius <span class="op">=</span> <span class="fl">400e3</span>
<span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="../reference/move.html">move</a></span><span class="op">(</span>
    trajectory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">40</span>, <span class="fl">30</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">30</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">60</span>, <span class="fl">40</span><span class="op">)</span><span class="op">)</span>,
    start <span class="op">=</span> <span class="fl">50000</span>, end <span class="op">=</span> <span class="fl">40000</span>, snapshots <span class="op">=</span> <span class="fl">20</span>
  <span class="op">)</span>

<span class="co"># Eastern hunter-gatherers</span>
<span class="va">ehg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span>
  <span class="st">"EHG"</span>, parent <span class="op">=</span> <span class="va">ooa</span>, time <span class="op">=</span> <span class="fl">28000</span>, N <span class="op">=</span> <span class="fl">1000</span>, remove <span class="op">=</span> <span class="fl">6000</span>,
  polygon <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">26</span>, <span class="fl">55</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">38</span>, <span class="fl">53</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">48</span>, <span class="fl">53</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">60</span>, <span class="fl">53</span><span class="op">)</span>,
    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">60</span>, <span class="fl">60</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">48</span>, <span class="fl">63</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">38</span>, <span class="fl">63</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">26</span>, <span class="fl">60</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># European population</span>
<span class="va">eur</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"EUR"</span>, parent <span class="op">=</span> <span class="va">ehg</span>, time <span class="op">=</span> <span class="fl">25000</span>, N <span class="op">=</span> <span class="fl">2000</span>, polygon <span class="op">=</span> <span class="va">europe</span><span class="op">)</span>

<span class="co"># Anatolian farmers</span>
<span class="va">ana</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span>
  name <span class="op">=</span> <span class="st">"ANA"</span>, time <span class="op">=</span> <span class="fl">28000</span>, N <span class="op">=</span> <span class="fl">3000</span>, parent <span class="op">=</span> <span class="va">ooa</span>, remove <span class="op">=</span> <span class="fl">4000</span>,
  center <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">34</span>, <span class="fl">38</span><span class="op">)</span>, radius <span class="op">=</span> <span class="fl">500e3</span>, polygon <span class="op">=</span> <span class="va">anatolia</span>
<span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="../reference/expand_range.html">expand_range</a></span><span class="op">(</span>
    by <span class="op">=</span> <span class="fl">2500e3</span>, start <span class="op">=</span> <span class="fl">10000</span>, end <span class="op">=</span> <span class="fl">7000</span>,
    polygon <span class="op">=</span> <span class="fu"><a href="../reference/join.html">join</a></span><span class="op">(</span><span class="va">europe</span>, <span class="va">anatolia</span><span class="op">)</span>, snapshots <span class="op">=</span> <span class="fl">20</span>
  <span class="op">)</span> <span class="co"># expand the range by 2.500 km</span>

<span class="co"># Yamnaya steppe population</span>
<span class="va">yam</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span>
  name <span class="op">=</span> <span class="st">"YAM"</span>, time <span class="op">=</span> <span class="fl">7000</span>, N <span class="op">=</span> <span class="fl">500</span>, parent <span class="op">=</span> <span class="va">ehg</span>, remove <span class="op">=</span> <span class="fl">2500</span>,
  polygon <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">26</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">38</span>, <span class="fl">49</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">48</span>, <span class="fl">50</span><span class="op">)</span>,
                 <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">48</span>, <span class="fl">56</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">38</span>, <span class="fl">59</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">26</span>, <span class="fl">56</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="../reference/move.html">move</a></span><span class="op">(</span>trajectory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span>, start <span class="op">=</span> <span class="fl">5000</span>, end <span class="op">=</span> <span class="fl">3000</span>, snapshots <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>

<span class="co"># geneflow events</span>
<span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
  <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">ana</span>, to <span class="op">=</span> <span class="va">yam</span>, rate <span class="op">=</span> <span class="fl">0.5</span>, start <span class="op">=</span> <span class="fl">6000</span>, end <span class="op">=</span> <span class="fl">5000</span>, overlap <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,
  <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">ana</span>, to <span class="op">=</span> <span class="va">eur</span>, rate <span class="op">=</span> <span class="fl">0.5</span>, start <span class="op">=</span> <span class="fl">8000</span>, end <span class="op">=</span> <span class="fl">6000</span><span class="op">)</span>,
  <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">yam</span>, to <span class="op">=</span> <span class="va">eur</span>, rate <span class="op">=</span> <span class="fl">0.75</span>, start <span class="op">=</span> <span class="fl">4000</span>, end <span class="op">=</span> <span class="fl">3000</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># compile the spatial model</span>
<span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span>
  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="va">ooa</span>, <span class="va">ehg</span>, <span class="va">eur</span>, <span class="va">ana</span>, <span class="va">yam</span><span class="op">)</span>,
  gene_flow <span class="op">=</span> <span class="va">gf</span>,
  generation_time <span class="op">=</span> <span class="fl">30</span>, resolution <span class="op">=</span> <span class="fl">10e3</span>,
  competition <span class="op">=</span> <span class="fl">130e3</span>, mating <span class="op">=</span> <span class="fl">100e3</span>, dispersal <span class="op">=</span> <span class="fl">70e3</span>,
  path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"_spatial-ts"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>As a sanity check that we defined the demography correctly, you can
plot a graph summarizing population divergences and geneflow events by
calling <code>plot_model(model)</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_model.html">plot_model</a></span><span class="op">(</span><span class="va">model</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-06-locations_files/figure-html/demographic_model-1.png" width="480"></p>
<p>And for completeness, here is a (slightly busy) overview of the
spatial population ranges that we defined above:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_map.html">plot_map</a></span><span class="op">(</span><span class="va">afr</span>, <span class="va">ooa</span>, <span class="va">ehg</span>, <span class="va">eur</span>, <span class="va">ana</span>, <span class="va">yam</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-06-locations_files/figure-html/spatial_maps-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="scheduling-sampling-events-and-simulation">Scheduling sampling events and simulation<a class="anchor" aria-label="anchor" href="#scheduling-sampling-events-and-simulation"></a>
</h2>
<p>Now we will schedule the sampling of a single individual from each
population every two thousand years, starting from 40 thousand years ago
all the way to the present (this is a feature discussed in the basic <a href="vignette-05-tree-sequences.html#scheduling-of-sampling-events-1">tree
sequence overview</a>):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># one ancient individual every two thousand years</span>
<span class="va">ancient</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>,
                    times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">40000</span>, <span class="fl">1</span>, by <span class="op">=</span> <span class="op">-</span><span class="fl">1000</span><span class="op">)</span>,
                    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ooa</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ehg</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">eur</span>, <span class="fl">1</span><span class="op">)</span>,
                    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ana</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">yam</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>

<span class="co"># present-day Africans and Europeans</span>
<span class="va">present</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">eur</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span>

<span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">ancient</span>, <span class="va">present</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span>
  <span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">100e3</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, burnin <span class="op">=</span> <span class="fl">200e3</span>,
  sampling <span class="op">=</span> <span class="va">samples</span>, method <span class="op">=</span> <span class="st">"batch"</span>, random_seed <span class="op">=</span> <span class="fl">314159</span>, max_attempts <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
<p>After the simulation is done, we can load the data in the same way we
did in our <a href="vignette-05-tree-sequences.html#loading-and-processing-tree-sequence-output-files-1">first
exploration</a> of tree sequence features in <em>slendr</em>. We will
also immediately simplify the tree sequence data structure to only those
genealogies which involve individuals that we explicitly scheduled for
sampling, and perform recapitation to ensure the coalescence of all
genealogies along the genome:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="va">model</span>, recapitate <span class="op">=</span> <span class="cn">TRUE</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, Ne <span class="op">=</span> <span class="fl">10000</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extracting-spatial-tree-sequence-infrmation">Extracting spatial tree sequence infrmation<a class="anchor" aria-label="anchor" href="#extracting-spatial-tree-sequence-infrmation"></a>
</h2>
<p>As we showed in the <a href="vignette-05-tree-sequences.html">basic
tutorial</a>, the most important function for data exploration is
<code><a href="../reference/ts_nodes.html">ts_nodes()</a></code>. This function extracts all information about
individuals and nodes recorded in a tree sequence object loaded and
annotated by <em>slendr</em> :</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_nodes.html">ts_nodes</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span></code></pre></div>
<p>For completeness, we have also functions such as
<code>ts_individuals()</code>, <code><a href="../reference/ts_nodes.html">ts_nodes()</a></code> and
<code><a href="../reference/ts_edges.html">ts_edges()</a></code> which extract <a href="https://tskit.dev/tskit/docs/stable/data-model.html#table-definitions" class="external-link">tree
sequence tables</a> in their “raw” unprocessed form, but
<code><a href="../reference/ts_nodes.html">ts_nodes()</a></code> is much more convenient for data exploration and
analyses. First, it combined information in the low-level tables of
individuals and nodes into a single table but more importantly, if the
model which generated this data was a spatial model,
<code><a href="../reference/ts_nodes.html">ts_nodes()</a></code> automatically annotates the node/individual
tables with the position of each node in space (in real projected
coordinates) and time. This means that we can do spatial data analysis
directly on the table returned by <code><a href="../reference/ts_nodes.html">ts_nodes()</a></code>.</p>
<p>Even better, although we can see below that the returned object
belongs to <em>slendr</em>’s own class <code>slendr_ts_nodes</code>, it
is internally stored as a spatial <code>sf</code> object. This means
that we can use all the functionality of the powerful <a href="https://r-spatial.github.io/sf/" class="external-link">R package <em>sf</em></a> as well
as many other packages for geospatial analyses <em>directly on the
data</em>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">#&gt; [1] "slendr"       "slendr_nodes" "sf"           "tbl_df"       "tbl"         </span>
<span class="co">#&gt; [6] "data.frame"</span></code></pre>
<p>Typing the object into the R console presents a user-friendly summary
of the spatio-temporal data extracted from the tree sequence:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span></code></pre></div>
<pre><code><span class="co">#&gt; slendr 'nodes' object </span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; times are expressed in a backward time direction</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; summary of the table data contents:</span>
<span class="co">#&gt;   AFR - 5 'sampled', 5 'remembered', 5 'retained', 5 'alive' individuals</span>
<span class="co">#&gt;   EUR - 54 'sampled', 54 'remembered', 30 'retained', 30 'alive' individuals</span>
<span class="co">#&gt;   OOA - 16 'sampled', 16 'remembered', NA 'retained', 0 'alive' individuals</span>
<span class="co">#&gt;   ANA - 24 'sampled', 24 'remembered', NA 'retained', 0 'alive' individuals</span>
<span class="co">#&gt;   EHG - 22 'sampled', 22 'remembered', NA 'retained', 0 'alive' individuals</span>
<span class="co">#&gt;   YAM - 4 'sampled', 4 'remembered', NA 'retained', 0 'alive' individuals</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; total:</span>
<span class="co">#&gt;   - 125 'sampled' individuals</span>
<span class="co">#&gt;   - 125 'remembered' individuals</span>
<span class="co">#&gt;   - 300 'retained' individuals</span>
<span class="co">#&gt;   - 35 'alive' individuals</span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; oldest sampled individual: 40000 time units 'before present' </span>
<span class="co">#&gt; youngest sampled individual: 0 time units 'before present' </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; oldest node: 2702925 time units 'before present' </span>
<span class="co">#&gt; youngest node: 0 time units 'before present' </span>
<span class="co">#&gt; --------------------- </span>
<span class="co">#&gt; overview of the underlying sf object:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 577 × 12</span></span>
<span class="co">#&gt;    name  pop   node_id  time time_tskit           location sampled</span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;POINT [m]&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> AFR_1 AFR       180     0          0 (2998019 580072.3) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> AFR_1 AFR       181     0          0 (2998019 580072.3) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> AFR_2 AFR       182     0          0  (3214580 1133048) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> AFR_2 AFR       183     0          0  (3214580 1133048) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> AFR_3 AFR       184     0          0 (3458466 600740.2) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> AFR_3 AFR       185     0          0 (3458466 600740.2) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> AFR_4 AFR       186     0          0  (3240591 1109739) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> AFR_4 AFR       187     0          0  (3240591 1109739) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AFR_5 AFR       188     0          0  (3311509 1357603) TRUE   </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> AFR_5 AFR       189     0          0  (3311509 1357603) TRUE   </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 567 more rows, and 5 more variables: remembered &lt;lgl&gt;, retained &lt;lgl&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   alive &lt;lgl&gt;, pedigree_id &lt;dbl&gt;, ind_id &lt;dbl&gt;</span></span></code></pre>
<p>In the first part of the summary, we see how many individuals
(sampled or retained) and nodes are present in the tree sequence
together with additional useful information, including a section of the
internally stored <code>sf</code> object. And this is a crucial
point—**we can always use the internal <code>sf</code> object with the
<a href="https://r-spatial.github.io/sf/articles/sf1.html#how-simple-features-in-r-are-organized-1" class="external-link">spatial
data</a> directly*.</p>
<p>Because the data returned by <code><a href="../reference/ts_nodes.html">ts_nodes()</a></code> is internally
transformed to the projected CRS used by the model, we can use the
returned object as any other data of the class <code>sf</code>. For
instance, at the beginning of this vignette, we specified the world map
of the model to be represented in projected CRS (EPSG 3035) which we can
verify by typing:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">map</span></code></pre></div>
<pre><code><span class="co">#&gt; slendr 'map' object </span>
<span class="co">#&gt; ------------------- </span>
<span class="co">#&gt; map: internal coordinate reference system EPSG 3035 </span>
<span class="co">#&gt; spatial limits (in degrees longitude and latitude):</span>
<span class="co">#&gt;   - vertical -15 ... 60</span>
<span class="co">#&gt;   - horizontal 20 ... 65</span></code></pre>
<p>The fact that the <code><a href="../reference/ts_nodes.html">ts_nodes()</a></code> result is just another
<code>sf</code> object makes it easy to visualize overlay contents on
this map, as we will see below.</p>
</div>
<div class="section level2">
<h2 id="using-the-simple-features-interface">Using the <em>simple features</em> interface<a class="anchor" aria-label="anchor" href="#using-the-simple-features-interface"></a>
</h2>
<p>It’s hard to overstate how powerful the <a href="https://r-spatial.github.io/sf/" class="external-link">R ecosystem around the
<em>sf</em> package</a> is. However, getting familiar with this package
and geospatial analysis in general can be a bit of a hurdle, especially
for novice users because it takes time to get familiar with many new
concepts.</p>
<p>Although many <em>slendr</em> features for encoding and programming
spatial models and handling simulated tree sequence data discussed so
far are designed to abstract away most of the complexities of the
underlying low-level details to let you focus on the problem at hand,
spatial data analysis is unfortunately whole another matter. Luckily,
because the data generated by <em>slendr</em> is no different from any
other source of spatial data out there and you have great <a href="https://keen-swartz-3146c4.netlify.app/intro.html" class="external-link">free
resources</a> at your disposal.</p>
<p>The bottom line is: the spatio-temporal data extracted from tree
sequences by <em>slendr</em> is no different than an any other normal <a href="https://r-spatial.github.io/sf/" class="external-link"><code>sf</code></a> object. Any
resource that you find for manipulating, plotting, and analysing
<code>sf</code> data can be applied to <em>slendr</em> results as
well.</p>
<p>In the remainder of this vignette we will look at a couple of
examples.</p>
</div>
<div class="section level2">
<h2 id="plotting-locations-of-simulated-sampled-individuals">Plotting locations of simulated sampled individuals<a class="anchor" aria-label="anchor" href="#plotting-locations-of-simulated-sampled-individuals"></a>
</h2>
<p>Every spatial object in <em>slendr</em> is internally of the class
<code>sf</code>. The flexibility of of <a href="https://ggplot2.tidyverse.org" class="external-link"><em>ggplot2</em></a> and <a href="https://r-spatial.github.io/sf/" class="external-link"><em>sf</em></a> packages means
that we can overlay the locations of sampled individuals (saved in a
<code>sf</code> format by <code><a href="../reference/ts_nodes.html">ts_nodes()</a></code>) on top of our world
map (also an <code>sf</code> object):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sampled_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_nodes.html">ts_nodes</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">sampled</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">map</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span>, color <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sampled_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">pop</span>, color <span class="op">=</span> <span class="va">time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Locations of simulated sampled individuals"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_continuous.html" class="external-link">scale_color_continuous</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-06-locations_files/figure-html/slendr_map_ts-1.png" width="480"></p>
<p>Because <code>sf</code> <em>simple features</em> objects (and, by
extension, even <code>slendr_spatial</code> objects) are internally
stored as <a href="https://r-spatial.github.io/sf/articles/sf1.html#how-simple-features-in-r-are-organized" class="external-link">normal
data frames</a> with couple more bells and whistles on top of them, we
have all the <a href="https://www.tidyverse.org" class="external-link">powerful tools</a> for
manipulating tabular data at our disposal.</p>
<p>As an example, let’s say we wanted to split the sampled individuals
in the tree sequence into epochs and plot those individually using
standard <em>ggplot2</em> features. We could simply first do this,
adding a new column specifying to which epoch does each simulated
individual belong:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">epochs</span> <span class="op">&lt;-</span> <span class="va">sampled_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>epoch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">time</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">40000</span>, <span class="fl">30000</span>, <span class="fl">10000</span>, <span class="fl">4000</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span>,
         epoch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">epoch</span><span class="op">)</span>, <span class="fl">0</span>, <span class="va">epoch</span><span class="op">)</span>,
         epoch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">epoch</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"present"</span>, <span class="st">"(present, 4 ky]"</span>, <span class="st">"(4 ky, 10 ky]"</span>,
                                          <span class="st">"(10 ky, 30 y]"</span>, <span class="st">"(30 ky, 40 ky]"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>This chunk of code simply adds a new column <code>epoch</code> to the
<code>sf</code> spatial data frame object called <code>epochs</code>
here.</p>
<p>Then we can use the <em>ggplot2</em> function <code>geom_sf</code> to
plot the locations of sampled individuals on the map, with each facet
corresponding to one epoch (the warning can be safely ignored):</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">map</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span>, color <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">epochs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">pop</span>, color <span class="op">=</span> <span class="va">pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">epoch</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Locations of simulated sampled individuals in different epochs"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-06-locations_files/figure-html/map_epochs-1.png" width="480"></p>
<p>We hope this little excursion to handling <em>slendr</em> spatial
objects (and, by extension, <code>sf</code> objects) with standard data
frame manipulation functions and <em>ggplot2</em> visualisation
convinced you that you have a great flexibility in analysing spatial
<em>slendr</em> data. For best introduction into so-called “tidy” data
analysis, we encourage you to read the freely-available book <a href="https://r4ds.had.co.nz" class="external-link">R for Data Science</a>.</p>
</div>
<div class="section level2">
<h2 id="extracting-spatio-temporal-ancestral-relationships">Extracting spatio-temporal ancestral relationships<a class="anchor" aria-label="anchor" href="#extracting-spatio-temporal-ancestral-relationships"></a>
</h2>
<p>Perhaps even more useful than plotting the locations of simulated
individuals is accessing the locations (and times) of <em>all
ancestors</em> of a particular tree sequence node (a “focal node”).
Starting from the focal node or individual, we can trace the
geographical location of nodes in its lineage going back all the way to
the root with the function <code><a href="../reference/ts_ancestors.html">ts_ancestors()</a></code>.</p>
<p>Because we record the time and the location of every individual that
happens to be the ancestor of at least one sampled individual, this
means that we know the true location of every node of the tree
sequence.</p>
<p>The simplest use case is determining the locations and times of every
single node in the genealogical history of an individual along the tree
sequence (it is possible to recover ancestral relationships for multiple
samples at once too):</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ind</span> <span class="op">&lt;-</span> <span class="st">"EUR_51"</span>

<span class="va">lineages</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_ancestors.html">ts_ancestors</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">ind</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">#&gt; Collecting ancestors of EUR_51 [1/1]...</span></code></pre>
<pre><code><span class="co">#&gt; </span>
<span class="co">#&gt; Generating data about spatial relationships of nodes...</span></code></pre>
<p>The function starts at a given node (or, if a name of a sampled
diploid individual is provided, two nodes), extracts information about
all the parent nodes of that node in the entire tree sequence, records
their locations and times, then proceeds one level “higher” in the
genealogical history to gather information about the parents of those
parent nodes, etc., until it reaches a root node. The result of this
process is another <code>sf</code> object in which each row of the table
encodes information about single branch in the genealogy of the “focal”
node or individual (in our example, <code>"EUR_25"</code>):</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lineages</span></code></pre></div>
<pre><code><span class="co">#&gt; Simple feature collection with 124 features and 12 fields</span>
<span class="co">#&gt; Active geometry column: connection</span>
<span class="co">#&gt; Geometry type: LINESTRING</span>
<span class="co">#&gt; Dimension:     XY</span>
<span class="co">#&gt; Bounding box:  xmin: 3343176 ymin: 368874.9 xmax: 8450665 ymax: 4574569</span>
<span class="co">#&gt; Projected CRS: ETRS89-extended / LAEA Europe</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 124 × 15</span></span>
<span class="co">#&gt;    name  pop   node_id level child_id parent_id child_time parent_time child_pop</span>
<span class="co">#&gt;  <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> EUR_… EUR       242 1          242       263          0         520 EUR      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> EUR_… EUR       242 2          263       267        520         790 EUR      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> EUR_… EUR       242 3          267       315        790        <span style="text-decoration: underline;">8</span>200 EUR      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> EUR_… EUR       242 3          267       320        790        <span style="text-decoration: underline;">8</span>860 EUR      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> EUR_… EUR       242 4          315       323       <span style="text-decoration: underline;">8</span>200        <span style="text-decoration: underline;">9</span>220 ANA      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> EUR_… EUR       242 4          320       341       <span style="text-decoration: underline;">8</span>860       <span style="text-decoration: underline;">13</span>780 ANA      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> EUR_… EUR       242 5          323       351       <span style="text-decoration: underline;">9</span>220       <span style="text-decoration: underline;">15</span>490 ANA      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> EUR_… EUR       242 5          341       349      <span style="text-decoration: underline;">13</span>780       <span style="text-decoration: underline;">15</span>250 ANA      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> EUR_… EUR       242 6          351       375      <span style="text-decoration: underline;">15</span>490       <span style="text-decoration: underline;">19</span>540 ANA      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> EUR_… EUR       242 6          351       413      <span style="text-decoration: underline;">15</span>490       <span style="text-decoration: underline;">23</span>920 ANA      </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 114 more rows, and 6 more variables: parent_pop &lt;fct&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   child_location &lt;POINT [m]&gt;, parent_location &lt;POINT [m]&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   connection &lt;LINESTRING [m]&gt;, left_pos &lt;dbl&gt;, right_pos &lt;dbl&gt;</span></span></code></pre>
<p>In each row of the table, two columns <code>location</code> and
<code>parent_location</code> carry the spatial location of a node
(<code>node_id</code>) and its parent node (<code>parent_id</code>),
respectively, same with the columns <code>time</code> and
<code>parent_time</code> (times of nodes) and <code>pop</code> and
<code>parent_pop</code> (populations to which the nodes belong). The
column <code>connection</code> contains an <code>sf</code> geometry
object of the line connecting the two nodes in the coordinate reference
system of the “model world”. The column <code>focal_id</code> tells us
to which focal node’s genealogy the rows of the table belong to, and the
<code>level</code> column shows how deep in the genealogical past does
each branch (i.e. row of the table) belong to.</p>
<p>This table contains a complete information about spatio-temporal
relationships between the nodes in the genealogy of the given focal
sample. In the spirit of demonstrating of <em>slendr</em> tree sequence
tables interact with the <em>sf</em> and <em>ggplot2</em> environments,
let’s look at the most immediate parent nodes of the two nodes in the
sampled individual (i.e. nodes at the level 1) using the
<code>filter</code> function from the R package <a href="https://r4ds.had.co.nz/transform.html#filter-rows-with-filter" class="external-link">dplyr</a>:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">lineages</span>, <span class="va">level</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">#&gt; Simple feature collection with 2 features and 12 fields</span>
<span class="co">#&gt; Active geometry column: connection</span>
<span class="co">#&gt; Geometry type: LINESTRING</span>
<span class="co">#&gt; Dimension:     XY</span>
<span class="co">#&gt; Bounding box:  xmin: 5071132 ymin: 2745273 xmax: 5563191 ymax: 3114111</span>
<span class="co">#&gt; Projected CRS: ETRS89-extended / LAEA Europe</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 15</span></span>
<span class="co">#&gt;   name   pop   node_id level child_id parent_id child_time parent_time child_pop</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">1</span> EUR_51 EUR       242 1          242       263          0         520 EUR      </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">2</span> EUR_51 EUR       243 1          243       274          0        <span style="text-decoration: underline;">1</span>420 EUR      </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 6 more variables: parent_pop &lt;fct&gt;, child_location &lt;POINT [m]&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   parent_location &lt;POINT [m]&gt;, connection &lt;LINESTRING [m]&gt;, left_pos &lt;dbl&gt;,</span></span>
<span class="co">#&gt; <span style="color: #949494;">#   right_pos &lt;dbl&gt;</span></span></code></pre>
<p>As we mentioned above, there are three columns encoding spatial
information: <code>location</code> and <code>parent_location</code>
carry information about the location of the child and parent node
(<code>POINT</code> class), and the <code>connection</code> object
(<code>LINESTRING</code> class) contains the line connecting the two
nodes (both a branch in the tree sequence and also the spatial
connection). We can plot all three spatial features (two points and a
line) individually on a map:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">level1_branches</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_ancestors.html">ts_ancestors</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"EUR_10"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">level</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">map</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span>, color <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">level1_branches</span><span class="op">[</span>, <span class="op">]</span><span class="op">$</span><span class="va">child_location</span>, shape <span class="op">=</span> <span class="fl">13</span>, size <span class="op">=</span> <span class="fl">3</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">level1_branches</span><span class="op">[</span>, <span class="op">]</span><span class="op">$</span><span class="va">parent_location</span>, shape <span class="op">=</span> <span class="fl">20</span>, color <span class="op">=</span> <span class="va">level1_branches</span><span class="op">[</span>, <span class="op">]</span><span class="op">$</span><span class="va">node_id</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">level1_branches</span><span class="op">[</span>, <span class="op">]</span><span class="op">$</span><span class="va">connection</span>, linetype <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Parent nodes (blue) of a focal individual (red)"</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-06-locations_files/figure-html/level1_branches-1.png" width="480"></p>
<p>In the figure above we can see the red focal node and its immediate
parents in the tree sequence genealogy (in the coalescent sense, not
immediate parents of that individual!).</p>
<p>A more convenient way to do this analysis is a companion function to
<code><a href="../reference/ts_ancestors.html">ts_ancestors()</a></code> called <code>plot_ancestors()</code>. This
function accepts an <code>sf</code> object with the spatial branching
data created by <code><a href="../reference/ts_ancestors.html">ts_ancestors()</a></code> and plots the paths between
nodes on a map leading from a focal node up to the root(s) of the tree
sequence (instead of just paths to immediate parents shown in the
previous figure). In this case, because we are working with a single
diploid individual, we get two sets of paths for each of its nodes
(chromosomes) and plot them in two facets:</p>
<!-- ```{r, eval = FALSE, echo = FALSE} -->
<!-- pdf("~/Desktop/x.pdf") -->
<!-- for (i in ts_samples(ts) %>% filter(pop == "EUR") %>% pull(name)) { -->
<!-- lineages <- ts_ancestors(ts, i, verbose = TRUE) -->
<!-- { plot_ancestors(lineages, i) + ggtitle(i) } %>% print -->
<!-- } -->
<!-- dev.off() -->
<!-- ``` -->
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">map</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">lineages</span>, size <span class="op">=</span> <span class="fl">0.5</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="va">parent_time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_set_geometry</a></span><span class="op">(</span><span class="va">lineages</span>, <span class="st">"parent_location"</span><span class="op">)</span>,
          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">parent_pop</span>, color <span class="op">=</span> <span class="va">parent_pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/ts_nodes.html">ts_nodes</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span>, <span class="va">name</span> <span class="op">==</span> <span class="va">ind</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"longitude"</span>, y <span class="op">=</span> <span class="st">"latitude"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">node_id</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-06-locations_files/figure-html/plot_ancestors_time-1.png" width="480"></p>
<p>You can compare this result to the animation which recapitulates the
simulation, presented in <a href="vignette-01-tutorial.html">the first
vignette</a>.</p>
<p>After comparing our spatial tree sequence figure with the animation,
we can immediately notice several things:</p>
<ol style="list-style-type: decimal">
<li><p>All spatial tree sequence paths trace the ancestry of our single
European individual back to Africa. In fact, we also see a cluster of
past ancestral nodes (i.e. concentrated coalescent events) in the place
where the Out of Africa (OOA) migrant population settled around 40,000
thousand years ago (yellow population in the the animation).</p></li>
<li><p>The first chromosome traces its ancestry to the Yamnaya
population indicated by a star. This is expected because we programmed
Yamnaya to contribute significant part of ancestry to present-day
Europeans (compare to the demographic graph on top of this vignette). In
turn, the Yamnaya node traces its ancestry to EHG (which is its parental
population). We can see a cluster of EHG nodes in one location because
we did not specify any spatial dynamic for this population in our
model</p></li>
<li><p>We also see that the other chromosome traces of its ancestry to
the Anatolia (triangles). This makes sense, because we have simulated
European ancestry as being part Anatolian.</p></li>
</ol>
<p>Let’s look at the spatial ancestry of another sample. For instance,
we know that the simulated history of the Anatolian population in our
model is much simpler. According to the demographic graph above,
Anatolians split from the ancestral population of Eurasians in Anatolia
and then expanded in a wave to Europe. We have sampled the following
individuals:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ts_samples.html">ts_samples</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pop</span> <span class="op">==</span> <span class="st">"ANA"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">#&gt; <span style="color: #949494;"># A tibble: 24 × 3</span></span>
<span class="co">#&gt;    name    time pop  </span>
<span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ANA_1  <span style="text-decoration: underline;">27</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ANA_2  <span style="text-decoration: underline;">26</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ANA_3  <span style="text-decoration: underline;">25</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ANA_4  <span style="text-decoration: underline;">24</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> ANA_5  <span style="text-decoration: underline;">23</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> ANA_6  <span style="text-decoration: underline;">22</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ANA_7  <span style="text-decoration: underline;">21</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> ANA_8  <span style="text-decoration: underline;">20</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> ANA_9  <span style="text-decoration: underline;">19</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span> ANA_10 <span style="text-decoration: underline;">18</span>000 ANA  </span>
<span class="co">#&gt; <span style="color: #949494;"># … with 14 more rows</span></span></code></pre>
<p>Can we see a hint of the spatial dynamics of Anatolians in the
spatio-temporal distribution of ancestral node locations of one of the
sampled individuals? Let’s pick the last individual and immediately plot
its spatial ancestry tidyverse-style using the pipe operator
<code>%&gt;%</code>:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lineages</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_ancestors.html">ts_ancestors</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"ANA_21"</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">map</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">lineages</span>, size <span class="op">=</span> <span class="fl">0.5</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="va">parent_time</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_set_geometry</a></span><span class="op">(</span><span class="va">lineages</span>, <span class="st">"parent_location"</span><span class="op">)</span>,
          <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="va">parent_pop</span>, color <span class="op">=</span> <span class="va">parent_pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="../reference/ts_nodes.html">ts_nodes</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span>, <span class="va">name</span> <span class="op">==</span> <span class="st">"ANA_21"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"longitude"</span>, y <span class="op">=</span> <span class="st">"latitude"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">.</span> <span class="op">~</span> <span class="va">node_id</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-06-locations_files/figure-html/plot_ancestors_levels_ana-1.png" width="480"></p>
<p>As we might expect given the late age of the sample, its position in
the map above (red crossed circle) is not in Anatolia but in Europe
because it represents one of the descendants of migrants who moved from
Anatolia into Europe. This can be clearly seen in the position of its
parental nodes in the tree sequence: these nodes represent real
individuals who lived at some point in the past, and we can see that
they did, indeed, lived in Anatolia.</p>
</div>
<div class="section level2">
<h2 id="calculating-distances-and-other-statistics-using-the-sf-package">Calculating distances and other statistics using the <em>sf</em>
package<a class="anchor" aria-label="anchor" href="#calculating-distances-and-other-statistics-using-the-sf-package"></a>
</h2>
<p>How can we summarise the spatial ancestral dynamics in the figures
using some statistics?</p>
<p>Lets take one more look at the <code>sf</code> object with the
locations and times of ancestral nodes of sampled individuals, focusing
on the following subset of columns:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lineages</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="../reference/ts_samples.html">ts_samples</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="../reference/ts_ancestors.html">ts_ancestors</a></span><span class="op">(</span><span class="va">ts</span>, x <span class="op">=</span> <span class="va">.</span><span class="op">)</span>
  
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">lineages</span>, <span class="va">connection</span>, <span class="va">child_time</span>, <span class="va">parent_time</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">#&gt; Simple feature collection with 6863 features and 2 fields</span>
<span class="co">#&gt; Geometry type: LINESTRING</span>
<span class="co">#&gt; Dimension:     XY</span>
<span class="co">#&gt; Bounding box:  xmin: 2521712 ymin: 329589.1 xmax: 8559748 ymax: 4814330</span>
<span class="co">#&gt; Projected CRS: ETRS89-extended / LAEA Europe</span>
<span class="co">#&gt; <span style="color: #949494;"># A tibble: 6,863 × 3</span></span>
<span class="co">#&gt;                              connection child_time parent_time</span>
<span class="co">#&gt;                        <span style="color: #949494; font-style: italic;">&lt;LINESTRING [m]&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>   (8559748 3193326, 7794902 2294646)      <span style="text-decoration: underline;">40</span>000       <span style="text-decoration: underline;">42</span>640</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>   (8559748 3193326, 7044473 1221506)      <span style="text-decoration: underline;">40</span>000       <span style="text-decoration: underline;">48</span>610</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>   (7794902 2294646, 7044473 1221506)      <span style="text-decoration: underline;">42</span>640       <span style="text-decoration: underline;">48</span>610</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>   (7044473 1221506, 4132744 1097930)      <span style="text-decoration: underline;">48</span>610       <span style="text-decoration: underline;">85</span>690</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>  (7044473 1221506, 3459546 578367.8)      <span style="text-decoration: underline;">48</span>610      <span style="text-decoration: underline;">103</span>420</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>  (7044473 1221506, 3844983 368874.9)      <span style="text-decoration: underline;">48</span>610      <span style="text-decoration: underline;">104</span>440</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>  (7044473 1221506, 3844983 368874.9)      <span style="text-decoration: underline;">48</span>610      <span style="text-decoration: underline;">104</span>440</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>  (4132744 1097930, 3844983 368874.9)      <span style="text-decoration: underline;">85</span>690      <span style="text-decoration: underline;">104</span>440</span>
<span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> (3459546 578367.8, 3844983 368874.9)     <span style="text-decoration: underline;">103</span>420      <span style="text-decoration: underline;">104</span>440</span>
<span class="co">#&gt; <span style="color: #BCBCBC;">10</span>   (8559748 3193326, 7305165 1650688)      <span style="text-decoration: underline;">40</span>000       <span style="text-decoration: underline;">46</span>990</span>
<span class="co">#&gt; <span style="color: #949494;"># … with 6,853 more rows</span></span></code></pre>
<p>We can use standard <em>dplyr</em> table manipulation functions to
compute the distances between connected notes and the times that
separate them (i.e. branch lengths in the traditional phylogenetic
sense). We can then use those two quantities to compute how fast (if any
at all) was the movement between ancestral individuals in different time
periods of the history of a sample:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">distances</span> <span class="op">&lt;-</span> <span class="va">lineages</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>branch_length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">parent_time</span> <span class="op">-</span> <span class="va">child_time</span><span class="op">)</span> <span class="op">/</span> <span class="va">model</span><span class="op">$</span><span class="va">generation_time</span>,
         distance <span class="op">=</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_measures.html" class="external-link">st_length</a></span><span class="op">(</span><span class="va">connection</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">units</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/units/man/units.html" class="external-link">set_units</a></span><span class="op">(</span><span class="va">km</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="op">)</span>,
         speed <span class="op">=</span> <span class="va">distance</span> <span class="op">/</span> <span class="va">branch_length</span>,
         epoch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">parent_time</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">Inf</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">60000</span>, <span class="fl">0</span>, by <span class="op">=</span> <span class="op">-</span><span class="fl">3000</span><span class="op">)</span><span class="op">)</span>, dig.lab <span class="op">=</span> <span class="fl">10</span>, include.lowest <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co"># strip away the spatial annotation</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">pop</span>, <span class="va">node_id</span>, <span class="va">branch_length</span>, <span class="va">distance</span>, <span class="va">speed</span>, <span class="va">parent_pop</span>, <span class="va">parent_time</span>, <span class="va">child_pop</span>, <span class="va">child_time</span>, <span class="va">epoch</span><span class="op">)</span></code></pre></div>
<p>Let’s also convert the data (absolute distances or distance per
generation – i.e., the “speed”) into a long format for easier plotting
side by side:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">distances_long</span> <span class="op">&lt;-</span> <span class="va">distances</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">child_time</span> <span class="op">&lt;</span> <span class="fl">60000</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">pop</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR"</span>, <span class="st">"OOA"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">distance</span>, <span class="va">speed</span><span class="op">)</span>,
                      names_to <span class="op">=</span> <span class="st">"stat"</span>,
                      values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>facet <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span>
    <span class="va">stat</span> <span class="op">==</span> <span class="st">"distance"</span> <span class="op">~</span> <span class="st">"absolute distance of a node from parent"</span>,
    <span class="va">stat</span> <span class="op">==</span> <span class="st">"speed"</span> <span class="op">~</span> <span class="st">"distance traveled by a node per generation"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Let’s try to summarise the information about distances “traveled” by
nodes in different time period by fitting a spline (rather than plotting
the raw data with individual nodes):</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">distances_long</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">child_time</span>, <span class="va">value</span>, color <span class="op">=</span> <span class="va">child_pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"loess"</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>group <span class="op">=</span> <span class="va">child_pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"kilometers"</span>, x <span class="op">=</span> <span class="st">"time [years ago]"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">1</span>, angle <span class="op">=</span> <span class="fl">45</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">facet</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span><span class="st">"ancestral node population"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">#&gt; `geom_smooth()` using formula 'y ~ x'</span></code></pre>
<p><img src="vignette-06-locations_files/figure-html/smooth_distance_fits-1.png" width="480"></p>
<!-- # Directional movement -->
<!-- Rather then looking at absolute distances traveled (or the speed of movement), we might be interested in knowing something about the *direction* of movement. -->
<!-- ```{r} -->
<!-- directions <- lineages %>% -->
<!--   mutate(direction = sf::st_coordinates(child_location - parent_location) / 1e3 / model$generation_time, -->
<!--          branch_length = abs(child_time - parent_time), -->
<!--          horizontal = as.vector(direction[, "X"]), -->
<!--          vertical = as.vector(direction[, "Y"])) %>% -->
<!--   mutate(epoch = cut(child_time, breaks = c(0, 3000, 4000, 5000, 20000, 25000, 30000, 40000, 50000, Inf), include.lowest = TRUE, dig.lab = 10)) %>% -->
<!-- #  mutate(epoch = cut(child_time, breaks = c(Inf, seq(60000, 0, by = -5000)), include.lowest = TRUE, dig.lab = 10)) %>%  -->
<!--   as_tibble() %>% -->
<!--   distinct(child_id, parent_id, .keep_all = TRUE) %>% -->
<!--   select(name, pop, node_id, horizontal, vertical, branch_length, -->
<!--          parent_pop, parent_time, child_pop, child_time, epoch, horizontal) -->
<!-- ``` -->
<!-- ```{r plot_directions, fig.width = 9, fig.height = 6} -->
<!-- # compute axes limits -->
<!-- limits <- max(abs(c(directions$horizontal, directions$vertical))) %>% c(-., .) -->
<!-- directions %>% -->
<!--   ggplot(aes(horizontal, vertical, color = child_pop)) + -->
<!--   geom_point() + -->
<!--   geom_hline(yintercept = 0, linetype = 2, alpha = 0.5) + -->
<!--   geom_vline(xintercept = 0, linetype = 2, alpha = 0.5) + -->
<!--   facet_grid(pop ~ epoch) + -->
<!--   labs(x = "West-East movement [kilometers / generation]", -->
<!--        y = "North-South movement [kilometers / generation]") + -->
<!-- #  coord_cartesian(xlim = limits, ylim = limits) + -->
<!--   theme(axis.text.x = element_text(hjust = 1, angle = 45)) -->
<!-- ``` -->
<!-- ## Densities -->
<!-- A couple of ideas: -->
<!-- - Heatmaps of ancestry locations (similar to the discrete node location maps) -->
<!-- - Effective population size heatmaps? -->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a895c93f0e9eda05eb10b29a9921d586',
    indexName: 'slendr',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
