<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tree-sequences, tskit interface, f-statistics • slendr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Tree-sequences, tskit interface, f-statistics">
<meta property="og:description" content="slendr">
<meta property="og:image" content="https://bodkan.net/slendr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slendr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    </li>
    <li>
      <a href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    </li>
    <li>
      <a href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    </li>
    <li>
      <a href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    </li>
    <li>
      <a href="../articles/vignette-05-tree-sequences.html">Tree-sequences, tskit interface, f-statistics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/slendr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/fleventy5">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="vignette-05-tree-sequences_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tree-sequences, tskit interface, f-statistics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/master/vignettes/vignette-05-tree-sequences.Rmd"><code>vignettes/vignette-05-tree-sequences.Rmd</code></a></small>
      <div class="hidden name"><code>vignette-05-tree-sequences.Rmd</code></div>

    </div>

    
    
<p>In this vignette, we will show how to specify sampling events to record individuals in the tree-sequence output file (a procedure which is called “remembering” of individuals in the SLiM context) and how to perform simple analyses using <em>slendr</em>’s interface to the <a href="http://tskit.dev"><em>tskit</em></a> Python library. We will demonstrate these features on a simulation of Neanderthal introgression into anatomically modern humans. Specifically, we will show how to estimate the amount of Neanderthal ancestry using <span class="math inline">\(f\)</span>-statistics calculated directly on the tree-sequence data structure generated by a <em>slendr</em> model, all entirely from R.</p>
<div id="model-of-neanderthal-introgression-into-eurasians" class="section level3">
<h3 class="hasAnchor">
<a href="#model-of-neanderthal-introgression-into-eurasians" class="anchor"></a>Model of Neanderthal introgression into Eurasians</h3>
<p>First, let’s set up a simple <em>non-spatial</em> model of Neanderthal introgression using <em>slendr.</em> This is essentially the same procedure which we have shown in another vignette introducing <a href="vignette-04-nonspatial-models.html">non-spatial <em>slendr</em> models</a>. This is no different from a spatial model, except that we left out the <code>map</code> argument in calling <code><a href="../reference/population.html">population()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr">slendr</a></span><span class="op">)</span>

<span class="co"># create the ancestor of everyone, and a chimpanzee outgroup</span>
<span class="co"># (both N = 1 to reduce the computational time for this model, as this</span>
<span class="co"># doesn't affect the results for this demonstration)</span>
<span class="va">anc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"ANC"</span>, time <span class="op">=</span> <span class="fl">6.5e6</span>, N <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">chimp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"CH"</span>, parent <span class="op">=</span> <span class="va">anc</span>, time <span class="op">=</span> <span class="fl">6e6</span>, N <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>

<span class="co"># two populations of anatomically modern humans: Africans and Europeans</span>
<span class="va">afr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"AFR"</span>, parent <span class="op">=</span> <span class="va">anc</span>, time <span class="op">=</span> <span class="fl">600e3</span>, N <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span>
<span class="va">eur</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"EUR"</span>, parent <span class="op">=</span> <span class="va">afr</span>, time <span class="op">=</span> <span class="fl">70e3</span>, N <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span>

<span class="co"># Neanderthal population splitting at 600 ky ago from the common ancestor with</span>
<span class="co"># anatomically modern humans (and extinct by 40 ky ago)</span>
<span class="va">nea</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"NEA"</span>, parent <span class="op">=</span> <span class="va">anc</span>, time <span class="op">=</span> <span class="fl">600e3</span>, N <span class="op">=</span> <span class="fl">1000</span>, remove <span class="op">=</span> <span class="fl">40e3</span><span class="op">)</span>

<span class="co"># 3% Neanderthal introgression into Europeans between 55-50 ky ago</span>
<span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneflow.html">geneflow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">nea</span>, to <span class="op">=</span> <span class="va">eur</span>, rate <span class="op">=</span> <span class="fl">0.03</span>, start <span class="op">=</span> <span class="fl">55000</span>, end <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span>

<span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile.html">compile</a></span><span class="op">(</span>
  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">anc</span>, <span class="va">chimp</span>, <span class="va">nea</span>, <span class="va">afr</span>, <span class="va">eur</span><span class="op">)</span>, geneflow <span class="op">=</span> <span class="va">gf</span>,
  generation_time <span class="op">=</span> <span class="fl">30</span>,
  dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"introgression"</span><span class="op">)</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
<p>Here’s our toy model visualized as a graph. Not particularly illuminating in this simple example, but it’s always worth keeping in mind that such graph is embedded within every <em>slendr</em> model and can be always invoked to make sure the model you’re setting up is correct:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/graph.html">graph</a></span><span class="op">(</span><span class="va">model</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-05-tree-sequences_files/figure-html/introgression_graph-1.png" width="480"></p>
</div>
<div id="scheduling-of-sampling-events" class="section level2">
<h2 class="hasAnchor">
<a href="#scheduling-of-sampling-events" class="anchor"></a>Scheduling of sampling events</h2>
<p>Now that we have defined a model, how do we sample data from it? Ideally, we would like to to schedule sampling events at a given time, sampling a defined number of individuals from a given population. This is why <em>slendr</em> provides a function <code><a href="../reference/sampling.html">sampling()</a></code> which allows defining such sampling schedule automatically while, at the same time, enforcing that only populations which are already (i.e. after their appearance in the simulation) or still (before they are removed from the simulation) will be sampled from.</p>
<p>In our example, we want to sample two Neanderthal individuals (the older one being the <a href="">Altai Neanderthal</a>, then younger one <a href="">Vindija Neanderthal</a>). This is what is required to estimate Neanderthal ancestyr proportion using a so-called <span class="math inline">\(f_4\)</span>-ratio statistics (more on that below):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nea_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampling.html">sampling</a></span><span class="op">(</span>times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">70000</span>, <span class="fl">40000</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">nea</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="va">nea_samples</span>
<span class="co">#&gt; # A tibble: 2 x 3</span>
<span class="co">#&gt;    time pop       n</span>
<span class="co">#&gt;   &lt;int&gt; &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1 70000 NEA       1</span>
<span class="co">#&gt; 2 40000 NEA       1</span></code></pre></div>
<p>The <code><a href="../reference/sampling.html">sampling()</a></code> function simply accepts the vector of times at which remembering should be schedule, and then a list of pairs <code>(&lt;slendr population&gt;, &lt;number of individuals&gt;)</code> encoding from which populations should how many individuals be remembered at times given in the <code>times</code> vector. Easy.</p>
<p>Next, we want to sample some present-day individuals: an outgroup representing a chimpanzee, and a couple of Africans and Europeans:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">present_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampling.html">sampling</a></span><span class="op">(</span>times <span class="op">=</span> <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">chimp</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">eur</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>
<span class="va">present_samples</span>
<span class="co">#&gt; # A tibble: 3 x 3</span>
<span class="co">#&gt;    time pop       n</span>
<span class="co">#&gt;   &lt;int&gt; &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1     0 CH        1</span>
<span class="co">#&gt; 2     0 AFR       5</span>
<span class="co">#&gt; 3     0 EUR       5</span></code></pre></div>
<p>As you can see above, the <code><a href="../reference/sampling.html">sampling()</a></code> function returns a plain old data frame with a very simple structure with three columns: time, population name, and the number of individuals. This means that you can define sampling events using whatever input data you might already have available (such as radiocarbon-dated ancient DNA samples from an Excel sheet from some publication). For instance, there has been a <a href="https://www.nature.com/articles/nature17993">lot</a> of <a href="https://www.pnas.org/content/116/5/1639">interest</a> to estimate the trajectory of Neanderthal ancestry in Europe over time using ancient DNA data from anatomically modern human individuals (also called early modern humans, EMH) across the last couple of tens of thousands of years. We can simulate something close to the available <a href="https://www.nature.com/articles/nature17993">EMH ancient DNA data set</a> over the last 50 thousand years by running something like this:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">emh_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampling.html">sampling</a></span><span class="op">(</span>times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">50</span>, min <span class="op">=</span> <span class="fl">10000</span>, max <span class="op">=</span> <span class="fl">40000</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">eur</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="va">emh_samples</span>
<span class="co">#&gt; # A tibble: 50 x 3</span>
<span class="co">#&gt;     time pop       n</span>
<span class="co">#&gt;    &lt;int&gt; &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt;  1 16361 EUR       1</span>
<span class="co">#&gt;  2 29322 EUR       1</span>
<span class="co">#&gt;  3 16979 EUR       1</span>
<span class="co">#&gt;  4 22342 EUR       1</span>
<span class="co">#&gt;  5 37738 EUR       1</span>
<span class="co">#&gt;  6 25728 EUR       1</span>
<span class="co">#&gt;  7 28594 EUR       1</span>
<span class="co">#&gt;  8 18357 EUR       1</span>
<span class="co">#&gt;  9 20367 EUR       1</span>
<span class="co">#&gt; 10 24655 EUR       1</span>
<span class="co">#&gt; # … with 40 more rows</span></code></pre></div>
<p>This samples a single ancient European individuals at randomly chosen times between 40 and 10 ky ago.</p>
<p>One nice feature of the <code><a href="../reference/sampling.html">sampling()</a></code> function is that it only schedules sampling events for populations if that population is present in the simulation at a given time. This makes it possible to simply specify the whole time range for sampling, specify all populations and sizes of the samples, and let the function generate sampling events only for populations present at each time. If for some reason a stricter control over sampling is required, this behavior can be switched off by setting <code>strict = TRUE</code> like this:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this attempts to sample a Neanderthal individual at a point when Neanderthals</span>
<span class="co"># are already extinct, resulting in an error</span>
<span class="fu"><a href="../reference/sampling.html">sampling</a></span><span class="op">(</span>times <span class="op">=</span> <span class="fl">10000</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">nea</span>, <span class="fl">1</span><span class="op">)</span>, strict <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>Error: Cannot schedule sampling for 'NEA' at time 10000 because the population will not be present in the simulation at that point. Consider running this function with `strict = FALSE` which will automatically retain only valid sampling events.</code></pre>
<p>Now that we already have the <code>model</code> object ready, we can simulate data from it, sampling individuals according to our sampling schedule. We do this by calling the <code><a href="../reference/slim.html">slim()</a></code> function as usual, but this time we set <code>ts_recording = TRUE</code> (switching of tree-sequence recording in SLiM) and we specify the sampling events with the <code>sampling =</code> argument. Note that we bind the individual sampling schedule data frames using the <code>rbind</code> function provided by base R (as we show above, the sampling schedule really is just a data frame and we can manipulate it as such):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span>
  <span class="va">model</span>, seq_length <span class="op">=</span> <span class="fl">10e6</span>, recomb_rate <span class="op">=</span> <span class="fl">1e-8</span>,
  ts_recording <span class="op">=</span> <span class="cn">TRUE</span>,
  sampling <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">nea_samples</span>, <span class="va">present_samples</span>, <span class="va">emh_samples</span><span class="op">)</span>,
  method <span class="op">=</span> <span class="st">"batch"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,
  seed <span class="op">=</span> <span class="fl">314159</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="r-interface-for-tskit-pyslim-and-msprime" class="section level2">
<h2 class="hasAnchor">
<a href="#r-interface-for-tskit-pyslim-and-msprime" class="anchor"></a>R interface for <em>tskit</em>, <em>pyslim</em> and <em>msprime</em>
</h2>
<p>The result of switching on <code>ts_recording = TRUE</code> in the <code><a href="../reference/slim.html">slim()</a></code> call above is that SLiM will save the output of the simulation as a tree sequence file. By default, the file is stored in the model directory:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">model</span><span class="op">$</span><span class="va">path</span>, <span class="st">"output_tree_seq.trees"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">ts_file</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<p>Tree-sequences are one of the most revolutionary developments in population genetics in the last couple of decades for a number of reasons. One of them is the possibility to store extremely large data sets succinctly by encoding the entire evolutionary history of a sample of individuals as a series of correlated tree genealogies along the genome.</p>
<p>Going into too much detail on this topic is clearly beyond the scope of this tutorial, especially because everything is explain much better <a href="https://tskit.dev/learn.html">elsewhere</a>. Instead, what we will demonstrate in the rest of this vignette is how you can access and manipulate tree-sequence outputs generated by <em>slendr</em> models and perform various statistics on them using Python modules <a href="https://tskit.dev/tskit/docs/stable/">tskit</a> and <a href="https://pyslim.readthedocs.io/en/latest/index.html">pyslim</a> (used to access and manipulate tree-sequence files produced by SLiM) from <em>slendr,</em> without having to leave R! The key is a magical R package <a href="https://rstudio.github.io/reticulate/index.html">reticulate</a> which creates seamless binding Python modules with R. This means that even if you don’t know Python, <em>slendr</em> allows you to do do quite a lot with tree-sequences in R.</p>
<p>Of course, if you are a proficient Python user, it needs to be said that once you have the tree-sequence file generated by <em>slendr</em> &amp; SLiM, you can easily perform every conceivable analysis directly using <em>tskit</em>. The intention here is to show how you can continue working on the tree-sequence files in R even after you have run the entire <em>slendr</em> simulation.</p>
<p>First, in order to be able to interface with <em>tskit</em> and <em>pyslim</em> using the <em>reticulate</em> package, you will need a working Python environment with the required Python modules already installed. I personally manage Python installations and separate package environments using <a href="https://github.com/pyenv/pyenv">pyenv</a> and <a href="https://github.com/pyenv/pyenv-virtualenv">pyenv-virtualenv</a>. However, this is only my own personal preference. If you prefer using conda or any <a href="https://xkcd.com/1987/">other Python management solution</a>, feel free to use that. In any case, I recommend using tooling which is supported by <a href="https://rstudio.github.io/reticulate/index.html">reticulate</a> itself, because it will be easy to find support should you run into any trouble.</p>
<p>For completeness, here is how I have <u>just</u> installed <em>tskit</em> and <em>pyslim</em> into a brand new environment on my Mac using pyenv:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. install the most recent version of Python (enforcing the compilation of</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Python as a shared library is important for the reticulate package to work,</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># at least on a Mac)</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">env</span> PYTHON_CONFIGURE_OPTS=<span class="st">"--enable-shared"</span> pyenv install 3.9.6</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># create a separate Python environment dedicated to our R/Python interface</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># and call it 'retipy'</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="ex">pyenv</span> virtualenv 3.9.6 retipy</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># install required Python modules into the environment</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="ex">pyenv</span> activate retipy</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install tskit pyslim</span></code></pre></div>
<p>Now we can instruct the <em>reticulate</em> package to use the newly created Python environment:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_virtualenv</a></span><span class="op">(</span><span class="st">"~/.pyenv/versions/retipy"</span>, required <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>We can make sure that <em>reticulate</em> is indeed using our environment by calling:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/py_config.html">py_config</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; python:         /Users/martin_petr/.pyenv/versions/3.9.6/envs/retipy/bin/python</span>
<span class="co">#&gt; libpython:      /Users/martin_petr/.pyenv/versions/3.9.6/lib/libpython3.9.dylib</span>
<span class="co">#&gt; pythonhome:     /Users/martin_petr/.pyenv/versions/3.9.6/envs/retipy:/Users/martin_petr/.pyenv/versions/3.9.6/envs/retipy</span>
<span class="co">#&gt; version:        3.9.6 (default, Jul 27 2021, 16:57:27)  [Clang 12.0.0 (clang-1200.0.32.29)]</span>
<span class="co">#&gt; numpy:          /Users/martin_petr/.pyenv/versions/3.9.6/envs/retipy/lib/python3.9/site-packages/numpy</span>
<span class="co">#&gt; numpy_version:  1.21.1</span>
<span class="co">#&gt; tskit:          /Users/martin_petr/.pyenv/versions/3.9.6/envs/retipy/lib/python3.9/site-packages/tskit</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; NOTE: Python version was forced by use_python function</span></code></pre></div>
</div>
<div id="loading-and-processing-tree-sequence-output-files" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-and-processing-tree-sequence-output-files" class="anchor"></a>Loading and processing tree-sequence output files</h2>
<p>With the technicalities out of the way, we can now load the tree-sequence file saved by SLiM using the <em>slendr</em> function <code><a href="../reference/ts_load.html">ts_load()</a></code>. Optionally, we can also instruct this function to simplify the tree-sequence to only the individuals that we explicitly sampled (using the sampling schedule set up by <code><a href="../reference/sampling.html">sampling()</a></code> above). What <em>slendr</em> does here is provide an R-friendly interface to <em>tskit</em> and <em>pyslim</em>, essentially reproducing steps in <a href="https://pyslim.readthedocs.io/en/stable/tutorial.html">this tutorial</a> in a single step. Note that we have to provide the <code>model</code> object generated by <code><a href="../reference/compile.html">compile()</a></code> above:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="va">ts_file</span>, <span class="va">model</span><span class="op">)</span>
<span class="va">ts</span>
<span class="co">#&gt; ╔══════════════════════════╗</span>
<span class="co">#&gt; ║TreeSequence              ║</span>
<span class="co">#&gt; ╠═══════════════╤══════════╣</span>
<span class="co">#&gt; ║Trees          │    101652║</span>
<span class="co">#&gt; ╟───────────────┼──────────╢</span>
<span class="co">#&gt; ║Sequence Length│10000000.0║</span>
<span class="co">#&gt; ╟───────────────┼──────────╢</span>
<span class="co">#&gt; ║Sample Nodes   │     30108║</span>
<span class="co">#&gt; ╟───────────────┼──────────╢</span>
<span class="co">#&gt; ║Total Size     │   2.1 GiB║</span>
<span class="co">#&gt; ╚═══════════════╧══════════╝</span>
<span class="co">#&gt; ╔═══════════╤════════╤═════════╤════════════╗</span>
<span class="co">#&gt; ║Table      │Rows    │Size     │Has Metadata║</span>
<span class="co">#&gt; ╠═══════════╪════════╪═════════╪════════════╣</span>
<span class="co">#&gt; ║Edges      │49989051│  1.3 GiB│          No║</span>
<span class="co">#&gt; ╟───────────┼────────┼─────────┼────────────╢</span>
<span class="co">#&gt; ║Individuals│   15054│942.3 KiB│         Yes║</span>
<span class="co">#&gt; ╟───────────┼────────┼─────────┼────────────╢</span>
<span class="co">#&gt; ║Migrations │       0│  4 Bytes│          No║</span>
<span class="co">#&gt; ╟───────────┼────────┼─────────┼────────────╢</span>
<span class="co">#&gt; ║Mutations  │       0│  1.2 KiB│          No║</span>
<span class="co">#&gt; ╟───────────┼────────┼─────────┼────────────╢</span>
<span class="co">#&gt; ║Nodes      │12687318│411.4 MiB│         Yes║</span>
<span class="co">#&gt; ╟───────────┼────────┼─────────┼────────────╢</span>
<span class="co">#&gt; ║Populations│       5│  3.1 KiB│         Yes║</span>
<span class="co">#&gt; ╟───────────┼────────┼─────────┼────────────╢</span>
<span class="co">#&gt; ║Provenances│       1│ 28.5 KiB│          No║</span>
<span class="co">#&gt; ╟───────────┼────────┼─────────┼────────────╢</span>
<span class="co">#&gt; ║Sites      │       0│  8 Bytes│          No║</span>
<span class="co">#&gt; ╚═══════════╧════════╧═════════╧════════════╝</span></code></pre></div>
<p>We can see that the tree-sequence data is very large and contains many more individuals than we need (i.e. that we explicitly specified for sampling): there are in total 15054 individuals in the tree-sequence which is a lot more than the 63 we scheduled for sampling! To avoid this issue, <code><a href="../reference/ts_load.html">ts_load()</a></code> accepts an optional argument <code>simplify</code> which triggers the <a href="https://pyslim.readthedocs.io/en/latest/tutorial.html#simplification">simplification process</a> to narrow down the set of recorded individuals just to those that we explicitly remembered:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="va">ts_file</span>, <span class="va">model</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Similarly, <em>slendr</em> provides a function <code><a href="../reference/ts_recapitate.html">ts_recapitate()</a></code> which performs <a href="https://pyslim.readthedocs.io/en/latest/tutorial.html#recapitation">recapitation</a>. Again, we can do this in one go, by specifying <code>recapitate = TRUE</code> in the call to <code><a href="../reference/ts_load.html">ts_load()</a></code> after specifying a couple of additional arguments required for recapitation (see the pyslim documentation in the <a href="https://pyslim.readthedocs.io/en/latest/tutorial.html#recapitation">recapitation</a> section for more detail):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="va">ts_file</span>, <span class="va">model</span>, recapitate <span class="op">=</span> <span class="cn">TRUE</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span>,
              recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, Ne <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></code></pre></div>
<p>In this case we get a message informing that the recapitation was not necessary because all genealogies in the tree-sequence are already coalesced (i.e. have each only one root node). We can do this check ourselves by another <em>slendr</em> function <code><a href="../reference/ts_coalesced.html">ts_coalesced()</a></code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ts_coalesced.html">ts_coalesced</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
<p>For computational efficiency, we did not simulate any mutations during the SLiM run. In order to be able to calculate statistics from genetic variation, we can add mutations to the tree-sequence at a given rate by running:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span><span class="va">ts</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span></code></pre></div>
<p>Having done that, we can calculate some basic statistics on our simulated data. We note that everything that we do here (every function with the prefix <code>ts_*()</code> in <em>slendr</em>) is, in fact, interfacing with the <em>tskit</em> Python module. Our goal is to capture the most basic of analyses one might want to perform on tree-sequences in R and wrap them in a neat interface indistinguishable from any other R function—this is, after all, the reason why <em>reticulate</em> has been created in the first place (making various Python data science modules appear as if they were regular R packages).</p>
<p>That said, <u><em>slendr</em> doesn’t try to capture everything that is possible to do in <em>tskit</em></u>. I personally love R and prefer to use it for as much of my work as possible, but for anything that does not involve calculating population genetics statistics on simulated tree genealogies it is better to use <em>tskit</em> directly. In other words, if you’re developing a novel population genetic inference method on which operates on tree-sequence data, you should probably use <em>tskit</em> directly (either its C API or the Python interface), not R. This is why we haven’t implemented <em>tskit</em>’s functionality for iterating over trees in the tree-sequence, traversing nodes, etc. It’s likely that the number of <em>tskit</em> functions implemented in <em>slendr</em> will slowly grow (especially those related to population genetics statistics) but this is where we currently draw the line: <em>slendr</em> is an R package for simulating data (tree-sequences or other) in SLiM and for calculating statistics on it, nothing less, nothing more.</p>
</div>
<div id="calculating-f-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#calculating-f-statistics" class="anchor"></a>Calculating <em>f</em>-statistics</h2>
<p>In addition to being a revolutionary breakthrough in terms of computation efficiency, many population genetics statistics we are often interested in are a natural consequence of having a direct access to tree sequence genealogies simply because those capture the true demographic history of a sample. Again, we can’t go into too much detail here but I encourage you to take a look at a paper by <a href="https://www.genetics.org/content/215/3/779">Ralph <em>et al.</em></a> on the duality between statistics expressed in terms of branch lengths and the traditional summaries based on samples of genetic variation.</p>
<p>For instance, we have functions such as <code><a href="../reference/ts_f2.html">ts_f2()</a></code>, <code><a href="../reference/ts_f2.html">ts_f3()</a></code>, <code><a href="../reference/ts_f2.html">ts_f4()</a></code> and <code><a href="../reference/ts_f2.html">ts_f4ratio()</a></code> which calculate the well-known set of Patterson’s <span class="math inline">\(f\)</span>-statistics:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ts_f2.html">ts_f2</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"EUR1"</span>, <span class="st">"EUR2"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   A     B            f2</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt; 1 EUR1  EUR2  0.0000107</span>

<span class="fu"><a href="../reference/ts_f2.html">ts_f3</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"EUR1"</span>, <span class="st">"AFR1"</span>, <span class="st">"CH1"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 4</span>
<span class="co">#&gt;   A     B     C             f3</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1 EUR1  AFR1  CH1   0.00000275</span>

<span class="fu"><a href="../reference/ts_f2.html">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"AFR1"</span>, <span class="st">"AFR2"</span>, <span class="st">"NEA1"</span>, <span class="st">"CH1"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 5</span>
<span class="co">#&gt;   W     X     Y     Z        f4</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 AFR1  AFR2  NEA1  CH1       0</span>
<span class="fu"><a href="../reference/ts_f2.html">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"AFR1"</span>, <span class="st">"EUR1"</span>, <span class="st">"NEA1"</span>, <span class="st">"CH1"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 1 x 5</span>
<span class="co">#&gt;   W     X     Y     Z             f4</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1 AFR1  EUR1  NEA1  CH1   -0.0000110</span></code></pre></div>
<p>These functions accept a <code>mode =</code> argument, specifying whether the statistics should be calculated using mutation site patterns (<code>mode = "site"</code>, the default) or branch lengths (<code>mode = "branch"</code>). See the <a href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.write_vcf">relevant section</a> of the official <em>tskit</em> documentation for more on this topic.</p>
</div>
<div id="estimating-neanderthal-ancestry-proportions" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-neanderthal-ancestry-proportions" class="anchor"></a>Estimating Neanderthal ancestry proportions</h2>
<p>Let’s try to put these new tools to practice and estimate the proportion of Neanderthal ancestry in Africans and Europeans in our simulated data. We can do this using the Patterson’s <span class="math inline">\(f_4\)</span>-ratio statistic implemented in the <code><a href="../reference/ts_f2.html">ts_f4ratio()</a></code> function in <em>slendr</em> (you can find more information about this particular version of the statistic <a href="https://www.pnas.org/content/116/5/1639">in this paper</a>):</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get a table of simulated African and European individuals in the tree-sequence</span>
<span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_individuals.html">ts_individuals</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">pop</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AFR"</span>, <span class="st">"EUR"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># estimate the amounts of Neanderthal ancestry in these individuals</span>
<span class="va">inds</span><span class="op">$</span><span class="va">ancestry</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_f2.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">ts</span>, X <span class="op">=</span> <span class="va">inds</span><span class="op">$</span><span class="va">name</span>, <span class="st">"NEA1"</span>, <span class="st">"NEA2"</span>, <span class="st">"AFR1"</span>, <span class="st">"CH1"</span><span class="op">)</span><span class="op">$</span><span class="va">alpha</span></code></pre></div>
<p>If we now summarise the inferred Neanderthal distribution in both populations, we see that there is no Neanderthal ancestry in Africans (as expected by our model–Africans did not receive a Neanderthal introgression pulse) but there is around 3% Neanderthal ancestry in Europeans (consistent with the 3% introgression pulse we simulated between 50-45 thousand years [ky] ago):</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggplot</span><span class="op">(</span><span class="va">inds</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">pop</span>, <span class="va">ancestry</span>, fill <span class="op">=</span> <span class="va">pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Neanderthal ancestry proportion"</span>, x <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">coord_cartesian</span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-05-tree-sequences_files/figure-html/ts_nea_distributions-1.png" width="480"></p>
<p>This is exactly as we specified in the model configuration above, suggesting that our simulations work as they should. You can see that there is quite a bit of noise but that’s because we simulated only 10 Mb of sequence.</p>
<p>We can also plot the trajectory of Neanderthal ancestry in Europe over the last 50 thousand years:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">inds</span><span class="op">[</span><span class="va">inds</span><span class="op">$</span><span class="va">pop</span> <span class="op">==</span> <span class="st">"EUR"</span>, <span class="op">]</span> <span class="op">%&gt;%</span>
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">time</span>, <span class="va">ancestry</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"red"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">xlim</a></span><span class="op">(</span><span class="fl">40000</span>, <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> <span class="fu">coord_cartesian</span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"time [years ago]"</span>, y <span class="op">=</span> <span class="st">"Neanderthal ancestry proportion"</span><span class="op">)</span>
<span class="co">#&gt; `geom_smooth()` using formula 'y ~ x'</span></code></pre></div>
<p><img src="vignette-05-tree-sequences_files/figure-html/ts_nea_trajectory-1.png" width="480"></p>
<p>Again, this is a result consistent with empirical estimates of Neanderthal ancestry using <a href="https://www.pnas.org/content/116/5/1639">ancient DNA data</a>.</p>
</div>
<div id="admixtools-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#admixtools-analyses" class="anchor"></a>ADMIXTOOLS analyses</h2>
<p>In case you would like to verify some <em>f</em>-statistics results using the venerable <a href="https://academic.oup.com/genetics/article/192/3/1065/5935193">ADMIXTOOLS</a> software (see the linked paper which formally introduced these statistics in the first place), you can convert the tree-sequence data to a file format called EIGENSTRAT using the <code><a href="../reference/ts_eigenstrat.html">ts_eigenstrat()</a></code> function. The file conversion is internally handled by the R package <a href="http://bodkan.net/admixr"><em>admixr</em></a> and returns an <code>EIGENSTRAT</code> object which ties all individual <code>EIGENSTRAT</code> file components together (see the <a href="https://bodkan.net/admixr/articles/tutorial.html">tutorial</a> to <em>admixr</em> for an extensive overview). <em>admixr</em> is an R package for running automated ADMIXTOOLS analyses entirely from R and makes these types of analyses very convenient.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_eigenstrat.html">ts_eigenstrat</a></span><span class="op">(</span><span class="va">ts</span>, prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"eigenstrat"</span>, <span class="st">"data"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 183 mutations (out of 58998) have been removed because they appeared on a position already occupied by another mutation. This is a consequence of mutations positions in tskit being in floating-point values but normal genomic locations being integer values.</span>
<span class="va">snps</span>
<span class="co">#&gt; EIGENSTRAT object</span>
<span class="co">#&gt; =================</span>
<span class="co">#&gt; components:</span>
<span class="co">#&gt;   ind file: /var/folders/hr/_t1b0f5n7c76yrfsg8yk9l100000gn/T//RtmpwDRNKP/eigenstrat/data.ind</span>
<span class="co">#&gt;   snp file: /var/folders/hr/_t1b0f5n7c76yrfsg8yk9l100000gn/T//RtmpwDRNKP/eigenstrat/data.snp</span>
<span class="co">#&gt;   geno file: /var/folders/hr/_t1b0f5n7c76yrfsg8yk9l100000gn/T//RtmpwDRNKP/eigenstrat/data.geno</span></code></pre></div>
<p>Running an <em>admixr</em> analysis is then as easy as plugging the object into an <em>admixr</em> function. For instance, we can estimate the proportion of Neanderthal ancestry in a couple of individuals <span class="math inline">\(X\)</span> like this (<em>admixr</em> calls this proportion <code>alpha</code>):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/admixr">admixr</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/admixr/man/f4ratio.html">f4ratio</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">snps</span>, X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"EUR1"</span>, <span class="st">"EUR2"</span>, <span class="st">"AFR2"</span><span class="op">)</span>,
        A <span class="op">=</span> <span class="st">"NEA1"</span>, B <span class="op">=</span> <span class="st">"NEA2"</span>, C <span class="op">=</span> <span class="st">"AFR1"</span>, O <span class="op">=</span> <span class="st">"CH1"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 x 8</span>
<span class="co">#&gt;   A     B     X     C     O       alpha  stderr Zscore</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 NEA1  NEA2  EUR1  AFR1  CH1   0.0712  0.0698    1.02</span>
<span class="co">#&gt; 2 NEA1  NEA2  EUR2  AFR1  CH1   0.00390 0.00382   1.02</span>
<span class="co">#&gt; 3 NEA1  NEA2  AFR2  AFR1  CH1   0       0       NaN</span></code></pre></div>
<p>In fact, lets compare the values obtained by both <em>tskit</em> and <em>admixr</em>/ADMIXTOOLS for all individuals:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># tskit result</span>
<span class="va">result_ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_f2.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">ts</span>, X <span class="op">=</span> <span class="va">inds</span><span class="op">$</span><span class="va">name</span>, A <span class="op">=</span> <span class="st">"NEA1"</span>, B <span class="op">=</span> <span class="st">"NEA2"</span>, C <span class="op">=</span> <span class="st">"AFR1"</span>, O <span class="op">=</span> <span class="st">"CH1"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span>alpha_ts <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span>

<span class="co"># result obtained by admixr/ADMIXTOOLS</span>
<span class="va">result_admixr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/admixr/man/f4ratio.html">f4ratio</a></span><span class="op">(</span><span class="va">snps</span>, X <span class="op">=</span> <span class="va">inds</span><span class="op">$</span><span class="va">name</span>, A <span class="op">=</span> <span class="st">"NEA1"</span>, B <span class="op">=</span> <span class="st">"NEA2"</span>, C <span class="op">=</span> <span class="st">"AFR1"</span>, O <span class="op">=</span> <span class="st">"CH1"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span>alpha_admixr <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span>

<span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_cols</a></span><span class="op">(</span><span class="va">result_ts</span>, <span class="va">result_admixr</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">alpha_ts</span>, <span class="va">alpha_admixr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_abline</span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"red"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"f4-ratio statistic calculated with admixr/ADMIXTOOLS"</span>,
       y <span class="op">=</span> <span class="st">"f4-ratio statistic calculated with tskit"</span><span class="op">)</span></code></pre></div>
<p><img src="vignette-05-tree-sequences_files/figure-html/ts_vs_admixr-1.png" width="480"></p>
<p>The correspondence between the two looks good! 🎉 Again, note that the large amount of variance around the expected value of 3% ancestry is due to an extremely small data simulated here (only 10 Mb of sequence).</p>
</div>
<div id="vcf-output" class="section level2">
<h2 class="hasAnchor">
<a href="#vcf-output" class="anchor"></a>VCF output</h2>
<p>In case you need to process simulated data in some other software, you can use the function <code><a href="../reference/ts_vcf.html">ts_vcf()</a></code> to save the simulated genotypes in a VCF format:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ts_vcf.html">ts_vcf</a></span><span class="op">(</span><span class="va">ts</span>, vcf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"output.vcf.gz"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>You can also specify only a subset of individuals to be saved in the VCF:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ts_vcf.html">ts_vcf</a></span><span class="op">(</span><span class="va">ts</span>, vcf <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"output_subset.vcf.gz"</span><span class="op">)</span>,
       individuals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"CH1"</span>, <span class="st">"NEA1"</span>, <span class="st">"EUR1"</span>, <span class="st">"AFR1"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a895c93f0e9eda05eb10b29a9921d586',
    indexName: 'slendr',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
