<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="slendr">
<title>Tree-sequence processing and statistics • slendr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tree-sequence processing and statistics">
<meta property="og:description" content="slendr">
<meta property="og:image" content="https://www.slendr.net/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">slendr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--articles">
    <span class="fa fa-book"></span>
     
    Articles
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--articles">
    <a class="dropdown-item" href="../articles/vignette-00-installation.html">Installation instructions</a>
    <a class="dropdown-item" href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    <a class="dropdown-item" href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    <a class="dropdown-item" href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    <a class="dropdown-item" href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    <a class="dropdown-item" href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a>
    <a class="dropdown-item" href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a>
    <a class="dropdown-item" href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a>
    <a class="dropdown-item" href="../articles/vignette-08-nonslendr-tskit.html">Analyzing non-slendr tree sequences</a>
    <a class="dropdown-item" href="../articles/vignette-09-paper.html">Examples from the slendr paper</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bodkan/slendr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/fleventy5">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Tree-sequence processing and statistics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/HEAD/vignettes/vignette-05-tree-sequences.Rmd" class="external-link"><code>vignettes/vignette-05-tree-sequences.Rmd</code></a></small>
      <div class="d-none name"><code>vignette-05-tree-sequences.Rmd</code></div>
    </div>

    
    
<p>In this vignette, we will show how to specify sampling events to
record individuals in the tree-sequence output file (a procedure which
is called “remembering” of individuals in the SLiM context) and how to
perform simple analyses using <em>slendr</em>’s interface to the <a href="https://tskit.dev" class="external-link"><em>tskit</em></a> Python library. We will
demonstrate these features on a simulation of Neanderthal introgression
into anatomically modern humans. Specifically, we will show how to
estimate the amount of Neanderthal ancestry using <span class="math inline">\(f\)</span>-statistics calculated directly on the
tree-sequence data structure generated by a <em>slendr</em> model, all
entirely from R.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="setting-up-python-environment">Setting up Python environment<a class="anchor" aria-label="anchor" href="#setting-up-python-environment"></a>
</h2>
<p>First, in order to be able to interface with <em>tskit</em> and
<em>pyslim</em> using the <em>reticulate</em> package (and run
simulations using <code>msprime</code>, as we do below), we will need a
working Python environment with the required Python modules
<code>pyslim</code>, <code>tskit</code> and <code>msprime</code> already
installed.</p>
<p>Because setting up Python environments can be quite a <a href="https://xkcd.com/1987/" class="external-link">hassle</a>, <em>slendr</em> provides a
single function <code><a href="../reference/setup_env.html">setup_env()</a></code> to make things easier. If you
call it without any arguments, <em>slendr</em> will automatically
download, install, and setup a <strong>completely separate</strong>
Python environment (based on the “miniconda” distribution) just for
<em>slendr</em> and activate it in the background.</p>
<p>It is important to stress that <code><a href="../reference/setup_env.html">setup_env()</a></code> will not
interfere in any way with any of the Python installations you might
already have on your computer. The Python installation and environment
will be entirely isolated and used just for the purpose of
<em>slendr</em> workflows.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setup_env.html">setup_env</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Once we have the Python environment set up, we can activate it by
calling:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/init_env.html">init_env</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; The interface to all required Python modules has been activated.</span></span></code></pre>
<p>We can use another built-in function <code><a href="../reference/check_env.html">check_env()</a></code> to make
sure that <em>slendr</em> installed and configured the correct
environment for us:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/check_env.html">check_env</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Summary of the currently active Python environment:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Python binary: /Users/mp/Library/r-miniconda-arm64/envs/Python-3.11_msprime-1.2.0_tskit-0.5.4_pyslim-1.0.1/bin/python </span></span>
<span><span class="co">#&gt; Python version: 3.11.0 | packaged by conda-forge | (main, Jan 14 2023, 12:26:40) [Clang 14.0.6 ] </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; slendr requirements:</span></span>
<span><span class="co">#&gt;  - tskit: version 0.5.4 ✓ </span></span>
<span><span class="co">#&gt;  - msprime: version 1.2.0 ✓ </span></span>
<span><span class="co">#&gt;  - pyslim: version 1.0.1 ✓</span></span></code></pre>
<p>Now we’re good to go and ready to simulate and analyse tree sequence
outputs in R!</p>
<div class="section level3">
<h3 id="model-of-neanderthal-introgression-into-eurasians">Model of Neanderthal introgression into Eurasians<a class="anchor" aria-label="anchor" href="#model-of-neanderthal-introgression-into-eurasians"></a>
</h3>
<p>First, let’s set up a simple <em>non-spatial</em> model of
Neanderthal introgression using <em>slendr.</em> This is essentially the
same procedure which we have shown in another vignette introducing <a href="vignette-04-nonspatial-models.html">non-spatial <em>slendr</em>
models</a>. This is no different from a spatial model, except that we
left out the <code>map</code> argument in calling
<code><a href="../reference/population.html">population()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314159</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create the ancestor of everyone and a chimpanzee outgroup</span></span>
<span><span class="co"># (we set both N = 1 to reduce the computational time for this model)</span></span>
<span><span class="va">chimp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"CH"</span>, time <span class="op">=</span> <span class="fl">6.5e6</span>, N <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># two populations of anatomically modern humans: Africans and Europeans</span></span>
<span><span class="va">afr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"AFR"</span>, parent <span class="op">=</span> <span class="va">chimp</span>, time <span class="op">=</span> <span class="fl">6e6</span>, N <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span><span class="va">eur</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"EUR"</span>, parent <span class="op">=</span> <span class="va">afr</span>, time <span class="op">=</span> <span class="fl">70e3</span>, N <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Neanderthal population splitting at 600 ky ago from modern humans</span></span>
<span><span class="co"># (becomes extinct by 40 ky ago)</span></span>
<span><span class="va">nea</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"NEA"</span>, parent <span class="op">=</span> <span class="va">afr</span>, time <span class="op">=</span> <span class="fl">600e3</span>, N <span class="op">=</span> <span class="fl">1000</span>, remove <span class="op">=</span> <span class="fl">40e3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3% Neanderthal introgression into Europeans between 55-50 ky ago</span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">nea</span>, to <span class="op">=</span> <span class="va">eur</span>, rate <span class="op">=</span> <span class="fl">0.03</span>, start <span class="op">=</span> <span class="fl">55000</span>, end <span class="op">=</span> <span class="fl">45000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span></span>
<span>  populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">chimp</span>, <span class="va">nea</span>, <span class="va">afr</span>, <span class="va">eur</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>,</span>
<span>  generation_time <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"_introgression"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here’s our toy model visualized as a “demographic graph” of sorts
(i.e., a tree-like structure specifying population splits with
additional edges representing gene flow events). Not particularly
illuminating in this simple example, but it’s always worth keeping in
mind that such graph is embedded within every <em>slendr</em> model and
can be always invoked to make sure the model you’re setting up is
correct:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/plot_model.html">plot_model</a></span><span class="op">(</span><span class="va">model</span>, sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/plot_model.html">plot_model</a></span><span class="op">(</span><span class="va">model</span>, sizes <span class="op">=</span> <span class="cn">FALSE</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-tree-sequences_files/figure-html/introgression_graph-1.png" width="640"></p>
</div>
</div>
<div class="section level2">
<h2 id="scheduling-of-sampling-events">Scheduling of sampling events<a class="anchor" aria-label="anchor" href="#scheduling-of-sampling-events"></a>
</h2>
<p>Now that we have defined a model, how do we sample data from it?
Ideally, we would like to to schedule sampling events at a given time,
sampling a defined number of individuals from a given population. This
is why <em>slendr</em> provides a function
<code><a href="../reference/schedule_sampling.html">schedule_sampling()</a></code> which serves to define such sampling
schedule automatically and enforces that only populations which are
already (i.e. after their appearance in the simulation) or still (before
they are removed from the simulation) present will be sampled from.</p>
<p>In our example, we want to sample two Neanderthal individuals (the
older one being the Altai Neanderthal published by Pruefer <em>et
al.</em> 2014, the younger one Vindija Neanderthal published by Pruefer
<em>et al.</em>, 2017). These two genomes are what we need to estimate
Neanderthal ancestry proportion using a so-called <span class="math inline">\(f_4\)</span>-ratio statistic (more on that below,
but also see Petr <em>et al.</em>, PNAS 2019):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nea_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">70000</span>, <span class="fl">40000</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">nea</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nea_samples</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 7</span></span></span>
<span><span class="co">#&gt;    time pop       n y_orig x_orig y     x    </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> <span style="text-decoration: underline;">40</span>000 NEA       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> <span style="text-decoration: underline;">70</span>000 NEA       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span></span></span></code></pre>
<p>As you can see, the <code><a href="../reference/schedule_sampling.html">schedule_sampling()</a></code> function simply
accepts the vector of times at which remembering should be schedule, and
then a list of pairs
<code>(&lt;slendr population&gt;, &lt;number of individuals&gt;)</code>
encoding from which populations should how many individuals be
remembered at time points given in the <code>times</code> vector.</p>
<p>Next, we want to sample some present-day individuals: an outgroup
representing a chimpanzee, and a couple of Africans and Europeans:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">present_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">chimp</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">afr</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">eur</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">present_samples</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 7</span></span></span>
<span><span class="co">#&gt;    time pop       n y_orig x_orig y     x    </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     0 CH        1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     0 AFR       5 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>     0 EUR      10 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span></span></span></code></pre>
<p>As you can see above, the <code><a href="../reference/schedule_sampling.html">schedule_sampling()</a></code> function
returns a plain old data frame with a very simple structure with three
columns: time, population name, and the number of individuals. This
means that you can define sampling events using whatever input data you
might already have available (such as radiocarbon-dated ancient DNA
samples from an Excel sheet from some publication). For instance, there
has been a <a href="https://www.nature.com/articles/nature17993" class="external-link">lot</a>
of interest to estimate the trajectory of Neanderthal ancestry in Europe
over time using ancient DNA data from anatomically modern human
individuals (also called early modern humans, EMH) across the last
couple of tens of thousands of years. We can simulate something close to
the available <a href="https://www.nature.com/articles/nature17993" class="external-link">EMH
ancient DNA data set</a> over the last 50 thousand years by running
doing this:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">emh_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">40</span>, min <span class="op">=</span> <span class="fl">10000</span>, max <span class="op">=</span> <span class="fl">40000</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">eur</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">emh_samples</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 40 × 7</span></span></span>
<span><span class="co">#&gt;     time pop       n y_orig x_orig y     x    </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> <span style="text-decoration: underline;">10</span>320 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> <span style="text-decoration: underline;">11</span>188 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> <span style="text-decoration: underline;">11</span>396 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> <span style="text-decoration: underline;">11</span>529 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> <span style="text-decoration: underline;">11</span>927 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> <span style="text-decoration: underline;">12</span>675 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> <span style="text-decoration: underline;">13</span>689 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> <span style="text-decoration: underline;">13</span>744 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> <span style="text-decoration: underline;">14</span>775 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> <span style="text-decoration: underline;">16</span>362 EUR       1 <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 30 more rows</span></span></span></code></pre>
<p>This samples a single ancient European individuals at randomly chosen
times between 40 and 10 ky ago.</p>
<p>One nice feature of the <code><a href="../reference/schedule_sampling.html">schedule_sampling()</a></code> function is
that it only schedules sampling events for a population, if that
population is present in the simulation at a given time. This makes it
possible to simply take a wide time range for sampling, specify all
populations and sizes of the samples, and let the function generate
sampling events only for populations present at each time. If for some
reason a stricter control over sampling is required, this behavior can
be switched off by setting <code>strict = TRUE</code> like this:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this attempts to sample a Neanderthal individual at a point when Neanderthals</span></span>
<span><span class="co"># are already extinct, resulting in an error</span></span>
<span><span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="fl">10000</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">nea</span>, <span class="fl">1</span><span class="op">)</span>, strict <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code>Error: Cannot schedule sampling for 'NEA' at time 10000 because the population will not be present in the simulation at that point. Consider running this function with `strict = FALSE` which will automatically retain only valid sampling events.</code></pre>
<p>Now that we already have the <code>model</code> object ready, we can
simulate data from it, sampling individuals according to our sampling
schedule. Although we could use the <code><a href="../reference/slim.html">slim()</a></code> function shown
in previous vignettes, in this case we will run the simulation with the
<code><a href="../reference/msprime.html">msprime()</a></code> coalescent back end. After all, our model is
non-spatial and using a coalescent simulator will be much more efficient
than the forward simulation. Switching between the msprime and SLiM back
ends of slendr is demonstrated in much more detail in a <a href="vignette-07-backends.html">dedicated vignette</a>.</p>
<p>The simulation back end utilized by the <code><a href="../reference/msprime.html">msprime()</a></code>
function (as well as the <code><a href="../reference/slim.html">slim()</a></code> function) produces a
tree-sequence output which is immediately loaded and ready for a
downstream analysis.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span></span>
<span>  <span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">100e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>,</span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">nea_samples</span>, <span class="va">present_samples</span>, <span class="va">emh_samples</span><span class="op">)</span>,</span>
<span>  random_seed <span class="op">=</span> <span class="fl">314159</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; --------------------------------------------------</span></span>
<span><span class="co">#&gt; msprime command:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; /Users/mp/Library/r-miniconda-arm64/envs/Python-3.11_msprime-1.2.0_tskit-0.5.4_pyslim-1.0.1/bin/python /var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T//RtmpJmVorf/file4c492cb0963f_introgression/script.py --seed 314159 --model /var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T//RtmpJmVorf/file4c492cb0963f_introgression --output /var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T//RtmpJmVorf/file4c493b2ee5de.trees --sequence-length 100000000 --recombination-rate 1e-08 --sampling-schedule /var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T//RtmpJmVorf/file4c4913aaeef8 --verbose  </span></span>
<span><span class="co">#&gt; --------------------------------------------------</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tree sequence was saved to:</span></span>
<span><span class="co">#&gt;  /var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T//RtmpJmVorf/file4c493b2ee5de.trees </span></span>
<span><span class="co">#&gt; Loading the tree-sequence file...</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔═══════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence               ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═══════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │     237615║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│  100000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │generations║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │        116║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │   36.5 MiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═══════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤══════╤═════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows  │Size     │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪══════╪═════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │859795│ 26.2 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│    58│  1.6 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │     0│  8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │     0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │138806│  3.7 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│     4│338 Bytes│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│     1│  7.0 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │     0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧══════╧═════════╧════════════╝</span></span></code></pre>
<p>Note that we bind the individual sampling schedule data frames using
the <code>rbind</code> function provided by base R (as we show above,
the sampling schedule really is just a data frame and we can manipulate
it as such).</p>
</div>
<div class="section level2">
<h2 id="r-interface-for-tskit-and-pyslim">R interface for <em>tskit</em> and <em>pyslim</em><a class="anchor" aria-label="anchor" href="#r-interface-for-tskit-and-pyslim"></a>
</h2>
<p>Tree-sequences are one of the most revolutionary developments in
population genetics in the last couple of decades for a number of
reasons. One of them is the possibility to store extremely large data
sets succinctly by encoding the entire evolutionary history of a sample
of individuals as a series of correlated tree genealogies along the
genome.</p>
<p>Going into too much detail on this topic is clearly beyond the scope
of this tutorial, especially because everything is explain much better
<a href="https://tskit.dev/learn/" class="external-link">elsewhere</a>. Instead, what we will
demonstrate in the rest of this vignette is how you can access and
manipulate tree-sequence outputs generated by <em>slendr</em> models and
perform various statistics on them using Python modules <a href="https://tskit.dev/tskit/docs/stable/" class="external-link"><em>tskit</em></a> and <a href="https://pyslim.readthedocs.io/en/latest/index.html" class="external-link"><em>pyslim</em></a>
directly from <em>slendr,</em> without having to leave R! The key is a
magical R package <a href="https://rstudio.github.io/reticulate/index.html" class="external-link"><em>reticulate</em></a>
which creates a seamless binding of Python modules with R. This means
that even if you don’t know Python, <em>slendr</em> allows you to do do
quite a lot with tree-sequences in R.</p>
<p>Of course, if you are a proficient Python user, it needs to be said
that once you have the tree-sequence file generated by <em>slendr</em>
&amp; SLiM, you can easily perform every conceivable analysis directly
using <em>tskit</em>. The intention here is to show how you can continue
working on the tree-sequence files in R even after you have run the
entire <em>slendr</em> simulation.</p>
</div>
<div class="section level2">
<h2 id="loading-and-processing-tree-sequence-output-files">Loading and processing tree-sequence output files<a class="anchor" aria-label="anchor" href="#loading-and-processing-tree-sequence-output-files"></a>
</h2>
<p>By default, any <code><a href="../reference/msprime.html">msprime()</a></code> or <code><a href="../reference/slim.html">slim()</a></code> run
will produce a tree-sequence object (saved to a temporary file) and
immediately load it. Thus, in most use-cases, explicit loading of a
simulated tree sequence is not needed. That said, for the sake of
completeness, let’s run the simulation again but specify a custom path
to the tree-sequence file ourselves.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span></span>
<span>  <span class="va">model</span>, sequence_length <span class="op">=</span> <span class="fl">100e6</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>,</span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">nea_samples</span>, <span class="va">present_samples</span>, <span class="va">emh_samples</span><span class="op">)</span>,</span>
<span>  output <span class="op">=</span> <span class="va">output_file</span>, random_seed <span class="op">=</span> <span class="fl">314159</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">output_file</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "/var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T//RtmpJmVorf/file4c4966979c62"</span></span></code></pre>
<p>In case have the tree-sequence output saved in a custom location on
disk, we can load the tree sequence using the <em>slendr</em> function
<code><a href="../reference/ts_load.html">ts_load()</a></code>. If we’re dealing with a tree sequence produced
by the SLiM back end (which is not the case here), we can also instruct
this function to simplify the tree-sequence to only the individuals that
we explicitly sampled (recall the sampling schedule we set up with the
<code><a href="../reference/schedule_sampling.html">schedule_sampling()</a></code> function above). Note that we have to
provide the <code>model</code> object generated by
<code><a href="../reference/compile_model.html">compile_model()</a></code> above in order to have all model annotation
information for the simulated tree-sequence data (we have to do this
only once, and only during loading):</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_load.html">ts_load</a></span><span class="op">(</span><span class="va">output_file</span>, <span class="va">model</span><span class="op">)</span></span>
<span><span class="va">ts</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔═══════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence               ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═══════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │     237615║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│  100000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │generations║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │        116║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │   36.5 MiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═══════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤══════╤═════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows  │Size     │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪══════╪═════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │859795│ 26.2 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│    58│  1.6 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │     0│  8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │     0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │138806│  3.7 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│     4│338 Bytes│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│     1│  7.0 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │     0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧══════╧═════════╧════════════╝</span></span></code></pre>
<p>Not surprisingly, we get the same output which we got above when we
printed the tree sequence returned by the <code><a href="../reference/msprime.html">msprime()</a></code>
function. This shows that under normal circumstances, loading the output
manually via <code><a href="../reference/ts_load.html">ts_load()</a></code> is not needed.</p>
<p>If we try to simplify an <em>msprime</em>-generated tree sequence, we
get a warning. This is because such tree sequence is already in a
simplified form, by definition coming from a coalescent simulator.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_simplify.html">ts_simplify</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Warning: If you want to simplify an msprime tree sequence, you must specify</span></span>
<span><span class="co">#&gt; the names of individuals to simplify to via the `simplify_to = `</span></span>
<span><span class="co">#&gt; function argument.</span></span></code></pre>
<pre><code><span><span class="co">#&gt; ╔═══════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence               ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═══════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │     237615║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│  100000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │generations║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │        116║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │   36.5 MiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═══════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤══════╤═════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows  │Size     │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪══════╪═════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │859795│ 26.2 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│    58│  1.6 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │     0│  8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │     0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │138806│  3.7 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│     4│338 Bytes│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│     1│  7.0 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │     0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧══════╧═════════╧════════════╝</span></span></code></pre>
<p>By default, simplification trims the tree sequence down to only
remembered individuals (i.e. those we explicitly scheduled for
sampling), which is true for every <em>msprime</em> tree sequence.
Alternatively, we can also narrow down the simplification to a defined
set of individuals using the <code>simplify_to =</code> argument.
Internally, simplification is implemented in a dedicated function
<code><a href="../reference/ts_simplify.html">ts_simplify()</a></code> which we can always call explicitly, like
this:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_simplify.html">ts_simplify</a></span><span class="op">(</span><span class="va">ts</span>, simplify_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CH_1"</span>, <span class="st">"NEA_1"</span>, <span class="st">"NEA_2"</span>, <span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"EUR_20"</span>, <span class="st">"EUR_50"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ts_small</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔═══════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence               ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═══════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │     132530║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│  100000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │generations║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │         14║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │   19.0 MiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═══════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤══════╤═════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows  │Size     │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪══════╪═════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │440810│ 13.5 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│     7│220 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │     0│  8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │     0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │ 79569│  2.1 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│     4│338 Bytes│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│     2│  7.5 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │     0│ 16 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧══════╧═════════╧════════════╝</span></span></code></pre>
<p>Similarly, <em>slendr</em> provides a function
<code><a href="../reference/ts_recapitate.html">ts_recapitate()</a></code> which performs [recapitation]<a href="https://tskit.dev/pyslim/docs/latest/tutorial.html#recapitation" class="external-link uri">https://tskit.dev/pyslim/docs/latest/tutorial.html#recapitation</a>).
Again, this is not needed for an <em>msprime</em> tree sequence, which
is fully coalesced (and recapitated) by definition. Still, if we were
dealing with a SLiM tree sequence, which is the case for some of our
other vignetted, we could do this in one go, by specifying
<code>recapitate = TRUE</code> in the call to <code><a href="../reference/ts_load.html">ts_load()</a></code>
after specifying a couple of additional arguments required for
recapitation (see the pyslim documentation in the <a href="https://tskit.dev/pyslim/docs/latest/tutorial.html#recapitation" class="external-link">recapitation</a>
section for more detail). If we do this on our current tree sequence
object, we simply get a warning informing us that we’re attempting to do
something that has no effect:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_recapitate.html">ts_recapitate</a></span><span class="op">(</span><span class="va">ts</span>, recombination_rate <span class="op">=</span> <span class="fl">1e-8</span>, Ne <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<p>We can make sure that our tree sequence is fully coalesced by calling
another <em>slendr</em> function <code><a href="../reference/ts_coalesced.html">ts_coalesced()</a></code>. This is
useful when dealing with <code><a href="../reference/slim.html">slim()</a></code>-produced tree
sequences:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_coalesced.html">ts_coalesced</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] TRUE</span></span></code></pre>
<p>You might have noticed that we did not simulate any mutations during
the SLiM run. This is for computational efficiency. Luckily, the
tree-sequence contains the complete history of a sample of individuals
which makes it very easy to sprinkle mutations on the genealogies after
this simulation is over. We can add mutations a given rate by
running:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span><span class="va">ts</span>, mutation_rate <span class="op">=</span> <span class="fl">1e-8</span>, random_seed <span class="op">=</span> <span class="fl">314159</span><span class="op">)</span></span>
<span><span class="va">ts</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔═══════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence               ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═══════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │     237615║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│  100000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │generations║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │        116║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │   73.3 MiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═══════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤══════╤═════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows  │Size     │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪══════╪═════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │859795│ 26.2 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│    58│  1.6 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │     0│  8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │622315│ 22.0 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │138806│  3.7 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│     4│338 Bytes│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│     2│  7.7 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │620386│ 14.8 MiB│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧══════╧═════════╧════════════╝</span></span></code></pre>
<p>Having processed the simulated tree sequence, we can calculate some
basic statistics on our simulated data.</p>
<p>However, before we do that, we would first like to note that
everything that we do in the rest of this vignette (i.e. whenever we
call a function with the prefix <code>ts_*()</code> in <em>slendr</em>),
we are interfacing with the <em>tskit</em> Python module under the hood.
Our goal is to capture most of the analyses one might want to perform on
tree-sequences in R and wrap them in a neat interface indistinguishable
from any other R function—this is, after all, the reason why
<em>reticulate</em> has been created in the first place (making various
Python data science modules appear as if they were regular R
packages).</p>
</div>
<div class="section level2">
<h2 id="visualisation-of-trees-and-tree-sequences">Visualisation of trees and tree-sequences<a class="anchor" aria-label="anchor" href="#visualisation-of-trees-and-tree-sequences"></a>
</h2>
<p>Now we introduce a function <code><a href="../reference/ts_phylo.html">ts_phylo()</a></code> which can be used
to extract one tree from the tree-sequence (either an <em>i</em>-th tree
in the sequence, or a tree overlapping an <em>i</em>-th position of the
simulated genome, depending on the value of its <code>mode</code>
argument) and convert it to a <code>phylo</code> class, which is a
standard format for phylogenetic trees in the R world. For more on the
<code>phylo</code> format, see packages <a href="https://cran.r-project.org/package=ape" class="external-link">ape</a>, <a href="https://cran.r-project.org/package=phangorn" class="external-link">phangorn</a>, and <a href="https://cran.r-project.org/package=phytools" class="external-link">phytools</a>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract the 42nd tree in the tree sequence</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_phylo.html">ts_phylo</a></span><span class="op">(</span><span class="va">ts_small</span>, <span class="fl">42</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 14 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 13 </span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre>
<p>When we type the tree object to an R console, we can verify that we
got an ordinary <code>phylo</code> object:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Phylogenetic tree with 14 tips and 13 internal nodes.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tip labels:</span></span>
<span><span class="co">#&gt;   13 (EUR_50), 12 (EUR_50), 11 (CH_1), 10 (CH_1), 9 (AFR_2), 8 (AFR_2), ...</span></span>
<span><span class="co">#&gt; Node labels:</span></span>
<span><span class="co">#&gt;   77849, 70, 3964, 11736, 12966, 16714, ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Rooted; includes branch lengths.</span></span></code></pre>
<p>This means that we have the whole R phylogenetic ecosystem at our
disposal to analyze such trees. For instance we can use the powerful
package <em>ggtree</em> to plot the tree we have just extracted:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/" class="external-link">ggtree</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;">ggtree v3.6.2 For help: https://yulab-smu.top/treedata-book/</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; If you use the ggtree package suite in published research, please cite</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; the appropriate paper(s):</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam.</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; ggtree: an R package for visualization and annotation of phylogenetic</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; trees with their covariates and other associated data. Methods in</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Guangchuang Yu.  Data Integration, Manipulation and Visualization of</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Phylogenetic Trees (1st edition). Chapman and Hall/CRC. 2022,</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; doi:10.1201/9781003279242</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; </span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; G Yu. Data Integration, Manipulation and Visualization of Phylogenetic</span></span></span>
<span><span class="co"><span style="color: #949494;">#&gt; Trees (1st ed.). Chapman and Hall/CRC. 2022. ISBN: 9781032233574 </span></span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_point2.html" class="external-link">geom_point2</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>subset <span class="op">=</span> <span class="op">!</span><span class="va">isTip</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="co"># points for internal nodes</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html" class="external-link">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="co"># sample labels for tips</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggexpand.html" class="external-link">hexpand</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span>    <span class="co"># make more space for the tip labels</span></span></code></pre></div>
<p><img src="vignette-05-tree-sequences_files/figure-html/plot_ggtree-1.png" width="480"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">nodelabels</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="accessing-tskit-functionality-directly">Accessing <em>tskit</em> functionality directly<a class="anchor" aria-label="anchor" href="#accessing-tskit-functionality-directly"></a>
</h2>
<p>As we mentioned in the previous section, the goal of this vignette is
to show how to use <em>slendr</em> to perform the main tree-sequence
operations using a convenient R interface to <em>tskit</em>. However,
you should always keep in mind that you are not restricted to a subset
of <em>tskit</em> functionality that <em>slendr</em> translated to R
(i.e. <a href="../reference/index.html">all functions</a> with the
prefix <code>ts_</code>). Thanks to the incredible R package <a href="https://rstudio.github.io/reticulate/index.html" class="external-link"><em>reticulate</em></a>,
you can access Python methods and object variables directly, using the
<code>$</code> operator.</p>
<p>As an example, instead of calling the function
<code><a href="../reference/ts_coalesced.html">ts_coalesced()</a></code> on the tree-sequence as we did above, we
could check that all trees are coalesced by running the following
snippet instead (note that this is very inefficient and we’re only doing
the operation for the first one hundred trees):</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># iterate over all trees in the tree-sequence and check if each</span></span>
<span><span class="co"># has only one root (i.e. is fully coalesced) - note that Python</span></span>
<span><span class="co"># lists are 0-based, which is something we need to take care of</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">ts</span><span class="op">$</span><span class="va">num_trees</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>,</span>
<span>           <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">ts</span><span class="op">$</span><span class="fu">at_index</span><span class="op">(</span><span class="va">i</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">$</span><span class="va">num_roots</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We believe it makes sense to use the R interface whenever possible
(even if only because it makes many operations a little bit more
convenient). However, if there is some functionality in <em>slendr</em>
missing, you can always resort to accessing the Python objects directly
as we have just demonstrated. You can verify that all <a href="https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence" class="external-link">methods
and attributes</a> of a Python tree-sequence object is still accessible
in R:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   [1] "alignments"                      "allele_frequency_spectrum"      </span></span>
<span><span class="co">#&gt;   [3] "as_fasta"                        "as_nexus"                       </span></span>
<span><span class="co">#&gt;   [5] "as_vcf"                          "aslist"                         </span></span>
<span><span class="co">#&gt;   [7] "at"                              "at_index"                       </span></span>
<span><span class="co">#&gt;   [9] "breakpoints"                     "check_index"                    </span></span>
<span><span class="co">#&gt;  [11] "coalescence_time_distribution"   "coiterate"                      </span></span>
<span><span class="co">#&gt;  [13] "count_topologies"                "decapitate"                     </span></span>
<span><span class="co">#&gt;  [15] "delete_intervals"                "delete_sites"                   </span></span>
<span><span class="co">#&gt;  [17] "diffs"                           "discrete_genome"                </span></span>
<span><span class="co">#&gt;  [19] "discrete_time"                   "divergence"                     </span></span>
<span><span class="co">#&gt;  [21] "diversity"                       "draw_svg"                       </span></span>
<span><span class="co">#&gt;  [23] "draw_text"                       "dump"                           </span></span>
<span><span class="co">#&gt;  [25] "dump_tables"                     "dump_text"                      </span></span>
<span><span class="co">#&gt;  [27] "edge"                            "edge_diffs"                     </span></span>
<span><span class="co">#&gt;  [29] "edges"                           "edges_child"                    </span></span>
<span><span class="co">#&gt;  [31] "edges_left"                      "edges_parent"                   </span></span>
<span><span class="co">#&gt;  [33] "edges_right"                     "edgesets"                       </span></span>
<span><span class="co">#&gt;  [35] "equals"                          "f2"                             </span></span>
<span><span class="co">#&gt;  [37] "f3"                              "f4"                             </span></span>
<span><span class="co">#&gt;  [39] "file_uuid"                       "first"                          </span></span>
<span><span class="co">#&gt;  [41] "Fst"                             "genealogical_nearest_neighbours"</span></span>
<span><span class="co">#&gt;  [43] "general_stat"                    "genetic_relatedness"            </span></span>
<span><span class="co">#&gt;  [45] "genotype_matrix"                 "get_ll_tree_sequence"           </span></span>
<span><span class="co">#&gt;  [47] "get_num_mutations"               "get_num_nodes"                  </span></span>
<span><span class="co">#&gt;  [49] "get_num_records"                 "get_num_sites"                  </span></span>
<span><span class="co">#&gt;  [51] "get_num_trees"                   "get_pairwise_diversity"         </span></span>
<span><span class="co">#&gt;  [53] "get_population"                  "get_sample_size"                </span></span>
<span><span class="co">#&gt;  [55] "get_samples"                     "get_sequence_length"            </span></span>
<span><span class="co">#&gt;  [57] "get_time"                        "haplotypes"                     </span></span>
<span><span class="co">#&gt;  [59] "has_reference_sequence"          "ibd_segments"                   </span></span>
<span><span class="co">#&gt;  [61] "indexes_edge_insertion_order"    "indexes_edge_removal_order"     </span></span>
<span><span class="co">#&gt;  [63] "individual"                      "individual_locations"           </span></span>
<span><span class="co">#&gt;  [65] "individual_populations"          "individual_times"               </span></span>
<span><span class="co">#&gt;  [67] "individuals"                     "individuals_flags"              </span></span>
<span><span class="co">#&gt;  [69] "individuals_location"            "individuals_population"         </span></span>
<span><span class="co">#&gt;  [71] "individuals_time"                "kc_distance"                    </span></span>
<span><span class="co">#&gt;  [73] "keep_intervals"                  "last"                           </span></span>
<span><span class="co">#&gt;  [75] "ll_tree_sequence"                "load"                           </span></span>
<span><span class="co">#&gt;  [77] "load_tables"                     "ltrim"                          </span></span>
<span><span class="co">#&gt;  [79] "max_root_time"                   "max_time"                       </span></span>
<span><span class="co">#&gt;  [81] "mean_descendants"                "metadata"                       </span></span>
<span><span class="co">#&gt;  [83] "metadata_schema"                 "migration"                      </span></span>
<span><span class="co">#&gt;  [85] "migrations"                      "migrations_dest"                </span></span>
<span><span class="co">#&gt;  [87] "migrations_left"                 "migrations_node"                </span></span>
<span><span class="co">#&gt;  [89] "migrations_right"                "migrations_source"              </span></span>
<span><span class="co">#&gt;  [91] "migrations_time"                 "min_time"                       </span></span>
<span><span class="co">#&gt;  [93] "mutation"                        "mutations"                      </span></span>
<span><span class="co">#&gt;  [95] "mutations_node"                  "mutations_parent"               </span></span>
<span><span class="co">#&gt;  [97] "mutations_site"                  "mutations_time"                 </span></span>
<span><span class="co">#&gt;  [99] "nbytes"                          "newick_trees"                   </span></span>
<span><span class="co">#&gt; [101] "node"                            "nodes"                          </span></span>
<span><span class="co">#&gt; [103] "nodes_flags"                     "nodes_individual"               </span></span>
<span><span class="co">#&gt; [105] "nodes_population"                "nodes_time"                     </span></span>
<span><span class="co">#&gt; [107] "num_edges"                       "num_individuals"                </span></span>
<span><span class="co">#&gt; [109] "num_migrations"                  "num_mutations"                  </span></span>
<span><span class="co">#&gt; [111] "num_nodes"                       "num_populations"                </span></span>
<span><span class="co">#&gt; [113] "num_provenances"                 "num_samples"                    </span></span>
<span><span class="co">#&gt; [115] "num_sites"                       "num_trees"                      </span></span>
<span><span class="co">#&gt; [117] "pairwise_diversity"              "parse_windows"                  </span></span>
<span><span class="co">#&gt; [119] "population"                      "populations"                    </span></span>
<span><span class="co">#&gt; [121] "provenance"                      "provenances"                    </span></span>
<span><span class="co">#&gt; [123] "records"                         "reference_sequence"             </span></span>
<span><span class="co">#&gt; [125] "rtrim"                           "sample_count_stat"              </span></span>
<span><span class="co">#&gt; [127] "sample_size"                     "samples"                        </span></span>
<span><span class="co">#&gt; [129] "segregating_sites"               "sequence_length"                </span></span>
<span><span class="co">#&gt; [131] "simplify"                        "site"                           </span></span>
<span><span class="co">#&gt; [133] "sites"                           "sites_position"                 </span></span>
<span><span class="co">#&gt; [135] "split_edges"                     "subset"                         </span></span>
<span><span class="co">#&gt; [137] "table_metadata_schemas"          "tables"                         </span></span>
<span><span class="co">#&gt; [139] "tables_dict"                     "Tajimas_D"                      </span></span>
<span><span class="co">#&gt; [141] "time_units"                      "to_macs"                        </span></span>
<span><span class="co">#&gt; [143] "to_nexus"                        "trait_correlation"              </span></span>
<span><span class="co">#&gt; [145] "trait_covariance"                "trait_linear_model"             </span></span>
<span><span class="co">#&gt; [147] "trait_regression"                "trees"                          </span></span>
<span><span class="co">#&gt; [149] "trim"                            "union"                          </span></span>
<span><span class="co">#&gt; [151] "variants"                        "write_fasta"                    </span></span>
<span><span class="co">#&gt; [153] "write_nexus"                     "write_vcf"                      </span></span>
<span><span class="co">#&gt; [155] "Y1"                              "Y2"                             </span></span>
<span><span class="co">#&gt; [157] "Y3"</span></span></code></pre>
<p>In fact, you will recognize some of the elements in the output above
from examples involving <code>ts_</code> functions in this vignette! In
short, there is no blackbox—<em>slendr</em> only provides a slightly
more convenient layer over <em>tskit</em> for R users.</p>
</div>
<div class="section level2">
<h2 id="calculating-f-statistics">Calculating <em>f</em>-statistics<a class="anchor" aria-label="anchor" href="#calculating-f-statistics"></a>
</h2>
<p>In addition to being a revolutionary breakthrough in terms of
computation efficiency, many statistics that we are often interested in
population genetics are a natural consequence of having a direct access
to tree sequence genealogies, simply because those genealogies capture
the true demographic history of a sample. Again, we can’t go into too
much detail here but we encourage you to take a look at a paper by <a href="https://doi.org/10.1534/genetics.120.303253" class="external-link">Ralph <em>et
al.</em></a> on the duality between statistics expressed in terms of
branch lengths and the traditional summaries based on samples of genetic
variation.</p>
<p>For instance, we have functions such as <code><a href="../reference/ts_f4ratio.html">ts_f2()</a></code>,
<code><a href="../reference/ts_f4ratio.html">ts_f3()</a></code>, <code><a href="../reference/ts_f4ratio.html">ts_f4()</a></code> and <code><a href="../reference/ts_f4ratio.html">ts_f4ratio()</a></code>
which calculate the well-known set of Patterson’s <span class="math inline">\(f\)</span>-statistics:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># f2 is a measure of the branch length connecting A and B</span></span>
<span><span class="fu"><a href="../reference/ts_f4ratio.html">ts_f2</a></span><span class="op">(</span><span class="va">ts</span>, A <span class="op">=</span> <span class="st">"EUR_1"</span>, B <span class="op">=</span> <span class="st">"AFR_1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 3</span></span></span>
<span><span class="co">#&gt;   A     B            f2</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> EUR_1 AFR_1 0.000<span style="text-decoration: underline;">039</span>3</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># f4 is a measure of the drift shared between A and B after their split from C</span></span>
<span><span class="fu"><a href="../reference/ts_f4ratio.html">ts_f3</a></span><span class="op">(</span><span class="va">ts</span>, A <span class="op">=</span> <span class="st">"EUR_1"</span>, B <span class="op">=</span> <span class="st">"AFR_1"</span>, C <span class="op">=</span> <span class="st">"CH_1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">#&gt;   A     B     C            f3</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> EUR_1 AFR_1 CH_1  0.000<span style="text-decoration: underline;">010</span>7</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this value should be very close to zero (no introgression in Africans)</span></span>
<span><span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"NEA_1"</span>, <span class="st">"CH_1"</span>, mode <span class="op">=</span> <span class="st">"branch"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 5</span></span></span>
<span><span class="co">#&gt;   W     X     Y     Z        f4</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> AFR_1 AFR_2 NEA_1 CH_1  -<span style="color: #BB0000;">16.9</span></span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this value should be significantly negative (many more ABBA sites</span></span>
<span><span class="co"># compared to BABA site due to the introgression into Europeans)</span></span>
<span><span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4</a></span><span class="op">(</span><span class="va">ts</span>, <span class="st">"AFR_1"</span>, <span class="st">"EUR_1"</span>, <span class="st">"NEA_1"</span>, <span class="st">"CH_1"</span>, mode <span class="op">=</span> <span class="st">"branch"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 5</span></span></span>
<span><span class="co">#&gt;   W     X     Y     Z        f4</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> AFR_1 EUR_1 NEA_1 CH_1  -<span style="color: #BB0000;">565.</span></span></span></code></pre>
<p>These functions accept a <code>mode =</code> argument, specifying
whether the statistics should be calculated using mutation site patterns
(<code>mode = "site"</code>, the default), branch lengths
(<code>mode = "branch"</code>), or for each node
(<code>mode = "node"</code>), as well as the <code>windows</code>
argument, similarly to other “multiway” statistics implemented by tskit.
See the <a href="https://tskit.dev/tskit/docs/stable/stats.html#sec-stats-mode" class="external-link">relevant</a>
<a href="https://tskit.dev/tskit/docs/stable/stats.html#sec-stats-sample-sets-multi-way" class="external-link">sections</a>
of the official <em>tskit</em> documentation for more on this topic.</p>
<p>Note that in the previous chunk we referred to individuals by their
names (not numeric IDs of nodes as you would do it with <em>tskit</em>
in Python). We allow this for readability and to make it easier to see
which individuals are which based on the specified sampling schedule
(the names are assigned to individuals based on the order of their
sampling). We can get an overview of the individuals scheduled for
sampling (i.e. permanently remembered) and their names with a helper
function <code><a href="../reference/ts_samples.html">ts_samples()</a></code>:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_samples.html">ts_samples</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 58 × 3</span></span></span>
<span><span class="co">#&gt;    name   time pop  </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> NEA_1 <span style="text-decoration: underline;">70</span>000 NEA  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> NEA_2 <span style="text-decoration: underline;">40</span>000 NEA  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> EUR_1 <span style="text-decoration: underline;">39</span>171 EUR  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> EUR_2 <span style="text-decoration: underline;">39</span>134 EUR  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> EUR_3 <span style="text-decoration: underline;">37</span>739 EUR  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> EUR_4 <span style="text-decoration: underline;">36</span>531 EUR  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> EUR_5 <span style="text-decoration: underline;">36</span>362 EUR  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> EUR_6 <span style="text-decoration: underline;">35</span>945 EUR  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> EUR_7 <span style="text-decoration: underline;">33</span>900 EUR  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> EUR_8 <span style="text-decoration: underline;">33</span>854 EUR  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 48 more rows</span></span></span></code></pre>
<p>That said, if you would like to run some statistics on nodes rather
than on individuals, you can do it simply by using integer IDs instead
of character names in each function’s interface.</p>
</div>
<div class="section level2">
<h2 id="estimating-neanderthal-ancestry-proportions">Estimating Neanderthal ancestry proportions<a class="anchor" aria-label="anchor" href="#estimating-neanderthal-ancestry-proportions"></a>
</h2>
<p>Let’s try to put these new tools to practice and estimate the
proportion of Neanderthal ancestry in Africans and Europeans in our
simulated data. We can do this using the Patterson’s <span class="math inline">\(f_4\)</span>-ratio statistic implemented in the
<code><a href="../reference/ts_f4ratio.html">ts_f4ratio()</a></code> function in <em>slendr</em> (you can find more
information about this particular version of the statistic in Petr
<em>et al.</em>, PNAS 2019):</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first get a table of simulated African and European individuals in the tree-sequence</span></span>
<span><span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_samples.html">ts_samples</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pop</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR"</span>, <span class="st">"EUR"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># estimate the amounts of Neanderthal ancestry in these individuals and add</span></span>
<span><span class="co"># these values to the table</span></span>
<span><span class="va">inds</span><span class="op">$</span><span class="va">ancestry</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">ts</span>, X <span class="op">=</span> <span class="va">inds</span><span class="op">$</span><span class="va">name</span>, <span class="st">"NEA_1"</span>, <span class="st">"NEA_2"</span>, <span class="st">"AFR_1"</span>, <span class="st">"CH_1"</span><span class="op">)</span><span class="op">$</span><span class="va">alpha</span></span></code></pre></div>
<p>If we now summarise the inferred Neanderthal distribution in both
populations, we see that there is no Neanderthal ancestry in Africans
(as expected by our model–Africans did not receive a Neanderthal
introgression pulse) but there is a small proportion of Neanderthal
ancestry in Europeans (consistent with the 3% introgression pulse we
simulated between):</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">inds</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">pop</span>, <span class="va">ancestry</span>, fill <span class="op">=</span> <span class="va">pop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Neanderthal ancestry proportion"</span>, x <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-tree-sequences_files/figure-html/ts_nea_distributions-1.png" width="480"></p>
<p>This is exactly as we specified in the model configuration above,
suggesting that our simulations work as they should. You can see that
there is quite a bit of noise but that’s because we simulated only a
small amount of sequence.</p>
<p>We can also plot the trajectory of Neanderthal ancestry in Europe
during the time-window for which we have simulated ancient and
present-day DNA samples:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">inds</span>, <span class="va">pop</span> <span class="op">==</span> <span class="st">"EUR"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">ancestry</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"red"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">40000</span>, <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html" class="external-link">coord_cartesian</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"time [years ago]"</span>, y <span class="op">=</span> <span class="st">"Neanderthal ancestry proportion"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<p><img src="vignette-05-tree-sequences_files/figure-html/ts_nea_trajectory-1.png" width="480"></p>
<p>Again, this is a result consistent with empirical estimates of
Neanderthal ancestry using ancient DNA data (see Petr <em>et al.</em>,
PNAS 2019).</p>
</div>
<div class="section level2">
<h2 id="admixtools-analyses">ADMIXTOOLS analyses<a class="anchor" aria-label="anchor" href="#admixtools-analyses"></a>
</h2>
<p>In case you would like to verify some <em>f</em>-statistics results
using the venerable <a href="https://doi.org/10.1534/genetics.112.145037" class="external-link">ADMIXTOOLS</a>
software (see the linked paper which formally introduced these
statistics in the first place), you can convert the tree-sequence data
to a file format called EIGENSTRAT using the
<code><a href="../reference/ts_eigenstrat.html">ts_eigenstrat()</a></code> function. The file conversion is internally
handled by the R package <a href="https://www.bodkan.net/admixr/" class="external-link"><em>admixr</em></a> and returns an
<code>EIGENSTRAT</code> object which ties all individual
<code>EIGENSTRAT</code> file components together (see the <a href="https://www.bodkan.net/admixr/articles/tutorial.html" class="external-link">tutorial</a>
to <em>admixr</em> for an extensive overview). <em>admixr</em> is an R
package for running automated ADMIXTOOLS analyses entirely from R and
makes these types of analyses very convenient.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_eigenstrat.html">ts_eigenstrat</a></span><span class="op">(</span><span class="va">ts</span>, prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"eigenstrat"</span>, <span class="st">"data"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; 1287 multiallelic sites (0.207% out of 620386 total) detected and removed</span></span></code></pre>
<p>Running an <em>admixr</em> analysis is then as easy as plugging the
object into an <em>admixr</em> function. For instance, we can estimate
the proportion of Neanderthal ancestry in a couple of individuals <span class="math inline">\(X\)</span> like this (<em>admixr</em> calls this
proportion <code>alpha</code>):</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/admixr" class="external-link">admixr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/admixr/man/f4ratio.html" class="external-link">f4ratio</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">snps</span>, X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span>, <span class="st">"AFR_2"</span><span class="op">)</span>,</span>
<span>        A <span class="op">=</span> <span class="st">"NEA_1"</span>, B <span class="op">=</span> <span class="st">"NEA_2"</span>, C <span class="op">=</span> <span class="st">"AFR_1"</span>, O <span class="op">=</span> <span class="st">"CH_1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">3</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">12</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> " "</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (8): X1, X2, X3, X4, X5, X6, X7, X8</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (4): X9, X10, X11, X12</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 8</span></span></span>
<span><span class="co">#&gt;   A     B     X     C     O         alpha  stderr Zscore</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> NEA_1 NEA_2 EUR_1 AFR_1 CH_1   0.017<span style="text-decoration: underline;">7</span>   0.005<span style="text-decoration: underline;">67</span>   3.13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> NEA_1 NEA_2 EUR_2 AFR_1 CH_1   0.028<span style="text-decoration: underline;">6</span>   0.009<span style="text-decoration: underline;">67</span>   2.96</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> NEA_1 NEA_2 AFR_2 AFR_1 CH_1  -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">606</span> 0.002<span style="text-decoration: underline;">43</span>  -<span style="color: #BB0000;">0.25</span></span></span></code></pre>
<p>In fact, lets compare the values obtained by both <em>tskit</em> and
<em>admixr</em>/ADMIXTOOLS for all individuals:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">europeans</span> <span class="op">&lt;-</span> <span class="va">inds</span><span class="op">[</span><span class="va">inds</span><span class="op">$</span><span class="va">pop</span> <span class="op">==</span> <span class="st">"EUR"</span>, <span class="op">]</span><span class="op">$</span><span class="va">name</span></span>
<span></span>
<span><span class="co"># tskit result</span></span>
<span><span class="va">result_ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">ts</span>, X <span class="op">=</span> <span class="va">europeans</span>, A <span class="op">=</span> <span class="st">"NEA_1"</span>, B <span class="op">=</span> <span class="st">"NEA_2"</span>, C <span class="op">=</span> <span class="st">"AFR_1"</span>, O <span class="op">=</span> <span class="st">"CH_1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>alpha_ts <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># result obtained by admixr/ADMIXTOOLS</span></span>
<span><span class="va">result_admixr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/admixr/man/f4ratio.html" class="external-link">f4ratio</a></span><span class="op">(</span><span class="va">snps</span>, X <span class="op">=</span> <span class="va">europeans</span>, A <span class="op">=</span> <span class="st">"NEA_1"</span>, B <span class="op">=</span> <span class="st">"NEA_2"</span>, C <span class="op">=</span> <span class="st">"AFR_1"</span>, O <span class="op">=</span> <span class="st">"CH_1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>alpha_admixr <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">50</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">12</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> " "</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (8): X1, X2, X3, X4, X5, X6, X7, X8</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (4): X9, X10, X11, X12</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="va">result_admixr</span>, <span class="va">result_ts</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">alpha_ts</span>, <span class="va">alpha_admixr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"red"</span>, linewidth <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"f4-ratio statistic calculated with admixr/ADMIXTOOLS"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"f4-ratio statistic calculated with tskit"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-05-tree-sequences_files/figure-html/ts_vs_admixr-1.png" width="480"></p>
<p>The correspondence between the two looks good! 🎉 Again, note that
the large amount of variance around the expected value of 3% ancestry is
due to an extremely small amount of sequence data simulated here.</p>
</div>
<div class="section level2">
<h2 id="vcf-output">VCF output<a class="anchor" aria-label="anchor" href="#vcf-output"></a>
</h2>
<p>In case you need to process simulated data in some other software,
you can use the function <code><a href="../reference/ts_vcf.html">ts_vcf()</a></code> to save the simulated
genotypes in a VCF format:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_vcf.html">ts_vcf</a></span><span class="op">(</span><span class="va">ts</span>, path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"output.vcf.gz"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>You can also specify only a subset of individuals to be saved in the
VCF:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_vcf.html">ts_vcf</a></span><span class="op">(</span><span class="va">ts</span>, path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"output_subset.vcf.gz"</span><span class="op">)</span>,</span>
<span>       individuals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CH_1"</span>, <span class="st">"NEA_1"</span>, <span class="st">"EUR_1"</span>, <span class="st">"AFR_1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="other-statistics">Other statistics<a class="anchor" aria-label="anchor" href="#other-statistics"></a>
</h2>
<p>What follows is a very brief overview of other statistics which are
implemented in <em>tskit</em> and for which <em>slendr</em> provides an
easy-to-use R interface. As you will see, the goal of these functions is
to get you to a result using a single function call, making them very
convenient for quick interactive exploratory analyses on the simulated
data right in the R console.</p>
<p>We will continue to use our simulated Neanderthal introgression
tree-sequence data for these examples.</p>
<div class="section level3">
<h3 id="f_st">
<span class="math inline">\(F_{st}\)</span><a class="anchor" aria-label="anchor" href="#f_st"></a>
</h3>
<p>The <span class="math inline">\(F_{st}\)</span> statistic is
implemented by the function <code><a href="../reference/ts_fst.html">ts_fst()</a></code>.</p>
<p>If a single genome-wide <span class="math inline">\(F_{st}\)</span>
is to be calculated (i.e. not a window-based calculation), the
<code><a href="../reference/ts_fst.html">ts_fst()</a></code> returns a simple three-column data frame</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_fst.html">ts_fst</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>afr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"AFR_3"</span><span class="op">)</span>, eur <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 3</span></span></span>
<span><span class="co">#&gt;   x     y        Fst</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> afr   eur   0.047<span style="text-decoration: underline;">5</span></span></span></code></pre>
<p>In case a non-named list of sample sets was provided, set names are
generated automatically:</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_fst.html">ts_fst</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"AFR_3"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 3</span></span></span>
<span><span class="co">#&gt;   x     y        Fst</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> set_1 set_2 0.047<span style="text-decoration: underline;">5</span></span></span></code></pre>
<p>Of course, this is much less readable and we encourage you to name
the sample sets appropriately.</p>
<p>In case more than two sample sets are specified, all pairwise
statistics are computed:</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_fst.html">ts_fst</a></span><span class="op">(</span><span class="va">ts</span>, sample_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>afr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"AFR_3"</span><span class="op">)</span>,</span>
<span>                              eur <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span><span class="op">)</span>,</span>
<span>                              nea <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NEA_1"</span>, <span class="st">"NEA_2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 3</span></span></span>
<span><span class="co">#&gt;   x     y        Fst</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> afr   eur   0.047<span style="text-decoration: underline;">5</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> afr   nea   0.553 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> eur   nea   0.536</span></span></code></pre>
<p>As with many other statistics implemented by <a href="https://tskit.dev/tskit/docs/stable/python-api.html" class="external-link"><em>tskit</em></a><em>,</em>
<code><a href="../reference/ts_fst.html">ts_fst()</a></code> accepts a <code>windows</code> argument,
specifying the breakpoints between windows. In this case, the
<code>Fst</code> column in the resulting data frame is a so called
“list-column”, with each item in the column being a vector of <span class="math inline">\(F_{st}\)</span> values, one per each window.
List-columns can be a little confusing for new R users, but we highly
encourage you to get used to them as they allow extremely concise and
elegant handling of structured data within normal data frames (you can
start with <a href="https://jennybc.github.io/purrr-tutorial/ls13_list-columns.html" class="external-link">this</a>
introduction).</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define breakpoints between 20 windows</span></span>
<span><span class="va">breakpoints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">ts</span><span class="op">$</span><span class="va">sequence_length</span>, length.out <span class="op">=</span> <span class="fl">21</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate window-based Fst statistic</span></span>
<span><span class="va">win_fst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_fst.html">ts_fst</a></span><span class="op">(</span></span>
<span>  <span class="va">ts</span>, windows <span class="op">=</span> <span class="va">breakpoints</span>,</span>
<span>  sample_sets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>afr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"AFR_3"</span><span class="op">)</span>,</span>
<span>                     eur <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span><span class="op">)</span>,</span>
<span>                     nea <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NEA_1"</span>, <span class="st">"NEA_2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we get 20 values for each parwise calculation</span></span>
<span><span class="va">win_fst</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 3</span></span></span>
<span><span class="co">#&gt;   x     y     Fst         </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;named list&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> afr   eur   <span style="color: #949494;">&lt;dbl [20]&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> afr   nea   <span style="color: #949494;">&lt;dbl [20]&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> eur   nea   <span style="color: #949494;">&lt;dbl [20]&gt;</span></span></span></code></pre>
<p>For instance, here are window-based <span class="math inline">\(F_st\)</span> values for the
<code>afr-vs-eur</code> calculation (first row of the table above):</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">win_fst</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">$</span><span class="va">Fst</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $`1`</span></span>
<span><span class="co">#&gt;  [1] 0.05295365 0.06322827 0.04193092 0.04470269 0.06213078 0.04110568</span></span>
<span><span class="co">#&gt;  [7] 0.04707814 0.05186687 0.02966762 0.02858609 0.03569889 0.04908546</span></span>
<span><span class="co">#&gt; [13] 0.06360736 0.06657169 0.03077708 0.03133374 0.05636180 0.05387645</span></span>
<span><span class="co">#&gt; [19] 0.04000369 0.05913374</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="tajimas-d">Tajima’s <span class="math inline">\(D\)</span><a class="anchor" aria-label="anchor" href="#tajimas-d"></a>
</h3>
<p>The function <code><a href="../reference/ts_tajima.html">ts_tajima()</a></code> has nearly the same interface
as <code><a href="../reference/ts_fst.html">ts_fst()</a></code> shown above.</p>
<p>If a non-window version is to be calculated, we get a single
genome-wide values for each sample set (named or non-named list of
character vectors with individual names):</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_tajima.html">ts_tajima</a></span><span class="op">(</span><span class="va">ts</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>afr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"AFR_3"</span><span class="op">)</span>, eur <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   set         D</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> afr   0.034<span style="text-decoration: underline;">7</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> eur   0.003<span style="text-decoration: underline;">01</span></span></span></code></pre>
<p>For window-based version, the function returns the <code>D</code>
column as a list column of vectors with <span class="math inline">\(i\)</span>-th element being the Tajima’s D value
for the <span class="math inline">\(i\)</span>-th window:</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_tajima.html">ts_tajima</a></span><span class="op">(</span><span class="va">ts</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>afr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span><span class="op">)</span>, eur <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span><span class="op">)</span><span class="op">)</span>, windows <span class="op">=</span> <span class="va">breakpoints</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   set   D           </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;named list&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> afr   <span style="color: #949494;">&lt;dbl [20]&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> eur   <span style="color: #949494;">&lt;dbl [20]&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="diversity">Diversity<a class="anchor" aria-label="anchor" href="#diversity"></a>
</h3>
<p>We can calculate diversity within given groups of individuals with
the function <code><a href="../reference/ts_diversity.html">ts_diversity()</a></code>. For instance, even in our
extremely simplified example, we would expect the highest levels of
diversity in Africans, followed by Europeans, Neanderthals and the
“degenerate” single individual outgroup “chimpanzee”. Is this true?
Let’s find out.</p>
<p>First we extract individuals from all populations, creating a list of
character vectors for each group (which is what functions such as
<code><a href="../reference/ts_diversity.html">ts_diversity()</a></code> expects as an input):</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get sampled individuals from all populations</span></span>
<span><span class="va">sample_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_samples.html">ts_samples</a></span><span class="op">(</span><span class="va">ts</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">.</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">pop</span><span class="op">)</span> <span class="va">pop</span><span class="op">$</span><span class="va">name</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_sets</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; $AFR</span></span>
<span><span class="co">#&gt; [1] "AFR_1" "AFR_2" "AFR_3" "AFR_4" "AFR_5"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $CH</span></span>
<span><span class="co">#&gt; [1] "CH_1"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $EUR</span></span>
<span><span class="co">#&gt;  [1] "EUR_1"  "EUR_2"  "EUR_3"  "EUR_4"  "EUR_5"  "EUR_6"  "EUR_7"  "EUR_8" </span></span>
<span><span class="co">#&gt;  [9] "EUR_9"  "EUR_10" "EUR_11" "EUR_12" "EUR_13" "EUR_14" "EUR_15" "EUR_16"</span></span>
<span><span class="co">#&gt; [17] "EUR_17" "EUR_18" "EUR_19" "EUR_20" "EUR_21" "EUR_22" "EUR_23" "EUR_24"</span></span>
<span><span class="co">#&gt; [25] "EUR_25" "EUR_27" "EUR_26" "EUR_28" "EUR_29" "EUR_30" "EUR_31" "EUR_32"</span></span>
<span><span class="co">#&gt; [33] "EUR_33" "EUR_34" "EUR_35" "EUR_36" "EUR_37" "EUR_38" "EUR_39" "EUR_40"</span></span>
<span><span class="co">#&gt; [41] "EUR_41" "EUR_42" "EUR_43" "EUR_44" "EUR_45" "EUR_46" "EUR_47" "EUR_48"</span></span>
<span><span class="co">#&gt; [49] "EUR_49" "EUR_50"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $NEA</span></span>
<span><span class="co">#&gt; [1] "NEA_1" "NEA_2"</span></span></code></pre>
<p>Now we can calculate diversity in each population and sort the
results in an increasing order of diversity:</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_diversity.html">ts_diversity</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">sample_sets</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">diversity</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 2</span></span></span>
<span><span class="co">#&gt;   set   diversity</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> CH    0.000<span style="text-decoration: underline;">042</span>8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> NEA   0.000<span style="text-decoration: underline;">045</span>6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> EUR   0.000<span style="text-decoration: underline;">395</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> AFR   0.000<span style="text-decoration: underline;">402</span></span></span></code></pre>
<p>Great! This matches our expectations. We simulated chimp “population”
as only one individual, so we expect essentially no diversity after
millions of years of evolution.</p>
</div>
<div class="section level3">
<h3 id="divergence">Divergence<a class="anchor" aria-label="anchor" href="#divergence"></a>
</h3>
<p>We can calculate pairwise divergence between groups of individuals
using the function <code><a href="../reference/ts_divergence.html">ts_divergence()</a></code>. Given our model, we
would expect the lowest divergence between the two modern human groups
<code>AFR</code> and <code>EUR</code>, then between Neanderthals and the
two modern humans, and all three groups (<code>AFR</code>,
<code>EUR</code> and <code>NEA</code>) should have equal, much deeper
divergence from the outgroup chimpanzee <code>CH</code>.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ts_divergence.html">ts_divergence</a></span><span class="op">(</span><span class="va">ts</span>, <span class="va">sample_sets</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">divergence</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   x     y     divergence</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> AFR   EUR     0.000<span style="text-decoration: underline;">450</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> EUR   NEA     0.000<span style="text-decoration: underline;">757</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> AFR   NEA     0.000<span style="text-decoration: underline;">779</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> CH    NEA     0.004<span style="text-decoration: underline;">00</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> CH    EUR     0.004<span style="text-decoration: underline;">01</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> AFR   CH      0.004<span style="text-decoration: underline;">02</span></span></span></code></pre>
<p>After sorting the table based on the value in the divergence column,
we can see the results fit our expectations.</p>
</div>
</div>
<div class="section level2">
<h2 id="more-information">More information<a class="anchor" aria-label="anchor" href="#more-information"></a>
</h2>
<p>These were only a couple of examples of statistical functions
implemented in <em>tskit</em> for which we provide a native R interface
in <em>slendr</em>. You can find more tree-sequence statistics in the <a href="../reference/index.html#section-tree-sequence-statistics">reference
manual</a> on the project website. Not <a href="https://tskit.dev/tskit/docs/stable/stats.html#available-statistics" class="external-link">all
statistics</a> from the <em>tskit</em> library are implemented, but we
intend to expand the selection provided by <em>slendr</em> in the near
future. If there is some functionality that you would like to use in
your project missing in <em>slendr</em>, please don’t hesitate to let us
now by creating an issue on our <a href="https://github.com/bodkan/slendr/issues" class="external-link">GitHub page</a>.</p>
<p>Finally, if you would like to see more examples of <em>tskit</em>
interface in action, take a look at the vignette which describes <a href="vignette-07-backends.html">switching between the SLiM and
<em>msprime</em> back ends</a> of the <em>slendr</em> package.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
