<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="slendr">
<title>Simulating data with SLiM and msprime backends • slendr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating data with SLiM and msprime backends">
<meta property="og:description" content="slendr">
<meta property="og:image" content="https://www.slendr.net/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">slendr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--articles">
    <span class="fa fa-book"></span>
     
    Articles
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--articles">
    <a class="dropdown-item" href="../articles/vignette-00-installation.html">Installation instructions</a>
    <a class="dropdown-item" href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    <a class="dropdown-item" href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    <a class="dropdown-item" href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    <a class="dropdown-item" href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    <a class="dropdown-item" href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a>
    <a class="dropdown-item" href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a>
    <a class="dropdown-item" href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a>
    <a class="dropdown-item" href="../articles/vignette-08-nonslendr-tskit.html">Analyzing non-slendr tree sequences</a>
    <a class="dropdown-item" href="../articles/vignette-09-paper.html">Examples from the slendr paper</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bodkan/slendr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/fleventy5">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulating data with SLiM and msprime backends</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/HEAD/vignettes/vignette-07-backends.Rmd" class="external-link"><code>vignettes/vignette-07-backends.Rmd</code></a></small>
      <div class="d-none name"><code>vignette-07-backends.Rmd</code></div>
    </div>

    
    
<p>Even though the most distinguishing feature of the <em>slendr</em> R
package is programming spatial population genetic models across real and
abstract geographic landscapes, having the ability to run “standard”,
non-spatial models directly from an R environment could be often quite
useful. This is why <em>slendr</em> supports also non-spatial SLiM
models, which can be executed by <em>slendr</em> simply by excluding the
map component from a SLiM simulation (briefly discussed in <a href="vignette-04-nonspatial-models.html">this vignette</a>).</p>
<p>In some cases, the forward nature of a SLiM simulation (meaning that
many individuals need to be tracked in each generation) makes the
simulations too slow. For instance, running an <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation" class="external-link">Approximate
Bayesian Computation (ABC)</a> analysis often requires running thousands
or even millions of simulations. In cases like this, the faster the
simulations the better.</p>
<p>For these situations, the <em>slendr</em> R package provides an
alternative back end for simulating non-spatial population models using
the Python coalescent simulation framework <a href="https://tskit.dev/msprime/docs/stable/intro.html" class="external-link"><em>msprime</em></a>.
Unlike forward time simulators, coalescent simulations only reconstruct
the genealogical history of a sample of individuals of interest
backwards in time, which makes it much more computational efficient.</p>
<p>The goal of this vignette is to demonstrate that a single
<em>slendr</em> model configuration can be run using the built-in SLiM
and msprime backend engines <em>without any changes</em>!</p>
<p>To start off, let’s load the <em>slendr</em> package together with
some standard R data analysis packages, and instruct <em>reticulate</em>
where to look for a Python environment with all necessary Python
dependencies (see <a href="vignette-05-tree-sequences.html#r-interface-for-tskit-and-pyslim-1">this</a>
section of another vignette for instructions how to setup the Python
environment the easiest way possible):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bodkan/slendr" class="external-link">slendr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/init_env.html">init_env</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seed</span> <span class="op">&lt;-</span> <span class="fl">42</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="va">seed</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="detecting-gene-flow-from-msprime-and-slim-tree-sequences">Detecting gene flow from msprime and SLiM tree sequences<a class="anchor" aria-label="anchor" href="#detecting-gene-flow-from-msprime-and-slim-tree-sequences"></a>
</h2>
<div class="section level3">
<h3 id="defining-a-model">Defining a model<a class="anchor" aria-label="anchor" href="#defining-a-model"></a>
</h3>
<p>We now define up a simple <em>non-spatial</em> model of gene flow
between populations (also known as population admixture or
introgression). This will involve essentially the same procedure which
we have shown in another vignette introducing <a href="vignette-04-nonspatial-models.html">non-spatial <em>slendr</em>
models</a>. Note that this is no different from how we would normally
specify a spatial model, except that we left out the <code>map</code>
argument in making the <code><a href="../reference/population.html">population()</a></code> calls.</p>
<p>To demonstrate additional features of the <em>slendr</em> interface
to the <a href="https://tskit.dev/tskit/docs/stable/introduction.html" class="external-link"><em>tskit</em></a>
Python library, we will perform gene flow detection test using the
so-called <span class="math inline">\(f_4\)</span> and <span class="math inline">\(f_4\)</span>-ratio statistics (very briefly
introduces in <a href="vignette-05-tree-sequences.html">this
vignette</a>). To make things a little more interesting, we will define
two population models: one model without any gene flow, and another
which includes gene flow. In defining this model and expressing the
<span class="math inline">\(f\)</span>-statistics we will use the same
nomenclature used in the first study by <a href="https://academic.oup.com/genetics/article/192/3/1065/5935193" class="external-link">Patterson
<em>et al.</em></a> that described <span class="math inline">\(f_4-ratio\)</span> statistic in the first place.
if you’re not familiar with these statistical tests of gene flow, we
recommend that you take a look at the relevant sections of the linked
paper. shiny_graph <img src="images/f4_tree.jpg" style="width:90.0%"></p>
<p>Let’s start by first defining the populations and their splits as
established in the figure above:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_len</span> <span class="op">&lt;-</span> <span class="fl">100e6</span> <span class="co"># amount of sequence to simulate</span></span>
<span><span class="va">rec_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span> <span class="co"># uniform recombination rate</span></span>
<span><span class="va">mut_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span> <span class="co"># mutation rate</span></span>
<span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"outgroup"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"c"</span>, time <span class="op">=</span> <span class="fl">2500</span>, N <span class="op">=</span> <span class="fl">100</span>, parent <span class="op">=</span> <span class="va">o</span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"a"</span>, time <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="fl">100</span>, parent <span class="op">=</span> <span class="va">c</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"b"</span>, time <span class="op">=</span> <span class="fl">3500</span>, N <span class="op">=</span> <span class="fl">100</span>, parent <span class="op">=</span> <span class="va">a</span><span class="op">)</span></span>
<span><span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"x1"</span>, time <span class="op">=</span> <span class="fl">3800</span>, N <span class="op">=</span> <span class="fl">5000</span>, parent <span class="op">=</span> <span class="va">c</span><span class="op">)</span></span>
<span><span class="va">x2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"x2"</span>, time <span class="op">=</span> <span class="fl">4000</span>, N <span class="op">=</span> <span class="fl">5000</span>, parent <span class="op">=</span> <span class="va">x1</span><span class="op">)</span></span></code></pre></div>
<p>Note that the Ne of the populations <code>x1</code> and
<code>x2</code> is set to be much higher than the rest. We set the <span class="math inline">\(N_e\)</span> of the other populations to lower
values to speed up the forward SLiM simulations (they won’t affect the
results we’re interested in anyway). Higher values of the <span class="math inline">\(N_e\)</span> of <code>x1</code> and
<code>x2</code> populations will ensure that the effect of the drift
acting on those two populations will be much smaller. Because we will
simulate and later measure the proportion of ancestry from the
population <code>b</code> to <code>x1</code>, this will ensure that the
ancestry proportion will not drift too far away from the expectation and
it will make the interesting patterns stand out more clearly.</p>
<p>(Of course, this is all done for demonstration purposes only and to
speed up the SLiM simulations by making the <span class="math inline">\(N_e\)</span> of the other populations smaller. In
practice, we would be running these kinds of simulations only using the
<em>msprime</em> back end.)</p>
</div>
</div>
<div class="section level2">
<h2 id="compiling-the-model-and-simulating-data">Compiling the model and simulating data<a class="anchor" aria-label="anchor" href="#compiling-the-model-and-simulating-data"></a>
</h2>
<p>We will now use those populations to compile two models: the first
model will be without any gene flow (left panel in the figure above),
the second model will include 10% gene flow from <code>b</code> to
<code>x1</code> (right panel in the figure above). We will also schedule
sampling of a couple of individuals at the very end of the simulation:
50 individuals each from populations <code>x1</code> and <code>x2</code>
to capture a bit more of the natural variation of <code>b</code>
ancestry in <code>x1</code>, and only one individual from the rest.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># no gene flow model</span></span>
<span><span class="va">model_nogf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span>populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="va">b</span>, <span class="va">x1</span>, <span class="va">x2</span>, <span class="va">c</span>, <span class="va">o</span><span class="op">)</span>, generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">4500</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span></span>
<span>  <span class="va">model_nogf</span>, times <span class="op">=</span> <span class="fl">4500</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">b</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x1</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x2</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">c</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">o</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># model with gene flow</span></span>
<span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_flow.html">gene_flow</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">b</span>, to <span class="op">=</span> <span class="va">x1</span>, start <span class="op">=</span> <span class="fl">4100</span>, end <span class="op">=</span> <span class="fl">4400</span>, rate <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">model_gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span>populations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="va">b</span>, <span class="va">x1</span>, <span class="va">x2</span>, <span class="va">c</span>, <span class="va">o</span><span class="op">)</span>, gene_flow <span class="op">=</span> <span class="va">gf</span>, generation_time <span class="op">=</span> <span class="fl">1</span>, simulation_length <span class="op">=</span> <span class="fl">4500</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span></span>
<span>  <span class="va">model_gf</span>, times <span class="op">=</span> <span class="fl">4500</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">b</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x1</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">x2</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">c</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">o</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Before running the simulations, let’s first make sure both models are
set up correctly:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_model.html">plot_model</a></span><span class="op">(</span><span class="va">model_nogf</span>, sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-07-backends_files/figure-html/unnamed-chunk-5-1.png" width="640"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_model.html">plot_model</a></span><span class="op">(</span><span class="va">model_gf</span>, sizes <span class="op">=</span> <span class="cn">FALSE</span>, proportions <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-07-backends_files/figure-html/unnamed-chunk-5-2.png" width="640"></p>
<p>Now we will run both models (with and without gene flow) in the two
<em>slendr</em> backends, SLiM and msprime:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># model without gene flow</span></span>
<span><span class="va">slim_nogf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model_nogf</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, samples <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span>
<span><span class="va">msprime_nogf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span><span class="va">model_nogf</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, samples <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># model with b -&gt; x1 gene flow</span></span>
<span><span class="va">slim_gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model_gf</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, samples <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span>
<span><span class="va">msprime_gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span><span class="va">model_gf</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, samples <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span></code></pre></div>
<p>Note that we are using exactly the same model configuration object in
both simulation runs! In fact, even the function interface looks nearly
exactly the same. It doesn’t matter what the specific details of your
demographic models are, <em>slendr</em> will interpret them correctly
regardless of which back end simulation engine you choose to use.</p>
<div class="section level3">
<h3 id="comparing-the-outputs-of-msprime-and-slim-runs-of-a-slendr-model">Comparing the outputs of msprime and SLiM runs of a <em>slendr</em>
model<a class="anchor" aria-label="anchor" href="#comparing-the-outputs-of-msprime-and-slim-runs-of-a-slendr-model"></a>
</h3>
<p>Having run our introgression models with the two simulation back
ends, let’s load both SLiM and msprime tree sequence outputs and compare
their contents (hoping they will be the same!).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SLiM outputs -- we can use built-in slendr functions for those</span></span>
<span><span class="va">slim_nogf</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">slim_nogf</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ts_recapitate.html">ts_recapitate</a></span><span class="op">(</span>Ne <span class="op">=</span> <span class="fl">10</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span><span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span>
<span></span>
<span><span class="va">slim_gf</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">slim_gf</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ts_recapitate.html">ts_recapitate</a></span><span class="op">(</span>Ne <span class="op">=</span> <span class="fl">10</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span><span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># msprime outputs (note that recapitation and simplification doesn't make</span></span>
<span><span class="co"># sense here because we already have fully coalesced genealogies for our</span></span>
<span><span class="co"># individuals  of interest</span></span>
<span><span class="va">msprime_nogf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span><span class="va">msprime_nogf</span>, <span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span>
<span><span class="va">msprime_gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span><span class="va">msprime_gf</span>, <span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span></code></pre></div>
<p>As we can see below, the tree sequence summary of the data loaded and
processed by <em>slendr</em>’s <code><a href="../reference/ts_load.html">ts_load()</a></code> function (data
produced by the SLiM backend) is very similar to what we got “manually”
for the msprime-produced tree sequence using the custom defined
functions. This is somehow obvious but it is a nice and cheap sanity
check indicating that the tree sequence data structure produced by the
two backends from the same demographic model is almost the same.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">slim_nogf</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔═════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence             ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │   248158║</span></span>
<span><span class="co">#&gt; ╟───────────────┼─────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│100000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼─────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │    ticks║</span></span>
<span><span class="co">#&gt; ╟───────────────┼─────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │    21008║</span></span>
<span><span class="co">#&gt; ╟───────────────┼─────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │ 70.6 MiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤══════╤════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows  │Size    │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪══════╪════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │972815│29.7 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│131750│12.6 MiB│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │     0│ 8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │257348│ 9.1 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │157230│ 5.7 MiB│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│     7│ 2.8 KiB│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│     3│37.9 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │257040│ 6.1 MiB│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧══════╧════════╧════════════╝</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msprime_nogf</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; ╔═══════════════════════════╗</span></span>
<span><span class="co">#&gt; ║TreeSequence               ║</span></span>
<span><span class="co">#&gt; ╠═══════════════╤═══════════╣</span></span>
<span><span class="co">#&gt; ║Trees          │      62562║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sequence Length│  100000000║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Time Units     │generations║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Sample Nodes   │        208║</span></span>
<span><span class="co">#&gt; ╟───────────────┼───────────╢</span></span>
<span><span class="co">#&gt; ║Total Size     │   14.4 MiB║</span></span>
<span><span class="co">#&gt; ╚═══════════════╧═══════════╝</span></span>
<span><span class="co">#&gt; ╔═══════════╤══════╤═════════╤════════════╗</span></span>
<span><span class="co">#&gt; ║Table      │Rows  │Size     │Has Metadata║</span></span>
<span><span class="co">#&gt; ╠═══════════╪══════╪═════════╪════════════╣</span></span>
<span><span class="co">#&gt; ║Edges      │246660│  7.5 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Individuals│   104│  2.9 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Migrations │     0│  8 Bytes│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Mutations  │ 69284│  2.4 MiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Nodes      │ 32645│892.6 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Populations│     6│414 Bytes│         Yes║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Provenances│     2│  3.8 KiB│          No║</span></span>
<span><span class="co">#&gt; ╟───────────┼──────┼─────────┼────────────╢</span></span>
<span><span class="co">#&gt; ║Sites      │ 69263│  1.7 MiB│          No║</span></span>
<span><span class="co">#&gt; ╚═══════════╧══════╧═════════╧════════════╝</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="comparing-results-of-population-genetics-statistics">Comparing results of population genetics statistics<a class="anchor" aria-label="anchor" href="#comparing-results-of-population-genetics-statistics"></a>
</h3>
<p>Having loaded both msprime and SLiM tree sequences generated by the
same model, let’s see if we can get comparable results when we calculate
some statistics of interest.</p>
<p>Let’s first extract all individuals from populations <code>x1</code>
and <code>x2</code> and estimate the values of <span class="math inline">\(f_4(c, x1 \textrm{ or } x2; b, o)\)</span> (which
should be significantly negative for individuals in the population
<code>x1</code> due to it’s ancestry coming from <code>b</code> in the
introgression models but consistent with zero for <code>x2</code>) and
<span class="math inline">\(f_4\textrm{-ratio}(a, b, \textrm{ or } x2;
c, o)\)</span>, which estimates the <em>proportion</em> of
<code>b</code>-like ancestry in <code>x1</code> and <code>x2</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract vector of names of the "test individuals" in populations `x1` and `x2`</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_samples.html">ts_samples</a></span><span class="op">(</span><span class="va">slim_gf</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pop</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span></span>
<span><span class="va">X</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt;   [1] "x1_1"  "x1_2"  "x1_3"  "x1_4"  "x1_5"  "x1_6"  "x1_7"  "x1_8"  "x1_9" </span></span>
<span><span class="co">#&gt;  [10] "x1_10" "x1_11" "x1_12" "x1_13" "x1_14" "x1_15" "x1_16" "x1_17" "x1_18"</span></span>
<span><span class="co">#&gt;  [19] "x1_19" "x1_20" "x1_21" "x1_22" "x1_23" "x1_24" "x1_25" "x1_26" "x1_27"</span></span>
<span><span class="co">#&gt;  [28] "x1_28" "x1_29" "x1_30" "x1_31" "x1_32" "x1_33" "x1_34" "x1_35" "x1_36"</span></span>
<span><span class="co">#&gt;  [37] "x1_37" "x1_38" "x1_39" "x1_40" "x1_41" "x1_42" "x1_43" "x1_44" "x1_45"</span></span>
<span><span class="co">#&gt;  [46] "x1_46" "x1_47" "x1_48" "x1_49" "x1_50" "x2_1"  "x2_2"  "x2_3"  "x2_4" </span></span>
<span><span class="co">#&gt;  [55] "x2_5"  "x2_6"  "x2_7"  "x2_8"  "x2_9"  "x2_10" "x2_11" "x2_12" "x2_13"</span></span>
<span><span class="co">#&gt;  [64] "x2_14" "x2_15" "x2_16" "x2_17" "x2_18" "x2_19" "x2_20" "x2_21" "x2_22"</span></span>
<span><span class="co">#&gt;  [73] "x2_23" "x2_24" "x2_25" "x2_26" "x2_27" "x2_28" "x2_29" "x2_30" "x2_31"</span></span>
<span><span class="co">#&gt;  [82] "x2_32" "x2_33" "x2_34" "x2_35" "x2_36" "x2_37" "x2_38" "x2_39" "x2_40"</span></span>
<span><span class="co">#&gt;  [91] "x2_41" "x2_42" "x2_43" "x2_44" "x2_45" "x2_46" "x2_47" "x2_48" "x2_49"</span></span>
<span><span class="co">#&gt; [100] "x2_50"</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate f4-statistics on individuals of `x1` and `x2` populations using data</span></span>
<span><span class="co"># from the two models (a model with no gene flow and a gene flow model) -- we use</span></span>
<span><span class="co"># map_dfr to iterate across all individuals from `X_individuals` and binding all</span></span>
<span><span class="co"># resulting data frames into a single data frame</span></span>
<span><span class="va">df_slim_f4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_dfr</a></span><span class="op">(</span><span class="va">X</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4</a></span><span class="op">(</span><span class="va">slim_nogf</span>, <span class="st">"c_1"</span>, <span class="va">.x</span>, <span class="st">"b_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"no gene flow"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_dfr</a></span><span class="op">(</span><span class="va">X</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4</a></span><span class="op">(</span><span class="va">slim_gf</span>, <span class="st">"c_1"</span>, <span class="va">.x</span>, <span class="st">"b_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"gene flow"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">f4</span>, <span class="va">model</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"SLiM backend"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute the proportions of `b` ancestry in `x1` (expected 10%) and `x2`</span></span>
<span><span class="co"># (expected 0% because this population did not receive any gene flow from `b`)</span></span>
<span><span class="va">df_slim_f4ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">slim_nogf</span>, <span class="va">X</span>, <span class="st">"a_1"</span>, <span class="st">"b_1"</span>, <span class="st">"c_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"no gene flow"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">slim_gf</span>, <span class="va">X</span>, <span class="st">"a_1"</span>, <span class="st">"b_1"</span>, <span class="st">"c_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"gene flow"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">alpha</span>, <span class="va">model</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"SLiM backend"</span><span class="op">)</span></span></code></pre></div>
<p>Now, let’s repeat the same analysis on tree sequences produced by the
<em>msprime</em> backend from the same two <em>slendr</em> demographic
models:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_msprime_f4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_dfr</a></span><span class="op">(</span><span class="va">X</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4</a></span><span class="op">(</span><span class="va">msprime_nogf</span>, <span class="st">"c_1"</span>, <span class="va">.x</span>, <span class="st">"b_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"no gene flow"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_dfr</a></span><span class="op">(</span><span class="va">X</span>, <span class="op">~</span> <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4</a></span><span class="op">(</span><span class="va">msprime_gf</span>, <span class="st">"c_1"</span>, <span class="va">.x</span>, <span class="st">"b_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"gene flow"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">f4</span>, <span class="va">model</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"msprime backend"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute the proportions of `b` ancestry in `x1` (expected 10%) and `x2`</span></span>
<span><span class="co"># (expected 0% because this population did not receive any gene flow from `b`)</span></span>
<span><span class="va">df_msprime_f4ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">msprime_nogf</span>, <span class="va">X</span>, <span class="st">"a_1"</span>, <span class="st">"b_1"</span>, <span class="st">"c_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"no gene flow"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ts_f4ratio.html">ts_f4ratio</a></span><span class="op">(</span><span class="va">msprime_gf</span>, <span class="va">X</span>, <span class="st">"a_1"</span>, <span class="st">"b_1"</span>, <span class="st">"c_1"</span>, <span class="st">"outgroup_1"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="st">"gene flow"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">alpha</span>, <span class="va">model</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"msprime backend"</span><span class="op">)</span></span></code></pre></div>
<p>Finally we can proceed with plotting both SLiM and msprime backend
results of the two models and compare them together. As noted above,
despite the fact that SLiM is a forward simulator and msprime is a
coalescent simulator, population genetic expectations from both the gene
flow and no gene flow models should match very closely (barring a level
of uncertainty expected due to randomness of the population genetic
processes at play).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_f4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">df_slim_f4</span>, <span class="va">df_msprime_f4</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"x1_"</span>, <span class="va">X</span><span class="op">)</span>,</span>
<span>                             <span class="st">"x1 (received gene flow)"</span>,</span>
<span>                             <span class="st">"x2 (no gene flow)"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_f4</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">f4</span>, fill <span class="op">=</span> <span class="va">population</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">simulator</span> <span class="op">~</span> <span class="va">model</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"number of individuals"</span>, x <span class="op">=</span> <span class="st">"f4 statistic"</span>,</span>
<span>       title <span class="op">=</span> <span class="st">"f4(c, x1 or x2; b, outgroup)"</span>,</span>
<span>       subtitle <span class="op">=</span> <span class="st">"f4 ~0 is consistent with no gene flow, negative value indicates gene flow with 'b'"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-07-backends_files/figure-html/msprime_slim_f4_distributions-1.png" width="640"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_f4ratio</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">df_slim_f4ratio</span>, <span class="va">df_msprime_f4ratio</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>population <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"x1_"</span>, <span class="va">X</span><span class="op">)</span>,</span>
<span>                             <span class="st">"x1 (received gene flow)"</span>,</span>
<span>                             <span class="st">"x2 (no gene flow)"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_f4ratio</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">alpha</span>, fill <span class="op">=</span> <span class="va">population</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">simulator</span> <span class="op">~</span> <span class="va">model</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.1</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"number of individuals"</span>, x <span class="op">=</span> <span class="st">"ancestry proportion (f4-ratio statistic)"</span>,</span>
<span>       title <span class="op">=</span> <span class="st">"f4-ratio estimate of 'b' ancestry calculated from simulated data"</span>,</span>
<span>       subtitle <span class="op">=</span> <span class="st">"f4-ratio = f4(a, outgroup; x1 or x2, c) / f4(a, outgroup; b, c)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-07-backends_files/figure-html/msprime_slim_f4ratio_distributions-1.png" width="640"></p>
<p>Things seem to be looking good and behave as expected! A couple of
observations:</p>
<ul>
<li>In the model without gene flow, the <span class="math inline">\(f_4\)</span> statistic value is zero for both
<code>x1</code> and <code>x2</code> populations. This is expected,
because <span class="math inline">\(f_4\)</span> ~0 is consistent with
the relationship between lineages conforming to a standard phylogenetic
tree (no admixture edges). Essentially, <span class="math inline">\(f_4\)</span> is a hypothesis test of “tree-ness”
of the data.</li>
<li>In the model which includes gene flow from <code>b</code> to
<code>x1</code>, the <span class="math inline">\(f_4\)</span> value is
significantly negative. This means that a simple tree hypothesis can be
rejected and that gene flow must have occurred between <code>b</code>
and <code>x1</code>. Note that this test does not reveal the
<em>direction</em> of gene flow, only it’s presence.</li>
<li>Consistently with the <span class="math inline">\(f_4\)</span>
results, the <span class="math inline">\(f_4-ratio\)</span> estimates
match our expectations. Specifically, the statistic estimates around 10%
ancestry of <code>b</code> in <code>x1</code> in the gene flow model and
0% ancestry in all other cases.</li>
<li>Most importantly, the results obtained from the SLiM and msprime
back ends are nearly exactly the same. This means that (for non-spatial
models) you can use SLiM and <em>msprime</em> backend
interchangeably.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="computing-allele-frequency-spectra">Computing allele frequency spectra<a class="anchor" aria-label="anchor" href="#computing-allele-frequency-spectra"></a>
</h2>
<p>Let’s look at one more example. Imagine the following five
demographic models describing the changes in the population size of some
abstract population:</p>
<p><img src="images/demographic_models.jpg" style="width:90.0%"></p>
<p>These models and the population size change events could be,
depending on the specifics of our project, defined using different time
units. If you work with radiocarbon-dated samples and ancient DNA you
might be used to thinking about population history in units of “years
before present”—in this case, the “present” time (bottom of the figure)
would be at time 0 and earlier events would be specified as “5000 years
before present”, etc.</p>
<p>On the other hand, if you were interested in modelling some
theoretical population, you might not want to concern yourselves with
this and would be quite happy starting our simulation at “generation 1”
and continue until some specified “generation X”.</p>
<p>The <em>slendr</em> package makes all of this extremely easy, because
you can specify times in whatever direction and units you want, as long
as they are kept consistent for all populations and events in your
model. This is a rather unique feature among popgen simulation software
as most programs either make you express your model in “generations in
forward direction”, or you have to explicitly convert time in
generations going backwards.</p>
<p>In the previous section of the vignette we have shown that SLiM and
msprime can be both used as simulation backend. In the remainder of this
vignette we will present additional analysis, demonstrating the fact
that you can express time of your models in a forward or backward
direction and still run those models in <em>slendr</em>’s SLiM (forward
simulator) or <em>msprime</em> (backward simulator).</p>
<p>Specifically, we will define all five demographic models from the
figure above twice, in different orientations of time – forward or
backward. We will then compute allele frequency spectrum from each
simulation (five models, each forward or backward) and verify that both
forward and backward pairs give the same result regardless of whether we
use the SLiM back end or the <em>msprime</em> backend.</p>
<div class="section level3">
<h3 id="model-definition">Model definition<a class="anchor" aria-label="anchor" href="#model-definition"></a>
</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="va">N_factor</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co"># by what factor should Ne change</span></span>
<span></span>
<span><span class="va">seq_len</span> <span class="op">&lt;-</span> <span class="fl">50e6</span></span>
<span><span class="va">rec_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span></span>
<span><span class="va">mut_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-8</span></span></code></pre></div>
<p>First, let’s define the five population histories in a forward time
direction:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># constant Ne model</span></span>
<span><span class="va">forward_const</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"const"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># decreasing step Ne model</span></span>
<span><span class="va">forward_decr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"decr"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span>, map <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">/</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># increasing step Ne model</span></span>
<span><span class="va">forward_incr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"inc"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">*</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># exponential increase in size</span></span>
<span><span class="va">forward_exp_incr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"exp_inc"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">2000</span>, end <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">*</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># exponential decrease in size</span></span>
<span><span class="va">forward_exp_decr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"exp_decr"</span>, time <span class="op">=</span> <span class="fl">1</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">2000</span>, end <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">/</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span></span></code></pre></div>
<p>Now let’s flip the flow of time around and define the population
histories in units of “time ago”:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># constant Ne model</span></span>
<span><span class="va">backward_const</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"const"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># decreasing step Ne model</span></span>
<span><span class="va">backward_decr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"decr"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">/</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># increasing step Ne model</span></span>
<span><span class="va">backward_incr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"inc"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">3000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">*</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"step"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># exponential increase in size</span></span>
<span><span class="va">backward_exp_incr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"exp_inc"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">3000</span>, end <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">*</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># exponential decrease in size</span></span>
<span><span class="va">backward_exp_decr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/population.html">population</a></span><span class="op">(</span><span class="st">"exp_decr"</span>, time <span class="op">=</span> <span class="fl">5000</span>, N <span class="op">=</span> <span class="va">N</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/resize.html">resize</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fl">3000</span>, end <span class="op">=</span> <span class="fl">2000</span>, N <span class="op">=</span> <span class="va">N</span> <span class="op">/</span> <span class="va">N_factor</span>, how <span class="op">=</span> <span class="st">"exponential"</span><span class="op">)</span></span></code></pre></div>
<p>For convenience, we will write a helper function which will, for a
given population history object (ten objects just above), compile a
<em>slendr</em> model, run it in both SLiM and <em>msprime</em> back
ends, load the tree sequence, and compute allele frequency spectrum
(AFS) from it. This will save us lots of code repetition (always a good
thing), minimize the chance of copy-pasting bugs (definitely a good
thing!) and compare the AFS patterns across different models and the two
<em>slendr</em> simulation back ends (they should be the same according
to population genetics theory).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compile_run_afs</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">model_name</span>, <span class="va">pop</span>, <span class="va">seed</span> <span class="op">=</span> <span class="fl">42</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># maximum length of the simulation (necessary for forward models which start</span></span>
<span>  <span class="co"># in generation 1)</span></span>
<span>  <span class="va">simulation_length</span> <span class="op">&lt;-</span> <span class="fl">5000</span></span>
<span></span>
<span>  <span class="co"># define sampling times given the direction of time</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">pop</span>, <span class="st">"history"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sampling_time</span> <span class="op">&lt;-</span> <span class="va">simulation_length</span></span>
<span>    <span class="va">direction</span> <span class="op">&lt;-</span> <span class="st">"forward"</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">sampling_time</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>    <span class="va">direction</span> <span class="op">&lt;-</span> <span class="st">"backward"</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="co"># compile model</span></span>
<span>  <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compile_model.html">compile_model</a></span><span class="op">(</span><span class="va">pop</span>, generation_time <span class="op">=</span> <span class="fl">15</span>, direction <span class="op">=</span> <span class="va">direction</span>, simulation_length <span class="op">=</span> <span class="va">simulation_length</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/schedule_sampling.html">schedule_sampling</a></span><span class="op">(</span><span class="va">model</span>, times <span class="op">=</span> <span class="va">sampling_time</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">pop</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># run the model in SLiM</span></span>
<span>  <span class="va">ts_slim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/slim.html">slim</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>,</span>
<span>                  samples <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># run the same model in msprim</span></span>
<span>  <span class="va">ts_msprime</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/msprime.html">msprime</a></span><span class="op">(</span><span class="va">model</span>, sequence_length <span class="op">=</span> <span class="va">seq_len</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>,</span>
<span>                        samples <span class="op">=</span> <span class="va">samples</span>, random_seed <span class="op">=</span> <span class="va">seed</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># load the SLiM tree sequence</span></span>
<span>  <span class="va">ts_slim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_recapitate.html">ts_recapitate</a></span><span class="op">(</span><span class="va">ts_slim</span>, Ne <span class="op">=</span> <span class="va">N</span>, recombination_rate <span class="op">=</span> <span class="va">rec_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span><span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># load the msprime tree sequence</span></span>
<span>  <span class="va">ts_msprime</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_mutate.html">ts_mutate</a></span><span class="op">(</span><span class="va">ts_msprime</span>, <span class="va">mut_rate</span>, random_seed <span class="op">=</span> <span class="va">seed</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># compute the AFS from the SLiM and msprime tree sequences and bind the</span></span>
<span>  <span class="co"># results (derived allele counts per frequency bin) in a data frame</span></span>
<span>  <span class="va">msprime_afs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_afs.html">ts_afs</a></span><span class="op">(</span><span class="va">ts_msprime</span>, polarised <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="va">slim_afs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ts_afs.html">ts_afs</a></span><span class="op">(</span><span class="va">ts_slim</span>, polarised <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"msprime"</span>, model <span class="op">=</span> <span class="va">model_name</span>, f <span class="op">=</span> <span class="va">msprime_afs</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>simulator <span class="op">=</span> <span class="st">"SLiM"</span>, model <span class="op">=</span> <span class="va">model_name</span>, f <span class="op">=</span> <span class="va">slim_afs</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">simulator</span>, <span class="va">model</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, direction <span class="op">=</span> <span class="va">direction</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now that this function doesn’t do anything special. It simply wraps a
standard <em>slendr</em> simulation pipeline (1. create population(s);
2. compile a model object; 3. simulate the model; 4. load the results;
5. analyse the data) into a single function.</p>
<p>Now that we have the whole pipeline, we can generate all of our
simulated data and bind all the result into a single data frame:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">afs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span></span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"constant"</span>, <span class="va">forward_const</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"constant"</span>, <span class="va">backward_const</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"step contraction"</span>, <span class="va">forward_decr</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"step contraction"</span>, <span class="va">backward_decr</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"step increase"</span>, <span class="va">forward_incr</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"step increase"</span>, <span class="va">backward_incr</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"exponential decrease"</span>, <span class="va">forward_exp_decr</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"exponential decrease"</span>, <span class="va">backward_exp_decr</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"exponential increase"</span>, <span class="va">forward_exp_incr</span><span class="op">)</span>,</span>
<span>  <span class="fu">compile_run_afs</span><span class="op">(</span><span class="st">"exponential increase"</span>, <span class="va">backward_exp_incr</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>model <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>    <span class="va">model</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"step contraction"</span>, <span class="st">"constant"</span>, <span class="st">"step increase"</span>,</span>
<span>                      <span class="st">"exponential decrease"</span>, <span class="st">"exponential increase"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Finally, we can plot the allele frequency spectra from all models,
for the two time directions, and for the two simulation back ends:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">afs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">n</span>, <span class="va">f</span>, color <span class="op">=</span> <span class="va">direction</span>, linetype <span class="op">=</span> <span class="va">simulator</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">model</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"number of derived alleles"</span>, y <span class="op">=</span> <span class="st">"frequency"</span>,</span>
<span>       title <span class="op">=</span> <span class="st">"Site frequency spectra obtained from five demographic models"</span>,</span>
<span>       subtitle <span class="op">=</span> <span class="st">"Each model was specified in forward or backward direction of time and executed by</span></span>
<span><span class="st">two different backend scripts in slendr (SLiM and msprime)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span><span class="st">"direction of\ntime in slendr"</span><span class="op">)</span>,</span>
<span>         linetype <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span><span class="st">"slendr backend\nengine used"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_linetype_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">100</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vignette-07-backends_files/figure-html/msprime_slim_afs-1.png" width="640"></p>
<p>Again, a couple of things to note:</p>
<ul>
<li><p>The two models that involve population contraction (be it in a
single step or an exponential collapse), we see a decrease in low
frequency variants. On the other hand, models with population expansion
show a high proportion of low frequency variants. The model with
constant population size shows intermediate pattern. This is a nice
validation that the simulation works as expected as it is exactly what
is predicted by population genetics theory.</p></li>
<li><p>Both forward and backward time models give exactly the same
result. This is not unexpected as anything else would be a bug in
<em>slendr</em>. However, although unsurprising, we hope that it makes
it clear that you are free to use whatever time unit specification you
want to encode <em>slendr</em> models.</p></li>
<li><p>For non-spatial models, results obtained from <em>msprime</em>
and SLiM back ends are consistent with one another. Again, this makes
sense because although they are completely different software,
population genetic theory governing the shape of allele frequency
spectra applies equally to both. In fact, the same example as we have
just implemented is an important part of <em>slendr</em> unit test suite
which verifies that the simulations created with <em>slendr</em> are
correct.</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>This example shows that if you have a standard, non-spatial
demographic model and would like to implement and simulate it in R,
<em>slendr</em> gives you a way to do so efficiently using it’s
<em>msprime</em> back end in addition to the SLiM back end used
throughout most of the documentation.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
