<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ slendr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://www.slendr.net/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">slendr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-question-circle"></span>
     
    Reference
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-book"></span>
     
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a>
    </li>
    <li>
      <a href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a>
    </li>
    <li>
      <a href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a>
    </li>
    <li>
      <a href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a>
    </li>
    <li>
      <a href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a>
    </li>
    <li>
      <a href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a>
    </li>
    <li>
      <a href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/bodkan/slendr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://twitter.com/dr_bodkan" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="installation-and-configuration-of-python-dependencies">Installation and configuration of Python dependencies <a href="#installation-and-configuration-of-python-dependencies" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="setup_env.html">setup_env()</a></code> </p>
        </td>
        <td><p>Setup a dedicated Python virtual environment for slendr</p></td>
      </tr><tr><td>
          <p><code><a href="clear_env.html">clear_env()</a></code> </p>
        </td>
        <td><p>Remove the automatically created slendr Python environment</p></td>
      </tr><tr><td>
          <p><code><a href="check_env.html">check_env()</a></code> </p>
        </td>
        <td><p>Check that the active Python environment is correctly setup for slendr</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="model-components">Model components <a href="#model-components" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="population.html">population()</a></code> </p>
        </td>
        <td><p>Define a population</p></td>
      </tr><tr><td>
          <p><code><a href="world.html">world()</a></code> </p>
        </td>
        <td><p>Define a world map for all spatial operations</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="spatial-population-dynamics">Spatial population dynamics <a href="#spatial-population-dynamics" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="move.html">move()</a></code> </p>
        </td>
        <td><p>Move the population to a new location in a given amount of time</p></td>
      </tr><tr><td>
          <p><code><a href="expand.html">expand()</a></code> </p>
        </td>
        <td><p>Expand the population range</p></td>
      </tr><tr><td>
          <p><code><a href="shrink.html">shrink()</a></code> </p>
        </td>
        <td><p>Shrink the population range</p></td>
      </tr><tr><td>
          <p><code><a href="boundary.html">boundary()</a></code> </p>
        </td>
        <td><p>Update the population range</p></td>
      </tr><tr><td>
          <p><code><a href="dispersal.html">dispersal()</a></code> </p>
        </td>
        <td><p>Change dispersal parameters</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="non-spatial-population-dynamics">Non-spatial population dynamics <a href="#non-spatial-population-dynamics" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="geneflow.html">geneflow()</a></code> </p>
        </td>
        <td><p>Define a gene flow event between two populations</p></td>
      </tr><tr><td>
          <p><code><a href="resize.html">resize()</a></code> </p>
        </td>
        <td><p>Change the population size</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="compiling-and-running-population-genetic-models">Compiling and running population genetic models <a href="#compiling-and-running-population-genetic-models" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="slim.html">slim()</a></code> </p>
        </td>
        <td><p>Run a slendr model in SLiM</p></td>
      </tr><tr><td>
          <p><code><a href="msprime.html">msprime()</a></code> </p>
        </td>
        <td><p>Run a slendr model in msprime</p></td>
      </tr><tr><td>
          <p><code><a href="compile.html">compile()</a></code> </p>
        </td>
        <td><p>Compile the spatial demographic model</p></td>
      </tr><tr><td>
          <p><code><a href="sampling.html">sampling()</a></code> </p>
        </td>
        <td><p>Define sampling events for a given set of populations</p></td>
      </tr><tr><td>
          <p><code><a href="read.html">read()</a></code> </p>
        </td>
        <td><p>Read a previously serialized model configuration</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="manipulation-of-spatial-objects">Manipulation of spatial objects <a href="#manipulation-of-spatial-objects" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="region.html">region()</a></code> </p>
        </td>
        <td><p>Define a geographic region</p></td>
      </tr><tr><td>
          <p><code><a href="join.html">join()</a></code> </p>
        </td>
        <td><p>Merge two spatial <code>slendr</code> objects into one</p></td>
      </tr><tr><td>
          <p><code><a href="overlap.html">overlap()</a></code> </p>
        </td>
        <td><p>Generate the overlap of two <code>slendr</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="subtract.html">subtract()</a></code> </p>
        </td>
        <td><p>Generate the difference between two <code>slendr</code> objects</p></td>
      </tr><tr><td>
          <p><code><a href="reproject.html">reproject()</a></code> </p>
        </td>
        <td><p>Reproject coordinates between coordinate systems</p></td>
      </tr><tr><td>
          <p><code><a href="distance.html">distance()</a></code> </p>
        </td>
        <td><p>Calculate the distance between a pair of spatial boundaries</p></td>
      </tr><tr><td>
          <p><code><a href="area.html">area()</a></code> </p>
        </td>
        <td><p>Calculate the area covered by the given slendr object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="model-visualization-and-diagnostics">Model visualization and diagnostics <a href="#model-visualization-and-diagnostics" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="plot.slendr.html">plot(<i>&lt;slendr&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot <code>slendr</code> geographic features on a map</p></td>
      </tr><tr><td>
          <p><code><a href="plot_graph.html">plot_graph()</a></code> </p>
        </td>
        <td><p>Plot geneflow graph based on given model configuration</p></td>
      </tr><tr><td>
          <p><code><a href="animate.html">animate()</a></code> </p>
        </td>
        <td><p>Animate the simulated population dynamics</p></td>
      </tr><tr><td>
          <p><code><a href="explore.html">explore()</a></code> </p>
        </td>
        <td><p>Open an interactive browser of the spatial model</p></td>
      </tr><tr><td>
          <p><code><a href="print.slendr_pop.html">print(<i>&lt;slendr_pop&gt;</i>)</a></code> <code><a href="print.slendr_pop.html">print(<i>&lt;slendr_region&gt;</i>)</a></code> <code><a href="print.slendr_pop.html">print(<i>&lt;slendr_map&gt;</i>)</a></code> <code><a href="print.slendr_pop.html">print(<i>&lt;slendr_model&gt;</i>)</a></code> <code><a href="print.slendr_pop.html">print(<i>&lt;slendr_tsdata&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print a short summary of a <code>slendr</code> object</p></td>
      </tr><tr><td>
          <p><code><a href="plot_ancestors.html">plot_ancestors()</a></code> </p>
        </td>
        <td><p>Plot locations of ancestors of given individual or node on a map</p></td>
      </tr><tr><td>
          <p><code><a href="time_direction.html">time_direction()</a></code> </p>
        </td>
        <td><p>Get direction of time implied by the history of the population</p></td>
      </tr><tr><td>
          <p><code><a href="split_time.html">split_time()</a></code> </p>
        </td>
        <td><p>Get split time of the given population</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="tree-sequence-loading-and-processing">Tree sequence loading and processing <a href="#tree-sequence-loading-and-processing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ts_load.html">ts_load()</a></code> </p>
        </td>
        <td><p>Load a tree sequence file produced by a given model</p></td>
      </tr><tr><td>
          <p><code><a href="ts_save.html">ts_save()</a></code> </p>
        </td>
        <td><p>Save a tree sequence to a file</p></td>
      </tr><tr><td>
          <p><code><a href="ts_recapitate.html">ts_recapitate()</a></code> </p>
        </td>
        <td><p>Recapitate the tree sequence</p></td>
      </tr><tr><td>
          <p><code><a href="ts_simplify.html">ts_simplify()</a></code> </p>
        </td>
        <td><p>Simplify the tree sequence down to a given set of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="ts_mutate.html">ts_mutate()</a></code> </p>
        </td>
        <td><p>Add mutations to the given tree sequence</p></td>
      </tr><tr><td>
          <p><code><a href="ts_coalesced.html">ts_coalesced()</a></code> </p>
        </td>
        <td><p>Check that all trees in the tree sequence are fully coalesced</p></td>
      </tr><tr><td>
          <p><code><a href="ts_samples.html">ts_samples()</a></code> </p>
        </td>
        <td><p>Extract names and times of individuals scheduled for sampling</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="tree-sequence-format-conversion">Tree sequence format conversion <a href="#tree-sequence-format-conversion" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ts_phylo.html">ts_phylo()</a></code> </p>
        </td>
        <td><p>Convert a tree in the tree sequence to an object of the class <code>phylo</code></p></td>
      </tr><tr><td>
          <p><code><a href="ts_genotypes.html">ts_genotypes()</a></code> </p>
        </td>
        <td><p>Extract genotype table from the tree sequence</p></td>
      </tr><tr><td>
          <p><code><a href="ts_eigenstrat.html">ts_eigenstrat()</a></code> </p>
        </td>
        <td><p>Extract genotypes from the tree sequence in the EIGENSTRAT format</p></td>
      </tr><tr><td>
          <p><code><a href="ts_vcf.html">ts_vcf()</a></code> </p>
        </td>
        <td><p>Save genotypes from the tree sequenceas a VCF file</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="accessing-tree-sequence-components">Accessing tree sequence components <a href="#accessing-tree-sequence-components" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ts_data.html">ts_data()</a></code> </p>
        </td>
        <td><p>Extract combined annotated table of individuals and nodes</p></td>
      </tr><tr><td>
          <p><code><a href="ts_individuals.html">ts_individuals()</a></code> <code><a href="ts_individuals.html">ts_edges()</a></code> <code><a href="ts_individuals.html">ts_nodes()</a></code> <code><a href="ts_individuals.html">ts_mutations()</a></code> </p>
        </td>
        <td><p>Get the table of individuals/nodes/edges/mutations from the tree sequence</p></td>
      </tr><tr><td>
          <p><code><a href="ts_tree.html">ts_tree()</a></code> </p>
        </td>
        <td><p>Get a tree from a given tree sequence</p></td>
      </tr><tr><td>
          <p><code><a href="ts_draw.html">ts_draw()</a></code> </p>
        </td>
        <td><p>Plot a graphical representation of a single tree</p></td>
      </tr><tr><td>
          <p><code><a href="ts_metadata.html">ts_metadata()</a></code> </p>
        </td>
        <td><p>Extract list with tree sequence metadata saved by SLiM</p></td>
      </tr><tr><td>
          <p><code><a href="ts_ancestors.html">ts_ancestors()</a></code> </p>
        </td>
        <td><p>Infer spatio-temporal ancestral history for given nodes/individuals</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="tree-sequence-statistics">Tree sequence statistics <a href="#tree-sequence-statistics" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ts_f4ratio.html">ts_f2()</a></code> <code><a href="ts_f4ratio.html">ts_f3()</a></code> <code><a href="ts_f4ratio.html">ts_f4()</a></code> <code><a href="ts_f4ratio.html">ts_f4ratio()</a></code> </p>
        </td>
        <td><p>Calculate the f2, f3, f4, and f4-ratio statistics</p></td>
      </tr><tr><td>
          <p><code><a href="ts_afs.html">ts_afs()</a></code> </p>
        </td>
        <td><p>Compute the allele frequency spectrum (AFS)</p></td>
      </tr><tr><td>
          <p><code><a href="ts_fst.html">ts_fst()</a></code> <code><a href="ts_fst.html">ts_divergence()</a></code> </p>
        </td>
        <td><p>Calculate pairwise statistics between sets of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="ts_diversity.html">ts_diversity()</a></code> </p>
        </td>
        <td><p>Calculate diversity in given sets of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="ts_tajima.html">ts_tajima()</a></code> </p>
        </td>
        <td><p>Calculate Tajima's D for given sets of individuals</p></td>
      </tr><tr><td>
          <p><code><a href="ts_segregating.html">ts_segregating()</a></code> </p>
        </td>
        <td><p>Calculate the density of segregating sites for the given sets of individuals</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a895c93f0e9eda05eb10b29a9921d586',
    indexName: 'slendr',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

