<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • slendr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://slendr.net/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">slendr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-question-circle"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette-00-installation.html">Installation instructions</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-08-nonslendr-tskit.html">Analyzing non-slendr tree sequences</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-09-paper.html">Examples from the slendr paper</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-10-tracts.html">Extracting true ancestry tracts</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-11-extensions.html">Extending models with custom SLiM code</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/bodkan/slendr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="compiling-and-running-population-genetic-models">Compiling and running population genetic models<a class="anchor" aria-label="anchor" href="#compiling-and-running-population-genetic-models"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="slim.html">slim()</a></code>

        </dt>
        <dd>Run a slendr model in SLiM</dd>

        <dt>

          <code><a href="msprime.html">msprime()</a></code>

        </dt>
        <dd>Run a slendr model in msprime</dd>

        <dt>

          <code><a href="compile_model.html">compile_model()</a></code>

        </dt>
        <dd>Compile a slendr demographic model</dd>

        <dt>

          <code><a href="read_model.html">read_model()</a></code>

        </dt>
        <dd>Read a previously serialized model configuration</dd>

        <dt>

          <code><a href="schedule_sampling.html">schedule_sampling()</a></code>

        </dt>
        <dd>Define sampling events for a given set of populations</dd>

        <dt>

          <code><a href="substitute_values.html">substitute_values()</a></code>

        </dt>
        <dd>Substitute values of parameters in a SLiM extension template</dd>
      </dl></div><div class="section level2">
      <h2 id="installation-and-configuration-of-external-dependencies">Installation and configuration of external dependencies<a class="anchor" aria-label="anchor" href="#installation-and-configuration-of-external-dependencies"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="setup_env.html">setup_env()</a></code>

        </dt>
        <dd>Setup a dedicated Python virtual environment for slendr</dd>

        <dt>

          <code><a href="init_env.html">init_env()</a></code>

        </dt>
        <dd>Activate slendr's own dedicated Python environment</dd>

        <dt>

          <code><a href="check_env.html">check_env()</a></code>

        </dt>
        <dd>Check that the active Python environment is setup for slendr</dd>

        <dt>

          <code><a href="clear_env.html">clear_env()</a></code>

        </dt>
        <dd>Remove the automatically created slendr Python environment</dd>

        <dt>

          <code><a href="get_python.html">get_python()</a></code>

        </dt>
        <dd>Get a path to internal Python interpreter of slendr</dd>

        <dt>

          <code><a href="check_dependencies.html">check_dependencies()</a></code>

        </dt>
        <dd>Check that the required dependencies are available for slendr to work</dd>
      </dl></div><div class="section level2">
      <h2 id="model-components">Model components<a class="anchor" aria-label="anchor" href="#model-components"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="population.html">population()</a></code>

        </dt>
        <dd>Define a population</dd>

        <dt>

          <code><a href="world.html">world()</a></code>

        </dt>
        <dd>Define a world map for all spatial operations</dd>
      </dl></div><div class="section level2">
      <h2 id="spatial-population-dynamics">Spatial population dynamics<a class="anchor" aria-label="anchor" href="#spatial-population-dynamics"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="move.html">move()</a></code>

        </dt>
        <dd>Move the population to a new location in a given amount of time</dd>

        <dt>

          <code><a href="expand_range.html">expand_range()</a></code>

        </dt>
        <dd>Expand the population range</dd>

        <dt>

          <code><a href="shrink_range.html">shrink_range()</a></code>

        </dt>
        <dd>Shrink the population range</dd>

        <dt>

          <code><a href="set_range.html">set_range()</a></code>

        </dt>
        <dd>Update the population range</dd>

        <dt>

          <code><a href="set_dispersal.html">set_dispersal()</a></code>

        </dt>
        <dd>Change dispersal parameters</dd>
      </dl></div><div class="section level2">
      <h2 id="non-spatial-population-dynamics">Non-spatial population dynamics<a class="anchor" aria-label="anchor" href="#non-spatial-population-dynamics"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="gene_flow.html">gene_flow()</a></code>

        </dt>
        <dd>Define a gene-flow event between two populations</dd>

        <dt>

          <code><a href="resize.html">resize()</a></code>

        </dt>
        <dd>Change the population size</dd>
      </dl></div><div class="section level2">
      <h2 id="manipulation-of-spatial-objects">Manipulation of spatial objects<a class="anchor" aria-label="anchor" href="#manipulation-of-spatial-objects"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="region.html">region()</a></code>

        </dt>
        <dd>Define a geographic region</dd>

        <dt>

          <code><a href="join.html">join()</a></code>

        </dt>
        <dd>Merge two spatial <code>slendr</code> objects into one</dd>

        <dt>

          <code><a href="overlap.html">overlap()</a></code>

        </dt>
        <dd>Generate the overlap of two <code>slendr</code> objects</dd>

        <dt>

          <code><a href="subtract.html">subtract()</a></code>

        </dt>
        <dd>Generate the difference between two <code>slendr</code> objects</dd>

        <dt>

          <code><a href="reproject.html">reproject()</a></code>

        </dt>
        <dd>Reproject coordinates between coordinate systems</dd>

        <dt>

          <code><a href="distance.html">distance()</a></code>

        </dt>
        <dd>Calculate the distance between a pair of spatial boundaries</dd>

        <dt>

          <code><a href="area.html">area()</a></code>

        </dt>
        <dd>Calculate the area covered by the given slendr object</dd>
      </dl></div><div class="section level2">
      <h2 id="model-visualization-and-diagnostics">Model visualization and diagnostics<a class="anchor" aria-label="anchor" href="#model-visualization-and-diagnostics"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="plot_map.html">plot_map()</a></code>

        </dt>
        <dd>Plot <code>slendr</code> geographic features on a map</dd>

        <dt>

          <code><a href="plot_model.html">plot_model()</a></code>

        </dt>
        <dd>Plot demographic history encoded in a slendr model</dd>

        <dt>

          <code><a href="animate_model.html">animate_model()</a></code>

        </dt>
        <dd>Animate the simulated population dynamics</dd>

        <dt>

          <code><a href="explore_model.html">explore_model()</a></code>

        </dt>
        <dd>Open an interactive browser of the spatial model</dd>

        <dt>

          <code><a href="print.slendr_pop.html">print(<i>&lt;slendr_pop&gt;</i>)</a></code> <code><a href="print.slendr_pop.html">print(<i>&lt;slendr_region&gt;</i>)</a></code> <code><a href="print.slendr_pop.html">print(<i>&lt;slendr_map&gt;</i>)</a></code> <code><a href="print.slendr_pop.html">print(<i>&lt;slendr_model&gt;</i>)</a></code>

        </dt>
        <dd>Print a short summary of a <code>slendr</code> object</dd>

        <dt>

          <code><a href="print.slendr_ts.html">print(<i>&lt;slendr_ts&gt;</i>)</a></code>

        </dt>
        <dd>Print tskit's summary table of the Python tree-sequence object</dd>

        <dt>

          <code><a href="summary.slendr_nodes.html">summary(<i>&lt;slendr_nodes&gt;</i>)</a></code>

        </dt>
        <dd>Summarise the contents of a <code>ts_nodes</code> result</dd>

        <dt>

          <code><a href="extract_parameters.html">extract_parameters()</a></code>

        </dt>
        <dd>Extract information from a compiled model or a simulated tree sequence</dd>
      </dl></div><div class="section level2">
      <h2 id="processing-tree-sequences">Processing tree sequences<a class="anchor" aria-label="anchor" href="#processing-tree-sequences"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="ts_read.html">ts_read()</a></code>

        </dt>
        <dd>Read a tree sequence from a file</dd>

        <dt>

          <code><a href="ts_write.html">ts_write()</a></code>

        </dt>
        <dd>Save a tree sequence to a file</dd>

        <dt>

          <code><a href="ts_recapitate.html">ts_recapitate()</a></code>

        </dt>
        <dd>Recapitate the tree sequence</dd>

        <dt>

          <code><a href="ts_simplify.html">ts_simplify()</a></code>

        </dt>
        <dd>Simplify the tree sequence down to a given set of individuals</dd>

        <dt>

          <code><a href="ts_mutate.html">ts_mutate()</a></code>

        </dt>
        <dd>Add mutations to the given tree sequence</dd>

        <dt>

          <code><a href="ts_samples.html">ts_samples()</a></code>

        </dt>
        <dd>Extract names and times of individuals of interest in the current tree sequence (either all sampled individuals or those that the user simplified to)</dd>

        <dt>

          <code><a href="ts_names.html">ts_names()</a></code>

        </dt>
        <dd>Extract names of individuals in a tree sequence</dd>

        <dt>

          <code><a href="ts_coalesced.html">ts_coalesced()</a></code>

        </dt>
        <dd>Check that all trees in the tree sequence are fully coalesced</dd>

        <dt>

          <code><a href="ts_save.html">ts_save()</a></code>

        </dt>
        <dd>Write a tree sequence to a file</dd>

        <dt>

          <code><a href="ts_load.html">ts_load()</a></code>

        </dt>
        <dd>Read a tree sequence from a file</dd>
      </dl></div><div class="section level2">
      <h2 id="tree-sequence-format-conversion">Tree sequence format conversion<a class="anchor" aria-label="anchor" href="#tree-sequence-format-conversion"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="ts_genotypes.html">ts_genotypes()</a></code>

        </dt>
        <dd>Extract genotype table from the tree sequence</dd>

        <dt>

          <code><a href="ts_eigenstrat.html">ts_eigenstrat()</a></code>

        </dt>
        <dd>Convert genotypes to the EIGENSTRAT file format</dd>

        <dt>

          <code><a href="ts_vcf.html">ts_vcf()</a></code>

        </dt>
        <dd>Save genotypes from the tree sequence as a VCF file</dd>
      </dl></div><div class="section level2">
      <h2 id="accessing-tree-sequence-components">Accessing tree sequence components<a class="anchor" aria-label="anchor" href="#accessing-tree-sequence-components"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="ts_nodes.html">ts_nodes()</a></code>

        </dt>
        <dd>Extract combined annotated table of individuals and nodes</dd>

        <dt>

          <code><a href="ts_edges.html">ts_edges()</a></code>

        </dt>
        <dd>Extract spatio-temporal edge annotation table from a given tree or tree sequence</dd>

        <dt>

          <code><a href="ts_table.html">ts_table()</a></code>

        </dt>
        <dd>Get the table of individuals/nodes/edges/mutations/sites from the tree sequence</dd>

        <dt>

          <code><a href="ts_phylo.html">ts_phylo()</a></code>

        </dt>
        <dd>Convert a tree in the tree sequence to an object of the class <code>phylo</code></dd>

        <dt>

          <code><a href="ts_tree.html">ts_tree()</a></code>

        </dt>
        <dd>Get a tree from a given tree sequence</dd>

        <dt>

          <code><a href="ts_draw.html">ts_draw()</a></code>

        </dt>
        <dd>Plot a graphical representation of a single tree</dd>

        <dt>

          <code><a href="ts_metadata.html">ts_metadata()</a></code>

        </dt>
        <dd>Extract list with tree sequence metadata saved by SLiM</dd>

        <dt>

          <code><a href="ts_ancestors.html">ts_ancestors()</a></code>

        </dt>
        <dd>Extract (spatio-)temporal ancestral history for given nodes/individuals</dd>

        <dt>

          <code><a href="ts_descendants.html">ts_descendants()</a></code>

        </dt>
        <dd>Extract all descendants of a given tree-sequence node</dd>
      </dl></div><div class="section level2">
      <h2 id="tree-sequence-statistics">Tree sequence statistics<a class="anchor" aria-label="anchor" href="#tree-sequence-statistics"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="ts_f4ratio.html">ts_f2()</a></code> <code><a href="ts_f4ratio.html">ts_f3()</a></code> <code><a href="ts_f4ratio.html">ts_f4()</a></code> <code><a href="ts_f4ratio.html">ts_f4ratio()</a></code>

        </dt>
        <dd>Calculate the f2, f3, f4, and f4-ratio statistics</dd>

        <dt>

          <code><a href="ts_afs.html">ts_afs()</a></code>

        </dt>
        <dd>Compute the allele frequency spectrum (AFS)</dd>

        <dt>

          <code><a href="ts_divergence.html">ts_divergence()</a></code>

        </dt>
        <dd>Calculate pairwise divergence between sets of individuals</dd>

        <dt>

          <code><a href="ts_diversity.html">ts_diversity()</a></code>

        </dt>
        <dd>Calculate diversity in given sets of individuals</dd>

        <dt>

          <code><a href="ts_fst.html">ts_fst()</a></code>

        </dt>
        <dd>Calculate pairwise statistics between sets of individuals</dd>

        <dt>

          <code><a href="ts_tajima.html">ts_tajima()</a></code>

        </dt>
        <dd>Calculate Tajima's D for given sets of individuals</dd>

        <dt>

          <code><a href="ts_segregating.html">ts_segregating()</a></code>

        </dt>
        <dd>Calculate the density of segregating sites for the given sets of individuals</dd>

        <dt>

          <code><a href="ts_ibd.html">ts_ibd()</a></code>

        </dt>
        <dd>Collect Identity-by-Descent (IBD) segments (EXPERIMENTAL)</dd>

        <dt>

          <code><a href="ts_tracts.html">ts_tracts()</a></code>

        </dt>
        <dd>Extract ancestry tracts from a tree sequence (EXPERIMENTAL)</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

