<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Save genotypes from the tree sequence as a VCF file — ts_vcf • slendr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Save genotypes from the tree sequence as a VCF file — ts_vcf"><meta name="description" content="This function writes a VCF file with diploid genotypes from a given tree sequence."><meta property="og:description" content="This function writes a VCF file with diploid genotypes from a given tree sequence."><meta property="og:image" content="https://www.slendr.net/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">slendr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-question-circle"></span> Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true"><span class="fa fa-book"></span> Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette-00-installation.html">Installation instructions</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-01-tutorial.html">Introduction and basic tutorial</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-02-grid-model.html">Demes on a regular spatial grid</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-03-interactions.html">Programming dispersion dynamics</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-04-nonspatial-models.html">Traditional non-spatial models</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-05-tree-sequences.html">Tree sequence processing and statistics</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-06-locations.html">Spatially annotated tree sequences</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-07-backends.html">Simulating data with SLiM and msprime backends</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-08-nonslendr-tskit.html">Analyzing non-slendr tree sequences</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-09-paper.html">Examples from the slendr paper</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-10-tracts.html">Extracting true ancestry tracts</a></li>
    <li><a class="dropdown-item" href="../articles/vignette-11-extensions.html">Extending models with custom SLiM code</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/bodkan/slendr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://twitter.com/fleventy5"><span class="fa fa-twitter"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Save genotypes from the tree sequence as a VCF file</h1>
      <small class="dont-index">Source: <a href="https://github.com/bodkan/slendr/blob/HEAD/R/tree-sequences.R" class="external-link"><code>R/tree-sequences.R</code></a></small>
      <div class="d-none name"><code>ts_vcf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function writes a VCF file with diploid genotypes from a given tree sequence.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ts_vcf</span><span class="op">(</span></span>
<span>  <span class="va">ts</span>,</span>
<span>  <span class="va">path</span>,</span>
<span>  chrom <span class="op">=</span> <span class="st">"chr1"</span>,</span>
<span>  individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  position_transform <span class="op">=</span> <span class="st">"lambda x: np.fmax(1, x)"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ts">ts<a class="anchor" aria-label="anchor" href="#arg-ts"></a></dt>
<dd><p>Tree sequence object of the class <code>slendr_ts</code></p></dd>


<dt id="arg-path">path<a class="anchor" aria-label="anchor" href="#arg-path"></a></dt>
<dd><p>Path to a VCF file</p></dd>


<dt id="arg-chrom">chrom<a class="anchor" aria-label="anchor" href="#arg-chrom"></a></dt>
<dd><p>Chromosome name to be written in the CHROM column of the VCF
(default value will be "chr1").</p></dd>


<dt id="arg-individuals">individuals<a class="anchor" aria-label="anchor" href="#arg-individuals"></a></dt>
<dd><p>A vector of individuals in the tree sequence to extract
genotypes from. If missing, all individuals present in the tree sequence
will be saved. For a slendr-based tree sequence a character vector of
individual names is expected. For non-slendr tree sequences, a numeric
vector of IDs of individuals is expected.</p></dd>


<dt id="arg-position-transform">position_transform<a class="anchor" aria-label="anchor" href="#arg-position-transform"></a></dt>
<dd><p>How to transform coordinates in a tree sequence
to coordinates in a VCF file? By default, any site with coordinate 0
is converted to a position 1 to ensure that the resulting VCF file adheres
to the VCF specification. Setting this to <code>NULL</code> will disable this</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>No return value, called for side effects</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Users should note that, as with many other tskit-based slendr functions,
<code>ts_vcf</code> is intended to provide some convenient defaults. For instance,
even for non-slendr tree sequences, it will name each individual in the
genotype columns after their integer IDs. In other words, if the
<code>individuals</code> function argument is given as <code>c(1, 42, 123)</code>, the
individuals will be named as "ind_1", "ind_42", and "ind_123", instead of
"tsk_0", "tsk_1", and "tsk_2". That said, the reticulate-based Python
interface of slendr allows calling the <code>write_vcf</code> function of tskit
directly!</p>
<p>By default, simulating a tree sequence with msprime and exporting the
genotypes into VCF can cause issues with some downstream software because
the VCF specification does not allow sites with the position 0. By default
<code>ts_vcf</code> automatically transforms a site with a zero coordinate to
a coordinate 1. Setting <code>position_transform</code> to NULL will disable
this, and <code>tsv_vcf</code> will save coordinates in their original form.
See this discussion for more detail:
<a href="https://github.com/tskit-dev/tskit/issues/2838#issuecomment-1931796988" class="external-link">https://github.com/tskit-dev/tskit/issues/2838#issuecomment-1931796988</a>,
as well as relevant topics in the tskit documentation on this issue, like
here: <a href="https://tskit.dev/tskit/docs/latest/export.html#modifying-coordinates" class="external-link">https://tskit.dev/tskit/docs/latest/export.html#modifying-coordinates</a>.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

